// @vectojs/core/dist/chunk-XIEQHSBB.mjs
import {
  ArabicShaper,
  BidiResolver,
  LayoutWorkerManager
} from "./chunk-IESDTEJ4.mjs";

// src/math/SpringPhysics.ts
var MAX_FRAME_DT = 0.25;
var MAX_STEP_DT = 1 / 120;
var SpringPhysics = class {
  value;
  target;
  velocity = 0;
  stiffness = 180;
  damping = 12;
  mass = 1;
  valEpsilon = 5e-3;
  velEpsilon = 5e-3;
  constructor(initial) {
    this.value = initial;
    this.target = initial;
  }
  update(dt) {
    if (this.isAtRest()) {
      this.value = this.target;
      this.velocity = 0;
      return;
    }
    if (!(dt > 0)) return;
    let remaining = dt < MAX_FRAME_DT ? dt : MAX_FRAME_DT;
    while (remaining > 0) {
      const step = remaining < MAX_STEP_DT ? remaining : MAX_STEP_DT;
      const forceSpring = -this.stiffness * (this.value - this.target);
      const forceDamping = -this.damping * this.velocity;
      const acceleration = (forceSpring + forceDamping) / this.mass;
      this.velocity += acceleration * step;
      this.value += this.velocity * step;
      remaining -= step;
      if (this.isAtRest()) {
        this.value = this.target;
        this.velocity = 0;
        return;
      }
    }
  }
  isAtRest() {
    return Math.abs(this.value - this.target) < this.valEpsilon && Math.abs(this.velocity) < this.velEpsilon;
  }
};

// src/animation/easing.ts
var c1 = 1.70158;
var c3 = c1 + 1;
var Easing = {
  linear: (t) => t,
  easeInQuad: (t) => t * t,
  easeOutQuad: (t) => t * (2 - t),
  easeInOutQuad: (t) => t < 0.5 ? 2 * t * t : 1 - Math.pow(-2 * t + 2, 2) / 2,
  easeInCubic: (t) => t * t * t,
  easeOutCubic: (t) => 1 - Math.pow(1 - t, 3),
  easeInOutCubic: (t) => t < 0.5 ? 4 * t * t * t : 1 - Math.pow(-2 * t + 2, 3) / 2,
  easeOutBack: (t) => 1 + c3 * Math.pow(t - 1, 3) + c1 * Math.pow(t - 1, 2),
  easeInOutBack: (t) => {
    const c2 = c1 * 1.525;
    return t < 0.5 ? Math.pow(2 * t, 2) * ((c2 + 1) * 2 * t - c2) / 2 : (Math.pow(2 * t - 2, 2) * ((c2 + 1) * (t * 2 - 2) + c2) + 2) / 2;
  }
};

// src/animation/drivers.ts
function isTweenConfig(c) {
  return typeof c === "object" && "duration" in c;
}
var TweenDriver = class {
  value;
  from;
  to;
  elapsed = 0;
  duration;
  delay;
  ease;
  constructor(from, to, cfg) {
    this.value = from;
    this.from = from;
    this.to = to;
    this.duration = Math.max(1, cfg.duration);
    this.delay = cfg.delay ?? 0;
    this.ease = typeof cfg.easing === "function" ? cfg.easing : Easing[cfg.easing ?? "easeOutQuad"];
  }
  get target() {
    return this.to;
  }
  retarget(to) {
    this.from = this.value;
    this.to = to;
    this.elapsed = 0;
  }
  tick(dtMs) {
    this.elapsed += dtMs;
    const active = this.elapsed - this.delay;
    if (active <= 0) return;
    const p = Math.min(active / this.duration, 1);
    this.value = this.from + (this.to - this.from) * this.ease(p);
  }
  isDone() {
    return this.elapsed - this.delay >= this.duration;
  }
};
var SpringDriver = class {
  spring;
  constructor(from, to, cfg) {
    this.spring = new SpringPhysics(from);
    if (cfg.stiffness !== void 0) this.spring.stiffness = cfg.stiffness;
    if (cfg.damping !== void 0) this.spring.damping = cfg.damping;
    if (cfg.mass !== void 0) this.spring.mass = cfg.mass;
    this.spring.target = to;
  }
  get value() {
    return this.spring.value;
  }
  get target() {
    return this.spring.target;
  }
  retarget(to) {
    this.spring.target = to;
  }
  tick(dtMs) {
    this.spring.update(dtMs / 1e3);
  }
  isDone() {
    return this.spring.isAtRest();
  }
};

// src/tree/Entity.ts
var ANIMATABLE_PROPS = /* @__PURE__ */ new Set([
  "x",
  "y",
  "scaleX",
  "scaleY",
  "rotation",
  "opacity"
]);
var VectoJSEvent = class {
  /** The event name. */
  type;
  /** The entity the event originated on. */
  target;
  /** The entity whose listeners are currently running (updated per node). */
  currentTarget;
  /** The wrapped browser event, if any. */
  nativeEvent;
  /** Whether the event bubbles past its target (capture always runs). */
  bubbles;
  explicitScenePoint;
  stopped = false;
  stoppedImmediate = false;
  constructor(type, target, nativeEvent, bubbles = true, scenePoint) {
    this.type = type;
    this.target = target;
    this.currentTarget = target;
    this.nativeEvent = nativeEvent;
    this.bubbles = bubbles;
    this.explicitScenePoint = scenePoint;
  }
  /** Stop the event from reaching the next node in the propagation path. */
  stopPropagation() {
    this.stopped = true;
  }
  /** Stop propagation AND skip any remaining listeners on the current node. */
  stopImmediatePropagation() {
    this.stopped = true;
    this.stoppedImmediate = true;
  }
  /** Forward to the native event's `preventDefault` (e.g. stop page scroll). */
  preventDefault() {
    this.nativeEvent?.preventDefault?.();
  }
  /** Whether {@link stopPropagation} has been called. */
  get propagationStopped() {
    return this.stopped;
  }
  /** Whether {@link stopImmediatePropagation} has been called. */
  get immediatePropagationStopped() {
    return this.stoppedImmediate;
  }
  /** Whether the native event's default action was prevented. */
  get defaultPrevented() {
    return !!this.nativeEvent?.defaultPrevented;
  }
  /** Native horizontal wheel delta, if this wraps a `WheelEvent`. */
  get deltaX() {
    return this.nativeEvent?.deltaX;
  }
  /** Native vertical wheel delta, if this wraps a `WheelEvent`. */
  get deltaY() {
    return this.nativeEvent?.deltaY;
  }
  /** Native pointer X, if this wraps a pointer/mouse event. */
  get clientX() {
    return this.nativeEvent?.clientX;
  }
  /** Native pointer Y, if this wraps a pointer/mouse event. */
  get clientY() {
    return this.nativeEvent?.clientY;
  }
  get resolvedScenePoint() {
    if (this.explicitScenePoint) return this.explicitScenePoint;
    const native = this.nativeEvent;
    if (native?.vectoSceneX !== void 0 && native.vectoSceneY !== void 0) {
      return { x: native.vectoSceneX, y: native.vectoSceneY };
    }
    if (native?.clientX === void 0 || native.clientY === void 0) return void 0;
    const scene = this.target.scene;
    return scene?.clientToScene?.(native.clientX, native.clientY) ?? {
      x: native.clientX,
      y: native.clientY
    };
  }
  /** Pointer X in the Scene's logical coordinate space. */
  get sceneX() {
    return this.resolvedScenePoint?.x;
  }
  /** Pointer Y in the Scene's logical coordinate space. */
  get sceneY() {
    return this.resolvedScenePoint?.y;
  }
  /** Pointer X local to the entity whose listener is currently running. */
  get localX() {
    const point = this.resolvedScenePoint;
    if (!point) return void 0;
    return this.currentTarget.worldToLocal(point.x, point.y)?.x;
  }
  /** Pointer Y local to the entity whose listener is currently running. */
  get localY() {
    const point = this.resolvedScenePoint;
    if (!point) return void 0;
    return this.currentTarget.worldToLocal(point.x, point.y)?.y;
  }
  get shiftKey() {
    return !!this.nativeEvent?.shiftKey;
  }
  get ctrlKey() {
    return !!this.nativeEvent?.ctrlKey;
  }
  get altKey() {
    return !!this.nativeEvent?.altKey;
  }
  get metaKey() {
    return !!this.nativeEvent?.metaKey;
  }
  /** Native key, if this wraps a keyboard event. */
  get key() {
    return this.nativeEvent?.key;
  }
};
var Entity = class {
  id;
  children = [];
  parent = null;
  /**
   * Walk up the parent chain to find the scene this entity is currently attached to.
   */
  get scene() {
    if (this._scene) return this._scene;
    return this.parent ? this.parent.scene : null;
  }
  _x = 0;
  _y = 0;
  _scaleX = 1;
  _scaleY = 1;
  _rotation = 0;
  _opacity = 1;
  // Fast-path flag: false for the overwhelming majority of entities (incl. the
  // Danmaku hot loop), so a bare `entity.x = v` is one boolean check + field write.
  _hasTransitions = false;
  _transitions = null;
  _drivers = /* @__PURE__ */ new Map();
  _mounted = false;
  get x() {
    return this._x;
  }
  set x(v) {
    if (this._hasTransitions) this._animateProp("x", v);
    else this._x = v;
  }
  get y() {
    return this._y;
  }
  set y(v) {
    if (this._hasTransitions) this._animateProp("y", v);
    else this._y = v;
  }
  get scaleX() {
    return this._scaleX;
  }
  set scaleX(v) {
    if (this._hasTransitions) this._animateProp("scaleX", v);
    else this._scaleX = v;
  }
  get scaleY() {
    return this._scaleY;
  }
  set scaleY(v) {
    if (this._hasTransitions) this._animateProp("scaleY", v);
    else this._scaleY = v;
  }
  get rotation() {
    return this._rotation;
  }
  set rotation(v) {
    if (this._hasTransitions) this._animateProp("rotation", v);
    else this._rotation = v;
  }
  get opacity() {
    return this._opacity;
  }
  set opacity(v) {
    if (this._hasTransitions) this._animateProp("opacity", v);
    else this._opacity = v;
  }
  isDOMPortal = false;
  _interactive = false;
  get interactive() {
    return this._interactive;
  }
  set interactive(val) {
    if (this._interactive !== val) {
      this._interactive = val;
      const s = this.scene;
      if (s) {
        s.a11yNeedsReorder = true;
        s.markDirty();
      }
    }
  }
  width = 0;
  height = 0;
  a11yOffsetX = 0;
  a11yOffsetY = 0;
  /**
   * Opt in to a viewport-filling accessibility/automation shadow node even when
   * this entity has no intrinsic box (`width`/`height` of `0`). Use for
   * full-screen, boundless interaction surfaces (e.g. an infinite-canvas graph)
   * that need global pointer events. The node is mounted behind all other shadow
   * nodes, so on-top components stay clickable.
   */
  a11yFullViewport = false;
  /**
   * Clip this node's children to its local box (`[0,0]–[width,height]`) while
   * rendering. Combined with translating a content child, this is how
   * scroll/overflow containers (e.g. `ScrollView`) keep their content inside a
   * fixed viewport. Off by default (children render unclipped). Canvas2D only.
   */
  clipChildren = false;
  listeners = /* @__PURE__ */ new Map();
  /** Capture-phase listeners (fired root→target before bubble). */
  captureListeners = /* @__PURE__ */ new Map();
  animations = [];
  constructor(id) {
    this.id = id || `entity_${Math.random().toString(36).substring(2, 9)}`;
  }
  /**
   * Append a child entity to this node's children array.
   *
   * If `child` already has a parent (including `this`), it's detached from it
   * first — otherwise re-adding an already-added child duplicates it in
   * `children[]` (one `remove()` call only strips the first occurrence,
   * leaving a stale entry that keeps rendering/updating despite
   * `child.parent` reporting `null`), and re-parenting to a different entity
   * without an explicit `remove()` first leaves the old parent holding a
   * stale reference whose own `.parent` disagrees with where it now lives.
   * `child.parent` is only ever `null` or `this`'s ultimate owner, so this
   * check is O(1) for the overwhelming common case (a brand-new entity).
   *
   * Accepts multiple children in one call (`parent.add(a, b, c)`); each is
   * attached in argument order with the same detach-then-attach semantics.
   *
   * @param children - One or more entities to add as children.
   * @returns `this` for method chaining.
   */
  add(...children) {
    for (const child of children) this._addOne(child);
    return this;
  }
  /** Attach a single child (the O(1) common path). See {@link add}. */
  _addOne(child) {
    if (child.parent) child.parent.remove(child);
    child.parent = this;
    this.children.push(child);
    const s = this.scene;
    if (s) {
      s.a11yNeedsReorder = true;
      s.markDirty();
      child._notifyMounted();
    }
  }
  /** Called once when this entity becomes attached to a live Scene. Override to react. */
  onMounted() {
  }
  /** Fire onMounted for this node and its descendants, guarded against double-fire. */
  _notifyMounted() {
    if (this._mounted) return;
    this._mounted = true;
    this.onMounted();
    for (const c of this.children) c._notifyMounted();
  }
  /**
   * Remove a child entity from this node.
   *
   * @param child - The entity to remove.
   * @returns `this` for method chaining.
   */
  remove(child) {
    const index = this.children.indexOf(child);
    if (index !== -1) {
      this.children.splice(index, 1);
      child.parent = null;
      const s = this.scene;
      if (s) {
        s.detachA11y(child);
        s.a11yNeedsReorder = true;
        s.markDirty();
      }
    }
    return this;
  }
  /**
   * Assign several own properties in one call, each through its normal setter
   * (so a property with a configured {@link setTransition} still animates, and
   * `interactive` still flags the scene for a11y reorder). A construction-time
   * ergonomic only — it is a plain `for…in` over the given object and touches
   * no per-frame path.
   *
   * @param props - Partial set of this entity's own properties to assign.
   * @returns `this` for method chaining.
   * @example rect.set({ x: 40, y: 40, width: 120, fill: '#38bdf8' });
   */
  set(props) {
    for (const key in props) {
      const value = props[key];
      if (value !== void 0) this[key] = value;
    }
    return this;
  }
  /**
   * Set the local position of this entity.
   *
   * @param x - Horizontal position in local space.
   * @param y - Vertical position in local space.
   * @returns `this` for method chaining.
   * @example entity.setPosition(100, 200);
   */
  setPosition(x, y) {
    this.x = x;
    this.y = y;
    return this;
  }
  /**
   * Queue a tween animation toward the specified target property values.
   *
   * Multiple calls chain animations sequentially.  Only numeric properties
   * are interpolated; non-numeric values are ignored.
   *
   * @param targetProps - Partial set of numeric properties to tween to.
   * @param durationMs - Duration of the tween in milliseconds.
   * @returns `this` for method chaining.
   * @example entity.animate({ x: 400, opacity: 0 }, 500);
   */
  animate(targetProps, durationMs) {
    this.animations.push({
      target: targetProps,
      duration: durationMs,
      startTime: -1,
      startProps: {}
    });
    this.scene?.markDirty();
    return this;
  }
  /** Write a driver-computed value to a backing field without re-triggering the setter. */
  _applyAnimated(prop, v) {
    switch (prop) {
      case "x":
        this._x = v;
        break;
      case "y":
        this._y = v;
        break;
      case "scaleX":
        this._scaleX = v;
        break;
      case "scaleY":
        this._scaleY = v;
        break;
      case "rotation":
        this._rotation = v;
        break;
      case "opacity":
        this._opacity = v;
        break;
    }
  }
  _currentOf(prop) {
    switch (prop) {
      case "x":
        return this._x;
      case "y":
        return this._y;
      case "scaleX":
        return this._scaleX;
      case "scaleY":
        return this._scaleY;
      case "rotation":
        return this._rotation;
      case "opacity":
        return this._opacity;
    }
  }
  /**
   * Write a value immediately, bypassing any configured transition. For subclasses
   * that need to seed a starting state (e.g. the presence helper's enter `from`).
   */
  setImmediate(prop, v) {
    const existing = this._drivers.get(prop);
    if (existing) this._settleDriver(existing);
    this._drivers.delete(prop);
    this._applyAnimated(prop, v);
  }
  _settleDriver(driver) {
    const active = driver;
    const onDone = active.onDone;
    active.onDone = void 0;
    onDone?.();
  }
  _spawnDriver(prop, to, cfg) {
    if (prop !== "opacity" && this.scene?.prefersReducedMotion) {
      const existing2 = this._drivers.get(prop);
      if (existing2) this._settleDriver(existing2);
      this._drivers.delete(prop);
      this._applyAnimated(prop, to);
      return;
    }
    const existing = this._drivers.get(prop);
    if (existing) {
      this._settleDriver(existing);
      existing.retarget(to);
      return;
    }
    const from = this._currentOf(prop);
    const driver = isTweenConfig(cfg) ? new TweenDriver(from, to, cfg) : new SpringDriver(from, to, cfg === "spring" ? {} : cfg);
    this._drivers.set(prop, driver);
    this.scene?.markDirty();
  }
  /** Assignment path when a declarative transition is configured for `prop`. */
  _animateProp(prop, to) {
    const cfg = this._transitions?.get(prop);
    if (!cfg) {
      this._applyAnimated(prop, to);
      return;
    }
    this._spawnDriver(prop, to, cfg);
  }
  /** Declare which properties animate, and how. Subsequent assignment animates them. */
  setTransition(config) {
    this._transitions ??= /* @__PURE__ */ new Map();
    for (const [k, v] of Object.entries(config))
      this._transitions.set(k, v);
    this._hasTransitions = this._transitions.size > 0;
    return this;
  }
  /** Imperative tween toward targets; resolves when all reach their end. */
  animateTo(props, cfg) {
    return this._driveTo(props, cfg);
  }
  /** Imperative spring toward targets; resolves when all reach rest. */
  springTo(props, cfg = {}) {
    return this._driveTo(props, cfg);
  }
  _driveTo(props, cfg) {
    const entries = Object.entries(props);
    return Promise.all(
      entries.map(
        (e) => new Promise((resolve) => {
          this._spawnDriver(e[0], e[1], cfg);
          const d = this._drivers.get(e[0]);
          if (!d)
            resolve();
          else d.onDone = resolve;
        })
      )
    ).then(() => void 0);
  }
  /** Advance active property drivers one frame. Call from update(). */
  tickDrivers(dt) {
    if (this._drivers.size === 0) return;
    for (const [prop, driver] of this._drivers) {
      driver.tick(dt);
      if (driver.isDone()) {
        this._applyAnimated(prop, driver.target);
        this._settleDriver(driver);
        this._drivers.delete(prop);
      } else {
        this._applyAnimated(prop, driver.value);
      }
    }
    this.scene?.markDirty();
  }
  /**
   * Advance the entity's internal state for one frame.
   *
   * Called automatically by the {@link Scene} render loop — override in
   * subclasses to implement custom per-frame logic.
   *
   * @param dt - Elapsed time since the last frame in milliseconds.
   * @param time - Absolute timestamp from `performance.now()`.
   */
  update(dt, time) {
    this.tickDrivers(dt);
    if (this.animations.length > 0) {
      const anim = this.animations[0];
      if (anim.startTime === -1) {
        anim.startTime = time;
        for (const key in anim.target) {
          anim.startProps[key] = this[key];
        }
      }
      const progress = Math.min((time - anim.startTime) / anim.duration, 1);
      for (const key in anim.target) {
        const start = anim.startProps[key];
        const end = anim.target[key];
        if (typeof start === "number" && typeof end === "number") {
          const easeOut = progress * (2 - progress);
          const value = start + (end - start) * easeOut;
          if (ANIMATABLE_PROPS.has(key)) {
            this._applyAnimated(key, value);
          } else {
            this[key] = value;
          }
        }
      }
      if (progress >= 1) {
        this.animations.shift();
      }
    }
  }
  /**
   * Register a listener for a {@link VectoEvent}.
   *
   * Listeners run in the bubble phase by default; pass `{ capture: true }` for the
   * capture phase (root→target). Bubble listeners also fire for the legacy
   * {@link emit} (direct, self-only) path.
   *
   * @param event - The event name to listen for.
   * @param callback - Handler invoked when the event fires.
   * @param options - `{ capture }` to register for the capture phase.
   * @returns `this` for method chaining.
   * @example entity.on('click', (e) => console.log('clicked', e));
   */
  on(event, callback, options) {
    const map = options?.capture ? this.captureListeners : this.listeners;
    if (!map.has(event)) {
      map.set(event, []);
    }
    map.get(event).push(callback);
    return this;
  }
  /**
   * Remove a previously registered event listener.
   *
   * @param event - The event name to stop listening to.
   * @param callback - The exact handler reference passed to {@link on}.
   * @param options - Must match the phase the listener was registered with.
   * @returns `this` for method chaining.
   */
  off(event, callback, options) {
    const handlers = (options?.capture ? this.captureListeners : this.listeners).get(event);
    if (handlers) {
      const idx = handlers.indexOf(callback);
      if (idx !== -1) handlers.splice(idx, 1);
    }
    return this;
  }
  /**
   * Tear down this entity: clear all animations, event listeners, and detach
   * from parent. Call before discarding an entity to prevent memory leaks.
   */
  destroy() {
    this.animations = [];
    for (const driver of this._drivers.values()) {
      this._settleDriver(driver);
    }
    this._drivers.clear();
    this.listeners.clear();
    this.captureListeners.clear();
    if (this.parent) {
      this.parent.remove(this);
    }
  }
  /**
   * Dispatch a {@link VectoEvent} directly to this entity's bubble-phase listeners
   * only — no tree propagation. Kept for component-internal/self events (e.g. a
   * form control emitting its own `change`); use {@link dispatchEvent} for the
   * capture/bubble path.
   *
   * @param event - The event name to dispatch.
   * @param payload - Arbitrary data forwarded to each listener.
   */
  emit(event, payload) {
    const handlers = this.listeners.get(event);
    if (handlers) {
      handlers.forEach((h) => h(payload));
    }
  }
  /**
   * Programmatically focus the entity's projected a11y shadow element, if one
   * exists. After `scene.add(entity)` the shadow element is typically created
   * on the next a11y sync (within one animation frame), so this method retries
   * once on the next rAF if the element isn't immediately available — matching
   * the timing pattern the prior workaround described in findings.md achieved
   * with `requestAnimationFrame(() => document.getElementById(id)?.focus())`.
   */
  focus() {
    const el = this.scene?.getA11yElement(this.id);
    if (el) {
      el.focus();
      return;
    }
    requestAnimationFrame(() => {
      const retry = this.scene?.getA11yElement(this.id);
      if (retry) retry.focus();
    });
  }
  /** Run one node's listeners for the event, honoring stopImmediatePropagation. */
  fireListeners(node, map, event) {
    const handlers = map.get(event.type);
    if (!handlers) return;
    event.currentTarget = node;
    for (const h of handlers.slice()) {
      h(event);
      if (event.immediatePropagationStopped) return;
    }
  }
  /**
   * Dispatch a {@link VectoJSEvent} through the entity tree, DOM-style: a capture
   * phase from the root down to `event.target`, then a bubble phase back up to the
   * root. `event.stopPropagation()` halts the walk; `stopImmediatePropagation()`
   * also skips the remaining listeners on the current node. A non-bubbling event
   * only fires its target in the bubble phase (capture still runs).
   *
   * @param event - The event to propagate (its `target` defines the path).
   */
  dispatchEvent(event) {
    const path = [];
    for (let n = event.target; n; n = n.parent) path.push(n);
    for (let i = path.length - 1; i >= 0; i--) {
      if (event.propagationStopped) return;
      this.fireListeners(path[i], path[i].captureListeners, event);
    }
    for (let i = 0; i < path.length; i++) {
      if (event.propagationStopped) return;
      this.fireListeners(path[i], path[i].listeners, event);
      if (!event.bubbles) return;
    }
  }
  /**
   * Compute the entity's position in world/canvas space by accumulating
   * local offsets up the scene-graph hierarchy using affine transformations (scale and rotation).
   *
   * @returns World-space {@link Point} for this entity.
   */
  getGlobalPosition() {
    return this.localToWorld(0, 0);
  }
  /**
   * Return the exact accumulated Canvas `T * S * R` transform for this entity.
   */
  getWorldTransform() {
    const path = [this];
    let ancestor = this.parent;
    while (ancestor) {
      path.push(ancestor);
      ancestor = ancestor.parent;
    }
    let a = 1;
    let b = 0;
    let c = 0;
    let d = 1;
    let e = 0;
    let f = 0;
    for (let i = path.length - 1; i >= 0; i--) {
      const node = path[i];
      const cos = Math.cos(node.rotation);
      const sin = Math.sin(node.rotation);
      const la = node.scaleX * cos;
      const lb = node.scaleY * sin;
      const lc = -node.scaleX * sin;
      const ld = node.scaleY * cos;
      const le = node.x;
      const lf = node.y;
      const nextA = a * la + c * lb;
      const nextB = b * la + d * lb;
      const nextC = a * lc + c * ld;
      const nextD = b * lc + d * ld;
      const nextE = a * le + c * lf + e;
      const nextF = b * le + d * lf + f;
      a = nextA;
      b = nextB;
      c = nextC;
      d = nextD;
      e = nextE;
      f = nextF;
    }
    return { a, b, c, d, e, f };
  }
  /** Convert a point from this entity's local space to Scene/world space. */
  localToWorld(localX, localY) {
    const { a, b, c, d, e, f } = this.getWorldTransform();
    return {
      x: a * localX + c * localY + e,
      y: b * localX + d * localY + f
    };
  }
  /**
   * Convert a Scene/world point into this entity's local space.
   * Returns `null` when the accumulated transform is singular.
   */
  worldToLocal(worldX, worldY) {
    const { a, b, c, d, e, f } = this.getWorldTransform();
    const determinant = a * d - b * c;
    if (!Number.isFinite(determinant) || Math.abs(determinant) < 1e-12) return null;
    const x = worldX - e;
    const y = worldY - f;
    return {
      x: (d * x - c * y) / determinant,
      y: (-b * x + a * y) / determinant
    };
  }
  /**
   * Return the entity's local bounds transformed into a world-space AABB.
   * Falls back to the entity's `[0, 0, width, height]` box when `getBounds()`
   * does not provide a render-specific box.
   */
  getWorldBounds() {
    const bounds = this.getBounds() ?? { x: 0, y: 0, width: this.width, height: this.height };
    const { a, b, c, d, e, f } = this.getWorldTransform();
    let minX = Infinity;
    let minY = Infinity;
    let maxX = -Infinity;
    let maxY = -Infinity;
    for (let i = 0; i < 4; i++) {
      const localX = i & 1 ? bounds.x + bounds.width : bounds.x;
      const localY = i & 2 ? bounds.y + bounds.height : bounds.y;
      const worldX = a * localX + c * localY + e;
      const worldY = b * localX + d * localY + f;
      minX = Math.min(minX, worldX);
      minY = Math.min(minY, worldY);
      maxX = Math.max(maxX, worldX);
      maxY = Math.max(maxY, worldY);
    }
    return { x: minX, y: minY, width: maxX - minX, height: maxY - minY };
  }
  /**
   * Accumulated world scale factors: this entity's own `scaleX`/`scaleY` times
   * those of every ancestor. Useful for mapping a world-space point back into
   * local space for hit-testing.
   *
   * @returns The world scale `{ x, y }`.
   */
  getWorldScale() {
    let sx = this.scaleX;
    let sy = this.scaleY;
    let curr = this.parent;
    while (curr) {
      sx *= curr.scaleX;
      sy *= curr.scaleY;
      curr = curr.parent;
    }
    return { x: sx, y: sy };
  }
  /**
   * Accumulated world rotation: this entity's own `rotation` plus
   * that of every ancestor.
   *
   * @returns The accumulated world rotation in radians.
   */
  getWorldRotation() {
    let rot = this.rotation;
    let curr = this.parent;
    while (curr) {
      rot += curr.rotation;
      curr = curr.parent;
    }
    return rot;
  }
  /**
   * Return `true` when the given world-space point lies within this entity's
   * interactive hit area.
   *
   * @param globalX - World-space X coordinate.
   * @param globalY - World-space Y coordinate.
   * @returns Whether the point is inside this entity.
   */
  /**
   * Describe this entity's semantics for the accessibility / automation shadow
   * layer. Override in components to project a real `<button>`, `<a href>`, etc.
   *
   * The default returns `{}`, which `Scene.syncA11y` maps to a plain `div`
   * (preserving the historical behavior of interactive entities).
   *
   * @returns The {@link A11yAttributes} for this entity's shadow node.
   */
  getA11yAttributes() {
    return {};
  }
  /**
   * Local-space axis-aligned bounding box of what this entity's {@link render}
   * draws, used by {@link Scene} for viewport culling.
   *
   * Returns `null` by default, meaning "unknown bounds" — the entity is then
   * never culled (always rendered). Override to return a {@link Bounds} so the
   * scene can skip rendering it when it lies outside the viewport.
   *
   * @returns The local bounds, or `null` to opt out of culling.
   */
  getBounds() {
    return null;
  }
  /**
   * Opt into the renderer's draw-call batching fast-path for point-cloud /
   * particle entities that draw as a single filled circle at their local origin.
   *
   * When a leaf entity returns a {@link BatchCircle} and its accumulated
   * transform is representable by the selected batch backend, the {@link Scene}
   * skips its per-entity `save`/`translate`/`scale`/`rotate`/`restore` and
   * {@link render}. Canvas mode or an unsupported affine transform uses the
   * normal render path, so implementations must keep {@link render} correct.
   * Returns `null` by default. Read each frame, so animated color/radius works.
   *
   * @returns The circle to batch, or `null` to use the normal {@link render} path.
   */
  getBatchCircle() {
    return null;
  }
  /**
   * Opt into the GPU instanced-rectangle fast-path for a leaf entity that draws
   * as a single filled rectangle from its local origin. Only used when the
   * {@link Scene} runs a WebGL `pointBackend`; otherwise the entity renders
   * normally via {@link render}. Returns `null` by default. Read each frame.
   *
   * @returns The rectangle to batch, or `null` for the normal render path.
   */
  getBatchRect() {
    return null;
  }
  /**
   * Opt into DOM content projection for entities that render static text.
   * The {@link Scene} mirrors the returned text as a transparent DOM node
   * positioned over the drawn glyphs, making canvas text findable (Ctrl+F),
   * readable by screen readers and crawlers, translatable, and — when
   * `selectable` is set — natively selectable. Returns `null` by default.
   * Read on the a11y sync cadence, so text changes propagate automatically.
   *
   * @returns The projection descriptor, or `null` to project nothing.
   */
  getContentProjection() {
    return null;
  }
  /**
   * Whether this entity still has a queued/running tween animation, or an
   * active {@link setTransition}/{@link animateTo}/{@link springTo} property
   * driver.
   *
   * Used by {@link Scene} to keep rendering continuously while an animation
   * is in flight — both in `onDemand` render mode, and to hold off the
   * `always`-mode idle auto-throttle. Without checking `_drivers` here, a
   * property driver becomes invisible to that throttle: `markDirty()` called
   * from inside `update()`/`tickDrivers()` is wiped by the loop's own
   * `dirty = false` at the end of that same tick, so once the throttle
   * engages an in-flight spring/tween only advances one animation-frame per
   * external `markDirty()` trigger instead of every render frame.
   *
   * @returns `true` if at least one animation or property driver remains.
   */
  hasPendingAnimations() {
    return this.animations.length > 0 || this._drivers.size > 0;
  }
};

// src/text/Typography.ts
var typographyContext;
var baselineCache = /* @__PURE__ */ new Map();
function cssLineBoxBaseline(font, lineHeight) {
  if (typeof document === "undefined") return lineHeight * 0.8;
  const key = `${font}\0${lineHeight}`;
  const cached = baselineCache.get(key);
  if (cached !== void 0) return cached;
  if (typographyContext === void 0) {
    typographyContext = document.createElement("canvas").getContext("2d");
  }
  if (!typographyContext) return lineHeight * 0.8;
  typographyContext.font = font;
  const metrics = typographyContext.measureText("Mg");
  const ascent = metrics.fontBoundingBoxAscent || metrics.actualBoundingBoxAscent;
  const descent = metrics.fontBoundingBoxDescent || metrics.actualBoundingBoxDescent;
  if (!(ascent > 0) || !(descent >= 0)) return lineHeight * 0.8;
  const baseline = (lineHeight - ascent - descent) / 2 + ascent;
  baselineCache.set(key, baseline);
  return baseline;
}
function clearCssLineBoxMetrics() {
  baselineCache.clear();
}

// src/text/MSDFFont.ts
function kernKey(a, b) {
  return a * 1114112 + b;
}
var MSDFFont = class _MSDFFont {
  static idCounter = 0;
  id;
  data;
  byCode = /* @__PURE__ */ new Map();
  kern = /* @__PURE__ */ new Map();
  constructor(data) {
    this.id = `font-${_MSDFFont.idCounter++}`;
    this.data = data;
    for (const g of data.glyphs) this.byCode.set(g.unicode, g);
    for (const k of data.kerning ?? []) this.kern.set(kernKey(k.unicode1, k.unicode2), k.advance);
  }
  /** Parse the `msdf-atlas-gen` JSON (string or already-parsed object). */
  static parse(json) {
    return new _MSDFFont(typeof json === "string" ? JSON.parse(json) : json);
  }
  /** Get a glyph's definition by its unicode value in O(1) time. */
  getGlyph(unicode) {
    return this.byCode.get(unicode);
  }
  /** Distance field range in atlas pixels (for the shader's `u_distanceRange`). */
  get distanceRange() {
    return this.data.atlas.distanceRange;
  }
  get atlasWidth() {
    return this.data.atlas.width;
  }
  get atlasHeight() {
    return this.data.atlas.height;
  }
  /**
   * Lay `text` out at `fontSizePx`. Returns positioned quads (skipping glyphs the
   * font doesn't contain), the widest line's advance, and the total block height.
   * Honors `\n`, kerning pairs, and `letterSpacing`.
   */
  layout(text, fontSizePx, opts = {}) {
    const { x = 0, y = 0, letterSpacing = 0 } = opts;
    const { width: aw, height: ah, yOrigin } = this.data.atlas;
    const { lineHeight, ascender } = this.data.metrics;
    const glyphs = [];
    let penX = x;
    let line = 0;
    let maxAdvance = 0;
    let prevCode = -1;
    const chars = Array.from(text);
    for (const char of chars) {
      if (char === "\n") {
        maxAdvance = Math.max(maxAdvance, penX - x);
        penX = x;
        line++;
        prevCode = -1;
        continue;
      }
      const code = char.codePointAt(0);
      const def = this.byCode.get(code);
      if (!def) {
        prevCode = -1;
        continue;
      }
      if (prevCode >= 0) {
        const k = this.kern.get(kernKey(prevCode, code));
        if (k) penX += k * fontSizePx;
      }
      const baseline = y + (ascender + line * lineHeight) * fontSizePx;
      const pb = def.planeBounds;
      const ab = def.atlasBounds;
      if (pb && ab) {
        const v0 = yOrigin === "bottom" ? 1 - ab.top / ah : ab.top / ah;
        const v1 = yOrigin === "bottom" ? 1 - ab.bottom / ah : ab.bottom / ah;
        glyphs.push({
          char,
          x: penX + pb.left * fontSizePx,
          y: baseline - pb.top * fontSizePx,
          w: (pb.right - pb.left) * fontSizePx,
          h: (pb.top - pb.bottom) * fontSizePx,
          u0: ab.left / aw,
          v0,
          u1: ab.right / aw,
          v1
        });
      }
      penX += def.advance * fontSizePx + letterSpacing;
      prevCode = code;
    }
    maxAdvance = Math.max(maxAdvance, penX - x);
    return {
      glyphs,
      width: maxAdvance,
      height: (line + 1) * lineHeight * fontSizePx
    };
  }
};

// src/text/MSDFTextEntity.ts
var MSDFTextEntity = class extends Entity {
  font;
  texture;
  fallbackFont;
  fontSize;
  color;
  letterSpacing;
  lineHeight;
  maxWidth;
  maxHeight;
  text = "";
  lastRenderedSeqId = 0;
  rgbColorCache = /* @__PURE__ */ new Map();
  fontStringCache = [];
  layoutResult = null;
  constructor(text, options) {
    super();
    this.font = options.font;
    this.texture = options.texture;
    this.fallbackFont = options.fallbackFont ?? "sans-serif";
    this.fontSize = options.fontSize ?? 32;
    this.color = options.color ?? "#ffffff";
    this.letterSpacing = options.letterSpacing ?? 0;
    this.lineHeight = options.lineHeight;
    this.maxWidth = options.maxWidth ?? 1e3;
    this.maxHeight = options.maxHeight ?? 1e3;
    this.setText(text);
  }
  /** Change the wrap boundary and re-run layout for the current text. */
  setMaxWidth(maxWidth) {
    if (this.maxWidth === maxWidth) return;
    this.maxWidth = maxWidth;
    this.queueLayout();
  }
  setText(text) {
    if (this.text === text && this.layoutResult) return;
    this.text = text;
    this.queueLayout();
  }
  queueLayout() {
    LayoutWorkerManager.getInstance().queueLayout(this.id, this.text, {
      fontId: this.font.id,
      fontSize: this.fontSize,
      maxWidth: this.maxWidth,
      maxHeight: this.maxHeight,
      fontData: this.font.data,
      letterSpacing: this.letterSpacing,
      lineHeight: this.lineHeight,
      callback: (res) => {
        if (res.seqId < this.lastRenderedSeqId) return;
        this.lastRenderedSeqId = res.seqId;
        this.layoutResult = res;
        this.scene?.markDirty();
      }
    });
  }
  /**
   * Mirror the rendered text into the DOM content layer: find-in-page, screen
   * readers, crawlers, and translation see the same string the canvas draws.
   */
  getContentProjection() {
    if (!this.text) return null;
    return {
      text: this.text,
      font: `${this.fontSize}px ${this.fallbackFont}`,
      lineHeight: this.lineHeight
    };
  }
  isPointInside(globalX, globalY) {
    if (!this.layoutResult) return false;
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x >= 0 && local.x <= this.layoutResult.width && local.y >= 0 && local.y <= this.layoutResult.height;
  }
  render(renderer) {
    if (!this.layoutResult) return;
    const scene = this.scene;
    const world = this.getWorldTransform();
    const worldScaleX = Math.hypot(world.a, world.b);
    const worldScaleY = Math.hypot(world.c, world.d);
    const orthogonalTolerance = Math.max(1, worldScaleX * worldScaleY) * 1e-6;
    const canUsePointGlyphs = Number.isFinite(worldScaleX) && Number.isFinite(worldScaleY) && world.a * world.d - world.b * world.c >= 0 && Math.abs(world.a * world.c + world.b * world.d) <= orthogonalTolerance;
    if (scene && scene.pointRenderer && scene.glCanvas && canUsePointGlyphs) {
      scene.pointRenderer.setMSDFTexture(this.texture, this.font.distanceRange);
      const worldRot = Math.atan2(world.b, world.a);
      let worldOpacity = this.opacity;
      for (let p = this.parent; p; p = p.parent) worldOpacity *= p.opacity;
      const len2 = this.layoutResult.codePoints.length;
      for (let i = 0; i < len2; i++) {
        const code = this.layoutResult.codePoints[i];
        const nodeX = this.layoutResult.xCoords[i];
        const nodeY = this.layoutResult.yCoords[i];
        const packedStyle = this.layoutResult.packedStyles[i];
        const def = this.font.getGlyph(code);
        if (!def || !def.atlasBounds || !def.planeBounds) continue;
        const { atlasBounds: ab, planeBounds: pb } = def;
        const aw = this.font.atlasWidth;
        const ah = this.font.atlasHeight;
        const lx = nodeX + pb.left * this.fontSize;
        const ly = nodeY - pb.top * this.fontSize;
        const glyphX = world.a * lx + world.c * ly + world.e;
        const glyphY = world.b * lx + world.d * ly + world.f;
        const glyphW = (pb.right - pb.left) * this.fontSize * worldScaleX;
        const glyphH = (pb.top - pb.bottom) * this.fontSize * worldScaleY;
        const v0 = this.font.data.atlas.yOrigin === "bottom" ? 1 - ab.top / ah : ab.top / ah;
        const v1 = this.font.data.atlas.yOrigin === "bottom" ? 1 - ab.bottom / ah : ab.bottom / ah;
        const colorVal = packedStyle >>> 8;
        let runColor = this.rgbColorCache.get(colorVal);
        if (!runColor) {
          const r = colorVal >> 16 & 255;
          const g = colorVal >> 8 & 255;
          const b = colorVal & 255;
          runColor = `rgb(${r},${g},${b})`;
          this.rgbColorCache.set(colorVal, runColor);
        }
        scene.pointRenderer.addGlyph(
          glyphX,
          glyphY,
          glyphW,
          glyphH,
          ab.left / aw,
          v0,
          ab.right / aw,
          v1,
          runColor,
          worldOpacity,
          worldRot
        );
      }
      return;
    }
    if (this.fontStringCache.length === 0) {
      this.fontStringCache[0] = `${this.fontSize}px ${this.fallbackFont}`;
      this.fontStringCache[1] = `bold ${this.fontSize}px ${this.fallbackFont}`;
      this.fontStringCache[2] = `italic ${this.fontSize}px ${this.fallbackFont}`;
      this.fontStringCache[3] = `italic bold ${this.fontSize}px ${this.fallbackFont}`;
    }
    const len = this.layoutResult.codePoints.length;
    for (let i = 0; i < len; i++) {
      const code = this.layoutResult.codePoints[i];
      const nodeX = this.layoutResult.xCoords[i];
      const nodeY = this.layoutResult.yCoords[i];
      const packedStyle = this.layoutResult.packedStyles[i];
      const fontString = this.fontStringCache[packedStyle & 3];
      const colorVal = packedStyle >>> 8;
      let runColor = this.rgbColorCache.get(colorVal);
      if (!runColor) {
        const r = colorVal >> 16 & 255;
        const g = colorVal >> 8 & 255;
        const b = colorVal & 255;
        runColor = `rgb(${r},${g},${b})`;
        this.rgbColorCache.set(colorVal, runColor);
      }
      renderer.fillText(String.fromCodePoint(code), nodeX, nodeY, fontString, runColor);
    }
  }
  destroy() {
    LayoutWorkerManager.getInstance().cancelLayout(this.id);
    super.destroy();
  }
};

// src/text/SVGEntity.ts
function isSvgWhitespace(ch) {
  return ch === " " || ch === "	" || ch === "\n" || ch === "\r";
}
function readSvgAttribute(source, name) {
  const lowerSource = source.toLowerCase();
  const svgStart = lowerSource.indexOf("<svg");
  if (svgStart < 0) return null;
  const tagEnd = source.indexOf(">", svgStart + 4);
  if (tagEnd < 0) return null;
  const tag = source.slice(svgStart + 4, tagEnd);
  const lowerTag = tag.toLowerCase();
  const lowerName = name.toLowerCase();
  for (let i = 0; i < tag.length; i++) {
    const before = i === 0 ? " " : tag[i - 1];
    if (!isSvgWhitespace(before)) continue;
    if (!lowerTag.startsWith(lowerName, i)) continue;
    let cursor = i + lowerName.length;
    while (cursor < tag.length && isSvgWhitespace(tag[cursor])) cursor++;
    if (tag[cursor] !== "=") continue;
    cursor++;
    while (cursor < tag.length && isSvgWhitespace(tag[cursor])) cursor++;
    const quote = tag[cursor];
    if (quote !== '"' && quote !== "'") continue;
    const valueStart = cursor + 1;
    const valueEnd = tag.indexOf(quote, valueStart);
    if (valueEnd < 0) return null;
    return tag.slice(valueStart, valueEnd);
  }
  return null;
}
var SVGEntity = class extends Entity {
  svgSource = "";
  imageBitmap = null;
  imageElement = null;
  blobURL = null;
  currentImg = null;
  lodTimeout = null;
  cachedDoc = null;
  baseWidth = 100;
  baseHeight = 100;
  lastRasterizedScale = 1;
  targetScale = 1;
  constructor(svgSource, id) {
    super(id);
    this.setSVGSource(svgSource);
  }
  setSVGSource(svgSource) {
    if (this.svgSource === svgSource) return;
    this.svgSource = svgSource;
    this.cachedDoc = null;
    this.parseSVGDimensions();
    this.triggerRasterization(this.lastRasterizedScale);
  }
  parseSVGDimensions() {
    let width = 100;
    let height = 100;
    if (typeof window !== "undefined" && typeof DOMParser !== "undefined") {
      try {
        const parser = new DOMParser();
        const doc = parser.parseFromString(this.svgSource, "image/svg+xml");
        const parserError = doc.querySelector("parsererror");
        if (parserError) {
          console.error("SVG Parsing error:", parserError.textContent);
        } else {
          this.cachedDoc = doc;
          const svgEl = doc.documentElement;
          const wAttr = svgEl.getAttribute("width");
          const hAttr = svgEl.getAttribute("height");
          const vbAttr = svgEl.getAttribute("viewBox");
          if (wAttr && hAttr) {
            width = parseFloat(wAttr) || 100;
            height = parseFloat(hAttr) || 100;
          } else if (vbAttr) {
            const parts = vbAttr.split(/[\s,]+/).map(parseFloat);
            if (parts.length === 4) {
              width = parts[2];
              height = parts[3];
            }
          }
        }
      } catch (e) {
        console.error("Failed parsing SVG via DOMParser, falling back to attribute scan:", e);
      }
    } else {
      const wAttr = readSvgAttribute(this.svgSource, "width");
      const hAttr = readSvgAttribute(this.svgSource, "height");
      const vbAttr = readSvgAttribute(this.svgSource, "viewBox");
      if (wAttr && hAttr) {
        width = parseFloat(wAttr) || 100;
        height = parseFloat(hAttr) || 100;
      } else if (vbAttr) {
        const parts = vbAttr.split(/[\s,]+/).map(parseFloat);
        if (parts.length === 4) {
          width = parts[2];
          height = parts[3];
        }
      }
    }
    this.baseWidth = width;
    this.baseHeight = height;
    this.width = width;
    this.height = height;
  }
  triggerRasterization(scale) {
    if (typeof window === "undefined" || typeof Blob === "undefined") return;
    if (this.currentImg) {
      this.currentImg.onload = null;
      this.currentImg.onerror = null;
      this.currentImg = null;
    }
    if (this.blobURL) {
      URL.revokeObjectURL(this.blobURL);
      this.blobURL = null;
    }
    let processedSource = this.svgSource;
    try {
      let doc = this.cachedDoc;
      if (!doc) {
        const parser = new DOMParser();
        doc = parser.parseFromString(this.svgSource, "image/svg+xml");
        this.cachedDoc = doc;
      }
      const parserError = doc.querySelector("parsererror");
      if (parserError) {
        console.error(
          "SVG Parsing validation error in triggerRasterization:",
          parserError.textContent
        );
      } else {
        const clonedDoc = doc.cloneNode(true);
        const svgEl = clonedDoc.documentElement;
        if (svgEl.tagName.toLowerCase() === "svg") {
          const targetWidth = Math.max(1, Math.round(this.baseWidth * scale));
          const targetHeight = Math.max(1, Math.round(this.baseHeight * scale));
          svgEl.setAttribute("width", `${targetWidth}`);
          svgEl.setAttribute("height", `${targetHeight}`);
          if (!svgEl.hasAttribute("viewBox")) {
            svgEl.setAttribute("viewBox", `0 0 ${this.baseWidth} ${this.baseHeight}`);
          }
          const serializer = new XMLSerializer();
          processedSource = serializer.serializeToString(clonedDoc);
        }
      }
    } catch (e) {
      console.error("Failed to apply LOD scaling to SVG XML:", e);
    }
    const blob = new Blob([processedSource], { type: "image/svg+xml;charset=utf-8" });
    this.blobURL = URL.createObjectURL(blob);
    const img = new Image();
    img.crossOrigin = "anonymous";
    this.currentImg = img;
    img.onload = () => {
      if (this.currentImg !== img) return;
      this.imageElement = img;
      if (typeof createImageBitmap === "undefined") {
        this.currentImg = null;
        if (this.scene) this.scene.markDirty();
        return;
      }
      createImageBitmap(img).then((bitmap) => {
        if (this.currentImg !== img) {
          bitmap.close();
          return;
        }
        if (this.imageBitmap) {
          this.imageBitmap.close();
        }
        this.imageBitmap = bitmap;
        this.currentImg = null;
        if (this.scene) this.scene.markDirty();
      }).catch((e) => {
        console.error("Failed to create ImageBitmap from SVG:", e);
        this.currentImg = null;
      });
    };
    img.onerror = (e) => {
      if (this.currentImg !== img) return;
      console.error("Failed to load SVG Image element:", e);
      this.currentImg = null;
    };
    img.src = this.blobURL;
  }
  isPointInside(globalX, globalY) {
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x >= 0 && local.x <= this.width && local.y >= 0 && local.y <= this.height;
  }
  render(r) {
    const svgRenderer = r;
    if (typeof svgRenderer.drawSVG === "function") {
      svgRenderer.drawSVG(this.svgSource, 0, 0, this.width, this.height);
      return;
    }
    const scale = this.getWorldScale();
    const currentScale = Math.max(0.1, Math.max(scale.x, scale.y));
    if (Math.abs(currentScale - this.lastRasterizedScale) / this.lastRasterizedScale > 0.2) {
      this.targetScale = currentScale;
      if (this.lodTimeout) clearTimeout(this.lodTimeout);
      this.lodTimeout = setTimeout(() => {
        this.triggerRasterization(this.targetScale);
        this.lastRasterizedScale = this.targetScale;
        this.lodTimeout = null;
      }, 200);
    }
    if (this.imageBitmap) {
      r.drawImage(this.imageBitmap, 0, 0, this.width, this.height);
    } else if (this.imageElement) {
      r.drawImage(this.imageElement, 0, 0, this.width, this.height);
    }
  }
  destroy() {
    if (this.lodTimeout) {
      clearTimeout(this.lodTimeout);
      this.lodTimeout = null;
    }
    if (this.currentImg) {
      this.currentImg.onload = null;
      this.currentImg.onerror = null;
      this.currentImg = null;
    }
    if (this.imageBitmap) {
      this.imageBitmap.close();
      this.imageBitmap = null;
    }
    if (this.blobURL) {
      URL.revokeObjectURL(this.blobURL);
      this.blobURL = null;
    }
    this.imageElement = null;
    this.cachedDoc = null;
    super.destroy();
  }
};

// src/text/PreparedContentGrid.ts
var nextRevision = 1;
var graphemeSegmenter = typeof Intl !== "undefined" && "Segmenter" in Intl ? new Intl.Segmenter(void 0, { granularity: "grapheme" }) : null;
var MARK = /\p{Mark}/u;
var EXTENDED_PICTOGRAPHIC = /\p{Extended_Pictographic}/u;
var REGIONAL_INDICATOR = /\p{Regional_Indicator}/u;
var BIDI_CONTROL = /\p{Bidi_Control}/u;
var EAST_ASIAN_WIDE = /[ᄀ-ᅟ⌚-⌛⏩-⏬⏰⏳◽-◾⺀-〾ぁ-㏿㐀-䶿一-鿿ꀀ-꓏가-힣豈-﫿︰-﹏＀-｠￠-￦]/u;
function codePointAt(text, index) {
  const point = text.codePointAt(index);
  if (point === void 0) return { value: "", next: index };
  const value = String.fromCodePoint(point);
  return { value, next: index + value.length };
}
function fallbackGraphemes(text) {
  const parts = [];
  let index = 0;
  while (index < text.length) {
    const start = index;
    let current = codePointAt(text, index);
    let segment = current.value;
    index = current.next;
    let regionalCount = REGIONAL_INDICATOR.test(segment) ? 1 : 0;
    while (index < text.length) {
      current = codePointAt(text, index);
      const point = current.value.codePointAt(0) ?? 0;
      const isVariation = point >= 65024 && point <= 65039;
      const isEmojiModifier = point >= 127995 && point <= 127999;
      const isKeycap = point === 8419;
      if (MARK.test(current.value) || isVariation || isEmojiModifier || isKeycap) {
        segment += current.value;
        index = current.next;
        continue;
      }
      if (REGIONAL_INDICATOR.test(current.value) && regionalCount === 1) {
        segment += current.value;
        index = current.next;
        regionalCount++;
        continue;
      }
      if (point === 8205) {
        segment += current.value;
        index = current.next;
        if (index < text.length) {
          current = codePointAt(text, index);
          segment += current.value;
          index = current.next;
        }
        continue;
      }
      break;
    }
    parts.push({ segment, index: start });
  }
  return parts;
}
function graphemes(text) {
  if (!graphemeSegmenter) return fallbackGraphemes(text);
  return Array.from(graphemeSegmenter.segment(text), (part) => ({
    segment: part.segment,
    index: part.index
  }));
}
function lowerBound(values, target) {
  let low = 0;
  let high = values.length;
  while (low < high) {
    const middle = low + high >>> 1;
    if (values[middle] < target) low = middle + 1;
    else high = middle;
  }
  return low;
}
function isWideCluster(cluster) {
  if (EXTENDED_PICTOGRAPHIC.test(cluster) || REGIONAL_INDICATOR.test(cluster)) return true;
  if (cluster.includes("\u20E3")) return true;
  if (EAST_ASIAN_WIDE.test(cluster)) return true;
  const point = cluster.codePointAt(0) ?? 0;
  return point >= 131072 && point <= 262141;
}
function sourceLines(source) {
  const lines = [];
  let start = 0;
  while (true) {
    let end = start;
    while (end < source.length && source[end] !== "\r" && source[end] !== "\n") end++;
    if (end === source.length) {
      lines.push({
        sourceStart: start,
        sourceEnd: end,
        nextSourceStart: end,
        text: source.slice(start)
      });
      break;
    }
    const next = source[end] === "\r" && source[end + 1] === "\n" ? end + 2 : end + 1;
    lines.push({
      sourceStart: start,
      sourceEnd: end,
      nextSourceStart: next,
      text: source.slice(start, end)
    });
    start = next;
    if (start === source.length) {
      lines.push({ sourceStart: start, sourceEnd: start, nextSourceStart: start, text: "" });
      break;
    }
  }
  return lines;
}
function assertPositiveFinite(value, name) {
  if (!Number.isFinite(value) || value <= 0) {
    throw new RangeError(`${name} must be a positive finite number`);
  }
}
function prepareContentGrid(source, options) {
  assertPositiveFinite(options.cellWidth, "cellWidth");
  assertPositiveFinite(options.lineHeight, "lineHeight");
  if (!Number.isFinite(options.baseline)) throw new RangeError("baseline must be finite");
  const tabSize = options.tabSize ?? 4;
  if (!Number.isInteger(tabSize) || tabSize <= 0) {
    throw new RangeError("tabSize must be a positive integer");
  }
  const rawLines = sourceLines(source);
  const lines = [];
  for (let lineIndex = 0; lineIndex < rawLines.length; lineIndex++) {
    const sourceLine = rawLines[lineIndex];
    const rawLine = sourceLine.text;
    const { sourceStart: lineStart, sourceEnd, nextSourceStart } = sourceLine;
    const rawCaretBoundaries = [
      0,
      ...graphemes(rawLine).map((grapheme) => grapheme.index + grapheme.segment.length)
    ];
    const shaped = ArabicShaper.shapeArabic(rawLine);
    const shapedParts = graphemes(shaped.shapedText);
    const levels = BidiResolver.resolveLevels(shaped.shapedText);
    const cells = [];
    let column = 0;
    for (let index = 0; index < shapedParts.length; index++) {
      const part = shapedParts[index];
      const sourceOffset = shaped.indexMap[part.index] ?? part.index;
      const nextPart = shapedParts[index + 1];
      const sourceOffsetEnd = nextPart ? shaped.indexMap[nextPart.index] ?? nextPart.index : rawLine.length;
      const raw = rawLine.slice(sourceOffset, sourceOffsetEnd);
      const sourceCaretOffsets = [0];
      for (let caretIndex = lowerBound(rawCaretBoundaries, sourceOffset + 1); caretIndex < rawCaretBoundaries.length && rawCaretBoundaries[caretIndex] < sourceOffsetEnd; caretIndex++) {
        sourceCaretOffsets.push(rawCaretBoundaries[caretIndex] - sourceOffset);
      }
      if (sourceCaretOffsets.at(-1) !== sourceOffsetEnd - sourceOffset) {
        sourceCaretOffsets.push(sourceOffsetEnd - sourceOffset);
      }
      let columns;
      if (BIDI_CONTROL.test(raw)) columns = 0;
      else if (raw === "	") columns = tabSize - column % tabSize;
      else columns = isWideCluster(raw) ? 2 : 1;
      const advance = columns * options.cellWidth;
      cells.push({
        sourceStart: lineStart + sourceOffset,
        sourceEnd: lineStart + sourceOffsetEnd,
        sourceCaretOffsets: Object.freeze(sourceCaretOffsets),
        glyph: part.segment,
        x: 0,
        advance,
        level: levels[part.index] ?? 0,
        char: part.segment
      });
      column += columns;
    }
    const visualCells = [...cells];
    BidiResolver.reorderVisual(visualCells, BidiResolver.getBaseLevel(shaped.shapedText));
    let visualX = 0;
    for (const cell of visualCells) {
      cell.x = visualX;
      visualX += cell.advance;
    }
    const frozenCells = cells.map(({ char: _char, ...cell }) => Object.freeze(cell));
    lines.push(
      Object.freeze({
        sourceStart: lineStart,
        sourceEnd,
        nextSourceStart,
        width: visualX,
        cells: Object.freeze(frozenCells)
      })
    );
  }
  return Object.freeze({
    kind: "content-grid",
    revision: nextRevision++,
    source,
    font: options.font,
    cellWidth: options.cellWidth,
    lineHeight: options.lineHeight,
    baseline: options.baseline,
    tabSize,
    lines: Object.freeze(lines)
  });
}

export {
  SpringPhysics,
  Easing,
  isTweenConfig,
  TweenDriver,
  SpringDriver,
  VectoJSEvent,
  Entity,
  cssLineBoxBaseline,
  clearCssLineBoxMetrics,
  MSDFFont,
  MSDFTextEntity,
  SVGEntity,
  prepareContentGrid
};
