// @vectojs/core/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; } var _class; var _class2; var _class3; var _class4; var _class5; var _class6; var _class7;




var _chunkBA5HUUDFjs = require('./chunk-BA5HUUDF.js');








var _chunkORCPCIJ7js = require('./chunk-ORCPCIJ7.js');














var _chunk2Z23LTH3js = require('./chunk-2Z23LTH3.js');




var _chunk4AR425ARjs = require('./chunk-4AR425AR.js');

// src/tree/ComputeParticleEntity.ts
var PARTICLE_STRIDE_FLOATS = 8;
var PARTICLE_OFFSET_POSITION_X = 0;
var PARTICLE_OFFSET_POSITION_Y = 1;
var PARTICLE_OFFSET_VELOCITY_X = 2;
var PARTICLE_OFFSET_VELOCITY_Y = 3;
var PARTICLE_OFFSET_ORIGIN_X = 4;
var PARTICLE_OFFSET_ORIGIN_Y = 5;
var PARTICLE_OFFSET_SIZE = 6;
var PARTICLE_OFFSET_LIFE = 7;
var ComputeParticleEntity = (_class = class extends _chunk2Z23LTH3js.Entity {
  
  
  
  
  
  
  
  
  /** Flat array containing layout of all particles: position, velocity, origin, size, life. */
  
  /** Flag indicating whether the particle coordinates need to be initialized. */
  __init() {this.needsInit = true}
  /** Active explosion impulse to apply in the next simulation step. */
  __init2() {this.pendingExplosion = null}
  /** WebGPU storage buffer containing particle states. */
  __init3() {this.gpuStorageBuffer = null}
  /** WebGPU uniform buffer containing simulation parameters. */
  __init4() {this.gpuUniformBuffer = null}
  /** WebGPU bind group for the compute shader pass. */
  __init5() {this.computeBindGroup = null}
  /** WebGPU bind group for the render pass (usually same as compute). */
  __init6() {this.renderBindGroup = null}
  constructor(options = {}) {
    super();_class.prototype.__init.call(this);_class.prototype.__init2.call(this);_class.prototype.__init3.call(this);_class.prototype.__init4.call(this);_class.prototype.__init5.call(this);_class.prototype.__init6.call(this);;
    this.maxParticles = _nullishCoalesce(options.maxParticles, () => ( 1e4));
    this.springK = _nullishCoalesce(options.springK, () => ( 0.05));
    this.damping = _nullishCoalesce(options.damping, () => ( 0.95));
    this.bounceDamping = _nullishCoalesce(options.bounceDamping, () => ( 0.5));
    this.maxVelocity = _nullishCoalesce(options.maxVelocity, () => ( 500));
    this.size = _nullishCoalesce(options.size, () => ( 4));
    this.baseColor = _nullishCoalesce(options.color, () => ( "#00f0ff"));
    this.pointerEvents = _nullishCoalesce(options.pointerEvents, () => ( false));
    this.particleData = new Float32Array(this.maxParticles * PARTICLE_STRIDE_FLOATS);
    this.interactive = true;
  }
  /**
   * Disperses all particles randomly across the specified screen bounds.
   * Sets initial positions, velocities, origins, and sizes.
   *
   * @param width - Simulation zone width.
   * @param height - Simulation zone height.
   */
  initRandomParticles(width, height) {
    const safeW = Math.max(1, width);
    const safeH = Math.max(1, height);
    for (let i = 0; i < this.maxParticles; i++) {
      const idx = i * PARTICLE_STRIDE_FLOATS;
      const x = Math.random() * safeW;
      const y = Math.random() * safeH;
      this.particleData[idx + PARTICLE_OFFSET_POSITION_X] = x;
      this.particleData[idx + PARTICLE_OFFSET_POSITION_Y] = y;
      this.particleData[idx + PARTICLE_OFFSET_VELOCITY_X] = 0;
      this.particleData[idx + PARTICLE_OFFSET_VELOCITY_Y] = 0;
      this.particleData[idx + PARTICLE_OFFSET_ORIGIN_X] = x;
      this.particleData[idx + PARTICLE_OFFSET_ORIGIN_Y] = y;
      this.particleData[idx + PARTICLE_OFFSET_SIZE] = this.size;
      this.particleData[idx + PARTICLE_OFFSET_LIFE] = -1;
    }
    this.needsInit = true;
    _optionalChain([this, 'access', _ => _.scene, 'optionalAccess', _2 => _2.markDirty, 'call', _3 => _3()]);
  }
  /**
   * Sets the origins (ox, oy) for a subset or all particles.
   * Also sets position to origin if requestPositionReset is true.
   *
   * @param points - Flat Float32Array containing [x0, y0, x1, y1, ...]
   * @param requestPositionReset - Whether to set current positions to the new origins. Defaults to true.
   */
  setOrigins(points, requestPositionReset = true) {
    const len = Math.min(this.maxParticles, Math.floor(points.length / 2));
    for (let i = 0; i < len; i++) {
      const idx = i * PARTICLE_STRIDE_FLOATS;
      const ptIdx = i * 2;
      const ox = points[ptIdx];
      const oy = points[ptIdx + 1];
      this.particleData[idx + PARTICLE_OFFSET_ORIGIN_X] = ox;
      this.particleData[idx + PARTICLE_OFFSET_ORIGIN_Y] = oy;
      if (requestPositionReset) {
        this.particleData[idx + PARTICLE_OFFSET_POSITION_X] = ox;
        this.particleData[idx + PARTICLE_OFFSET_POSITION_Y] = oy;
        this.particleData[idx + PARTICLE_OFFSET_VELOCITY_X] = 0;
        this.particleData[idx + PARTICLE_OFFSET_VELOCITY_Y] = 0;
      }
    }
    this.needsInit = true;
    _optionalChain([this, 'access', _4 => _4.scene, 'optionalAccess', _5 => _5.markDirty, 'call', _6 => _6()]);
  }
  /**
   * Sets the current positions (x, y) for a subset or all particles.
   *
   * @param positions - Flat Float32Array containing [x0, y0, x1, y1, ...]
   */
  setPositions(positions) {
    const len = Math.min(this.maxParticles, Math.floor(positions.length / 2));
    for (let i = 0; i < len; i++) {
      const idx = i * PARTICLE_STRIDE_FLOATS;
      const ptIdx = i * 2;
      this.particleData[idx + PARTICLE_OFFSET_POSITION_X] = positions[ptIdx];
      this.particleData[idx + PARTICLE_OFFSET_POSITION_Y] = positions[ptIdx + 1];
    }
    this.needsInit = true;
    _optionalChain([this, 'access', _7 => _7.scene, 'optionalAccess', _8 => _8.markDirty, 'call', _9 => _9()]);
  }
  /**
   * Sets the current velocities (vx, vy) for a subset or all particles.
   *
   * @param velocities - Flat Float32Array containing [vx0, vy0, vx1, vy1, ...]
   */
  setVelocities(velocities) {
    const len = Math.min(this.maxParticles, Math.floor(velocities.length / 2));
    for (let i = 0; i < len; i++) {
      const idx = i * PARTICLE_STRIDE_FLOATS;
      const ptIdx = i * 2;
      this.particleData[idx + PARTICLE_OFFSET_VELOCITY_X] = velocities[ptIdx];
      this.particleData[idx + PARTICLE_OFFSET_VELOCITY_Y] = velocities[ptIdx + 1];
    }
    this.needsInit = true;
    _optionalChain([this, 'access', _10 => _10.scene, 'optionalAccess', _11 => _11.markDirty, 'call', _12 => _12()]);
  }
  /**
   * Triggers an explosion force center.
   *
   * @param x - Explosion center x-coordinate.
   * @param y - Explosion center y-coordinate.
   * @param force - Magnitude force scalar.
   */
  triggerExplosion(x, y, force) {
    this.pendingExplosion = { x, y, force };
  }
  isPointInside(_x, _y) {
    return this.pointerEvents;
  }
  render(_r) {
  }
  /**
   * True while any live particle still has a visually meaningful velocity or
   * sits meaningfully away from its spring target.
   *
   * The actual per-frame simulation (WebGPU compute pass / {@link updateCPU})
   * runs through a dedicated Scene-level pre-pass outside the normal
   * render-tree walk, but this entity is still a normal tree member that
   * walk visits — without this override, the base `Entity.hasPendingAnimations()`
   * (always `false`) is what Scene sees, so `renderMode: 'always'`'s idle
   * auto-throttle drops the WHOLE scene to ~2fps the instant nothing else in
   * the tree is animating, even while thousands of particles are visibly
   * drifting or spring-settling. A spring+damping system asymptotically
   * approaches zero velocity but never reaches it exactly, so this checks
   * against a small epsilon rather than `!== 0` — otherwise this would
   * always return `true` and defeat the idle throttle entirely.
   */
  hasPendingAnimations() {
    const EPS_VELOCITY = 0.5;
    const EPS_DISTANCE = 0.5;
    for (let i = 0; i < this.maxParticles; i++) {
      const idx = i * PARTICLE_STRIDE_FLOATS;
      if (this.particleData[idx + PARTICLE_OFFSET_LIFE] === 0) continue;
      const vx = this.particleData[idx + PARTICLE_OFFSET_VELOCITY_X];
      const vy = this.particleData[idx + PARTICLE_OFFSET_VELOCITY_Y];
      if (vx * vx + vy * vy > EPS_VELOCITY * EPS_VELOCITY) return true;
      const dx = this.particleData[idx + PARTICLE_OFFSET_POSITION_X] - this.particleData[idx + PARTICLE_OFFSET_ORIGIN_X];
      const dy = this.particleData[idx + PARTICLE_OFFSET_POSITION_Y] - this.particleData[idx + PARTICLE_OFFSET_ORIGIN_Y];
      if (dx * dx + dy * dy > EPS_DISTANCE * EPS_DISTANCE) return true;
    }
    return this.pendingExplosion !== null;
  }
  /**
   * Updates particle simulation on the CPU.
   * Handles spring forces, mouse repulsion, explosion impulses, velocity capping, and bounds bouncing/clamping.
   *
   * @param dt - Delta time in seconds.
   * @param mouseX - Mouse x-coordinate, or a value below -9000 if inactive.
   * @param mouseY - Mouse y-coordinate, or a value below -9000 if inactive.
   * @param width - Boundary width.
   * @param height - Boundary height.
   */
  updateCPU(dt, mouseX, mouseY, width, height) {
    const safeDt = isNaN(dt) ? 0.016 : Math.max(0, Math.min(dt, 0.1));
    const explosion = this.pendingExplosion;
    const safeWidth = Math.max(1, width);
    const safeHeight = Math.max(1, height);
    const springK = Math.max(0, Math.min(10, this.springK));
    const damping = Math.max(0, Math.min(1, this.damping));
    const bounceDamping = Math.max(0, Math.min(1, this.bounceDamping));
    const maxVelocity = Math.max(1, this.maxVelocity);
    for (let i = 0; i < this.maxParticles; i++) {
      const offset = i * 8;
      let px = this.particleData[offset];
      let py = this.particleData[offset + 1];
      let vx = this.particleData[offset + 2];
      let vy = this.particleData[offset + 3];
      const ox = this.particleData[offset + 4];
      const oy = this.particleData[offset + 5];
      const life = this.particleData[offset + 7];
      if (isNaN(px)) px = ox;
      if (isNaN(py)) py = oy;
      if (isNaN(vx)) vx = 0;
      if (isNaN(vy)) vy = 0;
      const fx_spring = (ox - px) * springK;
      const fy_spring = (oy - py) * springK;
      let fx_mouse = 0;
      let fy_mouse = 0;
      if (!isNaN(mouseX) && !isNaN(mouseY) && mouseX > -9e3 && mouseY > -9e3) {
        const dx = mouseX - px;
        const dy = mouseY - py;
        const dist = Math.hypot(dx, dy);
        if (dist < 120 && dist > 0.1) {
          const forceMag = (120 - dist) * 2;
          fx_mouse = -(dx / dist) * forceMag;
          fy_mouse = -(dy / dist) * forceMag;
        }
      }
      let fx_expl = 0;
      let fy_expl = 0;
      if (explosion) {
        const ex = explosion.x - px;
        const ey = explosion.y - py;
        const edist = Math.hypot(ex, ey);
        if (edist < 150 && edist > 0.1) {
          const forceMag = (150 - edist) * explosion.force;
          fx_expl = -(ex / edist) * forceMag;
          fy_expl = -(ey / edist) * forceMag;
        }
      }
      const ax = fx_spring + fx_mouse + fx_expl;
      const ay = fy_spring + fy_mouse + fy_expl;
      let nvx = (vx + ax * safeDt) * damping;
      let nvy = (vy + ay * safeDt) * damping;
      const speed = Math.hypot(nvx, nvy);
      if (speed > maxVelocity) {
        nvx = nvx / speed * maxVelocity;
        nvy = nvy / speed * maxVelocity;
      }
      let npx = px + nvx * safeDt;
      let npy = py + nvy * safeDt;
      if (npx <= 0 && nvx < 0) {
        nvx = -nvx * bounceDamping;
      } else if (npx >= safeWidth && nvx > 0) {
        nvx = -nvx * bounceDamping;
      }
      if (npy <= 0 && nvy < 0) {
        nvy = -nvy * bounceDamping;
      } else if (npy >= safeHeight && nvy > 0) {
        nvy = -nvy * bounceDamping;
      }
      npx = Math.max(0, Math.min(safeWidth, npx));
      npy = Math.max(0, Math.min(safeHeight, npy));
      let nlife = life;
      if (life >= 0) {
        nlife = Math.max(0, life - safeDt * 0.5);
      }
      this.particleData[offset] = npx;
      this.particleData[offset + 1] = npy;
      this.particleData[offset + 2] = nvx;
      this.particleData[offset + 3] = nvy;
      this.particleData[offset + 7] = nlife;
    }
    this.pendingExplosion = null;
  }
  destroy() {
    this.destroyGPUResources();
    super.destroy();
  }
  /**
   * Frees all GPU resources allocated for WebGPU simulation.
   */
  destroyGPUResources() {
    if (this.gpuStorageBuffer) {
      if (typeof this.gpuStorageBuffer.destroy === "function") {
        this.gpuStorageBuffer.destroy();
      }
      this.gpuStorageBuffer = null;
    }
    if (this.gpuUniformBuffer) {
      if (typeof this.gpuUniformBuffer.destroy === "function") {
        this.gpuUniformBuffer.destroy();
      }
      this.gpuUniformBuffer = null;
    }
    this.computeBindGroup = null;
    this.renderBindGroup = null;
  }
}, _class);

// src/tree/Scene.ts
var INTERACTIVE_A11Y_ROLES = /* @__PURE__ */ new Set([
  "button",
  "switch",
  "checkbox",
  "radio",
  "link",
  "tab",
  "menuitem",
  "slider",
  "combobox"
]);
function isNativelyFocusable(element) {
  return element instanceof HTMLButtonElement || element instanceof HTMLInputElement || element instanceof HTMLSelectElement || element instanceof HTMLTextAreaElement || element instanceof HTMLAnchorElement && element.hasAttribute("href");
}
var REDUCED_MOTION_FPS = 30;
function parseInlinePx(value) {
  if (!value || !value.endsWith("px")) return null;
  const n = parseFloat(value);
  return Number.isFinite(n) && n > 0 ? n : null;
}
function collectTextNodes(root) {
  const out = [];
  const walker = document.createTreeWalker(root, NodeFilter.SHOW_TEXT);
  for (let n = walker.nextNode(); n; n = walker.nextNode()) out.push(n);
  return out;
}
var caretGraphemeSegmenter = typeof Intl !== "undefined" && typeof Intl.Segmenter === "function" ? new Intl.Segmenter(void 0, { granularity: "grapheme" }) : null;
var caretWordSegmenter = typeof Intl !== "undefined" && typeof Intl.Segmenter === "function" ? new Intl.Segmenter(void 0, { granularity: "word" }) : null;
function graphemeBoundaries(text) {
  if (!caretGraphemeSegmenter) {
    const boundaries2 = [0];
    for (let offset = 0; offset < text.length; ) {
      const codePoint = _nullishCoalesce(text.codePointAt(offset), () => ( 0));
      offset += codePoint > 65535 ? 2 : 1;
      boundaries2.push(offset);
    }
    return boundaries2;
  }
  const boundaries = [0];
  for (const segment of caretGraphemeSegmenter.segment(text)) {
    const end = segment.index + segment.segment.length;
    if (end > boundaries[boundaries.length - 1]) boundaries.push(end);
  }
  return boundaries;
}
function distanceToRectSquared(rect, x, y) {
  const dx = x < rect.left ? rect.left - x : x > rect.right ? x - rect.right : 0;
  const dy = y < rect.top ? rect.top - y : y > rect.bottom ? y - rect.bottom : 0;
  return dx * dx + dy * dy;
}
function nearestOffsetForPoint(node, x, y) {
  const boundaries = graphemeBoundaries(node.data);
  const range = document.createRange();
  let nearest = { offset: _nullishCoalesce(boundaries[0], () => ( 0)), distance: Infinity };
  for (const offset of boundaries) {
    range.setStart(node, offset);
    range.collapse(true);
    const rect = range.getBoundingClientRect();
    const distance = distanceToRectSquared(rect, x, y);
    if (distance < nearest.distance) nearest = { offset, distance };
  }
  return nearest;
}
function gridCellCaret(cell, localX) {
  const node = cell.firstChild;
  if (!(node instanceof Text)) return null;
  const sourceLength = Number(_nullishCoalesce(cell.dataset.vectoGridSourceLength, () => ( node.data.length)));
  const level = Number(_nullishCoalesce(cell.dataset.vectoGridLevel, () => ( 0)));
  const cellX = Number(_nullishCoalesce(cell.dataset.vectoGridX, () => ( 0)));
  const advance = Number(_nullishCoalesce(cell.dataset.vectoGridAdvance, () => ( 0)));
  const caretOffsets = (_nullishCoalesce(cell.dataset.vectoGridCaretOffsets, () => ( `0,${sourceLength}`))).split(",").map(Number).filter((offset) => Number.isInteger(offset) && offset >= 0 && offset <= sourceLength);
  const visuallyRtl = (level & 1) !== 0;
  const visualFraction = advance > 0 ? Math.max(0, Math.min(1, (localX - cellX) / advance)) : 0;
  const sourceFraction = visuallyRtl ? 1 - visualFraction : visualFraction;
  const caretIndex = Math.round(sourceFraction * Math.max(0, caretOffsets.length - 1));
  return {
    node,
    offset: _nullishCoalesce(caretOffsets[caretIndex], () => ( 0))
  };
}
function nearestGridPositionInLine(line, localX) {
  let nearest = null;
  for (const cell of line.querySelectorAll("[data-vecto-grid-cell]")) {
    const x = Number(_nullishCoalesce(cell.dataset.vectoGridX, () => ( 0)));
    const advance = Number(_nullishCoalesce(cell.dataset.vectoGridAdvance, () => ( 0)));
    if (localX >= x && localX <= x + advance) return gridCellCaret(cell, localX);
    const distance = localX < x ? x - localX : localX - (x + advance);
    if (!nearest || distance < nearest.distance) nearest = { cell, distance };
  }
  if (!nearest) return null;
  return gridCellCaret(nearest.cell, localX);
}
function parseCssMatrix(transform) {
  if (!transform || transform === "none") return [1, 0, 0, 1];
  const values = transform.slice(transform.indexOf("(") + 1, transform.lastIndexOf(")")).split(",").map(Number);
  return values.length >= 4 && values.slice(0, 4).every(Number.isFinite) ? [values[0], values[1], values[2], values[3]] : [1, 0, 0, 1];
}
function clientToGridLocal(contentEl, canvas, clientX, clientY) {
  const line = contentEl.querySelector("[data-vecto-grid-line]");
  const originMarker = _optionalChain([line, 'optionalAccess', _13 => _13.querySelector, 'call', _14 => _14('[data-vecto-grid-basis="origin"]')]);
  const xMarker = _optionalChain([line, 'optionalAccess', _15 => _15.querySelector, 'call', _16 => _16('[data-vecto-grid-basis="x"]')]);
  const yMarker = _optionalChain([line, 'optionalAccess', _17 => _17.querySelector, 'call', _18 => _18('[data-vecto-grid-basis="y"]')]);
  if (line && originMarker && xMarker && yMarker) {
    const origin = originMarker.getBoundingClientRect();
    const xPoint = xMarker.getBoundingClientRect();
    const yPoint = yMarker.getBoundingClientRect();
    const xx = xPoint.left - origin.left;
    const xy = xPoint.top - origin.top;
    const yx = yPoint.left - origin.left;
    const yy = yPoint.top - origin.top;
    const determinant2 = xx * yy - xy * yx;
    if (Number.isFinite(determinant2) && Math.abs(determinant2) > 1e-9) {
      const dx2 = clientX - origin.left;
      const dy2 = clientY - origin.top;
      return {
        x: (Number.parseFloat(line.style.left) || 0) + (yy * dx2 - yx * dy2) / determinant2,
        y: (Number.parseFloat(line.style.top) || 0) + (-xy * dx2 + xx * dy2) / determinant2
      };
    }
  }
  const [a, b, c, d] = parseCssMatrix(getComputedStyle(contentEl).transform);
  const canvasRect = canvas.getBoundingClientRect();
  const logicalWidth = Number.parseFloat(canvas.style.width) || canvas.clientWidth || canvas.width;
  const logicalHeight = Number.parseFloat(canvas.style.height) || canvas.clientHeight || canvas.height;
  const scaleX = logicalWidth > 0 ? canvasRect.width / logicalWidth : 1;
  const scaleY = logicalHeight > 0 ? canvasRect.height / logicalHeight : 1;
  const worldX = (clientX - canvasRect.left) / scaleX;
  const worldY = (clientY - canvasRect.top) / scaleY;
  const dx = worldX - (Number.parseFloat(contentEl.style.left) || 0);
  const dy = worldY - (Number.parseFloat(contentEl.style.top) || 0);
  const determinant = a * d - b * c;
  if (!Number.isFinite(determinant) || Math.abs(determinant) <= 1e-9) return null;
  return {
    x: (d * dx - c * dy) / determinant,
    y: (-b * dx + a * dy) / determinant
  };
}
function nearestGridPosition(contentEl, canvas, clientX, clientY) {
  const lines = [...contentEl.querySelectorAll("[data-vecto-grid-line]")];
  if (lines.length === 0) return null;
  const [a, b, c, d] = parseCssMatrix(getComputedStyle(contentEl).transform);
  if (a > 0 && d > 0 && Math.abs(b) <= 1e-9 && Math.abs(c) <= 1e-9) {
    let nearest2 = null;
    for (const line of lines) {
      const rect = line.getBoundingClientRect();
      const dy = clientY < rect.top ? rect.top - clientY : clientY > rect.bottom ? clientY - rect.bottom : 0;
      const dx = clientX < rect.left ? rect.left - clientX : clientX > rect.right ? clientX - rect.right : 0;
      const distance = dy * 4096 + dx;
      if (!nearest2 || distance < nearest2.distance) nearest2 = { line, distance, rect };
    }
    if (!nearest2) return null;
    const localWidth = Number.parseFloat(nearest2.line.style.width) || 0;
    const scaleX = localWidth > 0 && nearest2.rect.width > 0 ? nearest2.rect.width / localWidth : 1;
    const localX = (clientX - nearest2.rect.left) / scaleX;
    return nearestGridPositionInLine(nearest2.line, localX);
  }
  const local = clientToGridLocal(contentEl, canvas, clientX, clientY);
  if (!local) return null;
  let nearest = null;
  for (const line of lines) {
    const left = Number.parseFloat(line.style.left) || 0;
    const top = Number.parseFloat(line.style.top) || 0;
    const width = Number.parseFloat(line.style.width) || 0;
    const height = Number.parseFloat(line.style.height) || 0;
    const dy = local.y < top ? top - local.y : local.y > top + height ? local.y - top - height : 0;
    const dx = local.x < left ? left - local.x : local.x > left + width ? local.x - left - width : 0;
    const distance = dy * 4096 + dx;
    if (!nearest || distance < nearest.distance) nearest = { line, distance };
  }
  if (!nearest) return null;
  const lineLeft = Number.parseFloat(nearest.line.style.left) || 0;
  return nearestGridPositionInLine(nearest.line, local.x - lineLeft);
}
function nearestTextPositionInLine(line, x, y) {
  const texts = collectTextNodes(line);
  if (texts.length === 0) return null;
  let nearest = null;
  for (const node of texts) {
    const candidate = nearestOffsetForPoint(node, x, y);
    if (!nearest || candidate.distance < nearest.distance) {
      let { offset } = candidate;
      while (offset > 0 && node.data[offset - 1] === "\n") offset--;
      nearest = { position: { node, offset }, distance: candidate.distance };
    }
  }
  return _nullishCoalesce(_optionalChain([nearest, 'optionalAccess', _19 => _19.position]), () => ( null));
}
function nearestTextPositionInProjection(contentEl, canvas, x, y, eventTarget) {
  if (contentEl.dataset.vectoContentGrid !== void 0) {
    return nearestGridPosition(contentEl, canvas, x, y);
  }
  let targetLine = eventTarget;
  while (targetLine && targetLine.parentElement !== contentEl) {
    if (!contentEl.contains(targetLine)) {
      targetLine = null;
      break;
    }
    targetLine = targetLine.parentElement;
  }
  if (_optionalChain([targetLine, 'optionalAccess', _20 => _20.parentElement]) === contentEl) {
    return nearestTextPositionInLine(targetLine, x, y);
  }
  let bestLine = null;
  let bestDist = Infinity;
  for (let i = 0; i < contentEl.children.length; i++) {
    const child = contentEl.children[i];
    const rect = child.getBoundingClientRect();
    if (rect.width <= 0 && rect.height <= 0) continue;
    const dy = y < rect.top ? rect.top - y : y > rect.bottom ? y - rect.bottom : 0;
    const dx = x < rect.left ? rect.left - x : x > rect.right ? x - rect.right : 0;
    const dist = dy * 4096 + dx;
    if (dist < bestDist) {
      bestDist = dist;
      bestLine = child;
    }
  }
  if (bestLine) return nearestTextPositionInLine(bestLine, x, y);
  return nearestTextPositionInLine(contentEl, x, y);
}
function projectionAbsoluteOffset(root, caret) {
  let offset = 0;
  for (const node of collectTextNodes(root)) {
    if (node === caret.node) return offset + Math.min(caret.offset, node.data.length);
    offset += node.data.length;
  }
  return null;
}
function projectionCaretAt(root, absoluteOffset, affinity) {
  const nodes = collectTextNodes(root);
  if (nodes.length === 0) return null;
  let remaining = Math.max(0, absoluteOffset);
  for (let index = 0; index < nodes.length; index++) {
    const node = nodes[index];
    if (remaining < node.data.length || remaining === node.data.length && affinity === "backward") {
      return { node, offset: remaining };
    }
    if (remaining === node.data.length && index === nodes.length - 1) {
      return { node, offset: remaining };
    }
    remaining -= node.data.length;
  }
  const last = nodes[nodes.length - 1];
  return { node: last, offset: last.data.length };
}
function selectProjectionUnit(selection, root, caret, unit) {
  const absoluteOffset = projectionAbsoluteOffset(root, caret);
  const text = _nullishCoalesce(root.textContent, () => ( ""));
  if (absoluteOffset === null || text.length === 0) return false;
  let start = absoluteOffset;
  let end = absoluteOffset;
  if (unit === "line") {
    for (let index = Math.max(0, absoluteOffset - 1); index >= 0; index--) {
      if (text[index] === "\n" || text[index] === "\r") {
        start = index + 1;
        if (text[index] === "\r" && text[index + 1] === "\n") start++;
        break;
      }
    }
    const cr = text.indexOf("\r", absoluteOffset);
    const lf = text.indexOf("\n", absoluteOffset);
    const separator = [cr, lf].filter((index) => index >= 0).sort((a, b) => a - b)[0];
    end = separator === void 0 ? text.length : separator;
  } else if (caretWordSegmenter) {
    const segments = [...caretWordSegmenter.segment(text)];
    const selected = _nullishCoalesce(_nullishCoalesce(segments.find(
      (segment) => segment.isWordLike && absoluteOffset >= segment.index && absoluteOffset <= segment.index + segment.segment.length
    ), () => ( segments.find((segment) => segment.isWordLike && segment.index >= absoluteOffset))), () => ( [...segments].reverse().find((segment) => segment.isWordLike)));
    if (selected) {
      start = selected.index;
      end = selected.index + selected.segment.length;
    }
  } else {
    const isWord = (character) => /[\p{L}\p{N}_]/u.test(character);
    while (start > 0 && isWord(text[start - 1])) start--;
    while (end < text.length && isWord(text[end])) end++;
  }
  const anchor = projectionCaretAt(root, start, "forward");
  const focus = projectionCaretAt(root, end, "backward");
  if (!anchor || !focus) return false;
  selection.setBaseAndExtent(anchor.node, anchor.offset, focus.node, focus.offset);
  return true;
}
function extendSelection(selection, anchor, focus) {
  try {
    selection.setBaseAndExtent(anchor.node, anchor.offset, focus.node, focus.offset);
    return;
  } catch (e3) {
  }
  try {
    selection.collapse(anchor.node, anchor.offset);
    selection.extend(focus.node, focus.offset);
    return;
  } catch (e4) {
  }
  const anchorRange = document.createRange();
  anchorRange.setStart(anchor.node, anchor.offset);
  anchorRange.collapse(true);
  const focusRange = document.createRange();
  focusRange.setStart(focus.node, focus.offset);
  focusRange.collapse(true);
  const range = document.createRange();
  if (anchorRange.compareBoundaryPoints(Range.START_TO_START, focusRange) <= 0) {
    range.setStart(anchor.node, anchor.offset);
    range.setEnd(focus.node, focus.offset);
  } else {
    range.setStart(focus.node, focus.offset);
    range.setEnd(anchor.node, anchor.offset);
  }
  selection.removeAllRanges();
  selection.addRange(range);
}
var Scene = (_class2 = class _Scene {
  static __initStatic() {this.webglCreator = null}
  static __initStatic2() {this.webgpuManagerClass = null}
  static registerWebGLPointRendererCreator(creator) {
    _Scene.webglCreator = creator;
  }
  static registerWebGPUParticleSystemManager(managerClass) {
    _Scene.webgpuManagerClass = managerClass;
  }
  
  
  
  __init7() {this.isRunning = false}
  __init8() {this.lastTime = 0}
  
  /**
   * Redraw strategy:
   * - `'always'` (default): re-render every animation frame (legacy behavior).
   * - `'onDemand'`: only re-render when the scene is marked dirty (via
   *   {@link markDirty}) or while an animation is pending. Ideal for static /
   *   event-driven UIs where idle frames should cost ~0.
   */
  __init9() {this.renderMode = "always"}
  __init10() {this.dirty = true}
  /** Whether to throttle rendering to 2 FPS when the scene is static to save power. */
  __init11() {this.autoThrottle = true}
  /**
   * Frame-rate cap (power saving). `0` = uncapped (native refresh). When set,
   * the loop renders at most `maxFPS` times per second; animations still run,
   * just less often. See {@link SceneOptions.maxFPS}.
   */
  __init12() {this.maxFPS = 60}
  /** Whether the OS prefers-reduced-motion setting auto-caps the loop. */
  __init13() {this.respectReducedMotion = true}
  /** Cached media-query list; `.matches` is read live each frame. */
  __init14() {this.reducedMotionQuery = null}
  /** True when the OS asks for reduced motion and we respect it. Read by the animation drivers. */
  get prefersReducedMotion() {
    return this.respectReducedMotion && !!_optionalChain([this, 'access', _21 => _21.reducedMotionQuery, 'optionalAccess', _22 => _22.matches]);
  }
  /**
   * Throttle interval (ms) for the a11y/automation shadow sync. `0` = every
   * frame. See {@link SceneOptions.a11ySyncInterval}.
   */
  __init15() {this.a11ySyncInterval = 0}
  /** Timestamp of the last a11y sync, for throttling. */
  __init16() {this.lastA11ySync = -Infinity}
  /** True if we skipped an a11y sync during animation and need to sync when at rest. */
  __init17() {this.a11yPendingSyncAfterAnimation = false}
  // A11y / Automation Layer. `null` in non-DOM (SSR/Node) environments — the
  // whole projection degrades to a no-op so the engine's logic stays usable
  // server-side (e.g. headless layout / vector export) without jsdom.
  
  __init18() {this.a11yElements = /* @__PURE__ */ new Map()}
  /** DOM nodes mirroring static text content, keyed by entity id. */
  __init19() {this.contentElements = /* @__PURE__ */ new Map()}
  /** Pending cold font-calibration frame per projected grid entity. */
  __init20() {this.contentGridCalibrationFrames = /* @__PURE__ */ new Map()}
  /** Detached, untransformed font probes used by the cold calibration pass. */
  __init21() {this.contentGridCalibrationProbes = /* @__PURE__ */ new Map()}
  /** Invalidates grid font calibration after browser font availability changes. */
  __init22() {this.contentFontEpoch = 0}
  /** Cached Canvas-to-client scale for the current font/viewport epoch. */
  __init23() {this.contentMetricScaleEpoch = -1}
  __init24() {this.contentMetricScaleX = 1}
  __init25() {this.contentProjectionEnabled = true}
  /**
   * True while a text-selection drag that started on a projection's blank
   * region (no text node under the press) is being driven manually — the
   * browser has no native anchor for it, so mousemove extends the Selection
   * from the position we resolved ourselves.
   */
  __init26() {this.blankRegionSelectionDrag = false}
  __init27() {this.contentSelectionAnchor = null}
  __init28() {this.contentSelectionEndListener = null}
  // Animation/interactive flags collected during the render walk (tree-walk
  // fusion): the loop reads last frame's answers instead of re-walking the
  // tree up to 4× per tick. Start true so the first tick stays conservative.
  __init29() {this.frameHadAnimation = true}
  __init30() {this.frameHadInteractive = true}
  
  __init31() {this.focusedA11yElement = null}
  __init32() {this.caretBlinkTimer = null}
  __init33() {this.a11yNeedsReorder = true}
  __init34() {this.portalRoot = null}
  __init35() {this.fullViewportElements = []}
  __init36() {this.normalElements = []}
  __init37() {this.activeIds = /* @__PURE__ */ new Set()}
  __init38() {this.activePortalsThisFrame = /* @__PURE__ */ new Set()}
  __init39() {this.activePortalsPrevFrame = /* @__PURE__ */ new Set()}
  __init40() {this.portalEntities = /* @__PURE__ */ new Map()}
  __init41() {this.renderOrderCounter = 0}
  /**
   * Authoritative paint order for semantic nodes discovered during the main
   * render. A node may not have a DOM projection until the following a11y
   * sync, so retaining the order prevents a newly opened overlay from spending
   * its first frame below previously projected controls.
   */
  __init42() {this.a11yRenderOrders = /* @__PURE__ */ new Map()}
  // Optional WebGL point-cloud layer (see SceneOptions.pointBackend).
  __init43() {this.pointRenderer = null}
  __init44() {this.glCanvas = null}
  
  
  
  __init45() {this.disableWindowResize = false}
  /** See {@link SceneOptions.maxDPR}. `undefined` = uncapped (real DPR). */
  
  // WebGPU properties
  __init46() {this.destroyed = false}
  __init47() {this.device = null}
  __init48() {this.deviceLost = false}
  __init49() {this.particleBackend = "auto"}
  __init50() {this._webgpuDisabled = false}
  get webgpuDisabled() {
    return this._webgpuDisabled || this.particleBackend === "cpu";
  }
  set webgpuDisabled(value) {
    this._webgpuDisabled = value;
  }
  __init51() {this.recoveryTimerId = null}
  __init52() {this.manager = null}
  __init53() {this.initializingWebGPU = false}
  __init54() {this.gpuCanvas = null}
  __init55() {this.gpuContext = null}
  /** True while the GPU canvas holds a presented particle frame (needs clearing when they leave). */
  __init56() {this.gpuHasContent = false}
  __init57() {this.mouseX = -9999}
  __init58() {this.mouseY = -9999}
  __init59() {this.pointerMoveListener = null}
  __init60() {this.pointerLeaveListener = null}
  __init61() {this.hasWarnedZeroSize = false}
  __init62() {this.fontLoadHandler = null}
  // ── Dev-mode warning infrastructure ──────────────────────────────
  //
  // Enable with `Scene.devMode = true` or by setting `globalThis.__DEV__`.
  // Auto-detected when `NODE_ENV === 'development'`.
  //
  // Checks run once every ~120 frames (~2s at 60fps) to keep overhead
  // negligible even when dev mode is on.
  /** Toggle development-mode runtime warnings globally. */
  static __initStatic3() {this.devMode = false}
  static _devModeDetected() {
    if (_Scene.devMode) return true;
    const gp = typeof globalThis !== "undefined" ? globalThis : void 0;
    if (_optionalChain([gp, 'optionalAccess', _23 => _23.__DEV__])) return true;
    if (_optionalChain([gp, 'optionalAccess', _24 => _24.process, 'optionalAccess', _25 => _25.env, 'optionalAccess', _26 => _26.NODE_ENV]) === "development") return true;
    return false;
  }
  
  __init63() {this._devFrameCount = 0}
  _devWarn(message) {
    if (!this._devActive) return;
    console.warn(`[vectojs/dev] ${message}`);
  }
  /** @internal Periodic dev checks — called once per frame in dev mode. */
  _devRunChecks() {
    this._devFrameCount++;
    if (this._devFrameCount % 120 !== 0) return;
    if (this.a11yElements) {
      let interactiveCount = 0;
      const walk = (node) => {
        if (node.interactive && node.width > 0) interactiveCount++;
        for (const c of node.children) walk(c);
      };
      walk(this.root);
      for (const c of this.overlayRoot.children) walk(c);
      const shadowCount = this.a11yElements.size;
      if (shadowCount > interactiveCount + 2) {
        this._devWarn(
          `a11yElements (${shadowCount}) exceeds interactive entities (${interactiveCount}). Call scene.detachA11y(entity) before removing interactive children from the tree, or their shadow nodes leak.`
        );
      }
    }
    let checked = 0;
    const walkProjections = (node) => {
      if (checked > 10) return;
      const proj = _optionalChain([node, 'access', _27 => _27.getContentProjection, 'optionalCall', _28 => _28()]);
      if (_optionalChain([proj, 'optionalAccess', _29 => _29.text]) && proj.selectable !== false) {
        const el = _optionalChain([this, 'access', _30 => _30.contentElements, 'optionalAccess', _31 => _31.get, 'call', _32 => _32(node.id)]);
        if (el) {
          const projectedText = el.textContent || "";
          if (projectedText !== "" && projectedText !== proj.text) {
            this._devWarn(
              `Content projection mismatch for entity "${node.id}": projection says "${proj.text.slice(0, 60)}" but DOM shows "${projectedText.slice(0, 60)}". Ensure getContentProjection() output matches what drawSelf renders.`
            );
          }
        }
      }
      checked++;
      for (const c of node.children) walkProjections(c);
    };
    walkProjections(this.root);
  }
  constructor(canvas, options = {}) {;_class2.prototype.__init7.call(this);_class2.prototype.__init8.call(this);_class2.prototype.__init9.call(this);_class2.prototype.__init10.call(this);_class2.prototype.__init11.call(this);_class2.prototype.__init12.call(this);_class2.prototype.__init13.call(this);_class2.prototype.__init14.call(this);_class2.prototype.__init15.call(this);_class2.prototype.__init16.call(this);_class2.prototype.__init17.call(this);_class2.prototype.__init18.call(this);_class2.prototype.__init19.call(this);_class2.prototype.__init20.call(this);_class2.prototype.__init21.call(this);_class2.prototype.__init22.call(this);_class2.prototype.__init23.call(this);_class2.prototype.__init24.call(this);_class2.prototype.__init25.call(this);_class2.prototype.__init26.call(this);_class2.prototype.__init27.call(this);_class2.prototype.__init28.call(this);_class2.prototype.__init29.call(this);_class2.prototype.__init30.call(this);_class2.prototype.__init31.call(this);_class2.prototype.__init32.call(this);_class2.prototype.__init33.call(this);_class2.prototype.__init34.call(this);_class2.prototype.__init35.call(this);_class2.prototype.__init36.call(this);_class2.prototype.__init37.call(this);_class2.prototype.__init38.call(this);_class2.prototype.__init39.call(this);_class2.prototype.__init40.call(this);_class2.prototype.__init41.call(this);_class2.prototype.__init42.call(this);_class2.prototype.__init43.call(this);_class2.prototype.__init44.call(this);_class2.prototype.__init45.call(this);_class2.prototype.__init46.call(this);_class2.prototype.__init47.call(this);_class2.prototype.__init48.call(this);_class2.prototype.__init49.call(this);_class2.prototype.__init50.call(this);_class2.prototype.__init51.call(this);_class2.prototype.__init52.call(this);_class2.prototype.__init53.call(this);_class2.prototype.__init54.call(this);_class2.prototype.__init55.call(this);_class2.prototype.__init56.call(this);_class2.prototype.__init57.call(this);_class2.prototype.__init58.call(this);_class2.prototype.__init59.call(this);_class2.prototype.__init60.call(this);_class2.prototype.__init61.call(this);_class2.prototype.__init62.call(this);_class2.prototype.__init63.call(this);
    this.canvas = canvas;
    this.debugA11y = _nullishCoalesce(options.debugA11y, () => ( false));
    this.disableWindowResize = _nullishCoalesce(options.disableWindowResize, () => ( false));
    this.maxDPR = options.maxDPR;
    if (this.disableWindowResize) {
      const styleWidth = parseInlinePx(_optionalChain([canvas, 'access', _33 => _33.style, 'optionalAccess', _34 => _34.width]));
      const styleHeight = parseInlinePx(_optionalChain([canvas, 'access', _35 => _35.style, 'optionalAccess', _36 => _36.height]));
      this.width = _nullishCoalesce(styleWidth, () => ( (canvas.width || canvas.clientWidth || 0)));
      this.height = _nullishCoalesce(styleHeight, () => ( (canvas.height || canvas.clientHeight || 0)));
    } else {
      this.width = typeof window !== "undefined" ? window.innerWidth : canvas.clientWidth || canvas.width || 800;
      this.height = typeof window !== "undefined" ? window.innerHeight : canvas.clientHeight || canvas.height || 600;
    }
    const globalProcess = typeof globalThis !== "undefined" ? globalThis.process : void 0;
    const isTest = globalProcess && (_optionalChain([globalProcess, 'access', _37 => _37.env, 'optionalAccess', _38 => _38.NODE_ENV]) === "test" || _optionalChain([globalProcess, 'access', _39 => _39.env, 'optionalAccess', _40 => _40.VITEST]) === "true");
    this.maxFPS = _nullishCoalesce(options.maxFPS, () => ( (isTest ? 0 : 60)));
    this.respectReducedMotion = _nullishCoalesce(options.respectReducedMotion, () => ( true));
    this.autoThrottle = _nullishCoalesce(options.autoThrottle, () => ( true));
    this.particleBackend = _nullishCoalesce(options.particleBackend, () => ( "auto"));
    this.a11ySyncInterval = _nullishCoalesce(options.a11ySyncInterval, () => ( 0));
    this.contentProjectionEnabled = _nullishCoalesce(options.contentProjection, () => ( true));
    this._devActive = _Scene._devModeDetected();
    this.reducedMotionQuery = typeof window !== "undefined" && typeof window.matchMedia === "function" ? window.matchMedia("(prefers-reduced-motion: reduce)") : null;
    this.root = new class RootEntity extends _chunk2Z23LTH3js.Entity {
      isPointInside() {
        return false;
      }
      // Root renders nothing itself — renderNode() handles all child traversal.
      render(_r) {
      }
    }("root");
    this.root._scene = this;
    this.overlayRoot = new class OverlayRoot extends _chunk2Z23LTH3js.Entity {
      isPointInside() {
        return false;
      }
      render() {
      }
    }("overlayRoot");
    this.overlayRoot._scene = this;
    if (options.renderer) {
      this.renderer = options.renderer;
    } else {
      this.renderer = new (0, _chunkORCPCIJ7js.CanvasRenderer)(
        canvas,
        this.disableWindowResize ? { width: this.width, height: this.height } : void 0,
        this.maxDPR
      );
    }
    if (typeof document !== "undefined") {
      this.a11yRoot = document.createElement("div");
      this.a11yRoot.style.position = "absolute";
      this.a11yRoot.style.top = "0";
      this.a11yRoot.style.left = "0";
      this.a11yRoot.style.width = "100vw";
      this.a11yRoot.style.height = "100vh";
      this.a11yRoot.style.pointerEvents = "none";
      this.a11yRoot.style.overflow = "hidden";
      this.a11yRoot.style.zIndex = "10";
      this.a11yRoot.style.userSelect = "text";
      this.a11yRoot.addEventListener("mousedown", (e) => {
        if (e.button !== 0) return;
        const target = e.target;
        const contentEl = target.closest("[data-vecto-content]");
        if (target === this.a11yRoot || !contentEl) return;
        if (getComputedStyle(contentEl).pointerEvents !== "auto") return;
        const selection = window.getSelection();
        if (!selection) return;
        this.a11yRoot.style.pointerEvents = "auto";
        const resolved = nearestTextPositionInProjection(
          contentEl,
          this.canvas,
          e.clientX,
          e.clientY,
          target
        );
        if (resolved) {
          if (e.detail >= 2) {
            selection.removeAllRanges();
            selectProjectionUnit(selection, contentEl, resolved, e.detail >= 3 ? "line" : "word");
            this.endContentSelectionDrag();
            e.preventDefault();
            return;
          }
          const existingAnchor = e.shiftKey && selection.anchorNode instanceof Text ? { node: selection.anchorNode, offset: selection.anchorOffset } : null;
          const anchor = existingAnchor && this.a11yRoot.contains(existingAnchor.node) ? existingAnchor : resolved;
          if (e.shiftKey && existingAnchor) extendSelection(selection, anchor, resolved);
          else selection.collapse(resolved.node, resolved.offset);
          this.contentSelectionAnchor = anchor;
          this.blankRegionSelectionDrag = true;
          e.preventDefault();
        }
      });
      this.a11yRoot.addEventListener("dblclick", (e) => {
        const target = e.target;
        const contentEl = target.closest("[data-vecto-content]");
        if (!contentEl || getComputedStyle(contentEl).pointerEvents !== "auto") return;
        const selection = window.getSelection();
        const caret = nearestTextPositionInProjection(
          contentEl,
          this.canvas,
          e.clientX,
          e.clientY,
          target
        );
        if (!selection || !caret) return;
        selection.removeAllRanges();
        selectProjectionUnit(selection, contentEl, caret, "word");
        this.endContentSelectionDrag();
        e.preventDefault();
      });
      this.a11yRoot.addEventListener("mousemove", (e) => {
        if (!this.blankRegionSelectionDrag) return;
        const selection = window.getSelection();
        if (!selection || selection.rangeCount === 0) return;
        const target = e.target;
        let contentEl = target.closest("[data-vecto-content]");
        if (!contentEl) {
          let bestDistance = Infinity;
          for (const candidate of this.contentElements.values()) {
            if (getComputedStyle(candidate).pointerEvents !== "auto") continue;
            const rect = candidate.getBoundingClientRect();
            const dx = e.clientX < rect.left ? rect.left - e.clientX : e.clientX > rect.right ? e.clientX - rect.right : 0;
            const dy = e.clientY < rect.top ? rect.top - e.clientY : e.clientY > rect.bottom ? e.clientY - rect.bottom : 0;
            const distance = dx * dx + dy * dy;
            if (distance < bestDistance) {
              bestDistance = distance;
              contentEl = candidate;
            }
          }
        }
        const focus = contentEl ? nearestTextPositionInProjection(contentEl, this.canvas, e.clientX, e.clientY, target) : null;
        const anchor = this.contentSelectionAnchor;
        if (focus && anchor) {
          extendSelection(selection, anchor, focus);
        }
      });
      const endDrag = () => this.endContentSelectionDrag();
      this.a11yRoot.addEventListener("mouseup", endDrag);
      this.a11yRoot.addEventListener("mouseleave", endDrag);
      window.addEventListener("mouseup", endDrag);
      window.addEventListener("blur", endDrag);
      this.contentSelectionEndListener = endDrag;
      if (canvas.parentElement) {
        canvas.parentElement.appendChild(this.a11yRoot);
      }
      this.portalRoot = document.createElement("div");
      this.portalRoot.style.position = "absolute";
      this.portalRoot.style.top = "0";
      this.portalRoot.style.left = "0";
      this.portalRoot.style.width = "100vw";
      this.portalRoot.style.height = "100vh";
      this.portalRoot.style.pointerEvents = "none";
      this.portalRoot.style.overflow = "hidden";
      this.portalRoot.style.zIndex = "9";
      if (canvas.parentElement) {
        canvas.parentElement.appendChild(this.portalRoot);
      }
    } else {
      this.a11yRoot = null;
      this.portalRoot = null;
    }
    if (options.pointBackend === "webgl" && typeof document !== "undefined") {
      const gl = document.createElement("canvas");
      gl.style.position = "absolute";
      gl.style.top = "0";
      gl.style.left = "0";
      gl.style.pointerEvents = "none";
      gl.style.zIndex = "5";
      if (canvas.parentElement) canvas.parentElement.appendChild(gl);
      const pr = _Scene.webglCreator ? _Scene.webglCreator(gl) : null;
      if (pr) {
        pr.maxDPR = this.maxDPR;
        pr.resize(this.width, this.height);
        this.glCanvas = gl;
        this.pointRenderer = pr;
      } else {
        gl.remove();
      }
    }
    this.resizeHandler = () => {
      this.resize(window.innerWidth, window.innerHeight);
    };
    if (typeof document !== "undefined" && document.fonts) {
      this.fontLoadHandler = () => {
        _chunk2Z23LTH3js.clearCssLineBoxMetrics.call(void 0, );
        this.contentFontEpoch++;
        this.markDirty();
      };
      document.fonts.ready.then(this.fontLoadHandler);
      document.fonts.addEventListener("loadingdone", this.fontLoadHandler);
    }
    this.setupEvents();
  }
  endContentSelectionDrag() {
    this.blankRegionSelectionDrag = false;
    this.contentSelectionAnchor = null;
    if (this.a11yRoot) this.a11yRoot.style.pointerEvents = "none";
  }
  releaseContentSelectionForRebuild(el) {
    const selection = typeof window !== "undefined" && typeof window.getSelection === "function" ? window.getSelection() : null;
    const ownsSelection = this.contentSelectionAnchor && el.contains(this.contentSelectionAnchor.node) || (_optionalChain([selection, 'optionalAccess', _41 => _41.anchorNode]) ? el.contains(selection.anchorNode) : false) || (_optionalChain([selection, 'optionalAccess', _42 => _42.focusNode]) ? el.contains(selection.focusNode) : false);
    if (!ownsSelection) return;
    this.endContentSelectionDrag();
    _optionalChain([selection, 'optionalAccess', _43 => _43.removeAllRanges, 'call', _44 => _44()]);
  }
  /**
   * Expose the underlying {@link IRenderer} for advanced direct-draw operations.
   *
   * @returns The active renderer instance.
   */
  getRenderer() {
    return this.renderer;
  }
  /** Convert browser viewport coordinates into this Scene's logical coordinates. */
  clientToScene(clientX, clientY) {
    const rect = _optionalChain([this, 'access', _45 => _45.canvas, 'access', _46 => _46.getBoundingClientRect, 'optionalCall', _47 => _47()]);
    if (!rect) return { x: clientX, y: clientY };
    const cssWidth = rect.width || this.canvas.clientWidth || this.width;
    const cssHeight = rect.height || this.canvas.clientHeight || this.height;
    return {
      x: (clientX - rect.left) * (cssWidth > 0 ? this.width / cssWidth : 1),
      y: (clientY - rect.top) * (cssHeight > 0 ? this.height / cssHeight : 1)
    };
  }
  /**
   * Add a top-level entity to the scene graph.
   *
   * @param entity - The entity to attach to the scene root.
   * @returns `this` for method chaining.
   * @example scene.add(new CircleEntity());
   */
  add(entity) {
    this.root.add(entity);
    return this;
  }
  clearContentGridState(entityId, el) {
    const calibrationFrame = this.contentGridCalibrationFrames.get(entityId);
    if (calibrationFrame !== void 0 && typeof cancelAnimationFrame === "function") {
      cancelAnimationFrame(calibrationFrame);
    }
    this.contentGridCalibrationFrames.delete(entityId);
    _optionalChain([this, 'access', _48 => _48.contentGridCalibrationProbes, 'access', _49 => _49.get, 'call', _50 => _50(entityId), 'optionalAccess', _51 => _51.remove, 'call', _52 => _52()]);
    this.contentGridCalibrationProbes.delete(entityId);
    delete el.dataset.vectoGridCalibrationPending;
    delete el.dataset.vectoGridCalibration;
    delete el.dataset.vectoGridReady;
    delete el.dataset.vectoContentGrid;
    delete el.dataset.vectoGridCarriers;
    delete el.dataset.vectoGridMaterializeMs;
    delete el.dataset.vectoGridCalibrationSamples;
    delete el.dataset.vectoGridCalibrationMs;
    this.releaseContentSelectionForRebuild(el);
  }
  removeA11yRecursively(node) {
    if (node.isDOMPortal) {
      node.domElement.remove();
      this.portalEntities.delete(node.id);
      this.activePortalsThisFrame.delete(node.id);
      this.activePortalsPrevFrame.delete(node.id);
    }
    const contentEl = this.contentElements.get(node.id);
    if (contentEl) {
      this.clearContentGridState(node.id, contentEl);
      contentEl.remove();
      this.contentElements.delete(node.id);
      this.a11yNeedsReorder = true;
    }
    const el = this.a11yElements.get(node.id);
    if (el) {
      if (el === this.focusedA11yElement) {
        this.focusedA11yElement = null;
        if (this.caretBlinkTimer) {
          clearInterval(this.caretBlinkTimer);
          this.caretBlinkTimer = null;
        }
      }
      el.remove();
      this.a11yElements.delete(node.id);
      this.a11yNeedsReorder = true;
    }
    for (const child of node.children) {
      this.removeA11yRecursively(child);
    }
  }
  /**
   * Remove a top-level entity from the scene graph and clean up its
   * accessibility shadow elements recursively.
   *
   * @param entity - The entity to detach from the scene root.
   * @returns `this` for method chaining.
   */
  remove(entity) {
    this.root.remove(entity);
    this.removeA11yRecursively(entity);
    return this;
  }
  /**
   * Tear down the a11y/automation shadow nodes for `entity` and its descendants
   * without removing it from the scene graph. Components that manage dynamic
   * interactive *child* entities (e.g. a {@link Entity}'s per-link hotspots) call
   * this before discarding those children so their shadow `<a>`/controls don't
   * leak (the per-frame `syncA11y` only creates/updates, it never prunes).
   *
   * @param entity - The subtree whose shadow nodes should be removed.
   */
  detachA11y(entity) {
    this.removeA11yRecursively(entity);
  }
  /**
   * Add an overlay entity to the overlay root, bypassing main tree clipping bounds.
   */
  showOverlay(overlay) {
    this.overlayRoot.add(overlay);
    this.markDirty();
  }
  /**
   * Remove an overlay entity from the overlay root.
   */
  hideOverlay(overlay) {
    this.overlayRoot.remove(overlay);
    this.removeA11yRecursively(overlay);
    this.markDirty();
  }
  destroyEntitySubtree(entity) {
    while (entity.children.length > 0) this.destroyEntitySubtree(entity.children.at(-1));
    entity.destroy();
  }
  /**
   * Tear down the Scene, halt the loop, and clean up event listeners and DOM elements.
   */
  destroy() {
    if (this.destroyed) return;
    this.destroyed = true;
    this.stop();
    if (typeof document !== "undefined" && document.fonts && this.fontLoadHandler) {
      document.fonts.removeEventListener("loadingdone", this.fontLoadHandler);
    }
    while (this.root.children.length > 0) this.destroyEntitySubtree(this.root.children.at(-1));
    while (this.overlayRoot.children.length > 0) {
      this.destroyEntitySubtree(this.overlayRoot.children.at(-1));
    }
    if (typeof window !== "undefined" && !this.disableWindowResize) {
      window.removeEventListener("resize", this.resizeHandler);
    }
    if (typeof window !== "undefined" && this.contentSelectionEndListener) {
      window.removeEventListener("mouseup", this.contentSelectionEndListener);
      window.removeEventListener("blur", this.contentSelectionEndListener);
      this.contentSelectionEndListener = null;
    }
    if (typeof window !== "undefined" && this.canvas && typeof this.canvas.removeEventListener === "function") {
      if (this.pointerMoveListener) {
        this.canvas.removeEventListener("pointermove", this.pointerMoveListener);
      }
      if (this.pointerLeaveListener) {
        this.canvas.removeEventListener("pointerleave", this.pointerLeaveListener);
      }
    }
    _optionalChain([this, 'access', _53 => _53.a11yRoot, 'optionalAccess', _54 => _54.remove, 'call', _55 => _55()]);
    _optionalChain([this, 'access', _56 => _56.portalRoot, 'optionalAccess', _57 => _57.remove, 'call', _58 => _58()]);
    this.a11yElements.clear();
    for (const el of this.contentElements.values()) el.remove();
    this.contentElements.clear();
    if (typeof cancelAnimationFrame === "function") {
      for (const frame of this.contentGridCalibrationFrames.values()) {
        cancelAnimationFrame(frame);
      }
    }
    this.contentGridCalibrationFrames.clear();
    for (const probe of this.contentGridCalibrationProbes.values()) probe.remove();
    this.contentGridCalibrationProbes.clear();
    this.endContentSelectionDrag();
    _optionalChain([this, 'access', _59 => _59.pointRenderer, 'optionalAccess', _60 => _60.destroy, 'call', _61 => _61()]);
    _optionalChain([this, 'access', _62 => _62.renderer, 'access', _63 => _63.dispose, 'optionalCall', _64 => _64()]);
    _optionalChain([this, 'access', _65 => _65.glCanvas, 'optionalAccess', _66 => _66.remove, 'call', _67 => _67()]);
    _optionalChain([this, 'access', _68 => _68.gpuCanvas, 'optionalAccess', _69 => _69.remove, 'call', _70 => _70()]);
    this.gpuCanvas = null;
    this.gpuContext = null;
    if (this.recoveryTimerId) {
      clearTimeout(this.recoveryTimerId);
      this.recoveryTimerId = null;
    }
    if (this.manager) {
      this.manager.destroy();
      this.manager = null;
    }
    if (this.device) {
      _optionalChain([this, 'access', _71 => _71.device, 'access', _72 => _72.destroy, 'optionalCall', _73 => _73()]);
      this.device = null;
    }
  }
  setupEvents() {
    if (typeof window !== "undefined" && !this.disableWindowResize) {
      window.addEventListener("resize", this.resizeHandler);
    }
    if (typeof window !== "undefined" && this.canvas && typeof this.canvas.addEventListener === "function") {
      this.pointerMoveListener = (e) => {
        const point = this.clientToScene(e.clientX, e.clientY);
        this.mouseX = point.x;
        this.mouseY = point.y;
      };
      this.pointerLeaveListener = () => {
        this.mouseX = -9999;
        this.mouseY = -9999;
      };
      this.canvas.addEventListener("pointermove", this.pointerMoveListener);
      this.canvas.addEventListener("pointerleave", this.pointerLeaveListener);
    }
  }
  /**
   * Begin the `requestAnimationFrame` render loop.
   *
   * Idempotent — calling `start()` on an already-running scene is a no-op.
   */
  start() {
    if (this.isRunning) return;
    if ((this.width === 0 || this.height === 0) && !this.hasWarnedZeroSize) {
      console.warn(
        `[VectoJS] Scene started with width or height set to 0 (width: ${this.width}, height: ${this.height}). Entities may not render or simulate correctly. Please call scene.resize(width, height) to set valid dimensions.`
      );
      this.hasWarnedZeroSize = true;
    }
    this.isRunning = true;
    this.lastTime = typeof performance !== "undefined" ? performance.now() : 0;
    this.scheduleFrame();
    const isTextFocused = this.focusedA11yElement instanceof HTMLInputElement || this.focusedA11yElement instanceof HTMLTextAreaElement;
    if (isTextFocused && this.renderMode === "onDemand" && !this.caretBlinkTimer) {
      this.caretBlinkTimer = setInterval(() => {
        this.markDirty();
      }, 500);
    }
  }
  /** Schedule the next frame, or no-op where `requestAnimationFrame` is absent (SSR). */
  scheduleFrame() {
    if (typeof requestAnimationFrame !== "undefined") {
      requestAnimationFrame((t) => this.loop(t));
    }
  }
  /**
   * Halt the render loop after the current frame completes.
   *
   * Call {@link start} again to resume rendering.
   */
  stop() {
    this.isRunning = false;
    if (this.caretBlinkTimer) {
      clearInterval(this.caretBlinkTimer);
      this.caretBlinkTimer = null;
    }
  }
  /**
   * Manually advance the scene clock by `dt` milliseconds and render synchronously.
   * Essential for deterministic rendering (e.g. video export).
   * Note: You should call `scene.stop()` before using this to avoid conflict with the rAF loop.
   */
  /**
   * The scene-graph root entity. Exposed read-only for tooling — the devtools
   * inspector walks it to build the Virtual Math Tree view. Mutate the graph
   * through {@link add}/{@link remove}, not by editing this node directly.
   */
  get rootEntity() {
    return this.root;
  }
  /** The overlay layer root (see {@link showOverlay}), read-only for tooling. */
  get overlayRootEntity() {
    return this.overlayRoot;
  }
  step(dt) {
    const time = this.lastTime + dt;
    this.lastTime = time;
    this.render(this.renderer, dt, time);
    this.dirty = false;
  }
  /**
   * Mark the scene as needing a redraw on the next frame.
   *
   * Only meaningful in `onDemand` {@link renderMode}: call it after mutating
   * entity state outside of {@link Entity.animate} so the change is rendered.
   */
  markDirty() {
    this.dirty = true;
  }
  /** True when any node in the subtree has a pending animation. */
  /** True when any node in the subtree is interactive (drives a11y sync). */
  syncOptionalAttribute(element, name, value) {
    if (value === void 0) {
      if (element.hasAttribute(name)) element.removeAttribute(name);
      return;
    }
    if (element.getAttribute(name) !== value) element.setAttribute(name, value);
  }
  syncA11y(node) {
    if (!this.a11yRoot) return;
    if (node.isDOMPortal) {
      return;
    }
    if (node.interactive && (node.width > 0 || node.a11yFullViewport)) {
      let el = this.a11yElements.get(node.id);
      const attrs = node.getA11yAttributes();
      const expectedTag = attrs.tag || "div";
      if (el && el.tagName.toLowerCase() !== expectedTag.toLowerCase()) {
        if (el === this.focusedA11yElement) {
          this.focusedA11yElement = null;
          if (this.caretBlinkTimer) {
            clearInterval(this.caretBlinkTimer);
            this.caretBlinkTimer = null;
          }
        }
        if (el.parentNode === this.a11yRoot) {
          this.a11yRoot.removeChild(el);
        }
        this.a11yElements.delete(node.id);
        el = void 0;
        this.a11yNeedsReorder = true;
      }
      if (!el) {
        el = document.createElement(expectedTag);
        el.id = node.id;
        el.setAttribute("data-vecto-id", node.id);
        el.style.position = "absolute";
        el.style.transformOrigin = "0 0";
        el.style.pointerEvents = "auto";
        el.style.touchAction = "pinch-zoom";
        el.style.margin = "0";
        el.style.padding = "0";
        el.style.outline = "none";
        el.style.cursor = node.a11yFullViewport ? "default" : "pointer";
        if (this.debugA11y) {
          el.style.backgroundColor = "rgba(56, 189, 248, 0.05)";
          el.style.border = "1px dashed rgba(56, 189, 248, 0.4)";
        } else {
          el.style.opacity = "0";
          el.style.border = "none";
          el.style.background = "transparent";
        }
        el.addEventListener("click", (e) => {
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("click", node, e));
        });
        el.addEventListener("dblclick", (e) => {
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("dblclick", node, e));
        });
        el.addEventListener("mouseenter", (e) => {
          if (this.debugA11y) el.style.backgroundColor = "rgba(56, 189, 248, 0.2)";
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("hover", node, e, false));
        });
        el.addEventListener("mouseleave", (e) => {
          if (this.debugA11y) el.style.backgroundColor = "rgba(56, 189, 248, 0.05)";
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("pointerleave", node, e, false));
        });
        const capEl = el;
        const releasePointer = (event) => {
          if (typeof capEl.releasePointerCapture !== "function") return;
          if (typeof capEl.hasPointerCapture === "function" && !capEl.hasPointerCapture(event.pointerId)) {
            return;
          }
          try {
            capEl.releasePointerCapture(event.pointerId);
          } catch (error) {
            if (!(error instanceof DOMException) || error.name !== "NotFoundError") throw error;
          }
        };
        el.addEventListener("pointerdown", (e) => {
          if (typeof capEl.setPointerCapture === "function") capEl.setPointerCapture(e.pointerId);
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("pointerdown", node, e));
        });
        el.addEventListener("pointerup", (e) => {
          releasePointer(e);
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("pointerup", node, e));
        });
        el.addEventListener("pointercancel", (e) => {
          releasePointer(e);
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("pointercancel", node, e));
        });
        el.addEventListener(
          "pointermove",
          (e) => node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("pointermove", node, e))
        );
        el.addEventListener(
          "wheel",
          (e) => {
            node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("wheel", node, e));
          },
          { passive: false }
        );
        el.addEventListener("keydown", (e) => {
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("keydown", node, e));
        });
        el.addEventListener("keyup", (e) => {
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("keyup", node, e));
        });
        if (el instanceof HTMLInputElement || el instanceof HTMLTextAreaElement) {
          const input = el;
          let composition = null;
          const forward = () => {
            input._lastSyncedValue = input.value;
            node.emit("change", {
              value: input.value,
              checked: input instanceof HTMLInputElement ? input.checked : void 0,
              selectionStart: _nullishCoalesce(input.selectionStart, () => ( input.value.length)),
              selectionEnd: _nullishCoalesce(input.selectionEnd, () => ( input.value.length)),
              composition
            });
            this.markDirty();
          };
          el.addEventListener("input", forward);
          el.addEventListener("change", forward);
          el.addEventListener("keyup", forward);
          el.addEventListener("click", forward);
          el.addEventListener("select", forward);
          el.addEventListener("compositionstart", () => {
            composition = { start: _nullishCoalesce(input.selectionStart, () => ( input.value.length)), length: 0 };
            forward();
          });
          el.addEventListener("compositionupdate", (e) => {
            const data = _nullishCoalesce(e.data, () => ( ""));
            composition = { start: _nullishCoalesce(_optionalChain([composition, 'optionalAccess', _74 => _74.start]), () => ( 0)), length: data.length };
            forward();
          });
          el.addEventListener("compositionend", () => {
            composition = null;
            forward();
          });
        }
        const isTextInput = el instanceof HTMLInputElement || el instanceof HTMLTextAreaElement;
        el.addEventListener("focus", () => {
          this.focusedA11yElement = el;
          node.emit("focus", {});
          if (isTextInput && this.renderMode === "onDemand" && this.isRunning && !this.caretBlinkTimer) {
            this.caretBlinkTimer = setInterval(() => {
              this.markDirty();
            }, 500);
          }
        });
        el.addEventListener("blur", () => {
          if (this.focusedA11yElement === el) {
            this.focusedA11yElement = null;
          }
          const isTextFocused = this.focusedA11yElement instanceof HTMLInputElement || this.focusedA11yElement instanceof HTMLTextAreaElement;
          if (!isTextFocused && this.caretBlinkTimer) {
            clearInterval(this.caretBlinkTimer);
            this.caretBlinkTimer = null;
          }
          node.emit("blur", {});
        });
        if (!isNativelyFocusable(el) && attrs.role && INTERACTIVE_A11Y_ROLES.has(attrs.role)) {
          el.addEventListener("keydown", (e) => {
            if (e.key === "Enter" || e.key === " ") {
              e.preventDefault();
              node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("click", node, e));
            }
          });
        }
        if (node.a11yFullViewport) {
          this.a11yRoot.insertBefore(el, this.a11yRoot.firstChild);
        } else {
          this.a11yRoot.appendChild(el);
        }
        this.a11yElements.set(node.id, el);
        this.a11yNeedsReorder = true;
      }
      this.syncOptionalAttribute(el, "role", attrs.role);
      this.syncOptionalAttribute(el, "aria-label", attrs.label);
      const semanticPointerEvents = _nullishCoalesce(attrs.pointerEvents, () => ( "auto"));
      if (el.style.pointerEvents !== semanticPointerEvents) {
        el.style.pointerEvents = semanticPointerEvents;
      }
      const renderOrder = this.a11yRenderOrders.get(node.id);
      if (renderOrder !== void 0 && el.style.zIndex !== String(renderOrder)) {
        el.style.zIndex = String(renderOrder);
      }
      const implicitTabIndex = !isNativelyFocusable(el) && attrs.role && INTERACTIVE_A11Y_ROLES.has(attrs.role) ? 0 : null;
      const desiredTabIndex = _nullishCoalesce(attrs.tabIndex, () => ( implicitTabIndex));
      if (desiredTabIndex === null) {
        if (el.hasAttribute("tabindex")) el.removeAttribute("tabindex");
      } else if (el.getAttribute("tabindex") !== String(desiredTabIndex)) {
        el.setAttribute("tabindex", String(desiredTabIndex));
      }
      this.syncOptionalAttribute(el, "type", attrs.inputType);
      if (el instanceof HTMLInputElement || el instanceof HTMLTextAreaElement) {
        const placeholder = _nullishCoalesce(attrs.placeholder, () => ( ""));
        if (el.placeholder !== placeholder) el.placeholder = placeholder;
      }
      if (el instanceof HTMLAnchorElement) {
        this.syncOptionalAttribute(
          el,
          "href",
          attrs.href === void 0 ? void 0 : _chunkORCPCIJ7js.sanitizeUrl.call(void 0, attrs.href)
        );
        this.syncOptionalAttribute(el, "target", attrs.target);
      }
      if (el instanceof HTMLImageElement) {
        this.syncOptionalAttribute(el, "src", attrs.src);
        this.syncOptionalAttribute(el, "alt", attrs.alt);
      }
      if (el instanceof HTMLInputElement) {
        const checked = _nullishCoalesce(attrs.checked, () => ( false));
        if (el.checked !== checked) el.checked = checked;
      } else {
        this.syncOptionalAttribute(
          el,
          "aria-checked",
          attrs.checked === void 0 ? void 0 : String(attrs.checked)
        );
      }
      if ("disabled" in el) {
        const disabled = _nullishCoalesce(attrs.disabled, () => ( false));
        if (el.disabled !== disabled) el.disabled = disabled;
      } else {
        this.syncOptionalAttribute(
          el,
          "aria-disabled",
          attrs.disabled === void 0 ? void 0 : String(attrs.disabled)
        );
      }
      this.syncOptionalAttribute(
        el,
        "aria-expanded",
        attrs.expanded === void 0 ? void 0 : String(attrs.expanded)
      );
      this.syncOptionalAttribute(el, "aria-controls", attrs.controls);
      this.syncOptionalAttribute(el, "aria-haspopup", attrs.haspopup);
      this.syncOptionalAttribute(
        el,
        "aria-selected",
        attrs.selected === void 0 ? void 0 : String(attrs.selected)
      );
      this.syncOptionalAttribute(el, "aria-activedescendant", attrs.activedescendant);
      this.syncOptionalAttribute(el, "aria-valuemin", attrs.valuemin);
      this.syncOptionalAttribute(el, "aria-valuemax", attrs.valuemax);
      if (attrs.value !== void 0) {
        if (el instanceof HTMLInputElement || el instanceof HTMLTextAreaElement) {
          if (el.value !== attrs.value) {
            const userTyped = el._lastSyncedValue;
            if (attrs.value !== userTyped || document.activeElement !== el) {
              el.value = attrs.value;
              el._lastSyncedValue = attrs.value;
            }
          }
        } else {
          this.syncOptionalAttribute(el, "aria-valuenow", attrs.value);
        }
      } else if (!(el instanceof HTMLInputElement || el instanceof HTMLTextAreaElement)) {
        this.syncOptionalAttribute(el, "aria-valuenow", void 0);
      }
      if (attrs.textInputStyle !== void 0 && (el instanceof HTMLInputElement || el instanceof HTMLTextAreaElement)) {
        const textStyle = attrs.textInputStyle;
        if (el.style.font !== textStyle.font) el.style.font = textStyle.font;
        const lineHeight = `${textStyle.lineHeight}px`;
        if (el.style.lineHeight !== lineHeight) el.style.lineHeight = lineHeight;
        const padding = `${textStyle.padding}px`;
        if (el.style.padding !== padding) el.style.padding = padding;
        if (el.style.boxSizing !== "border-box") el.style.boxSizing = "border-box";
        if (el instanceof HTMLTextAreaElement) el.style.resize = "none";
      }
      if (node.a11yFullViewport) {
        el.style.left = "0px";
        el.style.top = "0px";
        el.style.width = `${this.width}px`;
        el.style.height = `${this.height}px`;
        el.style.transform = "";
      } else {
        const { a, b, c, d, e, f } = node.getWorldTransform();
        el.style.left = `${e + node.a11yOffsetX}px`;
        el.style.top = `${f + node.a11yOffsetY}px`;
        el.style.width = `${node.width}px`;
        el.style.height = `${node.height}px`;
        el.style.transform = `matrix(${a}, ${b}, ${c}, ${d}, 0, 0)`;
      }
    }
    this.syncContentProjection(node);
    for (const child of node.children) this.syncA11y(child);
    if (node === this.root) {
      for (const overlay of this.overlayRoot.children) this.syncA11y(overlay);
    }
  }
  /**
   * Mirror one entity's static text ({@link Entity.getContentProjection}) as a
   * transparent DOM node positioned over the drawn glyphs. Runs on the a11y
   * sync cadence; all writes are dirty-checked. Off-viewport projections are
   * hidden (`display: none`) so text-heavy scenes only materialize what is
   * visible to the browser's text machinery anyway.
   */
  syncContentProjection(node) {
    if (!this.contentProjectionEnabled || !this.a11yRoot) return;
    const projection = node.getContentProjection();
    let el = this.contentElements.get(node.id);
    if (!projection || !projection.text) {
      if (el) {
        this.clearContentGridState(node.id, el);
        el.remove();
        this.contentElements.delete(node.id);
        this.a11yNeedsReorder = true;
      }
      return;
    }
    if (!el) {
      el = document.createElement("div");
      el.setAttribute("data-vecto-content", node.id);
      const s = el.style;
      s.position = "absolute";
      s.transformOrigin = "0 0";
      s.margin = "0";
      s.padding = "0";
      s.color = "transparent";
      s.forcedColorAdjust = "none";
      s.setProperty("-webkit-text-fill-color", "transparent");
      s.whiteSpace = "pre-wrap";
      s.zIndex = "0";
      el.addEventListener(
        "wheel",
        (e2) => {
          node.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)("wheel", node, e2));
        },
        { passive: false }
      );
      this.a11yRoot.appendChild(el);
      this.contentElements.set(node.id, el);
      this.a11yNeedsReorder = true;
    }
    const lines = projection.lines;
    if (!projection.grid && el.dataset.vectoContentGrid !== void 0) {
      this.clearContentGridState(node.id, el);
    }
    if (projection.grid) {
      this.syncContentGridProjection(node, el, projection, projection.grid);
    } else if (lines && lines.length > 0) {
      const signature = JSON.stringify({
        lines,
        fallbackFont: _nullishCoalesce(projection.font, () => ( "")),
        fallbackLineHeight: _nullishCoalesce(projection.lineHeight, () => ( 16))
      });
      if (el.dataset.vectoProjectionLines !== signature) {
        this.releaseContentSelectionForRebuild(el);
        el.replaceChildren();
        for (let index = 0; index < lines.length; index++) {
          const line = lines[index];
          const lineElement = document.createElement("span");
          const lineFont = _nullishCoalesce(_nullishCoalesce(line.font, () => ( projection.font)), () => ( ""));
          const lineHeight2 = _nullishCoalesce(_nullishCoalesce(line.lineHeight, () => ( projection.lineHeight)), () => ( 16));
          lineElement.style.position = "absolute";
          lineElement.dir = "auto";
          lineElement.style.left = `${line.x}px`;
          lineElement.style.top = `${line.y + line.baseline - _chunk2Z23LTH3js.cssLineBoxBaseline.call(void 0, lineFont, lineHeight2)}px`;
          lineElement.style.whiteSpace = "pre";
          if (lineFont) lineElement.style.font = lineFont;
          lineElement.style.lineHeight = `${lineHeight2}px`;
          const separator = _nullishCoalesce(line.separatorAfter, () => ( (index < lines.length - 1 ? "\n" : "")));
          if (line.runs && line.runs.length > 0) {
            for (let runIndex = 0; runIndex < line.runs.length; runIndex++) {
              const run = line.runs[runIndex];
              const runElement = document.createElement("span");
              runElement.textContent = run.text + (runIndex === line.runs.length - 1 ? separator : "");
              if (run.font) runElement.style.font = run.font;
              runElement.style.lineHeight = `${lineHeight2}px`;
              lineElement.appendChild(runElement);
            }
          } else {
            lineElement.textContent = line.text + separator;
          }
          el.appendChild(lineElement);
        }
        el.dataset.vectoProjectionLines = signature;
      }
    } else {
      if (el.textContent !== projection.text) {
        this.releaseContentSelectionForRebuild(el);
        el.textContent = projection.text;
      }
      delete el.dataset.vectoProjectionLines;
    }
    const font = _nullishCoalesce(projection.font, () => ( ""));
    if (el.style.font !== font) el.style.font = font;
    const lineHeight = projection.lineHeight !== void 0 ? `${projection.lineHeight}px` : "";
    if (el.style.lineHeight !== lineHeight) el.style.lineHeight = lineHeight;
    const ligatures = projection.ligatures === "none" ? "none" : "";
    if (el.style.getPropertyValue("font-variant-ligatures") !== ligatures) {
      el.style.setProperty("font-variant-ligatures", ligatures);
      el.style.setProperty("font-kerning", ligatures ? "none" : "");
    }
    const hidden = node.interactive ? "true" : null;
    if (el.getAttribute("aria-hidden") !== hidden) {
      if (hidden) el.setAttribute("aria-hidden", hidden);
      else el.removeAttribute("aria-hidden");
    }
    const selectable = projection.selectable === true;
    const pointerEvents = selectable ? "auto" : "none";
    if (el.style.pointerEvents !== pointerEvents) {
      el.style.pointerEvents = pointerEvents;
      el.style.userSelect = selectable ? "text" : "none";
      el.style.cursor = selectable ? "text" : "";
    }
    const { a, b, c, d, e, f } = node.getWorldTransform();
    const contentX = _nullishCoalesce(projection.contentX, () => ( 0));
    const contentY = _nullishCoalesce(projection.contentY, () => ( 0));
    const baselineOffset = lines && lines.length > 0 ? 0 : projection.baseline === void 0 ? 0 : projection.baseline - _chunk2Z23LTH3js.cssLineBoxBaseline.call(void 0, font, _nullishCoalesce(projection.lineHeight, () => ( 16)));
    const localY = contentY + baselineOffset;
    el.style.left = `${e + a * contentX + c * localY}px`;
    el.style.top = `${f + b * contentX + d * localY}px`;
    if (node.width > 0) el.style.width = `${node.width}px`;
    if (node.height > 0) el.style.height = `${node.height}px`;
    el.style.transform = `matrix(${a}, ${b}, ${c}, ${d}, 0, 0)`;
    let visible = true;
    if (node.width > 0 && node.height > 0) {
      const worldCorners = [];
      let minX = Infinity;
      let minY = Infinity;
      let maxX = -Infinity;
      let maxY = -Infinity;
      for (let i = 0; i < 4; i++) {
        const lx = i & 1 ? node.width : 0;
        const ly = i & 2 ? node.height : 0;
        const wx = a * lx + c * ly + e;
        const wy = b * lx + d * ly + f;
        worldCorners.push({ x: wx, y: wy });
        if (wx < minX) minX = wx;
        if (wx > maxX) maxX = wx;
        if (wy < minY) minY = wy;
        if (wy > maxY) maxY = wy;
      }
      visible = maxX >= 0 && minX <= this.width && maxY >= 0 && minY <= this.height;
      for (let ancestor = node.parent; visible && ancestor; ancestor = ancestor.parent) {
        if (!ancestor.clipChildren || ancestor.width <= 0 || ancestor.height <= 0) continue;
        let localMinX = Infinity;
        let localMinY = Infinity;
        let localMaxX = -Infinity;
        let localMaxY = -Infinity;
        for (const corner of worldCorners) {
          const local = ancestor.worldToLocal(corner.x, corner.y);
          if (!local) continue;
          localMinX = Math.min(localMinX, local.x);
          localMinY = Math.min(localMinY, local.y);
          localMaxX = Math.max(localMaxX, local.x);
          localMaxY = Math.max(localMaxY, local.y);
        }
        visible = localMaxX >= 0 && localMinX <= ancestor.width && localMaxY >= 0 && localMinY <= ancestor.height;
      }
    }
    const display = visible ? "" : "none";
    if (el.style.display !== display) el.style.display = display;
  }
  /**
   * Materialize a prepared grid in logical source order while positioning each
   * carrier from the shared canvas geometry. Browser font measurement happens
   * later in one cold read/write batch, never inside projection synchronization.
   */
  syncContentGridProjection(node, el, projection, grid) {
    if (grid.source !== projection.text) {
      throw new Error("ContentProjection.grid.source must equal ContentProjection.text");
    }
    const signature = `${grid.revision}`;
    if (el.dataset.vectoContentGrid !== signature) {
      const materializeStart = typeof performance !== "undefined" ? performance.now() : 0;
      this.clearContentGridState(node.id, el);
      el.replaceChildren();
      const projectionLines = _nullishCoalesce(projection.lines, () => ( []));
      for (let lineIndex = 0; lineIndex < grid.lines.length; lineIndex++) {
        const gridLine = grid.lines[lineIndex];
        const projectedLine = projectionLines[lineIndex];
        const lineHeight = _nullishCoalesce(_optionalChain([projectedLine, 'optionalAccess', _75 => _75.lineHeight]), () => ( grid.lineHeight));
        const baseline = _nullishCoalesce(_optionalChain([projectedLine, 'optionalAccess', _76 => _76.baseline]), () => ( grid.baseline));
        const lineFont = _nullishCoalesce(_optionalChain([projectedLine, 'optionalAccess', _77 => _77.font]), () => ( grid.font));
        const lineElement = document.createElement("span");
        lineElement.dir = "ltr";
        lineElement.dataset.vectoGridLine = `${lineIndex}`;
        lineElement.style.position = "absolute";
        lineElement.style.left = `${_nullishCoalesce(_optionalChain([projectedLine, 'optionalAccess', _78 => _78.x]), () => ( 0))}px`;
        lineElement.style.top = `${(_nullishCoalesce(_optionalChain([projectedLine, 'optionalAccess', _79 => _79.y]), () => ( lineIndex * grid.lineHeight))) + baseline - _chunk2Z23LTH3js.cssLineBoxBaseline.call(void 0, lineFont, lineHeight)}px`;
        lineElement.style.width = `${gridLine.width}px`;
        lineElement.style.height = `${lineHeight}px`;
        lineElement.style.whiteSpace = "pre";
        lineElement.style.font = lineFont;
        lineElement.style.lineHeight = `${lineHeight}px`;
        if (gridLine.cells.length === 0) {
          lineElement.textContent = grid.source.slice(gridLine.sourceEnd, gridLine.nextSourceStart);
        } else {
          let logicalX = 0;
          for (let cellIndex = 0; cellIndex < gridLine.cells.length; cellIndex++) {
            const cell = gridLine.cells[cellIndex];
            const cellElement = document.createElement("span");
            cellElement.dir = "ltr";
            const separator = cellIndex === gridLine.cells.length - 1 ? grid.source.slice(gridLine.sourceEnd, gridLine.nextSourceStart) : "";
            const sourceText = grid.source.slice(cell.sourceStart, cell.sourceEnd);
            cellElement.textContent = sourceText + separator;
            cellElement.dataset.vectoGridCell = `${cellIndex}`;
            cellElement.dataset.vectoGridSourceLength = `${sourceText.length}`;
            cellElement.dataset.vectoGridSourceStart = `${cell.sourceStart}`;
            cellElement.dataset.vectoGridSourceEnd = `${cell.sourceEnd}`;
            cellElement.dataset.vectoGridCaretOffsets = cell.sourceCaretOffsets.join(",");
            cellElement.dataset.vectoGridLevel = `${cell.level}`;
            cellElement.dataset.vectoGridAdvance = `${cell.advance}`;
            cellElement.dataset.vectoGridX = `${cell.x}`;
            cellElement.style.position = "relative";
            cellElement.style.display = "inline-block";
            cellElement.style.left = `${cell.x - logicalX}px`;
            cellElement.style.top = "0";
            cellElement.style.width = `${cell.advance}px`;
            cellElement.style.height = `${lineHeight}px`;
            cellElement.style.boxSizing = "border-box";
            cellElement.style.verticalAlign = "top";
            cellElement.style.whiteSpace = "pre";
            cellElement.style.font = lineFont;
            cellElement.style.lineHeight = `${lineHeight}px`;
            cellElement.style.transformOrigin = "0 50%";
            lineElement.appendChild(cellElement);
            logicalX += cell.advance;
          }
        }
        if (lineIndex === 0) {
          for (const [basis, left, top] of [
            ["origin", 0, 0],
            ["x", 1, 0],
            ["y", 0, 1]
          ]) {
            const marker = document.createElement("span");
            marker.dataset.vectoGridBasis = basis;
            marker.setAttribute("aria-hidden", "true");
            marker.style.position = "absolute";
            marker.style.left = `${left}px`;
            marker.style.top = `${top}px`;
            marker.style.width = "0";
            marker.style.height = "0";
            marker.style.pointerEvents = "none";
            marker.style.userSelect = "none";
            lineElement.appendChild(marker);
          }
        }
        el.appendChild(lineElement);
      }
      el.dataset.vectoProjectionLines = signature;
      el.dataset.vectoContentGrid = signature;
      el.dataset.vectoGridCarriers = `${el.querySelectorAll("[data-vecto-grid-cell]").length}`;
      if (typeof performance !== "undefined") {
        el.dataset.vectoGridMaterializeMs = `${performance.now() - materializeStart}`;
      }
      delete el.dataset.vectoGridCalibration;
      delete el.dataset.vectoGridReady;
    }
    const pageScaleX = this.getContentMetricScaleX();
    const calibrationKey = `${signature}:${this.contentFontEpoch}:${pageScaleX.toFixed(4)}`;
    if (el.dataset.vectoGridCalibration !== calibrationKey) {
      this.scheduleContentGridCalibration(node.id, el, calibrationKey, pageScaleX);
    }
  }
  getContentMetricScaleX() {
    if (this.contentMetricScaleEpoch === this.contentFontEpoch) {
      return this.contentMetricScaleX;
    }
    const rect = this.canvas.getBoundingClientRect();
    const inlineWidth = parseInlinePx(this.canvas.style.width);
    const logicalWidth = _nullishCoalesce(inlineWidth, () => ( (this.canvas.clientWidth || this.width)));
    const scale = logicalWidth > 0 ? rect.width / logicalWidth : 1;
    this.contentMetricScaleX = Number.isFinite(scale) && scale > 0 ? scale : 1;
    this.contentMetricScaleEpoch = this.contentFontEpoch;
    return this.contentMetricScaleX;
  }
  scheduleContentGridCalibration(entityId, el, calibrationKey, pageScaleX) {
    if (typeof requestAnimationFrame !== "function") return;
    if (el.dataset.vectoGridCalibrationPending === calibrationKey) return;
    const previous = this.contentGridCalibrationFrames.get(entityId);
    if (previous !== void 0 && typeof cancelAnimationFrame === "function") {
      cancelAnimationFrame(previous);
    }
    _optionalChain([this, 'access', _80 => _80.contentGridCalibrationProbes, 'access', _81 => _81.get, 'call', _82 => _82(entityId), 'optionalAccess', _83 => _83.remove, 'call', _84 => _84()]);
    this.contentGridCalibrationProbes.delete(entityId);
    const calibrationStart = typeof performance !== "undefined" ? performance.now() : 0;
    const probe = document.createElement("div");
    probe.setAttribute("aria-hidden", "true");
    probe.dataset.vectoGridProbe = entityId;
    probe.style.position = "absolute";
    probe.style.left = "-100000px";
    probe.style.top = "0";
    probe.style.width = "100000px";
    probe.style.height = "1px";
    probe.style.visibility = "hidden";
    probe.style.pointerEvents = "none";
    probe.style.whiteSpace = "pre";
    probe.style.contain = "layout style paint";
    const probeOrigin = document.createElement("span");
    probeOrigin.style.position = "absolute";
    probeOrigin.style.left = "0";
    probeOrigin.style.top = "0";
    const probeX = document.createElement("span");
    probeX.style.position = "absolute";
    probeX.style.left = "1px";
    probeX.style.top = "0";
    probe.append(probeOrigin, probeX);
    const measurements = [];
    const measurementsByKey = /* @__PURE__ */ new Map();
    for (const target of el.querySelectorAll("[data-vecto-grid-cell]")) {
      const sourceLength = Number(_nullishCoalesce(target.dataset.vectoGridSourceLength, () => ( 0)));
      const targetWidth = Number(_nullishCoalesce(target.dataset.vectoGridAdvance, () => ( 0)));
      if (sourceLength <= 0 || targetWidth <= 0) continue;
      const sourceText = _nullishCoalesce(_optionalChain([target, 'access', _85 => _85.textContent, 'optionalAccess', _86 => _86.slice, 'call', _87 => _87(0, sourceLength)]), () => ( ""));
      if (!sourceText) continue;
      const measurementKey = JSON.stringify([
        target.style.font,
        target.style.lineHeight,
        targetWidth,
        sourceText
      ]);
      const shared = measurementsByKey.get(measurementKey);
      if (shared) {
        shared.targets.push(target);
        continue;
      }
      const carrier = document.createElement("span");
      carrier.dir = "ltr";
      carrier.style.position = "absolute";
      carrier.style.left = "0";
      carrier.style.top = "0";
      carrier.style.whiteSpace = "pre";
      carrier.style.font = target.style.font;
      carrier.style.lineHeight = target.style.lineHeight;
      carrier.style.fontVariantLigatures = "none";
      carrier.style.fontKerning = "none";
      const source = document.createTextNode(sourceText);
      carrier.appendChild(source);
      probe.appendChild(carrier);
      const measurement = { targets: [target], targetWidth, sourceLength, source };
      measurements.push(measurement);
      measurementsByKey.set(measurementKey, measurement);
    }
    (_nullishCoalesce(_nullishCoalesce(this.a11yRoot, () => ( document.body)), () => ( document.documentElement))).appendChild(probe);
    el.dataset.vectoGridCalibrationSamples = `${measurements.length}`;
    this.contentGridCalibrationProbes.set(entityId, probe);
    el.dataset.vectoGridCalibrationPending = calibrationKey;
    delete el.dataset.vectoGridReady;
    const readFrame = requestAnimationFrame(() => {
      if (!el.isConnected || el.dataset.vectoGridCalibrationPending !== calibrationKey) {
        probe.remove();
        this.contentGridCalibrationProbes.delete(entityId);
        this.contentGridCalibrationFrames.delete(entityId);
        return;
      }
      const updates = [];
      const probeOriginRect = probeOrigin.getBoundingClientRect();
      const probeXRect = probeX.getBoundingClientRect();
      const basisScale = Math.abs(probeXRect.left - probeOriginRect.left);
      const projectionPageScaleX = Number.isFinite(basisScale) && basisScale > 0 ? basisScale : pageScaleX;
      let valid = true;
      for (const measurement of measurements) {
        const range = document.createRange();
        range.setStart(measurement.source, 0);
        range.setEnd(measurement.source, measurement.sourceLength);
        const natural = range.getBoundingClientRect().width;
        if (!Number.isFinite(natural) || natural <= 0) {
          valid = false;
          break;
        }
        const scale = measurement.targetWidth * projectionPageScaleX / natural;
        for (const element of measurement.targets) updates.push({ element, scale });
      }
      probe.remove();
      this.contentGridCalibrationProbes.delete(entityId);
      if (!valid) {
        delete el.dataset.vectoGridCalibrationPending;
        this.contentGridCalibrationFrames.delete(entityId);
        return;
      }
      const writeFrame = requestAnimationFrame(() => {
        if (!el.isConnected || el.dataset.vectoGridCalibrationPending !== calibrationKey) {
          this.contentGridCalibrationFrames.delete(entityId);
          return;
        }
        for (const { element, scale } of updates) {
          element.style.transform = Math.abs(scale - 1) <= 1e-3 ? "" : `scaleX(${scale})`;
        }
        el.dataset.vectoGridCalibration = calibrationKey;
        el.dataset.vectoGridReady = "true";
        if (typeof performance !== "undefined") {
          el.dataset.vectoGridCalibrationMs = `${performance.now() - calibrationStart}`;
        }
        delete el.dataset.vectoGridCalibrationPending;
        this.contentGridCalibrationFrames.delete(entityId);
      });
      this.contentGridCalibrationFrames.set(entityId, writeFrame);
    });
    this.contentGridCalibrationFrames.set(entityId, readFrame);
  }
  enforceA11yDomOrder() {
    if (!this.a11yRoot) return;
    this.fullViewportElements.length = 0;
    this.normalElements.length = 0;
    this.activeIds.clear();
    const collect = (node) => {
      if (node.isDOMPortal) return;
      const contentEl = this.contentElements.get(node.id);
      if (contentEl) {
        if (node.a11yFullViewport) this.fullViewportElements.push(contentEl);
        else this.normalElements.push(contentEl);
      }
      if (node.interactive && (node.width > 0 || node.a11yFullViewport)) {
        const el = this.a11yElements.get(node.id);
        if (el) {
          this.activeIds.add(node.id);
          if (node.a11yFullViewport) this.fullViewportElements.push(el);
          else this.normalElements.push(el);
        }
      }
      for (const child of node.children) collect(child);
      if (node === this.root) {
        for (const overlay of this.overlayRoot.children) collect(overlay);
      }
    };
    collect(this.root);
    let elementsPruned = false;
    for (const [id, el] of this.a11yElements.entries()) {
      if (!this.activeIds.has(id)) {
        elementsPruned = true;
        if (el === this.focusedA11yElement) {
          this.focusedA11yElement = null;
          if (this.caretBlinkTimer) {
            clearInterval(this.caretBlinkTimer);
            this.caretBlinkTimer = null;
          }
        }
        if (el.parentNode === this.a11yRoot) {
          this.a11yRoot.removeChild(el);
        }
        this.a11yElements.delete(id);
      }
    }
    if (elementsPruned) {
      this.a11yNeedsReorder = true;
    }
    if (!this.a11yNeedsReorder) return;
    const fullLen = this.fullViewportElements.length;
    const normalLen = this.normalElements.length;
    const totalLen = fullLen + normalLen;
    for (let i = 0; i < totalLen; i++) {
      const expected = i < fullLen ? this.fullViewportElements[i] : this.normalElements[i - fullLen];
      const current = this.a11yRoot.childNodes[i];
      if (current !== expected) {
        this.a11yRoot.insertBefore(expected, current || null);
      }
    }
    this.a11yNeedsReorder = false;
  }
  /** Keep DOM/WebGL overlay layers aligned with the canvas's CSS box. */
  syncOverlayGeometry() {
    const parent = this.canvas.parentElement;
    if (!parent) return;
    const canvasRect = _optionalChain([this, 'access', _88 => _88.canvas, 'access', _89 => _89.getBoundingClientRect, 'optionalCall', _90 => _90()]);
    const parentRect = _optionalChain([parent, 'access', _91 => _91.getBoundingClientRect, 'optionalCall', _92 => _92()]);
    const cssWidth = _optionalChain([canvasRect, 'optionalAccess', _93 => _93.width]) || this.canvas.clientWidth || this.width;
    const cssHeight = _optionalChain([canvasRect, 'optionalAccess', _94 => _94.height]) || this.canvas.clientHeight || this.height;
    const left = (_nullishCoalesce(_optionalChain([canvasRect, 'optionalAccess', _95 => _95.left]), () => ( 0))) - (_nullishCoalesce(_optionalChain([parentRect, 'optionalAccess', _96 => _96.left]), () => ( 0))) - (parent.clientLeft || 0) + parent.scrollLeft;
    const top = (_nullishCoalesce(_optionalChain([canvasRect, 'optionalAccess', _97 => _97.top]), () => ( 0))) - (_nullishCoalesce(_optionalChain([parentRect, 'optionalAccess', _98 => _98.top]), () => ( 0))) - (parent.clientTop || 0) + parent.scrollTop;
    const scaleX = this.width > 0 ? cssWidth / this.width : 1;
    const scaleY = this.height > 0 ? cssHeight / this.height : 1;
    for (const root of [this.a11yRoot, this.portalRoot]) {
      if (!root) continue;
      root.style.left = `${left}px`;
      root.style.top = `${top}px`;
      root.style.width = `${this.width}px`;
      root.style.height = `${this.height}px`;
      root.style.transformOrigin = "0 0";
      root.style.transform = `scale(${scaleX}, ${scaleY})`;
    }
    for (const canvas of [this.glCanvas, this.gpuCanvas]) {
      if (!canvas) continue;
      canvas.style.left = `${left}px`;
      canvas.style.top = `${top}px`;
      canvas.style.width = `${cssWidth}px`;
      canvas.style.height = `${cssHeight}px`;
    }
  }
  getA11yTree() {
    const map = /* @__PURE__ */ new Map();
    const roots = [];
    const traverse = (node, parentNode) => {
      if (node.isDOMPortal) return;
      let currentA11yNode = null;
      if (node.interactive && (node.width > 0 || node.a11yFullViewport)) {
        const el = this.a11yElements.get(node.id);
        if (el) {
          const attrs = node.getA11yAttributes();
          currentA11yNode = {
            id: node.id,
            tag: el.tagName.toLowerCase(),
            role: el.getAttribute("role") || void 0,
            label: el.getAttribute("aria-label") || void 0,
            value: attrs.value,
            checked: attrs.checked,
            expanded: attrs.expanded,
            valuemin: attrs.valuemin,
            valuemax: attrs.valuemax,
            children: []
          };
          map.set(node.id, currentA11yNode);
          const parentA11y = parentNode ? map.get(parentNode.id) : null;
          if (parentA11y) {
            parentA11y.children.push(currentA11yNode);
          } else {
            roots.push(currentA11yNode);
          }
        }
      }
      for (const child of node.children) {
        traverse(child, currentA11yNode ? node : parentNode);
      }
      if (node === this.root) {
        for (const overlay of this.overlayRoot.children) {
          traverse(overlay, currentA11yNode ? node : parentNode);
        }
      }
    };
    traverse(this.root, null);
    return roots;
  }
  renderPortalDOM(portal, te, tf, a, b, c, d, opacity) {
    if (!this.portalRoot) return;
    this.activePortalsThisFrame.add(portal.id);
    this.portalEntities.set(portal.id, portal);
    if (portal.domElement.parentElement !== this.portalRoot) {
      this.portalRoot.appendChild(portal.domElement);
    }
    if (!portal.domElement.hasAttribute("data-vecto-id")) {
      portal.domElement.setAttribute("data-vecto-id", portal.id);
    }
    const transformStr = `matrix(${a}, ${b}, ${c}, ${d}, ${te}, ${tf})`;
    let widthStr = "";
    let heightStr = "";
    if (portal.width > 0) widthStr = `${portal.width}px`;
    if (portal.height > 0) heightStr = `${portal.height}px`;
    const zIndexStr = String(this.renderOrderCounter++);
    if (portal.lastWidth !== widthStr) {
      portal.domElement.style.width = widthStr;
      portal.lastWidth = widthStr;
    }
    if (portal.lastHeight !== heightStr) {
      portal.domElement.style.height = heightStr;
      portal.lastHeight = heightStr;
    }
    if (portal.lastTransform !== transformStr) {
      portal.domElement.style.left = "0px";
      portal.domElement.style.top = "0px";
      portal.domElement.style.transform = transformStr;
      portal.lastTransform = transformStr;
    }
    if (portal.lastZIndex !== zIndexStr) {
      portal.domElement.style.zIndex = zIndexStr;
      portal.lastZIndex = zIndexStr;
    }
    const opacityStr = String(opacity);
    if (portal.lastOpacity !== opacityStr) {
      portal.domElement.style.opacity = opacityStr;
      portal.lastOpacity = opacityStr;
    }
  }
  reconcilePortals() {
    if (!this.portalRoot) return;
    for (const oldId of this.activePortalsPrevFrame) {
      if (!this.activePortalsThisFrame.has(oldId)) {
        const portal = this.portalEntities.get(oldId);
        if (portal) {
          if (portal.domElement.parentElement === this.portalRoot && (!portal.scene || portal.scene === this)) {
            portal.domElement.remove();
          }
          this.portalEntities.delete(oldId);
        }
      }
    }
    this.activePortalsPrevFrame = new Set(this.activePortalsThisFrame);
    this.activePortalsThisFrame.clear();
  }
  /**
   * The frame-rate cap actually in effect: the explicit {@link maxFPS}, further
   * lowered to {@link REDUCED_MOTION_FPS} when the OS requests reduced motion
   * (and {@link respectReducedMotion} is on). `0` means uncapped.
   */
  effectiveMaxFPS() {
    const reduced = this.respectReducedMotion && !!_optionalChain([this, 'access', _99 => _99.reducedMotionQuery, 'optionalAccess', _100 => _100.matches]);
    if (reduced)
      return this.maxFPS > 0 ? Math.min(this.maxFPS, REDUCED_MOTION_FPS) : REDUCED_MOTION_FPS;
    return this.maxFPS;
  }
  loop(time) {
    if (!this.isRunning) return;
    let cap = this.effectiveMaxFPS();
    const isIdle = !this.dirty && !this.frameHadAnimation;
    if (isIdle && this.autoThrottle && this.renderMode === "always" && this.maxFPS > 0) {
      cap = Math.min(cap, 2);
    }
    if (cap > 0 && time - this.lastTime < 1e3 / cap - 1) {
      this.scheduleFrame();
      return;
    }
    let dt = time - this.lastTime;
    if (cap > 0) {
      const nominal = 1e3 / cap;
      if (Math.abs(dt - nominal) < nominal * 0.3) dt = nominal;
    }
    this.lastTime = time;
    if (this.renderMode === "onDemand" && isIdle) {
      this.scheduleFrame();
      return;
    }
    this.dirty = false;
    this.render(this.renderer, dt, time);
    const hasActiveAnimation = this.frameHadAnimation;
    const hasInteractive = this.frameHadInteractive;
    const wantsContentSync = this.contentProjectionEnabled;
    const shouldSyncInterval = this.a11ySyncInterval <= 0 || time - this.lastA11ySync >= this.a11ySyncInterval;
    if ((hasInteractive || this.a11yElements.size > 0 || wantsContentSync) && (shouldSyncInterval || this.a11yPendingSyncAfterAnimation)) {
      this.lastA11ySync = time;
      if (hasInteractive || wantsContentSync) {
        this.syncA11y(this.root);
      }
      this.enforceA11yDomOrder();
      this.a11yPendingSyncAfterAnimation = hasActiveAnimation;
    } else if (hasActiveAnimation) {
      this.a11yPendingSyncAfterAnimation = true;
    }
    this.scheduleFrame();
  }
  /**
   * Render the entire scene graph onto the specified renderer.
   *
   * @param renderer - The renderer instance to draw to.
   * @param dt - Delta time in milliseconds (default 0).
   * @param time - Current absolute time in milliseconds (default 0).
   */
  render(renderer, dt = 0, time = 0) {
    const isMainRenderer = renderer === this.renderer;
    if (isMainRenderer && this.a11yRoot && this.canvas.parentElement) {
      const parentStyle = this.canvas.parentElement.style;
      if (!parentStyle.position || parentStyle.position === "static") {
        parentStyle.position = "relative";
      }
      this.syncOverlayGeometry();
    }
    if (isMainRenderer) {
      this.renderOrderCounter = 0;
      this.a11yRenderOrders.clear();
      this.activePortalsThisFrame.clear();
    }
    const computeEntities = [];
    const collectComputeEntities = (node) => {
      if (node instanceof ComputeParticleEntity) {
        computeEntities.push(node);
      }
      for (const child of node.children) {
        collectComputeEntities(child);
      }
    };
    collectComputeEntities(this.root);
    for (const overlay of this.overlayRoot.children) {
      collectComputeEntities(overlay);
    }
    if (computeEntities.length > 0) {
      const isMainRenderPath = renderer === this.renderer;
      if (isMainRenderPath && !this.device && !this.webgpuDisabled && !this.initializingWebGPU && !this.deviceLost) {
        this.initializingWebGPU = true;
        this.initWebGPUContext(computeEntities).then((newDevice) => {
          this.device = newDevice;
          this.initializingWebGPU = false;
          const format = navigator.gpu ? navigator.gpu.getPreferredCanvasFormat() : "rgba8unorm";
          if (_Scene.webgpuManagerClass) {
            this.manager = new _Scene.webgpuManagerClass(newDevice);
          } else if (this.particleBackend === "webgpu") {
            throw new Error(
              "WebGPU particle manager is not registered. Please call Scene.registerWebGPUParticleSystemManager(WebGPUParticleSystemManager) first."
            );
          }
          if (this.manager) {
            this.manager.initPipelines(format);
            for (const entity of computeEntities) {
              this.manager.setupEntityResources(entity);
              if (entity.gpuStorageBuffer) {
                newDevice.queue.writeBuffer(entity.gpuStorageBuffer, 0, entity.particleData);
              }
            }
          }
        }).catch((err) => {
          if (this.particleBackend === "webgpu") {
            console.error("Failed to initialize WebGPU:", err);
          } else {
            console.warn("WebGPU unavailable; using CPU particle fallback.", err);
          }
          this.webgpuDisabled = true;
          this.initializingWebGPU = false;
        });
      }
      if (isMainRenderPath && this.device && this.manager && !this.deviceLost && !this.webgpuDisabled) {
        try {
          const commandEncoder = this.device.createCommandEncoder();
          const computePass = commandEncoder.beginComputePass();
          for (const entity of computeEntities) {
            if (!entity.gpuStorageBuffer || entity.needsInit) {
              if (!entity.gpuStorageBuffer) {
                this.manager.setupEntityResources(entity);
              }
              this.device.queue.writeBuffer(entity.gpuStorageBuffer, 0, entity.particleData);
              entity.needsInit = false;
            }
            this.manager.recordComputePass(
              computePass,
              entity,
              dt / 1e3,
              this.mouseX,
              this.mouseY,
              this.width,
              this.height
            );
          }
          computePass.end();
          if (this.gpuContext) {
            const view = this.gpuContext.getCurrentTexture().createView();
            const renderPassDescriptor = {
              colorAttachments: [
                {
                  view,
                  clearValue: { r: 0, g: 0, b: 0, a: 0 },
                  loadOp: "clear",
                  storeOp: "store"
                }
              ]
            };
            const renderPass = commandEncoder.beginRenderPass(renderPassDescriptor);
            for (const entity of computeEntities) {
              this.manager.recordRenderPass(renderPass, entity);
            }
            renderPass.end();
          }
          this.device.queue.submit([commandEncoder.finish()]);
          if (this.gpuContext) this.gpuHasContent = true;
        } catch (e) {
          console.error("WebGPU frame execution failed. Falling back.", e);
          this.deviceLost = true;
          this.device = null;
          this.recreateWebGPUDeviceWithRetry(computeEntities);
        }
      } else if (isMainRenderPath) {
        for (const entity of computeEntities) {
          let mx = this.mouseX;
          let my = this.mouseY;
          if (mx > -9e3 && my > -9e3) {
            const local = entity.worldToLocal(mx, my);
            if (local) {
              mx = local.x;
              my = local.y;
            } else {
              mx = -9999;
              my = -9999;
            }
          }
          entity.updateCPU(dt / 1e3, mx, my, this.width, this.height);
        }
      }
    } else if (isMainRenderer) {
      this.clearGPUCanvasIfStale();
    }
    renderer.clear();
    if (isMainRenderer) {
      _optionalChain([this, 'access', _101 => _101.pointRenderer, 'optionalAccess', _102 => _102.begin, 'call', _103 => _103()]);
    }
    const vw = this.width;
    const vh = this.height;
    let walkHadAnimation = false;
    let walkHadInteractive = false;
    const renderNode = (node, pa, pb, pc, pd, pe, pf, parentOpacity) => {
      if (isMainRenderer) {
        node.update(dt, time);
        if (!walkHadAnimation && node.hasPendingAnimations()) walkHadAnimation = true;
        if (!walkHadInteractive && node.interactive) walkHadInteractive = true;
        if (this._devActive && this._devFrameCount % 120 === 0) {
          if (node.update !== _chunk2Z23LTH3js.Entity.prototype.update && node.hasPendingAnimations === _chunk2Z23LTH3js.Entity.prototype.hasPendingAnimations) {
            this._devWarn(
              `Entity "${node.id}" overrides update() but not hasPendingAnimations(). Custom motion in update() without overriding hasPendingAnimations() causes the idle throttle to drop the animation to ~2fps. Override hasPendingAnimations() to return true while motion is in flight.`
            );
          }
        }
      }
      const cos = Math.cos(node.rotation);
      const sin = Math.sin(node.rotation);
      const te = pa * node.x + pc * node.y + pe;
      const tf = pb * node.x + pd * node.y + pf;
      const sxCos = node.scaleX * cos;
      const sxSin = node.scaleX * sin;
      const syCos = node.scaleY * cos;
      const sySin = node.scaleY * sin;
      const a = pa * sxCos + pc * sySin;
      const b = pb * sxCos + pd * sySin;
      const c = pa * -sxSin + pc * syCos;
      const d = pb * -sxSin + pd * syCos;
      const worldScaleX = Math.hypot(a, b);
      const worldScaleY = Math.hypot(c, d);
      const worldOpacity = parentOpacity * node.opacity;
      const scaleTolerance = Math.max(1, worldScaleX, worldScaleY) * 1e-6;
      const orthogonalTolerance = Math.max(1, worldScaleX * worldScaleY) * 1e-6;
      const isSimilarityTransform = Number.isFinite(worldScaleX) && Number.isFinite(worldScaleY) && Math.abs(worldScaleX - worldScaleY) <= scaleTolerance && Math.abs(a * c + b * d) <= orthogonalTolerance;
      const a11yEl = isMainRenderer ? this.a11yElements.get(node.id) : void 0;
      const willProjectA11y = isMainRenderer && node.interactive && (node.width > 0 || node.a11yFullViewport);
      if (a11yEl || willProjectA11y) {
        const renderOrder = this.renderOrderCounter++;
        if (willProjectA11y) this.a11yRenderOrders.set(node.id, renderOrder);
        if (a11yEl) a11yEl.style.zIndex = String(renderOrder);
      }
      if (node.isDOMPortal) {
        if (isMainRenderer) {
          this.renderPortalDOM(node, te, tf, a, b, c, d, worldOpacity);
        }
        return;
      }
      let visible = true;
      const bounds = node.getBounds();
      if (bounds) {
        let minX = Infinity;
        let minY = Infinity;
        let maxX = -Infinity;
        let maxY = -Infinity;
        for (let i = 0; i < 4; i++) {
          const lx = i & 1 ? bounds.x + bounds.width : bounds.x;
          const ly = i & 2 ? bounds.y + bounds.height : bounds.y;
          const wx = a * lx + c * ly + te;
          const wy = b * lx + d * ly + tf;
          if (wx < minX) minX = wx;
          if (wx > maxX) maxX = wx;
          if (wy < minY) minY = wy;
          if (wy > maxY) maxY = wy;
        }
        visible = maxX >= 0 && minX <= vw && maxY >= 0 && minY <= vh;
      }
      if (!visible && node.children.length === 0) return;
      if (node.children.length === 0 && node.scaleX === node.scaleY) {
        const bc = node.getBatchCircle();
        if (bc) {
          if (!visible) return;
          if (isMainRenderer && this.pointRenderer) {
            if (isSimilarityTransform) {
              this.pointRenderer.addCircle(te, tf, bc.radius * worldScaleX, bc.color, worldOpacity);
              return;
            }
          } else {
            renderer.fillCircle(node.x, node.y, bc.radius * node.scaleX, bc.color, worldOpacity);
            return;
          }
        } else if (isMainRenderer && this.pointRenderer) {
          const br = node.getBatchRect();
          if (br && isSimilarityTransform && a * d - b * c >= 0) {
            if (visible) {
              this.pointRenderer.addRect(
                te,
                tf,
                br.width * worldScaleX,
                br.height * worldScaleX,
                br.color,
                worldOpacity,
                Math.atan2(b, a)
              );
            }
            return;
          }
        }
      }
      renderer.flush();
      renderer.save();
      renderer.translate(node.x, node.y);
      renderer.scale(node.scaleX, node.scaleY);
      renderer.rotate(node.rotation);
      renderer.setGlobalAlpha(worldOpacity);
      if (visible) {
        if (node instanceof ComputeParticleEntity) {
          if (this.deviceLost || this.webgpuDisabled || !this.device || !this.manager) {
            this.renderCPUParticles(
              renderer,
              node,
              worldOpacity,
              a,
              b,
              c,
              d,
              te,
              tf,
              worldScaleX,
              isSimilarityTransform
            );
          }
        } else {
          node.render(renderer);
        }
      }
      if (node.clipChildren) {
        renderer.clip(0, 0, node.width, node.height);
      }
      for (const child of node.children) {
        renderNode(child, a, b, c, d, te, tf, worldOpacity);
      }
      renderer.flush();
      renderer.restore();
    };
    renderNode(this.root, 1, 0, 0, 1, 0, 0, 1);
    for (const overlay of this.overlayRoot.children) {
      renderNode(overlay, 1, 0, 0, 1, 0, 0, 1);
    }
    if (isMainRenderer) {
      this.frameHadAnimation = walkHadAnimation;
      this.frameHadInteractive = walkHadInteractive;
      this.reconcilePortals();
    }
    renderer.flush();
    if (isMainRenderer) {
      _optionalChain([this, 'access', _104 => _104.pointRenderer, 'optionalAccess', _105 => _105.flush, 'call', _106 => _106()]);
    }
    _optionalChain([renderer, 'access', _107 => _107.present, 'optionalCall', _108 => _108()]);
    if (this._devActive) {
      this._devFrameCount++;
      this._devRunChecks();
    }
  }
  /**
   * Export the current scene state to a lightweight, flat SVG XML string.
   */
  toSVG() {
    const renderer = new (0, _chunkORCPCIJ7js.SVGRenderer)(this.width, this.height);
    this.render(renderer, 0, 0);
    return renderer.toXMLString();
  }
  /**
   * Manually resize the Scene's viewport.
   */
  resize(width, height) {
    this.width = width;
    this.height = height;
    this.contentFontEpoch++;
    if (typeof this.renderer.resize === "function") {
      if ("maxDPR" in this.renderer) this.renderer.maxDPR = this.maxDPR;
      this.renderer.resize(width, height);
    }
    if (this.pointRenderer) {
      this.pointRenderer.maxDPR = this.maxDPR;
      this.pointRenderer.resize(width, height);
    }
    if (this.gpuCanvas) {
      this.gpuCanvas.width = width;
      this.gpuCanvas.height = height;
    }
    this.markDirty();
  }
  /**
   * Gets the accessibility DOM element projected for the given entity ID.
   */
  getA11yElement(entityId) {
    return this.a11yElements.get(entityId);
  }
  /** Gets the static-content DOM projection for an entity ID, when materialized. */
  getContentElement(entityId) {
    return this.contentElements.get(entityId);
  }
  /**
   * Gets the root entity of the scene.
   */
  getRoot() {
    return this.root;
  }
  /**
   * Finds the topmost interactive entity at the given coordinates.
   */
  findEntityAt(x, y) {
    const overlayHit = this.findHitRecursively(this.overlayRoot, x, y);
    if (overlayHit) return overlayHit;
    return this.findHitRecursively(this.root, x, y);
  }
  /** Submit one transparent clear pass when particle content lingers on the GPU canvas. */
  clearGPUCanvasIfStale() {
    if (!this.gpuHasContent || !this.device || !this.gpuContext || this.deviceLost) return;
    try {
      const encoder = this.device.createCommandEncoder();
      const pass = encoder.beginRenderPass({
        colorAttachments: [
          {
            view: this.gpuContext.getCurrentTexture().createView(),
            clearValue: { r: 0, g: 0, b: 0, a: 0 },
            loadOp: "clear",
            storeOp: "store"
          }
        ]
      });
      pass.end();
      this.device.queue.submit([encoder.finish()]);
    } catch (e5) {
    }
    this.gpuHasContent = false;
  }
  async initWebGPUContext(entities) {
    if (!navigator.gpu) {
      throw new Error("WebGPU not supported on this platform.");
    }
    const adapter = await navigator.gpu.requestAdapter();
    if (!adapter) {
      throw new Error("No GPUAdapter found.");
    }
    const device = await adapter.requestDevice();
    if (typeof document !== "undefined" && !this.gpuCanvas) {
      const gpuCanvas = document.createElement("canvas");
      gpuCanvas.width = this.width;
      gpuCanvas.height = this.height;
      gpuCanvas.style.position = "absolute";
      gpuCanvas.style.top = "0";
      gpuCanvas.style.left = "0";
      gpuCanvas.style.pointerEvents = "none";
      gpuCanvas.style.zIndex = "6";
      if (this.canvas.parentElement) {
        this.canvas.parentElement.appendChild(gpuCanvas);
      }
      this.gpuCanvas = gpuCanvas;
      this.gpuContext = gpuCanvas.getContext("webgpu");
    }
    if (this.gpuContext) {
      this.gpuContext.configure({
        device,
        format: navigator.gpu.getPreferredCanvasFormat(),
        alphaMode: "premultiplied"
      });
    }
    this.setupDeviceLostHandler(device, entities);
    return device;
  }
  setupDeviceLostHandler(device, entities) {
    device.lost.then((info) => {
      if (info.reason === "destroyed") return;
      console.warn(`WebGPU device lost: ${info.message}`);
      this.deviceLost = true;
      this.device = null;
      this.recreateWebGPUDeviceWithRetry(entities);
    });
  }
  recreateWebGPUDeviceWithRetry(entities, attempt = 0) {
    if (this.destroyed) return;
    if (attempt >= 3) {
      console.error(
        "Failed to recover WebGPU device after 3 retries. Remaining on fallback renderer."
      );
      this.webgpuDisabled = true;
      this.deviceLost = true;
      return;
    }
    for (const entity of entities) {
      entity.destroyGPUResources();
    }
    if (this.manager) {
      this.manager.destroy();
      this.manager = null;
    }
    const backoff = Math.pow(2, attempt) * 1e3;
    if (this.recoveryTimerId) clearTimeout(this.recoveryTimerId);
    this.recoveryTimerId = setTimeout(() => {
      if (this.destroyed) return;
      this.initWebGPUContext(entities).then((newDevice) => {
        if (this.destroyed) {
          newDevice.destroy();
          return;
        }
        console.log("Successfully recovered WebGPU device.");
        this.device = newDevice;
        this.deviceLost = false;
        const format = navigator.gpu.getPreferredCanvasFormat();
        if (_Scene.webgpuManagerClass) {
          this.manager = new _Scene.webgpuManagerClass(newDevice);
        } else if (this.particleBackend === "webgpu") {
          throw new Error(
            "WebGPU particle manager is not registered. Please call Scene.registerWebGPUParticleSystemManager(WebGPUParticleSystemManager) first."
          );
        }
        if (this.manager) {
          this.manager.initPipelines(format);
          for (const entity of entities) {
            this.manager.setupEntityResources(entity);
            if (entity.gpuStorageBuffer) {
              newDevice.queue.writeBuffer(entity.gpuStorageBuffer, 0, entity.particleData);
            }
          }
        }
      }).catch(() => this.recreateWebGPUDeviceWithRetry(entities, attempt + 1));
    }, backoff);
  }
  renderCPUParticles(renderer, entity, worldOpacity, a, b, c, d, e, f, worldScale, isSimilarityTransform) {
    const data = entity.particleData;
    const size = entity.maxParticles;
    const useGL = renderer === this.renderer && !!this.pointRenderer && isSimilarityTransform;
    for (let i = 0; i < size; i++) {
      const idx = i * 8;
      const x = data[idx];
      const y = data[idx + 1];
      const pSize = data[idx + 6];
      const life = data[idx + 7];
      if (life === 0) continue;
      const opacity = life < 0 ? worldOpacity : worldOpacity * Math.min(1, life);
      const scale = life >= 0 ? Math.min(1, life) : 1;
      if (useGL) {
        this.pointRenderer.addCircle(
          a * x + c * y + e,
          b * x + d * y + f,
          pSize * scale * worldScale,
          entity.baseColor,
          opacity
        );
      } else {
        renderer.fillCircle(x, y, pSize * scale, entity.baseColor, opacity);
      }
    }
  }
  findHitRecursively(node, x, y) {
    for (let i = node.children.length - 1; i >= 0; i--) {
      const hit = this.findHitRecursively(node.children[i], x, y);
      if (hit) return hit;
    }
    if (node.isPointInside && node.isPointInside(x, y)) {
      return node;
    }
    return null;
  }
}, _class2.__initStatic(), _class2.__initStatic2(), _class2.__initStatic3(), _class2);

// src/components/TextEntity.ts
var sharedMeasurer;
function defaultMeasurer() {
  if (sharedMeasurer === void 0) sharedMeasurer = _chunkBA5HUUDFjs.createCanvasMeasurer.call(void 0, "sans-serif");
  return sharedMeasurer;
}
var TextEntity = (_class3 = class extends _chunk2Z23LTH3js.Entity {
  
  
  
  
  __init64() {this.nodes = []}
  
  __init65() {this.fillStyle = "#94a3b8"}
  __init66() {this.strokeStyle = null}
  __init67() {this.hoveredFillStyle = "#ffffff"}
  __init68() {this.lineWidth = 1}
  __init69() {this.isHovered = false}
  constructor(text, atlas, maxWidth, fontSize = 24) {
    super();_class3.prototype.__init64.call(this);_class3.prototype.__init65.call(this);_class3.prototype.__init66.call(this);_class3.prototype.__init67.call(this);_class3.prototype.__init68.call(this);_class3.prototype.__init69.call(this);;
    this.text = text;
    this.atlas = atlas;
    this.fontSize = fontSize;
    this.layout = new (0, _chunkBA5HUUDFjs.LayoutEngine)(maxWidth, 1e4, defaultMeasurer());
    this.prepared = this.layout.prepare(this.text, this.atlas, this.fontSize);
    this.applyLayout();
    this.interactive = true;
    this.on("hover", () => this.isHovered = true);
    this.on("pointerleave", () => this.isHovered = false);
  }
  /**
   * Mirror the rendered text into the DOM content layer: find-in-page, screen
   * readers, crawlers, and translation see the same string the canvas draws.
   */
  getContentProjection() {
    if (!this.text) return null;
    return { text: this.text, font: `${this.fontSize}px sans-serif` };
  }
  /**
   * Replace the text content. Runs the **cold** measurement pass (re-segment +
   * re-measure) since the glyphs changed, then re-lays out.
   *
   * @returns `this` for chaining.
   */
  setText(text) {
    this.text = text;
    this.prepared = this.layout.prepare(this.text, this.atlas, this.fontSize);
    this.applyLayout();
    return this;
  }
  /**
   * Change the wrap width and reflow. Cheap **hot** path only — reuses the
   * cached {@link PreparedText}, doing no re-segmentation or re-measurement.
   * Ideal for responsive resize.
   *
   * @returns `this` for chaining.
   */
  setMaxWidth(maxWidth) {
    this.layout.maxWidth = maxWidth;
    this.applyLayout();
    return this;
  }
  /**
   * Set horizontal alignment (`'justify'` stretches wrapped lines flush to
   * the wrap width; the last line stays ragged) and reflow.
   */
  setTextAlign(align) {
    this.layout.textAlign = align;
    this.applyLayout();
    return this;
  }
  /**
   * Plug a hyphenator (word → parts). Break opportunities are baked in during
   * the cold pass, so this re-prepares the current text. Soft hyphens
   * (U+00AD) in the text work without one.
   */
  setHyphenator(fn) {
    this.layout.hyphenate = fn;
    this.prepared = this.layout.prepare(this.text, this.atlas, this.fontSize);
    this.applyLayout();
    return this;
  }
  /** Hot pass: place the cached {@link PreparedText} and refresh the a11y box. */
  applyLayout() {
    const result = this.layout.layoutPrepared(this.prepared);
    this.nodes = result.nodes;
    this.width = result.totalWidth;
    this.height = result.totalHeight;
    this.a11yOffsetY = 0;
  }
  isPointInside(globalX, globalY) {
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x >= 0 && local.x <= this.width && local.y >= 0 && local.y <= this.height;
  }
  render(renderer) {
    const currentFill = this.isHovered ? this.hoveredFillStyle : this.fillStyle;
    for (const node of this.nodes) {
      const glyph = this.atlas[node.char];
      if (!glyph) {
        renderer.save();
        renderer.translate(node.x, node.y + this.fontSize * 0.8);
        renderer.fillText(node.char, 0, 0, `${this.fontSize}px sans-serif`, currentFill);
        renderer.restore();
        continue;
      }
      renderer.save();
      renderer.translate(node.x, node.y);
      const scale = this.fontSize / glyph.baseSize;
      renderer.scale(scale, scale);
      for (const path of glyph.ast.paths) {
        renderer.beginPath();
        for (const cmd of path.commands) {
          if (cmd.type === "M") renderer.moveTo(cmd.x, cmd.y);
          else if (cmd.type === "L") renderer.lineTo(cmd.x, cmd.y);
          else if (cmd.type === "C")
            renderer.bezierCurveTo(cmd.x1, cmd.y1, cmd.x2, cmd.y2, cmd.x, cmd.y);
          else if (cmd.type === "Z") renderer.closePath();
        }
        if (currentFill) {
          renderer.fill(currentFill);
        }
        if (this.strokeStyle) {
          renderer.stroke(this.strokeStyle, this.lineWidth / scale);
        }
      }
      renderer.restore();
    }
  }
}, _class3);

// src/components/GridTextEntity.ts
var GridTextEntity = (_class4 = class extends _chunk2Z23LTH3js.Entity {
  
  __init70() {this.fillStyle = "#ffffff"}
  __init71() {this.grid = []}
  // Array of rows
  __init72() {this.cols = 0}
  __init73() {this.rows = 0}
  
  
  constructor(_atlas, fontSize = 10) {
    super();_class4.prototype.__init70.call(this);_class4.prototype.__init71.call(this);_class4.prototype.__init72.call(this);_class4.prototype.__init73.call(this);;
    this.fontSize = fontSize;
    this.charWidth = fontSize * 1;
    this.charHeight = fontSize * 1.1;
    this.interactive = false;
  }
  updateGrid(ascii) {
    this.grid = ascii;
    this.rows = ascii.length;
    this.cols = _optionalChain([ascii, 'access', _109 => _109[0], 'optionalAccess', _110 => _110.length]) || 0;
  }
  isPointInside(_globalX, _globalY) {
    return false;
  }
  render(renderer) {
    if (this.rows === 0) return;
    for (let r = 0; r < this.rows; r++) {
      const row = this.grid[r];
      if (!row) continue;
      for (let c = 0; c < this.cols; c++) {
        const char = row[c];
        if (char === " ") continue;
        const x = c * this.charWidth;
        const y = r * this.charHeight;
        renderer.save();
        renderer.translate(x, y + this.fontSize * 0.8);
        renderer.fillText(char, 0, 0, `bold ${this.fontSize}px monospace`, this.fillStyle);
        renderer.restore();
      }
    }
  }
}, _class4);

// src/components/SplineEntity.ts
function polySegmentToBezier(seg) {
  const ax = seg.x_poly[0] || 0;
  const bx = seg.x_poly[1] || 0;
  const cx = seg.x_poly[2] || 0;
  const dx = seg.x_poly[3] || 0;
  const ay = seg.y_poly[0] || 0;
  const by = seg.y_poly[1] || 0;
  const cy = seg.y_poly[2] || 0;
  const dy = seg.y_poly[3] || 0;
  return {
    x0: ax,
    y0: ay,
    cp1x: ax + bx / 3,
    cp1y: ay + by / 3,
    cp2x: ax + 2 * bx / 3 + cx / 3,
    cp2y: ay + 2 * by / 3 + cy / 3,
    x3: ax + bx + cx + dx,
    y3: ay + by + cy + dy
  };
}
function rgbToCss(rgb) {
  return `rgb(${Math.round(rgb[0] * 255)}, ${Math.round(rgb[1] * 255)}, ${Math.round(rgb[2] * 255)})`;
}
var HIT_SAMPLES = 16;
function flattenBezier(b, samples) {
  const pts = new Float32Array((samples + 1) * 2);
  for (let i = 0; i <= samples; i++) {
    const t = i / samples;
    const mt = 1 - t;
    const a = mt * mt * mt;
    const c1 = 3 * mt * mt * t;
    const c2 = 3 * mt * t * t;
    const d = t * t * t;
    pts[i * 2] = a * b.x0 + c1 * b.cp1x + c2 * b.cp2x + d * b.x3;
    pts[i * 2 + 1] = a * b.y0 + c1 * b.cp1y + c2 * b.cp2y + d * b.y3;
  }
  return pts;
}
function distSqToSegment(px, py, x1, y1, x2, y2) {
  const dx = x2 - x1;
  const dy = y2 - y1;
  const lenSq = dx * dx + dy * dy;
  let t = lenSq > 0 ? ((px - x1) * dx + (py - y1) * dy) / lenSq : 0;
  t = t < 0 ? 0 : t > 1 ? 1 : t;
  const cx = x1 + t * dx;
  const cy = y1 + t * dy;
  const ex = px - cx;
  const ey = py - cy;
  return ex * ex + ey * ey;
}
var SplineEntity = (_class5 = class extends _chunk2Z23LTH3js.Entity {
  
  
  
  
  
  
  
  __init74() {this.offscreen = null}
  __init75() {this.baked = false}
  /** Logical (CSS-pixel) size of the baked bitmap — the blit destination size. */
  __init76() {this.bakedWidth = 0}
  __init77() {this.bakedHeight = 0}
  /** Gradient strokes can't be baked to a solid-color bitmap; they render per-frame. */
  
  /** Lazily-flattened polylines (one Float32Array of [x,y,...] per segment) for hit-testing. */
  __init78() {this.polylines = null}
  /**
   * When `true`, the renderer draws a rounded-rect outline of the entity's
   * local bounds after painting the curves. Useful for drag feedback and
   * debugging hit areas. Defaults to `false`.
   */
  __init79() {this.showBounds = false}
  constructor(doc, opts = {}) {
    super();_class5.prototype.__init74.call(this);_class5.prototype.__init75.call(this);_class5.prototype.__init76.call(this);_class5.prototype.__init77.call(this);_class5.prototype.__init78.call(this);_class5.prototype.__init79.call(this);;
    this.doc = doc;
    this.lineWidth = _nullishCoalesce(opts.lineWidth, () => ( 2));
    this.cache = _nullishCoalesce(opts.cache, () => ( true));
    this.defaultColor = _nullishCoalesce(opts.defaultColor, () => ( "#e2e8f0"));
    this.hitMode = _nullishCoalesce(opts.hitTest, () => ( "curve"));
    this.hitTolerance = _nullishCoalesce(opts.hitTolerance, () => ( 0));
    this.bounds = this.computeBounds();
    this.width = this.bounds.width;
    this.height = this.bounds.height;
    const isGradient = (c) => c !== null && !Array.isArray(c);
    this.containsGradient = (_nullishCoalesce(_optionalChain([this, 'access', _111 => _111.doc, 'access', _112 => _112.equations, 'optionalAccess', _113 => _113.some, 'call', _114 => _114((eq) => isGradient(eq.color_rgb))]), () => ( false))) || (_nullishCoalesce(_optionalChain([this, 'access', _115 => _115.doc, 'access', _116 => _116.paths, 'optionalAccess', _117 => _117.some, 'call', _118 => _118((p) => isGradient(p.color_rgb))]), () => ( false)));
    this.interactive = true;
  }
  computeBounds() {
    if (this.doc.bounding_box) {
      const [minX2, minY2, maxX2, maxY2] = this.doc.bounding_box;
      return { x: minX2, y: minY2, width: maxX2 - minX2, height: maxY2 - minY2 };
    }
    let minX = Infinity;
    let minY = Infinity;
    let maxX = -Infinity;
    let maxY = -Infinity;
    if (this.doc.equations) {
      for (const eq of this.doc.equations) {
        for (const seg of eq.data) {
          const b = polySegmentToBezier(seg);
          for (const [x, y] of [
            [b.x0, b.y0],
            [b.cp1x, b.cp1y],
            [b.cp2x, b.cp2y],
            [b.x3, b.y3]
          ]) {
            if (x < minX) minX = x;
            if (x > maxX) maxX = x;
            if (y < minY) minY = y;
            if (y > maxY) maxY = y;
          }
        }
      }
    }
    if (this.doc.paths) {
      for (const path of this.doc.paths) {
        for (const pt of path.data) {
          if (pt.x < minX) minX = pt.x;
          if (pt.x > maxX) maxX = pt.x;
          if (pt.y < minY) minY = pt.y;
          if (pt.y > maxY) maxY = pt.y;
        }
      }
    }
    if (minX === Infinity) return { x: 0, y: 0, width: 0, height: 0 };
    return { x: minX, y: minY, width: maxX - minX, height: maxY - minY };
  }
  /** @inheritdoc */
  getBounds() {
    return this.bounds;
  }
  /**
   * AABB hit-test against the document bounds in world space.
   *
   * Curve-accurate hit-testing can be layered on later via {@link hitTestCurve};
   * this method already calls it as a refinement when it is overridden.
   */
  isPointInside(globalX, globalY) {
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    const { x: lx, y: ly } = local;
    const inAabb = lx >= this.bounds.x && lx <= this.bounds.x + this.bounds.width && ly >= this.bounds.y && ly <= this.bounds.y + this.bounds.height;
    if (!inAabb) return false;
    const refined = this.hitTestCurve(lx, ly);
    return refined === null ? true : refined;
  }
  /**
   * Curve-accurate refinement of {@link isPointInside}: hit only when the local
   * point lies within `lineWidth/2 + hitTolerance` of an actual curve.
   *
   * Returns `null` in `hitTest: 'aabb'` mode (keep the bounding-box result).
   * Curves are flattened to polylines once and cached. Override for custom logic.
   *
   * @param localX - X in the entity's local space.
   * @param localY - Y in the entity's local space.
   * @returns `true`/`false`, or `null` to keep the AABB result.
   */
  hitTestCurve(localX, localY) {
    if (this.hitMode === "aabb") return null;
    const tol = this.lineWidth / 2 + this.hitTolerance;
    const tol2 = tol * tol;
    const polylines = this.getPolylines();
    for (const pts of polylines) {
      for (let i = 0; i + 3 < pts.length; i += 2) {
        if (distSqToSegment(localX, localY, pts[i], pts[i + 1], pts[i + 2], pts[i + 3]) <= tol2) {
          return true;
        }
      }
    }
    return false;
  }
  /** Flatten every Bézier segment into a sampled polyline once, then cache. */
  getPolylines() {
    if (this.polylines) return this.polylines;
    const out = [];
    if (this.doc.equations) {
      for (const eq of this.doc.equations) {
        for (const seg of eq.data) {
          out.push(flattenBezier(polySegmentToBezier(seg), HIT_SAMPLES));
        }
      }
    }
    if (this.doc.paths) {
      for (const path of this.doc.paths) {
        const pts = new Float32Array(path.data.length * 2);
        for (let i = 0; i < path.data.length; i++) {
          pts[i * 2] = path.data[i].x;
          pts[i * 2 + 1] = path.data[i].y;
        }
        out.push(pts);
      }
    }
    this.polylines = out;
    return out;
  }
  resolveColor(color, r) {
    if (color === null) return this.defaultColor;
    if (Array.isArray(color)) return rgbToCss(color);
    const w = this.bounds.width || 1;
    const h = this.bounds.height || 1;
    return r.createLinearGradient(
      color.start_pos[0] * w + this.bounds.x,
      color.start_pos[1] * h + this.bounds.y,
      color.end_pos[0] * w + this.bounds.x,
      color.end_pos[1] * h + this.bounds.y,
      color.stops.map(([stop, rgb]) => ({
        stop: Math.max(0, Math.min(1, stop)),
        color: rgbToCss(rgb)
      }))
    );
  }
  strokeEquations(r) {
    if (this.doc.equations) {
      for (const eq of this.doc.equations) {
        const stroke = this.resolveColor(eq.color_rgb, r);
        r.beginPath();
        for (const seg of eq.data) {
          const b = polySegmentToBezier(seg);
          r.moveTo(b.x0, b.y0);
          r.bezierCurveTo(b.cp1x, b.cp1y, b.cp2x, b.cp2y, b.x3, b.y3);
        }
        r.stroke(stroke, this.lineWidth);
      }
    }
    if (this.doc.paths) {
      for (const path of this.doc.paths) {
        const stroke = this.resolveColor(path.color_rgb, r);
        r.beginPath();
        if (path.data.length > 0) {
          r.moveTo(path.data[0].x, path.data[0].y);
          for (let i = 1; i < path.data.length; i++) {
            r.lineTo(path.data[i].x, path.data[i].y);
          }
        }
        r.stroke(stroke, this.lineWidth);
      }
    }
  }
  /** Bake all equations into an OffscreenCanvas once (when available). */
  bake() {
    this.baked = true;
    const pad = this.lineWidth + 2;
    const w = Math.max(1, Math.ceil(this.bounds.width) + pad * 2);
    const h = Math.max(1, Math.ceil(this.bounds.height) + pad * 2);
    this.bakedWidth = w;
    this.bakedHeight = h;
    const dpr = typeof window !== "undefined" && typeof window.devicePixelRatio === "number" ? window.devicePixelRatio || 1 : 1;
    let canvas;
    if (typeof OffscreenCanvas !== "undefined") {
      canvas = new OffscreenCanvas(w * dpr, h * dpr);
    } else if (typeof document !== "undefined") {
      canvas = document.createElement("canvas");
      canvas.width = w * dpr;
      canvas.height = h * dpr;
    } else {
      return;
    }
    const ctx = canvas.getContext("2d");
    if (!ctx) return;
    ctx.scale(dpr, dpr);
    ctx.translate(pad - this.bounds.x, pad - this.bounds.y);
    ctx.lineWidth = this.lineWidth;
    ctx.lineCap = "round";
    ctx.lineJoin = "round";
    if (this.doc.equations) {
      for (const eq of this.doc.equations) {
        ctx.strokeStyle = eq.color_rgb === null ? this.defaultColor : Array.isArray(eq.color_rgb) ? rgbToCss(eq.color_rgb) : this.defaultColor;
        ctx.beginPath();
        for (const seg of eq.data) {
          const b = polySegmentToBezier(seg);
          ctx.moveTo(b.x0, b.y0);
          ctx.bezierCurveTo(b.cp1x, b.cp1y, b.cp2x, b.cp2y, b.x3, b.y3);
        }
        ctx.stroke();
      }
    }
    if (this.doc.paths) {
      for (const path of this.doc.paths) {
        ctx.strokeStyle = path.color_rgb === null ? this.defaultColor : Array.isArray(path.color_rgb) ? rgbToCss(path.color_rgb) : this.defaultColor;
        ctx.beginPath();
        if (path.data.length > 0) {
          ctx.moveTo(path.data[0].x, path.data[0].y);
          for (let i = 1; i < path.data.length; i++) {
            ctx.lineTo(path.data[i].x, path.data[i].y);
          }
        }
        ctx.stroke();
      }
    }
    this.offscreen = canvas;
  }
  render(r) {
    let rendered = false;
    if (this.cache && !this.containsGradient) {
      if (!this.baked) this.bake();
      if (this.offscreen) {
        const pad = this.lineWidth + 2;
        r.drawImage(
          this.offscreen,
          this.bounds.x - pad,
          this.bounds.y - pad,
          // Logical size, not the (DPR-scaled) bitmap size.
          this.bakedWidth,
          this.bakedHeight
        );
        rendered = true;
      }
    }
    if (!rendered) {
      this.strokeEquations(r);
    }
    if (this.showBounds) {
      r.beginPath();
      r.roundRect(this.bounds.x, this.bounds.y, this.bounds.width, this.bounds.height, 4);
      r.stroke("rgba(0, 150, 255, 0.8)", 2);
    }
  }
}, _class5);
async function loadSpline(url) {
  const res = await fetch(url);
  return await res.json();
}

// src/components/Rect.ts
var Rect = class extends _chunk2Z23LTH3js.Entity {
  
  
  
  
  constructor(opts = {}) {
    super();
    this.width = _nullishCoalesce(opts.width, () => ( 0));
    this.height = _nullishCoalesce(opts.height, () => ( 0));
    this.fill = opts.fill === void 0 ? "#38bdf8" : opts.fill;
    this.stroke = _nullishCoalesce(opts.stroke, () => ( null));
    this.strokeWidth = _nullishCoalesce(opts.strokeWidth, () => ( 1));
    this.radius = _nullishCoalesce(opts.radius, () => ( 0));
  }
  getBounds() {
    return { x: 0, y: 0, width: this.width, height: this.height };
  }
  /**
   * Solid-fill, square-cornered, unstroked rectangles opt into the GPU
   * instanced-rect batch (WebGL `pointBackend` only). Any stroke or corner
   * radius needs the exact Canvas path, so return `null` to fall back to
   * {@link render}. Read each frame, so an animated `fill` still batches.
   */
  getBatchRect() {
    if (!this.fill || this.stroke || this.radius > 0) return null;
    return { width: this.width, height: this.height, color: this.fill };
  }
  isPointInside(globalX, globalY) {
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x >= 0 && local.x <= this.width && local.y >= 0 && local.y <= this.height;
  }
  render(renderer) {
    renderer.beginPath();
    if (this.radius > 0) {
      renderer.roundRect(0, 0, this.width, this.height, this.radius);
    } else {
      renderer.moveTo(0, 0);
      renderer.lineTo(this.width, 0);
      renderer.lineTo(this.width, this.height);
      renderer.lineTo(0, this.height);
      renderer.closePath();
    }
    if (this.fill) renderer.fill(this.fill);
    if (this.stroke) renderer.stroke(this.stroke, this.strokeWidth);
  }
};

// src/components/Circle.ts
var Circle = class extends _chunk2Z23LTH3js.Entity {
  
  
  
  
  constructor(opts = {}) {
    super();
    this._radius = _nullishCoalesce(opts.radius, () => ( 0));
    this.fill = opts.fill === void 0 ? "#38bdf8" : opts.fill;
    this.stroke = _nullishCoalesce(opts.stroke, () => ( null));
    this.strokeWidth = _nullishCoalesce(opts.strokeWidth, () => ( 1));
    this.syncBox();
  }
  get radius() {
    return this._radius;
  }
  set radius(v) {
    this._radius = v;
    this.syncBox();
  }
  /** Keep the a11y box (a square around the centered disc) in sync with radius. */
  syncBox() {
    this.width = this._radius * 2;
    this.height = this._radius * 2;
    this.a11yOffsetX = -this._radius;
    this.a11yOffsetY = -this._radius;
  }
  getBounds() {
    return {
      x: -this._radius,
      y: -this._radius,
      width: this._radius * 2,
      height: this._radius * 2
    };
  }
  /**
   * A solid-fill, unstroked circle opts into the renderer's circle batch
   * (center = entity origin, radius scaled by world scale). A stroke needs the
   * exact Canvas path, so return `null` there. Read each frame, so an animated
   * `fill`/`radius` still batches.
   */
  getBatchCircle() {
    if (!this.fill || this.stroke) return null;
    return { radius: this._radius, color: this.fill };
  }
  isPointInside(globalX, globalY) {
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x * local.x + local.y * local.y <= this._radius * this._radius;
  }
  render(renderer) {
    renderer.beginPath();
    renderer.arc(0, 0, this._radius, 0, Math.PI * 2);
    renderer.closePath();
    if (this.fill) renderer.fill(this.fill);
    if (this.stroke) renderer.stroke(this.stroke, this.strokeWidth);
  }
};

// src/components/Group.ts
var Group = class extends _chunk2Z23LTH3js.Entity {
  constructor(...children) {
    super();
    if (children.length > 0) this.add(...children);
  }
  isPointInside() {
    return false;
  }
  render(_renderer) {
  }
};

// src/math/SpatialHashGrid.ts
var SpatialHashGrid = (_class6 = class {
  
  __init80() {this.grid = /* @__PURE__ */ new Map()}
  __init81() {this.entityCells = /* @__PURE__ */ new Map()}
  constructor(cellSize = 64) {;_class6.prototype.__init80.call(this);_class6.prototype.__init81.call(this);
    this.cellSize = cellSize;
  }
  hash(cx, cy) {
    const x = cx < 0 ? -2 * cx - 1 : 2 * cx;
    const y = cy < 0 ? -2 * cy - 1 : 2 * cy;
    return (x + y) * (x + y + 1) / 2 + y;
  }
  cellsForAABB(x, y, w, h) {
    const minCx = Math.floor(x / this.cellSize);
    const minCy = Math.floor(y / this.cellSize);
    const maxCx = Math.floor((x + w) / this.cellSize);
    const maxCy = Math.floor((y + h) / this.cellSize);
    const keys = [];
    for (let cx = minCx; cx <= maxCx; cx++) {
      for (let cy = minCy; cy <= maxCy; cy++) {
        keys.push(this.hash(cx, cy));
      }
    }
    return keys;
  }
  /**
   * Insert or update an entity's axis-aligned bounding box in the grid.
   *
   * If the entity is already registered its old cell memberships are removed
   * before the new ones are computed, so this method is safe to call every
   * frame.
   *
   * @param id - Unique string identifier for the entity.
   * @param x - Left edge of the AABB in world space.
   * @param y - Top edge of the AABB in world space.
   * @param w - Width of the AABB.
   * @param h - Height of the AABB.
   */
  insert(id, x, y, w, h) {
    this.remove(id);
    const keys = this.cellsForAABB(x, y, w, h);
    this.entityCells.set(id, keys);
    for (const key of keys) {
      if (!this.grid.has(key)) this.grid.set(key, /* @__PURE__ */ new Set());
      this.grid.get(key).add(id);
    }
  }
  /**
   * Remove an entity from all grid cells it currently occupies.
   *
   * Silently does nothing if the entity is not registered.
   *
   * @param id - Unique string identifier of the entity to remove.
   */
  remove(id) {
    const keys = this.entityCells.get(id);
    if (!keys) return;
    for (const key of keys) {
      _optionalChain([this, 'access', _119 => _119.grid, 'access', _120 => _120.get, 'call', _121 => _121(key), 'optionalAccess', _122 => _122.delete, 'call', _123 => _123(id)]);
    }
    this.entityCells.delete(id);
  }
  /**
   * Return all entity IDs whose grid cells overlap the given AABB.
   *
   * Time complexity: O(k) where k is the number of cells the query AABB spans
   * plus the number of results — O(1) average for small, similarly-sized entities.
   *
   * @param x - Left edge of the query AABB.
   * @param y - Top edge of the query AABB.
   * @param w - Width of the query AABB.
   * @param h - Height of the query AABB.
   * @returns A `Set` of entity ID strings whose cells intersect the query region.
   */
  query(x, y, w, h) {
    const result = /* @__PURE__ */ new Set();
    for (const key of this.cellsForAABB(x, y, w, h)) {
      const cell = this.grid.get(key);
      if (cell) for (const id of cell) result.add(id);
    }
    return result;
  }
  /**
   * Clear all cells and entity registrations, resetting the grid to an empty state.
   *
   * Call once per frame before re-inserting all dynamic entities.
   */
  clear() {
    this.grid.clear();
    this.entityCells.clear();
  }
}, _class6);

// src/tree/DOMPortalEntity.ts
var DOMPortalEntity = (_class7 = class extends _chunk2Z23LTH3js.Entity {
  
  __init82() {this.isDOMPortal = true}
  __init83() {this.domListeners = []}
  __init84() {this.resizeObserver = null}
  __init85() {this.cachedWidth = 100}
  __init86() {this.cachedHeight = 100}
  __init87() {this.lastWidth = ""}
  __init88() {this.lastHeight = ""}
  __init89() {this.lastTransform = ""}
  __init90() {this.lastZIndex = ""}
  __init91() {this.lastOpacity = ""}
  constructor(domElement, width, height, id) {
    super(id);_class7.prototype.__init82.call(this);_class7.prototype.__init83.call(this);_class7.prototype.__init84.call(this);_class7.prototype.__init85.call(this);_class7.prototype.__init86.call(this);_class7.prototype.__init87.call(this);_class7.prototype.__init88.call(this);_class7.prototype.__init89.call(this);_class7.prototype.__init90.call(this);_class7.prototype.__init91.call(this);;
    this.domElement = domElement;
    this.width = _nullishCoalesce(width, () => ( 0));
    this.height = _nullishCoalesce(height, () => ( 0));
    if (typeof window !== "undefined") {
      this.domElement.style.position = "absolute";
      this.domElement.style.transformOrigin = "0 0";
      this.domElement.style.pointerEvents = "auto";
      this.cachedWidth = parseFloat(domElement.style.width) || domElement.offsetWidth || 100;
      this.cachedHeight = parseFloat(domElement.style.height) || domElement.offsetHeight || 100;
      if (typeof ResizeObserver !== "undefined") {
        this.resizeObserver = new ResizeObserver((entries) => {
          for (const entry of entries) {
            this.cachedWidth = entry.contentRect.width || entry.target.offsetWidth;
            this.cachedHeight = entry.contentRect.height || entry.target.offsetHeight;
          }
        });
        this.resizeObserver.observe(this.domElement);
      }
      const events = [
        "click",
        "pointerdown",
        "pointerup",
        "pointercancel",
        "pointermove",
        "wheel"
      ];
      for (const type of events) {
        const handler = (e) => {
          this.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)(type, this, e));
        };
        this.domElement.addEventListener(type, handler);
        this.domListeners.push({ type, handler, capture: false });
      }
      const hoverEvents = [
        { native: "mouseenter", vecto: "hover" },
        { native: "mouseleave", vecto: "pointerleave" }
      ];
      for (const { native, vecto } of hoverEvents) {
        const handler = (e) => {
          this.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)(vecto, this, e, false));
        };
        this.domElement.addEventListener(native, handler);
        this.domListeners.push({ type: native, handler, capture: false });
      }
      const focusEvents = ["focus", "blur"];
      for (const type of focusEvents) {
        const handler = (e) => {
          this.dispatchEvent(new (0, _chunk2Z23LTH3js.VectoJSEvent)(type, this, e, true));
        };
        this.domElement.addEventListener(type, handler, true);
        this.domListeners.push({ type, handler, capture: true });
      }
    }
  }
  isPointInside(globalX, globalY) {
    const w = this.width > 0 ? this.width : this.cachedWidth;
    const h = this.height > 0 ? this.height : this.cachedHeight;
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x >= 0 && local.x <= w && local.y >= 0 && local.y <= h;
  }
  add(..._children) {
    console.warn(`DOMPortalEntity (${this.id}) is a leaf node. Child entities are not supported.`);
    return this;
  }
  render() {
  }
  destroy() {
    if (typeof window !== "undefined") {
      if (this.resizeObserver) {
        this.resizeObserver.disconnect();
        this.resizeObserver = null;
      }
      if (this.domElement) {
        for (const { type, handler, capture } of this.domListeners) {
          this.domElement.removeEventListener(type, handler, capture);
        }
        this.domListeners = [];
        this.domElement.remove();
      }
    }
    super.destroy();
  }
}, _class7);

// src/index.ts
Scene.registerWebGLPointRendererCreator(_chunkORCPCIJ7js.createWebGLPointRenderer);
Scene.registerWebGPUParticleSystemManager(_chunkORCPCIJ7js.WebGPUParticleSystemManager);


















































exports.ArabicShaper = _chunk4AR425ARjs.ArabicShaper; exports.BidiResolver = _chunk4AR425ARjs.BidiResolver; exports.CanvasRenderer = _chunkORCPCIJ7js.CanvasRenderer; exports.Circle = Circle; exports.ComputeParticleEntity = ComputeParticleEntity; exports.DOMPortalEntity = DOMPortalEntity; exports.Easing = _chunk2Z23LTH3js.Easing; exports.Entity = _chunk2Z23LTH3js.Entity; exports.GridTextEntity = GridTextEntity; exports.Group = Group; exports.LayoutEngine = _chunkBA5HUUDFjs.LayoutEngine; exports.LayoutResultBuffer = _chunkBA5HUUDFjs.LayoutResultBuffer; exports.LayoutWorkerManager = _chunk4AR425ARjs.LayoutWorkerManager; exports.MSDFFont = _chunk2Z23LTH3js.MSDFFont; exports.MSDFTextEntity = _chunk2Z23LTH3js.MSDFTextEntity; exports.PARTICLE_OFFSET_LIFE = PARTICLE_OFFSET_LIFE; exports.PARTICLE_OFFSET_ORIGIN_X = PARTICLE_OFFSET_ORIGIN_X; exports.PARTICLE_OFFSET_ORIGIN_Y = PARTICLE_OFFSET_ORIGIN_Y; exports.PARTICLE_OFFSET_POSITION_X = PARTICLE_OFFSET_POSITION_X; exports.PARTICLE_OFFSET_POSITION_Y = PARTICLE_OFFSET_POSITION_Y; exports.PARTICLE_OFFSET_SIZE = PARTICLE_OFFSET_SIZE; exports.PARTICLE_OFFSET_VELOCITY_X = PARTICLE_OFFSET_VELOCITY_X; exports.PARTICLE_OFFSET_VELOCITY_Y = PARTICLE_OFFSET_VELOCITY_Y; exports.PARTICLE_STRIDE_FLOATS = PARTICLE_STRIDE_FLOATS; exports.REDUCED_MOTION_FPS = REDUCED_MOTION_FPS; exports.Rect = Rect; exports.SVGEntity = _chunk2Z23LTH3js.SVGEntity; exports.SVGRenderer = _chunkORCPCIJ7js.SVGRenderer; exports.Scene = Scene; exports.SpatialHashGrid = SpatialHashGrid; exports.SplineEntity = SplineEntity; exports.SpringDriver = _chunk2Z23LTH3js.SpringDriver; exports.SpringPhysics = _chunk2Z23LTH3js.SpringPhysics; exports.TextEntity = TextEntity; exports.TweenDriver = _chunk2Z23LTH3js.TweenDriver; exports.VectoJSEvent = _chunk2Z23LTH3js.VectoJSEvent; exports.WebGPUParticleSystemManager = _chunkORCPCIJ7js.WebGPUParticleSystemManager; exports.clearCssLineBoxMetrics = _chunk2Z23LTH3js.clearCssLineBoxMetrics; exports.computeLineSegments = _chunkBA5HUUDFjs.computeLineSegments; exports.createCanvasMeasurer = _chunkBA5HUUDFjs.createCanvasMeasurer; exports.createWebGLPointRenderer = _chunkORCPCIJ7js.createWebGLPointRenderer; exports.cssLineBoxBaseline = _chunk2Z23LTH3js.cssLineBoxBaseline; exports.isSafeUrl = _chunkORCPCIJ7js.isSafeUrl; exports.isTweenConfig = _chunk2Z23LTH3js.isTweenConfig; exports.loadSpline = loadSpline; exports.parseColorToRGBA = _chunkORCPCIJ7js.parseColorToRGBA; exports.polySegmentToBezier = polySegmentToBezier; exports.prepareContentGrid = _chunk2Z23LTH3js.prepareContentGrid; exports.sanitizeUrl = _chunkORCPCIJ7js.sanitizeUrl;
