// @vectojs/ui/dist/ContextMenu.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ContextMenu.ts
var ContextMenu_exports = {};
__export(ContextMenu_exports, {
  ContextMenu: () => ContextMenu
});
module.exports = __toCommonJS(ContextMenu_exports);
var import_core3 = require("@vectojs/core");

// src/UIComponent.ts
var import_core = require("@vectojs/core");
var UIComponent = class extends import_core.Entity {
  /** Inner padding in pixels, used by box-style components. */
  padding = 0;
  /** Played automatically when the component mounts to a live scene, if set. */
  enterMotion;
  /** Played by {@link dismiss} before the component removes itself, if set. */
  exitMotion;
  onMounted() {
    if (this.enterMotion) void this.playMotion(this.enterMotion);
  }
  /** Seed each prop to `from`, register the transition, then drive to `to`. */
  playMotion(spec) {
    const cfg = spec.config ?? "spring";
    const targets = {};
    const trans = {};
    for (const [k, pair] of Object.entries(spec.props)) {
      const prop = k;
      const [from, to] = pair;
      this.setImmediate(prop, from);
      targets[prop] = to;
      trans[prop] = cfg;
    }
    this.setTransition(trans);
    return this.driveMotion(targets, cfg);
  }
  driveMotion(targets, cfg) {
    if (typeof cfg === "object" && "duration" in cfg) {
      return this.animateTo(targets, cfg);
    }
    return this.springTo(targets, cfg === "spring" ? {} : cfg);
  }
  /** Play the exit motion (if any), then remove self from the tree. */
  async dismiss() {
    if (this.exitMotion) await this.playMotion(this.exitMotion);
    this.parent?.remove(this);
  }
  /** Wakes the render loop at the next caret-blink phase boundary, or `null`. */
  caretBlinkTimer = null;
  /**
   * Keep a focused text component's caret blink visible to the idle throttle.
   * The blink phase is derived from `Date.now()` inside `render()`, so nothing
   * marks the scene dirty when the phase flips — an idle `onDemand` scene
   * would freeze the caret solid (the ScrollView 0.2.x regression class).
   * One timeout per 500 ms phase boundary costs ~2 renders/s while focused;
   * reporting a permanent pending animation instead would pin the scene at
   * full frame rate for as long as the field holds focus.
   */
  startCaretBlinkWake() {
    this.stopCaretBlinkWake();
    const schedule = () => {
      this.caretBlinkTimer = setTimeout(
        () => {
          this.scene?.markDirty();
          schedule();
        },
        500 - Date.now() % 500
      );
    };
    schedule();
    this.scene?.markDirty();
  }
  /** Stop the caret-blink wake-up (on blur; destroy also clears it). */
  stopCaretBlinkWake() {
    if (this.caretBlinkTimer !== null) {
      clearTimeout(this.caretBlinkTimer);
      this.caretBlinkTimer = null;
    }
    this.scene?.markDirty();
  }
  destroy() {
    if (this.caretBlinkTimer !== null) {
      clearTimeout(this.caretBlinkTimer);
      this.caretBlinkTimer = null;
    }
    super.destroy();
  }
  /**
   * Axis-aligned hit-test against the component's box in global space.
   *
   * @param globalX - World-space X coordinate.
   * @param globalY - World-space Y coordinate.
   * @returns Whether the point lies within `[0, width] x [0, height]` locally.
   */
  isPointInside(globalX, globalY) {
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x >= 0 && local.x <= this.width && local.y >= 0 && local.y <= this.height;
  }
  /** Local-space box, enabling viewport culling by {@link Scene}. */
  getBounds() {
    return { x: 0, y: 0, width: this.width, height: this.height };
  }
};

// src/Overlay.ts
var Overlay = class extends UIComponent {
  placement;
  offset;
  visible = false;
  constructor(opts) {
    super();
    this.width = opts.width;
    this.height = opts.height;
    this.placement = opts.placement ?? "bottom";
    this.offset = opts.offset ?? 6;
    this.interactive = false;
    this.opacity = 0;
    this.scaleX = 0.92;
    this.scaleY = 0.92;
    this.setTransition({
      opacity: { duration: 160, easing: "easeOutQuad" },
      scaleX: "spring",
      scaleY: "spring"
    });
  }
  /**
   * Show the overlay anchored to a target entity.
   * Automatically mounts to `scene.overlayRoot` on first call.
   */
  showAt(target) {
    this._mount(target);
    this._position(target);
    this.visible = true;
    this.opacity = 1;
    this.scaleX = 1;
    this.scaleY = 1;
    this.scene?.markDirty();
  }
  /**
   * Show the overlay at an absolute canvas position.
   *
   * If the overlay has never been added to the scene tree (no `parent`), its
   * `scene` resolves to `null` (the `Entity.scene` getter walks the parent
   * chain). To make the documented "bare `new ContextMenu({...})` then
   * `showAtPoint`" pattern work without forcing every caller to pre-mount,
   * pass a `source` — either the `Scene` itself or any mounted `Entity` whose
   * `.scene` is set (e.g. the entity whose `pointerdown` listener is calling
   * you) — and the overlay will auto-mount to that scene's `overlayRoot`.
   * Without a `source` and no resolvable `scene`, this method silently
   * no-ops (preserved for backward compatibility; existing pre-mount-then-
   * showAtPoint callers are unchanged).
   */
  showAtPoint(x, y, source) {
    const scene = this.scene ?? this._sceneFromSource(source);
    if (!scene) return;
    if (!this.parent) scene.overlayRoot.add(this);
    this._placeAt(x, y);
    this.visible = true;
    this.opacity = 1;
    this.scaleX = 1;
    this.scaleY = 1;
    scene.markDirty();
  }
  /**
   * Resolve a {@link Scene} from the optional `showAtPoint` `source` arg.
   * Accepts a `Scene` passed directly, or any mounted `Entity` whose `.scene`
   * is the authoritative source (e.g. the entity whose `pointerdown` listener
   * is opening the menu).
   *
   * `protected`, not `private`: subclasses that override `showAtPoint` (e.g.
   * `ContextMenu`, to mount its outside-click backdrop) need this same
   * resolution — duplicating it drifts out of sync with the base
   * implementation, which is exactly how `ContextMenu.showAtPoint` silently
   * dropped the `source` arg it was handed (see its override for the fix).
   */
  _sceneFromSource(source) {
    if (!source) return null;
    if ("overlayRoot" in source && typeof source.markDirty === "function") {
      return source;
    }
    return source.scene ?? null;
  }
  /** Animate the overlay out (stays mounted; re-show via showAt). */
  hide() {
    this.visible = false;
    this.opacity = 0;
    this.scaleX = 0.92;
    this.scaleY = 0.92;
    this.scene?.markDirty();
  }
  _mount(target) {
    const scene = target.scene;
    if (!scene) return;
    if (this.parent) this.parent.remove(this);
    scene.overlayRoot.add(this);
  }
  _position(target) {
    const targetBounds = target.getWorldBounds();
    const gp = { x: targetBounds.x, y: targetBounds.y };
    const tw = targetBounds.width;
    const th = targetBounds.height;
    const sw = this.scene?.width ?? window.innerWidth;
    const sh = this.scene?.height ?? window.innerHeight;
    let pl = this.placement;
    if (pl === "auto") {
      const below = sh - (gp.y + th);
      const above = gp.y;
      pl = below >= this.height + this.offset ? "bottom" : "top";
      if (below < this.height + this.offset && above < this.height + this.offset)
        pl = below >= above ? "bottom" : "top";
    }
    let ax = gp.x;
    let ay = gp.y;
    switch (pl) {
      case "top":
        ax = gp.x + tw / 2 - this.width / 2;
        ay = gp.y - this.height - this.offset;
        break;
      case "top-start":
        ax = gp.x;
        ay = gp.y - this.height - this.offset;
        break;
      case "top-end":
        ax = gp.x + tw - this.width;
        ay = gp.y - this.height - this.offset;
        break;
      case "bottom":
        ax = gp.x + tw / 2 - this.width / 2;
        ay = gp.y + th + this.offset;
        break;
      case "bottom-start":
        ax = gp.x;
        ay = gp.y + th + this.offset;
        break;
      case "bottom-end":
        ax = gp.x + tw - this.width;
        ay = gp.y + th + this.offset;
        break;
      case "left":
        ax = gp.x - this.width - this.offset;
        ay = gp.y + th / 2 - this.height / 2;
        break;
      case "right":
        ax = gp.x + tw + this.offset;
        ay = gp.y + th / 2 - this.height / 2;
        break;
    }
    this._placeAt(ax, ay, sw, sh);
  }
  _placeAt(ax, ay, sw, sh) {
    const W = sw ?? this.scene?.width ?? window.innerWidth;
    const H = sh ?? this.scene?.height ?? window.innerHeight;
    this.x = Math.max(4, Math.min(ax, W - this.width - 4));
    this.y = Math.max(4, Math.min(ay, H - this.height - 4));
  }
  render(_r) {
  }
};

// src/measure.ts
var import_core2 = require("@vectojs/core");
var sharedCtx;
function getCtx() {
  if (sharedCtx !== void 0) return sharedCtx;
  sharedCtx = typeof document !== "undefined" ? document.createElement("canvas").getContext("2d") : null;
  return sharedCtx;
}
function fontSizePx(font) {
  const pxIndex = font.indexOf("px");
  if (pxIndex <= 0) return 16;
  let start = pxIndex - 1;
  while (start >= 0) {
    const ch = font[start];
    if (ch >= "0" && ch <= "9" || ch === ".") {
      start--;
    } else {
      break;
    }
  }
  const raw = font.slice(start + 1, pxIndex);
  if (raw === "") return 16;
  const size = Number.parseFloat(raw);
  return Number.isFinite(size) ? size : 16;
}
var MEASURE_CACHE_MAX = 1e3;
var measureCache = /* @__PURE__ */ new Map();
if (typeof document !== "undefined" && document.fonts) {
  const clearCache = () => measureCache.clear();
  document.fonts.ready.then(clearCache);
  document.fonts.addEventListener("loadingdone", clearCache);
}
function measureText(text, font) {
  const shaped = import_core2.ArabicShaper.shapeArabic(text).shapedText;
  const key = `${font} ${shaped}`;
  const cached = measureCache.get(key);
  if (cached !== void 0) {
    measureCache.delete(key);
    measureCache.set(key, cached);
    return cached;
  }
  const ctx = getCtx();
  let width;
  if (!ctx) {
    width = shaped.length * fontSizePx(font) * 0.5;
  } else {
    ctx.font = font;
    width = ctx.measureText(shaped).width;
  }
  measureCache.set(key, width);
  if (measureCache.size > MEASURE_CACHE_MAX) {
    measureCache.delete(measureCache.keys().next().value);
  }
  return width;
}

// src/ContextMenu.ts
var nextContextMenuId = 1;
var ContextMenu = class _ContextMenu extends Overlay {
  _items;
  _font;
  _textColor;
  _disColor;
  _bg;
  _hoverBg;
  _border;
  _iH;
  _sH;
  _hoverIdx = -1;
  _submenu = null;
  _parentMenu = null;
  /** Which item's `children` `_submenu` currently represents, if any. */
  _submenuFor = null;
  _opts;
  /** Full-screen invisible entity mounted behind the menu while open, so a
   * click anywhere outside it closes the menu — the way every native context
   * menu behaves. Without this the menu only ever closed by selecting one of
   * its own (non-disabled) items. */
  _backdrop = null;
  constructor(opts) {
    const iH = opts.itemHeight ?? 32;
    const sH = opts.separatorHeight ?? 9;
    const totalH = (opts.items ?? []).reduce((acc, it) => acc + (it.separator ? sH : iH), 0);
    super({ width: opts.width ?? 220, height: totalH + 8, placement: "auto", offset: 2 });
    this.id = `context-menu-${nextContextMenuId++}`;
    this._opts = opts;
    this._items = opts.items ?? [];
    this._font = opts.font ?? "13px sans-serif";
    this._textColor = opts.color ?? "#e2e8f0";
    this._disColor = opts.disabledColor ?? "rgba(255,255,255,0.3)";
    this._bg = opts.bg ?? "rgba(18,18,32,0.97)";
    this._hoverBg = opts.hoverBg ?? "rgba(0,240,255,0.14)";
    this._border = opts.borderColor ?? "rgba(255,255,255,0.12)";
    this._iH = iH;
    this._sH = sH;
    this.interactive = false;
    this.on("pointermove", (e) => {
      this._hoverIdx = e.localY === void 0 ? -1 : this._idxAt(e.localY);
      this.scene?.markDirty();
    });
    this.on("pointerleave", () => {
      this._hoverIdx = -1;
      this.scene?.markDirty();
    });
    this.on("pointerdown", (e) => {
      if (e.localY === void 0) return;
      const idx = this._idxAt(e.localY);
      const item = this._items[idx];
      if (!item || item.separator || item.disabled) return;
      if (item.children && item.children.length > 0) {
        if (!this._submenu || this._submenuFor !== item) {
          if (this._submenu) this._submenu.destroy();
          this._submenu = new _ContextMenu({ ...this._opts, items: item.children });
          this._submenu._parentMenu = this;
          this._submenuFor = item;
          if (this.scene) this.scene.overlayRoot.add(this._submenu);
        }
        this._submenu.showAtPoint(this.x + this.width, this.y + this._rowTop(idx));
      } else {
        item.onClick?.();
        this._rootMenu().hide();
      }
    });
  }
  showAtPoint(x, y, source) {
    const scene = this.scene ?? this._sceneFromSource(source);
    if (scene) this.interactive = true;
    if (this._parentMenu === null && !this._backdrop && scene) {
      const backdrop = new class ContextMenuBackdrop extends import_core3.Entity {
        isPointInside() {
          return true;
        }
        render() {
        }
      }(`${this.id}-backdrop`);
      backdrop.width = scene.width;
      backdrop.height = scene.height;
      backdrop.interactive = true;
      const dismiss = (e) => {
        e.stopPropagation();
        this.hide();
      };
      backdrop.on("pointerdown", dismiss);
      backdrop.on("click", dismiss);
      const parent = this.parent ?? scene.overlayRoot;
      if (this.parent) this.parent.remove(this);
      scene.overlayRoot.add(backdrop);
      parent.add(this);
      this._backdrop = backdrop;
    }
    super.showAtPoint(x, y, source);
  }
  hide() {
    const scene = this.scene;
    if (this._backdrop) {
      this._backdrop.destroy();
      this._backdrop = null;
    }
    if (this._submenu) this._submenu.hide();
    super.hide();
    this.interactive = false;
    scene?.detachA11y(this);
  }
  destroy() {
    if (this._backdrop) {
      this._backdrop.destroy();
      this._backdrop = null;
    }
    if (this._submenu) {
      this._submenu.destroy();
      this._submenu = null;
    }
    this._parentMenu = null;
    super.destroy();
  }
  _rootMenu() {
    return this._parentMenu?._rootMenu() ?? this;
  }
  _idxAt(localY) {
    let y = 4;
    for (let i = 0; i < this._items.length; i++) {
      const h = this._items[i].separator ? this._sH : this._iH;
      if (localY >= y && localY < y + h) return i;
      y += h;
    }
    return -1;
  }
  _rowTop(idx) {
    let y = 4;
    for (let i = 0; i < idx; i++) y += this._items[i].separator ? this._sH : this._iH;
    return y;
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 8);
    r.fill(this._bg);
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 8);
    r.stroke(this._border, 1);
    let y = 4;
    for (let i = 0; i < this._items.length; i++) {
      const item = this._items[i];
      if (item.separator) {
        const mid = y + this._sH / 2;
        r.beginPath();
        r.moveTo(8, mid);
        r.lineTo(this.width - 8, mid);
        r.stroke("rgba(255,255,255,0.1)", 1);
        y += this._sH;
        continue;
      }
      const col = item.disabled ? this._disColor : this._textColor;
      if (i === this._hoverIdx && !item.disabled) {
        r.beginPath();
        r.roundRect(4, y, this.width - 8, this._iH, 4);
        r.fill(this._hoverBg);
      }
      const ty = y + this._iH / 2 + 4;
      let lx = 12;
      if (item.icon) {
        r.fillText(item.icon, lx, ty, this._font, col);
        lx += 22;
      }
      r.fillText(item.label ?? "", lx, ty, this._font, col);
      if (item.shortcut) {
        r.fillText(
          item.shortcut,
          this.width - 12 - measureText(item.shortcut, this._font),
          ty,
          this._font,
          item.disabled ? this._disColor : "rgba(255,255,255,0.4)"
        );
      }
      if (item.children && item.children.length > 0) {
        r.fillText("\u25B8", this.width - 16, ty, "10px sans-serif", "rgba(255,255,255,0.5)");
      }
      y += this._iH;
    }
  }
  getA11yAttributes() {
    return { role: "menu", label: "Context menu" };
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ContextMenu
});
