// @vectojs/ui/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Button: () => Button,
  Card: () => Card,
  Checkbox: () => Checkbox,
  CodeBlock: () => CodeBlock,
  ContextMenu: () => ContextMenu,
  Dropdown: () => Dropdown,
  Flow: () => Flow,
  Image: () => Image,
  Input: () => Input,
  Link: () => Link,
  Markdown: () => Markdown,
  Modal: () => Modal,
  Overlay: () => Overlay,
  Panel: () => Panel,
  PanelGroup: () => PanelGroup,
  PanelResizeHandle: () => PanelResizeHandle,
  Popover: () => Popover,
  ProgressBar: () => ProgressBar,
  RadioGroup: () => RadioGroup,
  RichText: () => RichText,
  ScrollView: () => ScrollView,
  Slider: () => Slider,
  Stack: () => Stack,
  Table: () => Table,
  Tabs: () => Tabs,
  Text: () => Text,
  TextArea: () => TextArea,
  Toggle: () => Toggle,
  Tooltip: () => Tooltip,
  TreeView: () => TreeView,
  UIComponent: () => UIComponent,
  VirtualList: () => VirtualList,
  measureText: () => measureText,
  wrapLines: () => wrapLines,
  wrapText: () => wrapText
});
module.exports = __toCommonJS(index_exports);

// src/UIComponent.ts
var import_core = require("@vectojs/core");
var UIComponent = class extends import_core.Entity {
  /** Inner padding in pixels, used by box-style components. */
  padding = 0;
  /** Played automatically when the component mounts to a live scene, if set. */
  enterMotion;
  /** Played by {@link dismiss} before the component removes itself, if set. */
  exitMotion;
  onMounted() {
    if (this.enterMotion) void this.playMotion(this.enterMotion);
  }
  /** Seed each prop to `from`, register the transition, then drive to `to`. */
  playMotion(spec) {
    const cfg = spec.config ?? "spring";
    const targets = {};
    const trans = {};
    for (const [k, pair] of Object.entries(spec.props)) {
      const prop = k;
      const [from, to] = pair;
      this.setImmediate(prop, from);
      targets[prop] = to;
      trans[prop] = cfg;
    }
    this.setTransition(trans);
    return this.driveMotion(targets, cfg);
  }
  driveMotion(targets, cfg) {
    if (typeof cfg === "object" && "duration" in cfg) {
      return this.animateTo(targets, cfg);
    }
    return this.springTo(targets, cfg === "spring" ? {} : cfg);
  }
  /** Play the exit motion (if any), then remove self from the tree. */
  async dismiss() {
    if (this.exitMotion) await this.playMotion(this.exitMotion);
    this.parent?.remove(this);
  }
  /** Wakes the render loop at the next caret-blink phase boundary, or `null`. */
  caretBlinkTimer = null;
  /**
   * Keep a focused text component's caret blink visible to the idle throttle.
   * The blink phase is derived from `Date.now()` inside `render()`, so nothing
   * marks the scene dirty when the phase flips — an idle `onDemand` scene
   * would freeze the caret solid (the ScrollView 0.2.x regression class).
   * One timeout per 500 ms phase boundary costs ~2 renders/s while focused;
   * reporting a permanent pending animation instead would pin the scene at
   * full frame rate for as long as the field holds focus.
   */
  startCaretBlinkWake() {
    this.stopCaretBlinkWake();
    const schedule = () => {
      this.caretBlinkTimer = setTimeout(
        () => {
          this.scene?.markDirty();
          schedule();
        },
        500 - Date.now() % 500
      );
    };
    schedule();
    this.scene?.markDirty();
  }
  /** Stop the caret-blink wake-up (on blur; destroy also clears it). */
  stopCaretBlinkWake() {
    if (this.caretBlinkTimer !== null) {
      clearTimeout(this.caretBlinkTimer);
      this.caretBlinkTimer = null;
    }
    this.scene?.markDirty();
  }
  destroy() {
    if (this.caretBlinkTimer !== null) {
      clearTimeout(this.caretBlinkTimer);
      this.caretBlinkTimer = null;
    }
    super.destroy();
  }
  /**
   * Axis-aligned hit-test against the component's box in global space.
   *
   * @param globalX - World-space X coordinate.
   * @param globalY - World-space Y coordinate.
   * @returns Whether the point lies within `[0, width] x [0, height]` locally.
   */
  isPointInside(globalX, globalY) {
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x >= 0 && local.x <= this.width && local.y >= 0 && local.y <= this.height;
  }
  /** Local-space box, enabling viewport culling by {@link Scene}. */
  getBounds() {
    return { x: 0, y: 0, width: this.width, height: this.height };
  }
};

// src/Text.ts
var import_core3 = require("@vectojs/core");

// src/measure.ts
var import_core2 = require("@vectojs/core");
var sharedCtx;
function getCtx() {
  if (sharedCtx !== void 0) return sharedCtx;
  sharedCtx = typeof document !== "undefined" ? document.createElement("canvas").getContext("2d") : null;
  return sharedCtx;
}
function fontSizePx(font) {
  const pxIndex = font.indexOf("px");
  if (pxIndex <= 0) return 16;
  let start = pxIndex - 1;
  while (start >= 0) {
    const ch = font[start];
    if (ch >= "0" && ch <= "9" || ch === ".") {
      start--;
    } else {
      break;
    }
  }
  const raw = font.slice(start + 1, pxIndex);
  if (raw === "") return 16;
  const size = Number.parseFloat(raw);
  return Number.isFinite(size) ? size : 16;
}
var MEASURE_CACHE_MAX = 1e3;
var measureCache = /* @__PURE__ */ new Map();
if (typeof document !== "undefined" && document.fonts) {
  const clearCache = () => measureCache.clear();
  document.fonts.ready.then(clearCache);
  document.fonts.addEventListener("loadingdone", clearCache);
}
function measureText(text, font) {
  const shaped = import_core2.ArabicShaper.shapeArabic(text).shapedText;
  const key = `${font} ${shaped}`;
  const cached = measureCache.get(key);
  if (cached !== void 0) {
    measureCache.delete(key);
    measureCache.set(key, cached);
    return cached;
  }
  const ctx = getCtx();
  let width;
  if (!ctx) {
    width = shaped.length * fontSizePx(font) * 0.5;
  } else {
    ctx.font = font;
    width = ctx.measureText(shaped).width;
  }
  measureCache.set(key, width);
  if (measureCache.size > MEASURE_CACHE_MAX) {
    measureCache.delete(measureCache.keys().next().value);
  }
  return width;
}
function wrapLines(text, font, maxWidth) {
  const lines = [];
  for (const paragraph of text.split("\n")) {
    const words = paragraph.split(" ");
    let current = "";
    for (const word of words) {
      const candidate = current ? `${current} ${word}` : word;
      if (current && measureText(candidate, font) > maxWidth) {
        lines.push(current);
        current = word;
      } else {
        current = candidate;
      }
    }
    lines.push(current);
  }
  return lines;
}

// src/Text.ts
function fontMeasurer(font) {
  if (typeof document === "undefined") return null;
  const ctx = document.createElement("canvas").getContext("2d");
  if (!ctx) return null;
  const cache = /* @__PURE__ */ new Map();
  return {
    measure(char) {
      let w = cache.get(char);
      if (w === void 0) {
        ctx.font = font;
        w = ctx.measureText(char).width;
        cache.set(char, w);
      }
      return w;
    }
  };
}
var Text = class extends UIComponent {
  text;
  font;
  color;
  maxWidth;
  lineHeight;
  selectable;
  engine;
  prepared;
  fontSize;
  lines = [];
  lineSourceRanges = [];
  constructor(text, opts = {}) {
    super();
    this.text = text;
    this.font = opts.font ?? "16px sans-serif";
    this.color = opts.color ?? "#e2e8f0";
    this.maxWidth = opts.maxWidth;
    this.lineHeight = opts.lineHeight ?? 20;
    this.selectable = opts.selectable ?? true;
    this.fontSize = fontSizePx(this.font);
    this.engine = new import_core3.LayoutEngine(this.maxWidth ?? 1e9, 1e9, fontMeasurer(this.font));
    if (opts.preserveLeadingSpaces) {
      this.engine.preserveLeadingSpaces = true;
    }
    this.prepared = this.engine.prepare(this.text, {}, this.fontSize);
    this.interactive = false;
    this.applyLayout();
  }
  /**
   * Replace the text. Runs the **cold** pass (re-segment + re-measure), then re-lays out.
   *
   * @param text - The new text content.
   * @returns `this` for chaining.
   */
  /** Mirror the rendered text into the DOM content layer (find-in-page, SR, SEO). */
  getContentProjection() {
    if (!this.text) return null;
    const lines = this.lines.map((visualText, index) => {
      const range = this.lineSourceRanges[index] ?? { start: 0, end: 0 };
      const nextStart = this.lineSourceRanges[index + 1]?.start ?? this.text.length;
      return {
        // Canvas keeps its visual glyph order; the semantic layer keeps logical
        // source order so native copy and RTL text remain correct.
        text: this.text.slice(range.start, range.end) || visualText,
        separatorAfter: this.text.slice(range.end, Math.max(range.end, nextStart)),
        x: 0,
        y: index * this.lineHeight,
        baseline: this.lineHeight * 0.8,
        font: this.font,
        lineHeight: this.lineHeight
      };
    });
    return {
      text: this.text,
      font: this.font,
      lineHeight: this.lineHeight,
      // The canvas has already decided the exact line breaks. Project visual
      // rows independently so browser whitespace/wrapping can never create a
      // different selection grid at a fractional zoom level.
      lines,
      selectable: this.selectable
    };
  }
  /** Enable or disable browser-native drag selection without rebuilding the entity. */
  setSelectable(selectable) {
    this.selectable = selectable;
    this.scene?.markDirty();
    return this;
  }
  setText(text) {
    this.text = text;
    this.prepared = this.engine.prepare(this.text, {}, this.fontSize);
    this.applyLayout();
    this.scene?.markDirty();
    return this;
  }
  /**
   * Append text — the streaming / typewriter path. Goes through the same cold
   * pass as {@link setText}, but the engine's paragraph memo reuses every
   * untouched leading paragraph, so only the changed (last) paragraph is
   * re-segmented + re-measured.
   *
   * @returns `this` for chaining.
   */
  append(text) {
    return this.setText(this.text + text);
  }
  /**
   * Change the wrap width and reflow via the cheap **hot** path (reuses the cached
   * measured text — no re-segmentation or re-measurement).
   *
   * @returns `this` for chaining.
   */
  setMaxWidth(maxWidth) {
    this.maxWidth = maxWidth;
    this.engine.maxWidth = maxWidth;
    this.applyLayout();
    this.scene?.markDirty();
    return this;
  }
  /** Hot pass: place the cached prepared text and regroup glyphs into lines. */
  applyLayout() {
    const result = this.engine.layoutPrepared(this.prepared);
    const lineQuantum = this.fontSize * 1.5;
    const byLine = /* @__PURE__ */ new Map();
    const nodesByLine = /* @__PURE__ */ new Map();
    let maxIdx = -1;
    for (const node of result.nodes) {
      const idx = Math.round(node.y / lineQuantum);
      byLine.set(idx, (byLine.get(idx) ?? "") + node.char);
      const nodes = nodesByLine.get(idx) ?? [];
      nodes.push(node);
      nodesByLine.set(idx, nodes);
      if (idx > maxIdx) maxIdx = idx;
    }
    this.lines = [];
    this.lineSourceRanges = [];
    let previousEnd = 0;
    for (let i = 0; i <= maxIdx; i++) {
      this.lines.push(byLine.get(i) ?? "");
      const nodes = nodesByLine.get(i) ?? [];
      let start = nodes.length > 0 ? Math.min(...nodes.map((node) => node.sourceIndex ?? previousEnd)) : previousEnd;
      const end = nodes.length > 0 ? Math.max(
        ...nodes.map((node) => (node.sourceIndex ?? previousEnd) + (node.sourceLength ?? 0))
      ) : start;
      if (i === 0) start = 0;
      this.lineSourceRanges.push({ start, end: Math.max(start, end) });
      previousEnd = Math.max(previousEnd, end);
    }
    if (this.lines.length === 0 && this.text) {
      this.lines = [""];
      this.lineSourceRanges = [{ start: 0, end: 0 }];
    }
    this.width = result.totalWidth;
    this.height = Math.max(maxIdx + 1, 1) * this.lineHeight;
  }
  getA11yAttributes() {
    return { label: this.text };
  }
  render(r) {
    for (let i = 0; i < this.lines.length; i++) {
      if (this.lines[i])
        r.fillText(this.lines[i], 0, (i + 0.8) * this.lineHeight, this.font, this.color);
    }
  }
};

// src/Button.ts
var Button = class extends UIComponent {
  label;
  bg;
  hoverBg;
  color;
  font;
  radius;
  focused = false;
  hovered = false;
  textWidth;
  constructor(label, opts = {}) {
    super();
    this.label = label;
    this.bg = opts.bg ?? "#2563eb";
    this.hoverBg = opts.hoverBg ?? "#3b82f6";
    this.color = opts.color ?? "#ffffff";
    this.font = opts.font ?? "600 16px sans-serif";
    this.padding = opts.padding ?? 12;
    this.radius = opts.radius ?? 8;
    this.interactive = true;
    this.textWidth = measureText(this.label, this.font);
    this.width = opts.width ?? this.textWidth + this.padding * 2;
    this.height = opts.height ?? fontSizePx(this.font) + this.padding * 2;
    this.on("hover", () => {
      if (this.hovered) return;
      this.hovered = true;
      this.scene?.markDirty();
    });
    this.on("pointerleave", () => {
      if (!this.hovered) return;
      this.hovered = false;
      this.scene?.markDirty();
    });
    if (opts.onClick) this.on("click", opts.onClick);
  }
  getA11yAttributes() {
    return { tag: "button", role: "button", label: this.label };
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.fill(this.hovered ? this.hoverBg : this.bg);
    if (this.focused) {
      r.stroke("#00f0ff", 2);
    }
    const textX = (this.width - this.textWidth) / 2;
    r.fillText(this.label, textX, this.height * 0.66, this.font, this.color);
  }
};

// src/Link.ts
var import_core4 = require("@vectojs/core");
var Link = class extends UIComponent {
  label;
  href;
  color;
  font;
  underline;
  constructor(label, opts) {
    super();
    this.label = label;
    this.href = opts.href;
    this.color = opts.color ?? "#38bdf8";
    this.font = opts.font ?? "16px sans-serif";
    this.underline = opts.underline ?? true;
    this.interactive = true;
    this.width = measureText(this.label, this.font);
    this.height = fontSizePx(this.font);
    this.on("click", () => {
      const safe = (0, import_core4.sanitizeUrl)(this.href);
      if (safe && safe !== "#" && typeof window !== "undefined") {
        window.open(safe, "_blank", "noopener");
      }
    });
  }
  getA11yAttributes() {
    return { tag: "a", href: (0, import_core4.sanitizeUrl)(this.href), label: this.label, target: "_blank" };
  }
  render(r) {
    const baseline = this.height * 0.8;
    r.fillText(this.label, 0, baseline, this.font, this.color);
    if (this.underline) {
      r.beginPath();
      r.moveTo(0, baseline + 2);
      r.lineTo(this.width, baseline + 2);
      r.stroke(this.color, 1);
    }
  }
};

// src/Image.ts
var Image = class extends UIComponent {
  src;
  alt;
  placeholder;
  radius;
  bitmap = null;
  loaded = false;
  constructor(src, opts) {
    super();
    this.src = src;
    this.alt = opts.alt ?? "";
    this.placeholder = opts.placeholder ?? "#1e293b";
    this.radius = opts.radius ?? 0;
    this.width = opts.width;
    this.height = opts.height;
    this.interactive = true;
    if (typeof globalThis.Image !== "undefined") {
      const bmp = new globalThis.Image();
      bmp.onload = () => {
        this.loaded = true;
        opts.onLoad?.();
      };
      bmp.src = src;
      this.bitmap = bmp;
    }
  }
  getA11yAttributes() {
    return { tag: "img", src: this.src, alt: this.alt, label: this.alt };
  }
  render(r) {
    if (this.loaded && this.bitmap) {
      r.drawImage(this.bitmap, 0, 0, this.width, this.height);
      return;
    }
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.fill(this.placeholder);
  }
};

// src/Card.ts
var Card = class extends UIComponent {
  bg;
  border;
  borderWidth;
  radius;
  label;
  _content = null;
  _fitWidth = false;
  _fitHeight = false;
  constructor(opts) {
    super();
    this.width = opts.width;
    this.height = opts.height;
    this.bg = opts.bg ?? "#0f172a";
    this.border = opts.border ?? null;
    this.borderWidth = opts.borderWidth ?? 1;
    this.radius = opts.radius ?? 12;
    this.padding = opts.padding ?? 0;
    this.label = opts.label ?? null;
    if (this.label) this.interactive = true;
    if (opts.onClick) {
      if (!this.label) {
        throw new Error(
          "Card: onClick requires a label (the a11y projection needs an accessible name for the interactive region it creates)."
        );
      }
      this.on("click", opts.onClick);
    }
  }
  getA11yAttributes() {
    return this.label ? { role: "group", label: this.label } : {};
  }
  /**
   * Place a single content entity inside the card, sized to match by
   * default. Same `fitContent` contract as {@link import('./ResizablePanel').Panel.setContent}
   * — pass `false` (or a per-axis object) to keep the old position-only
   * `add()` behavior instead. Unlike `Panel`, a `Card` is not required to
   * host content this way; use plain {@link add} for cards whose children
   * are manually positioned decorations rather than a single sized region.
   */
  setContent(content, fit = true) {
    if (this._content) super.remove(this._content);
    this._content = content;
    content.x = 0;
    content.y = 0;
    if (fit === false) {
      this._fitWidth = false;
      this._fitHeight = false;
    } else if (fit === true) {
      this._fitWidth = true;
      this._fitHeight = true;
    } else {
      this._fitWidth = fit.width ?? true;
      this._fitHeight = fit.height ?? true;
    }
    this._applyFit();
    super.add(content);
    return this;
  }
  _applyFit() {
    if (!this._content) return;
    if (this._fitWidth) this._content.width = this.width;
    if (this._fitHeight) this._content.height = this.height;
  }
  update(dt, time) {
    super.update(dt, time);
    if (this._content) this._applyFit();
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.fill(this.bg);
    if (this.border) {
      r.beginPath();
      r.roundRect(0, 0, this.width, this.height, this.radius);
      r.stroke(this.border, this.borderWidth);
    }
  }
};

// src/Stack.ts
var Stack = class extends UIComponent {
  direction;
  gap;
  align;
  wrap;
  maxWidth;
  maxHeight;
  constructor(opts = {}) {
    super();
    this.direction = opts.direction ?? "vertical";
    this.gap = opts.gap ?? 0;
    this.align = opts.align ?? "start";
    this.wrap = opts.wrap ?? false;
    this.maxWidth = opts.maxWidth ?? Infinity;
    this.maxHeight = opts.maxHeight ?? Infinity;
  }
  /** Add a child and re-run layout. */
  add(child) {
    super.add(child);
    this.layout();
    return this;
  }
  /**
   * Position all children along the main axis and align them on the cross axis,
   * then size this container to fit.
   */
  layout() {
    const vertical = this.direction === "vertical";
    const limit = vertical ? this.maxHeight : this.maxWidth;
    const lines = [];
    let currentLine = [];
    let currentMain = 0;
    for (const c of this.children) {
      const childMain = vertical ? c.height : c.width;
      if (this.wrap && currentLine.length > 0 && currentMain + this.gap + childMain > limit) {
        lines.push(currentLine);
        currentLine = [c];
        currentMain = childMain;
      } else {
        currentLine.push(c);
        currentMain += currentLine.length > 1 ? this.gap + childMain : childMain;
      }
    }
    if (currentLine.length > 0) lines.push(currentLine);
    let totalCross = 0;
    let maxTotalMain = 0;
    for (const line of lines) {
      let lineCross = 0;
      let lineMain = 0;
      for (const c of line) {
        lineCross = Math.max(lineCross, vertical ? c.width : c.height);
        lineMain += vertical ? c.height : c.width;
      }
      lineMain += (line.length - 1) * this.gap;
      maxTotalMain = Math.max(maxTotalMain, lineMain);
      let currentMain2 = 0;
      for (const c of line) {
        const childCross = vertical ? c.width : c.height;
        let offset = totalCross;
        if (this.align === "center") offset += (lineCross - childCross) / 2;
        else if (this.align === "end") offset += lineCross - childCross;
        if (vertical) {
          c.x = offset;
          c.y = currentMain2;
          currentMain2 += c.height + this.gap;
        } else {
          c.x = currentMain2;
          c.y = offset;
          currentMain2 += c.width + this.gap;
        }
      }
      totalCross += lineCross + this.gap;
    }
    if (lines.length > 0) totalCross -= this.gap;
    this.width = vertical ? totalCross : maxTotalMain;
    this.height = vertical ? maxTotalMain : totalCross;
  }
  /** Structural container — draws nothing itself. */
  render(_r) {
  }
};

// src/Flow.ts
var Flow = class extends Stack {
  constructor(opts = {}) {
    super({
      ...opts,
      direction: opts.direction ?? "horizontal",
      wrap: true
    });
  }
};

// src/Input.ts
var import_core5 = require("@vectojs/core");
var Input = class extends UIComponent {
  value;
  placeholder;
  font;
  color;
  placeholderColor;
  bg;
  border;
  selectionColor;
  radius;
  /** Caret / selection anchor & focus offsets, mirrored from the real input. */
  selectionStart;
  selectionEnd;
  /** Active IME pre-edit range `[start, start+length)`, or `null`. */
  composition = null;
  /** Whether the shadow input currently holds focus (drives caret blink). */
  focused = false;
  /** Horizontal scroll offset so the caret stays in view (scroll-to-caret). */
  scrollLeft = 0;
  constructor(opts) {
    super();
    this.width = opts.width;
    this.height = opts.height ?? 40;
    this.value = opts.value ?? "";
    this.placeholder = opts.placeholder ?? "";
    this.font = opts.font ?? "16px sans-serif";
    this.color = opts.color ?? "#e2e8f0";
    this.placeholderColor = opts.placeholderColor ?? "#64748b";
    this.bg = opts.bg ?? "#0f172a";
    this.border = opts.border ?? "#334155";
    this.selectionColor = opts.selectionColor ?? "rgba(56, 189, 248, 0.35)";
    this.radius = opts.radius ?? 6;
    this.padding = opts.padding ?? 10;
    this.selectionStart = this.value.length;
    this.selectionEnd = this.value.length;
    this.interactive = true;
    this.on(
      "change",
      (e) => {
        this.value = e.value;
        this.selectionStart = e.selectionStart ?? this.value.length;
        this.selectionEnd = e.selectionEnd ?? this.value.length;
        this.composition = e.composition ?? null;
        opts.onChange?.(this.value);
      }
    );
    this.on("focus", () => {
      this.focused = true;
      this.startCaretBlinkWake();
    });
    this.on("blur", () => {
      this.focused = false;
      this.stopCaretBlinkWake();
    });
  }
  getA11yAttributes() {
    return {
      tag: "input",
      inputType: "text",
      placeholder: this.placeholder,
      value: this.value,
      label: this.placeholder,
      textInputStyle: {
        font: this.font,
        lineHeight: this.height - 2 * this.padding,
        padding: this.padding
      }
    };
  }
  cachedValue = "";
  cachedFont = "";
  cachedLayout = null;
  _rtlCacheValue = null;
  _rtlCacheResult = false;
  /**
   * Whether `value` contains any RTL-script character. Cached per distinct
   * `value` — otherwise this O(n) scan re-runs from scratch on every
   * `charOffset()` call, and a single render/caret-blink tick calls it several
   * times (caret, selection start, selection end, composition bounds).
   */
  hasRTL() {
    if (this._rtlCacheValue === this.value) return this._rtlCacheResult;
    let result = false;
    for (let j = 0; j < this.value.length; j++) {
      const code = this.value.charCodeAt(j);
      if (code >= 1424 && code <= 1535 || code >= 1536 && code <= 1791 || code >= 64336 && code <= 65279) {
        result = true;
        break;
      }
    }
    this._rtlCacheValue = this.value;
    this._rtlCacheResult = result;
    return result;
  }
  getLayout() {
    if (this.value === this.cachedValue && this.font === this.cachedFont && this.cachedLayout) {
      return this.cachedLayout;
    }
    const fSize = fontSizePx(this.font);
    const engine = new import_core5.LayoutEngine(1e6, 1e3, {
      measure: (char) => measureText(char, this.font)
    });
    engine.preserveLeadingSpaces = true;
    const layout = engine.layoutText(this.value, {}, fSize);
    this.cachedValue = this.value;
    this.cachedFont = this.font;
    this.cachedLayout = layout;
    return layout;
  }
  /** Measured x of the glyph boundary before char index `i` (text-relative). */
  charOffset(i) {
    if (this.value.length === 0) return 0;
    if (!this.hasRTL()) {
      return measureText(this.value.slice(0, i), this.font);
    }
    const layout = this.getLayout();
    if (layout.nodes.length === 0) return 0;
    let targetNode = null;
    let isRTL = false;
    for (const node of layout.nodes) {
      const start2 = node.sourceIndex ?? 0;
      const len2 = node.sourceLength ?? 0;
      if (i >= start2 && i <= start2 + len2) {
        targetNode = node;
        isRTL = !!node.isRTL;
        if (i > start2 && i < start2 + len2) {
          break;
        }
      }
    }
    if (!targetNode) {
      let maxNode = layout.nodes[0];
      for (const node of layout.nodes) {
        if ((node.sourceIndex ?? 0) + (node.sourceLength ?? 0) > (maxNode.sourceIndex ?? 0) + (maxNode.sourceLength ?? 0)) {
          maxNode = node;
        }
      }
      targetNode = maxNode;
      isRTL = !!maxNode.isRTL;
    }
    const start = targetNode.sourceIndex ?? 0;
    const len = targetNode.sourceLength ?? 0;
    const fraction = len > 0 ? (i - start) / len : 0;
    if (isRTL) {
      return targetNode.x + targetNode.width * (1 - fraction);
    } else {
      return targetNode.x + targetNode.width * fraction;
    }
  }
  /** Screen-space x of the caret (after scroll), in the component's box coords. */
  caretScreenX() {
    return this.padding - this.scrollLeft + this.charOffset(this.selectionStart);
  }
  /** Keep the caret within the padded inner box by adjusting `scrollLeft`. */
  updateScroll() {
    const innerWidth = this.width - 2 * this.padding;
    const caretAbs = this.charOffset(this.selectionStart);
    if (caretAbs - this.scrollLeft > innerWidth) this.scrollLeft = caretAbs - innerWidth;
    if (caretAbs - this.scrollLeft < 0) this.scrollLeft = caretAbs;
    if (this.scrollLeft < 0) this.scrollLeft = 0;
  }
  /** Caret blink "on" phase (500ms cadence). */
  caretOn() {
    return Math.floor(Date.now() / 500) % 2 === 0;
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.fill(this.bg);
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.stroke(this.border, 1);
    this.updateScroll();
    const innerWidth = this.width - 2 * this.padding;
    const editorLineHeight = this.height - 2 * this.padding;
    const baseline = this.padding + (0, import_core5.cssLineBoxBaseline)(this.font, editorLineHeight);
    const textOriginX = this.padding - this.scrollLeft;
    r.save();
    r.clip(this.padding, this.padding, innerWidth, editorLineHeight);
    if (this.selectionStart !== this.selectionEnd) {
      const a = Math.min(this.selectionStart, this.selectionEnd);
      const b = Math.max(this.selectionStart, this.selectionEnd);
      if (!this.hasRTL()) {
        const sx = textOriginX + this.charOffset(a);
        const ex = textOriginX + this.charOffset(b);
        r.beginPath();
        r.roundRect(sx, this.padding, ex - sx, editorLineHeight, 0);
        r.fill(this.selectionColor);
      } else {
        const layout = this.getLayout();
        r.beginPath();
        for (const node of layout.nodes) {
          const start = node.sourceIndex ?? 0;
          const len = node.sourceLength ?? 0;
          if (!(start + len <= a || start >= b)) {
            r.roundRect(textOriginX + node.x, this.padding, node.width, editorLineHeight, 0);
          }
        }
        r.fill(this.selectionColor);
      }
    }
    const text = this.value || this.placeholder;
    const color = this.value ? this.color : this.placeholderColor;
    r.fillText(text, textOriginX, baseline, this.font, color);
    if (this.composition && this.composition.length > 0) {
      const ux0 = textOriginX + this.charOffset(this.composition.start);
      const ux1 = textOriginX + this.charOffset(this.composition.start + this.composition.length);
      const uy = baseline + 2;
      r.beginPath();
      r.moveTo(ux0, uy);
      r.lineTo(ux1, uy);
      r.stroke(this.color, 1);
    }
    if (this.focused && this.caretOn()) {
      const cx = this.caretScreenX();
      r.beginPath();
      r.moveTo(cx, this.padding + 2);
      r.lineTo(cx, this.padding + editorLineHeight - 2);
      r.stroke(this.color, 1);
    }
    r.restore();
  }
};

// src/TextArea.ts
var import_core6 = require("@vectojs/core");
function wrapText(value, maxWidth, measure) {
  const lines = [];
  let paraStart = 0;
  const paragraphs = [];
  for (let i = 0; i < value.length; i++) {
    if (value[i] === "\n") {
      paragraphs.push({ start: paraStart, text: value.slice(paraStart, i) });
      paraStart = i + 1;
    }
  }
  paragraphs.push({ start: paraStart, text: value.slice(paraStart) });
  for (const para of paragraphs) {
    if (para.text === "") {
      lines.push({ text: "", start: para.start, end: para.start });
      continue;
    }
    softWrap(para.text, para.start, maxWidth, measure, lines);
  }
  return lines;
}
function softWrap(text, base, maxWidth, measure, lines) {
  let lineText = "";
  let lineStart = base;
  let lineEnd = base;
  let started = false;
  const startWith = (word, absStart) => {
    if (measure(word) <= maxWidth || word.length <= 1) {
      lineText = word;
      lineStart = absStart;
      lineEnd = absStart + word.length;
      return;
    }
    let chunkStart = absStart;
    let chunk = "";
    for (let k = 0; k < word.length; k++) {
      const c = word[k];
      if (chunk !== "" && measure(chunk + c) > maxWidth) {
        lines.push({ text: chunk, start: chunkStart, end: chunkStart + chunk.length });
        chunkStart += chunk.length;
        chunk = c;
      } else {
        chunk += c;
      }
    }
    lineText = chunk;
    lineStart = chunkStart;
    lineEnd = chunkStart + chunk.length;
  };
  let i = 0;
  while (i < text.length) {
    const ws = i;
    while (i < text.length && text[i] !== " ") i++;
    const word = text.slice(ws, i);
    while (i < text.length && text[i] === " ") i++;
    if (word === "") continue;
    if (!started) {
      startWith(word, base + ws);
      started = true;
    } else if (measure(`${lineText} ${word}`) <= maxWidth) {
      lineText = `${lineText} ${word}`;
      lineEnd = base + ws + word.length;
    } else {
      lines.push({ text: lineText, start: lineStart, end: lineEnd });
      startWith(word, base + ws);
    }
  }
  if (started) {
    lines.push({ text: lineText, start: lineStart, end: lineEnd });
  } else {
    lines.push({ text: "", start: base, end: base });
  }
}
var TextArea = class extends UIComponent {
  value;
  placeholder;
  font;
  lineHeightFactor;
  color;
  placeholderColor;
  bg;
  border;
  selectionColor;
  radius;
  /** Caret / selection anchor & focus offsets, mirrored from the real textarea. */
  selectionStart;
  selectionEnd;
  /** Active IME pre-edit range `[start, start+length)`, or `null`. */
  composition = null;
  /** Whether the shadow textarea currently holds focus (drives caret blink). */
  focused = false;
  /** Vertical scroll offset so the caret line stays in view. */
  scrollTop = 0;
  constructor(opts) {
    super();
    this.width = opts.width;
    this.height = opts.height ?? 120;
    this.value = opts.value ?? "";
    this.placeholder = opts.placeholder ?? "";
    this.font = opts.font ?? "16px sans-serif";
    this.lineHeightFactor = opts.lineHeight ?? 1.4;
    this.color = opts.color ?? "#e2e8f0";
    this.placeholderColor = opts.placeholderColor ?? "#64748b";
    this.bg = opts.bg ?? "#0f172a";
    this.border = opts.border ?? "#334155";
    this.selectionColor = opts.selectionColor ?? "rgba(56, 189, 248, 0.35)";
    this.radius = opts.radius ?? 6;
    this.padding = opts.padding ?? 10;
    this.selectionStart = this.value.length;
    this.selectionEnd = this.value.length;
    this.interactive = true;
    this.on(
      "change",
      (e) => {
        this.value = e.value;
        this.selectionStart = e.selectionStart ?? this.value.length;
        this.selectionEnd = e.selectionEnd ?? this.value.length;
        this.composition = e.composition ?? null;
        opts.onChange?.(this.value);
      }
    );
    this.on("focus", () => {
      this.focused = true;
      this.startCaretBlinkWake();
    });
    this.on("blur", () => {
      this.focused = false;
      this.stopCaretBlinkWake();
    });
  }
  getA11yAttributes() {
    return {
      tag: "textarea",
      placeholder: this.placeholder,
      value: this.value,
      label: this.placeholder,
      textInputStyle: {
        font: this.font,
        lineHeight: this.lineHeight(),
        padding: this.padding
      }
    };
  }
  innerWidth() {
    return this.width - 2 * this.padding;
  }
  lineHeight() {
    return fontSizePx(this.font) * this.lineHeightFactor;
  }
  cachedValue = "";
  cachedFont = "";
  cachedWidth = 0;
  cachedLines = [];
  /** Wrap the current value to the inner box width. */
  computeLines() {
    const innerW = this.innerWidth();
    if (this.value === this.cachedValue && this.font === this.cachedFont && innerW === this.cachedWidth && this.cachedLines.length > 0) {
      return this.cachedLines;
    }
    const lines = [];
    const paragraphs = this.value.split("\n");
    let pStart = 0;
    const fSize = fontSizePx(this.font);
    const engine = new import_core6.LayoutEngine(innerW, 1e6, {
      measure: (char) => measureText(char, this.font)
    });
    engine.preserveLeadingSpaces = true;
    for (let pIdx = 0; pIdx < paragraphs.length; pIdx++) {
      const paragraphText = paragraphs[pIdx];
      if (paragraphText === "") {
        lines.push({ text: "", start: pStart, end: pStart });
        pStart += 1;
        continue;
      }
      const layout = engine.layoutText(paragraphText, {}, fSize);
      const lineGroups = [];
      let currentGroup = [];
      let currentY = -1;
      for (const node of layout.nodes) {
        if (currentY === -1) {
          currentY = node.y;
        }
        if (Math.abs(node.y - currentY) > 0.1 * fSize) {
          lineGroups.push(currentGroup);
          currentGroup = [node];
          currentY = node.y;
        } else {
          currentGroup.push(node);
        }
      }
      if (currentGroup.length > 0) {
        lineGroups.push(currentGroup);
      }
      if (lineGroups.length === 0) {
        lines.push({ text: "", start: pStart, end: pStart + paragraphText.length });
      } else {
        for (const group of lineGroups) {
          group.sort((a, b) => a.x - b.x);
          const text = group.map((n) => n.char).join("");
          let start = paragraphText.length;
          let end = 0;
          for (const node of group) {
            const nodeStart = node.sourceIndex ?? 0;
            const nodeLen = node.sourceLength ?? 0;
            if (nodeStart < start) start = nodeStart;
            if (nodeStart + nodeLen > end) end = nodeStart + nodeLen;
          }
          lines.push({
            text,
            start: pStart + start,
            end: pStart + end,
            nodes: group
          });
        }
      }
      pStart += paragraphText.length + 1;
    }
    this.cachedValue = this.value;
    this.cachedFont = this.font;
    this.cachedWidth = innerW;
    this.cachedLines = lines;
    return lines;
  }
  /**
   * The visual line index containing caret `offset` (0-based). Boundary offsets
   * resolve to the earliest containing line; out-of-range clamps to the last.
   *
   * @param offset - A linear character offset into the value.
   * @returns The wrapped-line index.
   */
  lineOfOffset(offset) {
    const lines = this.computeLines();
    for (let i = 0; i < lines.length; i++) {
      if (offset >= lines[i].start && offset <= lines[i].end) return i;
    }
    return lines.length - 1;
  }
  /** X (text-relative) of `offset` within its line. */
  offsetX(line, offset) {
    if (!line.nodes || line.nodes.length === 0) return 0;
    const pStart = line.start - (line.nodes[0].sourceIndex ?? 0);
    let targetNode = null;
    let isRTL = false;
    for (const node of line.nodes) {
      const nodeAbsStart2 = pStart + (node.sourceIndex ?? 0);
      const len2 = node.sourceLength ?? 0;
      if (offset >= nodeAbsStart2 && offset <= nodeAbsStart2 + len2) {
        targetNode = node;
        isRTL = !!node.isRTL;
        if (offset > nodeAbsStart2 && offset < nodeAbsStart2 + len2) {
          break;
        }
      }
    }
    if (!targetNode) {
      let maxNode = line.nodes[0];
      for (const node of line.nodes) {
        const nodeAbsStart2 = pStart + (node.sourceIndex ?? 0);
        const maxAbsStart = pStart + (maxNode.sourceIndex ?? 0);
        if (nodeAbsStart2 + (node.sourceLength ?? 0) > maxAbsStart + (maxNode.sourceLength ?? 0)) {
          maxNode = node;
        }
      }
      targetNode = maxNode;
      isRTL = !!maxNode.isRTL;
    }
    const nodeAbsStart = pStart + (targetNode.sourceIndex ?? 0);
    const len = targetNode.sourceLength ?? 0;
    const fraction = len > 0 ? (offset - nodeAbsStart) / len : 0;
    if (isRTL) {
      return targetNode.x + targetNode.width * (1 - fraction);
    } else {
      return targetNode.x + targetNode.width * fraction;
    }
  }
  /** Keep the caret line within the padded box by adjusting `scrollTop`. */
  updateScroll(caretLine, lh, innerH) {
    const caretY = caretLine * lh;
    if (caretY - this.scrollTop > innerH - lh) this.scrollTop = caretY - (innerH - lh);
    if (caretY - this.scrollTop < 0) this.scrollTop = caretY;
    if (this.scrollTop < 0) this.scrollTop = 0;
  }
  caretOn() {
    return Math.floor(Date.now() / 500) % 2 === 0;
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.fill(this.bg);
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.stroke(this.border, 1);
    const lines = this.computeLines();
    const lh = this.lineHeight();
    const innerW = this.innerWidth();
    const innerH = this.height - 2 * this.padding;
    const caretLine = this.lineOfOffset(this.selectionStart);
    this.updateScroll(caretLine, lh, innerH);
    const originX = this.padding;
    const originY = this.padding - this.scrollTop;
    const baselineOffset = (0, import_core6.cssLineBoxBaseline)(this.font, lh);
    r.save();
    r.clip(this.padding, this.padding, innerW, innerH);
    if (!this.value && this.placeholder) {
      r.fillText(
        this.placeholder,
        originX,
        originY + baselineOffset,
        this.font,
        this.placeholderColor
      );
    }
    if (this.selectionStart !== this.selectionEnd) {
      const a = Math.min(this.selectionStart, this.selectionEnd);
      const b = Math.max(this.selectionStart, this.selectionEnd);
      for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        const lo = Math.max(a, line.start);
        const hi = Math.min(b, line.end);
        if (hi <= lo) continue;
        const y = originY + i * lh;
        if (!line.nodes || line.nodes.length === 0) {
          const sx = originX + this.offsetX(line, lo);
          const ex = originX + this.offsetX(line, hi);
          r.beginPath();
          r.roundRect(sx, y, Math.max(1, ex - sx), lh, 0);
          r.fill(this.selectionColor);
        } else {
          r.beginPath();
          const pStart = line.start - (line.nodes[0].sourceIndex ?? 0);
          for (const node of line.nodes) {
            const start = pStart + (node.sourceIndex ?? 0);
            const len = node.sourceLength ?? 0;
            if (!(start + len <= lo || start >= hi)) {
              r.roundRect(originX + node.x, y, node.width, lh, 0);
            }
          }
          r.fill(this.selectionColor);
        }
      }
    }
    if (this.value) {
      for (let i = 0; i < lines.length; i++) {
        r.fillText(
          lines[i].text,
          originX,
          originY + i * lh + baselineOffset,
          this.font,
          this.color
        );
      }
    }
    if (this.focused && this.caretOn()) {
      const line = lines[caretLine];
      const cx = originX + this.offsetX(line, this.selectionStart);
      const cy = originY + caretLine * lh;
      r.beginPath();
      r.moveTo(cx, cy + 2);
      r.lineTo(cx, cy + lh - 2);
      r.stroke(this.color, 1);
    }
    r.restore();
  }
};

// src/RichText.ts
var import_core7 = require("@vectojs/core");
var LinkHotspot = class extends UIComponent {
  href;
  constructor(href, onClick) {
    super();
    this.href = href;
    this.interactive = true;
    this.on("click", () => {
      const safe = (0, import_core7.sanitizeUrl)(this.href);
      if (safe && safe !== "#") onClick?.(safe);
    });
  }
  getA11yAttributes() {
    return { tag: "a", href: (0, import_core7.sanitizeUrl)(this.href), label: this.href, target: "_blank" };
  }
  render() {
  }
};
function familyOf(font) {
  const pxIndex = font.indexOf("px");
  if (pxIndex < 0) return font.trim() || "sans-serif";
  let rest = font.slice(pxIndex + 2).trimStart();
  if (rest.startsWith("/")) {
    let i = 1;
    while (i < rest.length && rest[i] !== " " && rest[i] !== "	") i++;
    rest = rest.slice(i).trimStart();
  }
  return rest || "sans-serif";
}
function baseMeasurer(font) {
  if (typeof document === "undefined") return null;
  const ctx = document.createElement("canvas").getContext("2d");
  if (!ctx) return null;
  const cache = /* @__PURE__ */ new Map();
  return {
    measure(char, fontSize) {
      const key = `${fontSize} ${char}`;
      let w = cache.get(key);
      if (w === void 0) {
        ctx.font = `${fontSize}px ${familyOf(font)}`;
        w = ctx.measureText(char).width;
        cache.set(key, w);
      }
      return w;
    }
  };
}
var RichText = class extends UIComponent {
  spans;
  font;
  color;
  maxWidth;
  selectable;
  linkColor;
  exclusions;
  engine;
  baseFontSize;
  baseStyle;
  result;
  onLinkClick;
  /** One transparent `<a>` hotspot child per link run (kept in sync with layout). */
  hotspots = [];
  constructor(spans, opts = {}) {
    super();
    this.spans = spans;
    this.font = opts.font ?? "16px sans-serif";
    this.color = opts.color ?? "#e2e8f0";
    this.maxWidth = opts.maxWidth;
    this.selectable = opts.selectable ?? true;
    this.baseStyle = opts.baseStyle;
    this.linkColor = opts.linkColor ?? "#38bdf8";
    this.onLinkClick = opts.onLinkClick;
    this.exclusions = opts.exclusions;
    this.baseFontSize = fontSizePx(this.font);
    this.engine = new import_core7.LayoutEngine(this.maxWidth ?? 1e9, 1e9, baseMeasurer(this.font));
    this.interactive = false;
    this.result = this.layout();
  }
  /** Replace the styled runs and re-lay out. */
  setSpans(spans) {
    this.spans = spans;
    this.result = this.layout();
    this.scene?.markDirty();
    return this;
  }
  /** Change the wrap width and re-lay out. */
  setMaxWidth(maxWidth) {
    this.maxWidth = maxWidth;
    this.engine.maxWidth = maxWidth;
    this.result = this.layout();
    this.scene?.markDirty();
    return this;
  }
  /** Set the rect regions the text flows around (exclusion shapes) and re-lay out. */
  setExclusions(exclusions) {
    this.exclusions = exclusions;
    this.result = this.layout();
    this.scene?.markDirty();
    return this;
  }
  /**
   * Append styled runs and re-lay out — the streaming / typewriter path. The
   * engine's rich paragraph memo reuses every untouched leading paragraph, so a
   * token-by-token stream re-prepares in O(changed paragraph), not O(document).
   */
  appendSpans(spans) {
    this.spans = [...this.spans, ...spans];
    this.result = this.layout();
    this.scene?.markDirty();
    return this;
  }
  /** Enable or disable browser-native drag selection without rebuilding the entity. */
  setSelectable(selectable) {
    this.selectable = selectable;
    this.scene?.markDirty();
    return this;
  }
  layout() {
    const prepared = this.engine.prepareRich(this.spans, {}, this.baseFontSize, this.baseStyle);
    const result = this.engine.layoutPrepared(prepared, void 0, this.exclusions);
    this.width = result.totalWidth;
    this.height = result.totalHeight;
    this.result = result;
    this.syncHotspots();
    return result;
  }
  /** One box per contiguous link run, sized to its first wrapped line. */
  computeLinks() {
    const out = [];
    const nodes = this.result.nodes;
    let i = 0;
    while (i < nodes.length) {
      const href = nodes[i].style?.href;
      if (!href) {
        i++;
        continue;
      }
      let j = i;
      while (j < nodes.length && nodes[j].style?.href === href) j++;
      const run = nodes.slice(i, j);
      const y0 = Math.min(...run.map((n) => n.y));
      const firstLine = run.filter((n) => n.y === y0);
      const x = Math.min(...firstLine.map((n) => n.x));
      const right = Math.max(...firstLine.map((n) => n.x + n.width));
      const height = Math.max(...firstLine.map((n) => n.height));
      out.push({ href, x, y: y0, width: right - x, height });
      i = j;
    }
    return out;
  }
  /**
   * Reconcile the `<a>` hotspot children with the current link runs. Stable
   * across re-wrap (one hotspot per run), so positions update in place; only a
   * change in link *count* rebuilds (pruning old shadow nodes via the scene).
   */
  syncHotspots() {
    const links = this.computeLinks();
    if (links.length !== this.hotspots.length) {
      for (const old of this.hotspots) {
        this.remove(old);
        this.scene?.detachA11y(old);
      }
      this.hotspots = links.map((l) => {
        const h = new LinkHotspot(l.href, this.onLinkClick);
        this.add(h);
        return h;
      });
    }
    for (let k = 0; k < links.length; k++) {
      const l = links[k];
      const h = this.hotspots[k];
      h.href = l.href;
      h.setPosition(l.x, l.y);
      h.width = l.width;
      h.height = l.height;
    }
  }
  /** The full text content, used as the accessible name. */
  fullText() {
    return this.spans.map((s) => s.text).join("");
  }
  /** Rebuild styled DOM runs from a logical UTF-16 source interval. */
  logicalRuns(start, end) {
    const runs = [];
    let offset = 0;
    for (const span of this.spans) {
      const spanEnd = offset + span.text.length;
      const from = Math.max(start, offset);
      const to = Math.min(end, spanEnd);
      if (from < to) {
        const style = span.style || this.baseStyle ? { ...this.baseStyle, ...span.style } : void 0;
        const font = this.nodeFont(style, style?.fontSize ?? this.baseFontSize);
        const text = span.text.slice(from - offset, to - offset);
        const previous = runs.at(-1);
        if (previous?.font === font) previous.text += text;
        else runs.push({ text, font });
      }
      offset = spanEnd;
    }
    return runs;
  }
  /**
   * Group the laid-out glyphs into the same visual lines the canvas draws.
   * Each line keeps its real local origin and run fonts, so the semantic DOM
   * never has to re-flow mixed-size markdown differently from the canvas.
   */
  visualLineGroups() {
    const groups = [];
    for (const node of this.result.nodes) {
      const previous = groups.at(-1);
      const baseline = node.y + node.height * 0.8;
      const previousBaseline = previous ? previous[0].y + previous[0].height * 0.8 : Number.NaN;
      if (!previous || Math.abs(previousBaseline - baseline) > 0.01) groups.push([node]);
      else previous.push(node);
    }
    const source = this.fullText();
    return groups.map((nodes, index) => {
      const largest = Math.max(...nodes.map((node) => node.height));
      const font = this.nodeFont(void 0, largest);
      let sourceStart = Math.min(...nodes.map((node) => node.sourceIndex ?? 0));
      const sourceEnd = Math.max(
        ...nodes.map((node) => (node.sourceIndex ?? 0) + (node.sourceLength ?? 0))
      );
      if (index === 0) sourceStart = 0;
      const nextNodes = groups[index + 1];
      const nextStart = nextNodes ? Math.min(...nextNodes.map((node) => node.sourceIndex ?? sourceEnd)) : source.length;
      const runs = this.logicalRuns(sourceStart, sourceEnd);
      const y = Math.min(...nodes.map((node) => node.y));
      const baseline = nodes[0].y + nodes[0].height * 0.8 - y;
      return {
        nodes,
        projection: {
          text: source.slice(sourceStart, sourceEnd),
          separatorAfter: source.slice(sourceEnd, Math.max(sourceEnd, nextStart)),
          x: Math.min(...nodes.map((node) => node.x)),
          y,
          baseline,
          font,
          // LayoutEngine advances a paragraph by its largest run, not the
          // component default. Keep the native selection box on that same
          // rhythm when a heading or inline large text shares a line.
          lineHeight: largest * 1.5,
          runs
        }
      };
    });
  }
  projectedLines() {
    return this.visualLineGroups().map(({ projection }) => projection);
  }
  /** Build the CSS font shorthand for a node's style. */
  nodeFont(style, size) {
    const italic = style?.italic ? "italic " : "";
    const bold = style?.bold ? "bold " : "";
    return `${italic}${bold}${size}px ${familyOf(this.font)}`;
  }
  getA11yAttributes() {
    return { label: this.fullText() };
  }
  /** Mirror the concatenated span text into the DOM content layer. */
  getContentProjection() {
    const text = this.fullText();
    if (!text) return null;
    return {
      text,
      font: this.font,
      lineHeight: this.baseFontSize * 1.5,
      lines: this.projectedLines(),
      selectable: this.selectable
    };
  }
  render(r) {
    for (const { nodes, projection: line } of this.visualLineGroups()) {
      for (const node of nodes) {
        if (node.char.trim().length === 0) continue;
        const size = node.height;
        const font = this.nodeFont(node.style, size);
        const isLink = !!node.style?.href;
        const color = node.style?.color ?? (isLink ? this.linkColor : this.color);
        const baseline = line.y + line.baseline;
        r.fillText(node.char, node.x, baseline, font, color);
        if (isLink) {
          const uy = baseline + 2;
          r.beginPath();
          r.moveTo(node.x, uy);
          r.lineTo(node.x + node.width, uy);
          r.stroke(color, 1);
        }
      }
    }
  }
};

// src/Checkbox.ts
var Checkbox = class extends UIComponent {
  checked;
  label;
  size;
  font;
  color;
  accent;
  border;
  constructor(opts) {
    super();
    this.checked = opts.checked ?? false;
    this.label = opts.label ?? "";
    this.size = opts.size ?? 20;
    this.font = opts.font ?? "16px sans-serif";
    this.color = opts.color ?? "#e2e8f0";
    this.accent = opts.accent ?? "#2563eb";
    this.border = opts.border ?? "#475569";
    this.interactive = true;
    this.height = this.size;
    this.width = this.size + (this.label ? 8 + measureText(this.label, this.font) : 0);
    this.on("click", () => this.emit("change", { checked: !this.checked }));
    this.on("change", (e) => {
      if (e.checked === this.checked) return;
      this.checked = e.checked;
      opts.onChange?.(this.checked);
      this.scene?.markDirty();
    });
  }
  getA11yAttributes() {
    return { tag: "input", inputType: "checkbox", checked: this.checked, label: this.label };
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.size, this.size, 4);
    if (this.checked) {
      r.fill(this.accent);
      r.beginPath();
      r.moveTo(this.size * 0.25, this.size * 0.5);
      r.lineTo(this.size * 0.45, this.size * 0.7);
      r.lineTo(this.size * 0.75, this.size * 0.3);
      r.stroke("#ffffff", 2);
    } else {
      r.stroke(this.border, 2);
    }
    if (this.label) {
      r.fillText(this.label, this.size + 8, this.size * 0.75, this.font, this.color);
    }
  }
};

// src/Toggle.ts
var ToggleKnob = class extends UIComponent {
  knobR;
  constructor(knobR) {
    super();
    this.knobR = knobR;
    this.interactive = false;
  }
  render(r) {
    r.beginPath();
    r.arc(0, 0, this.knobR, 0, Math.PI * 2);
    r.fill("#ffffff");
  }
};
var Toggle = class extends UIComponent {
  checked;
  label;
  trackW;
  trackH;
  font;
  color;
  accent;
  track;
  knobEntity;
  constructor(opts) {
    super();
    this.checked = opts.checked ?? false;
    this.label = opts.label ?? "";
    this.trackW = opts.width ?? 44;
    this.trackH = opts.height ?? 24;
    this.font = opts.font ?? "16px sans-serif";
    this.color = opts.color ?? "#e2e8f0";
    this.accent = opts.accent ?? "#2563eb";
    this.track = opts.track ?? "#475569";
    this.interactive = true;
    this.height = this.trackH;
    this.width = this.trackW + (this.label ? 8 + measureText(this.label, this.font) : 0);
    const radius = this.trackH / 2;
    const knobR = radius - 3;
    const initialX = this.checked ? this.trackW - radius : radius;
    this.knobEntity = new ToggleKnob(knobR);
    this.knobEntity.setPosition(initialX, radius);
    this.knobEntity.setTransition({
      x: {
        stiffness: 210,
        damping: 14,
        mass: 0.6
      }
    });
    this.add(this.knobEntity);
    this.on("click", () => {
      this.emit("change", { checked: !this.checked });
    });
    this.on("change", (e) => {
      if (e.checked === this.checked) return;
      this.checked = e.checked;
      const targetX = this.checked ? this.trackW - radius : radius;
      this.knobEntity.x = targetX;
      opts.onChange?.(this.checked);
      this.scene?.markDirty();
    });
  }
  getA11yAttributes() {
    return { role: "switch", checked: this.checked, label: this.label };
  }
  render(r) {
    const radius = this.trackH / 2;
    r.beginPath();
    r.roundRect(0, 0, this.trackW, this.trackH, radius);
    r.fill(this.checked ? this.accent : this.track);
    if (this.label) {
      r.fillText(this.label, this.trackW + 8, this.trackH * 0.7, this.font, this.color);
    }
  }
};

// src/ScrollView.ts
var import_core8 = require("@vectojs/core");
var ScrollView = class extends UIComponent {
  content;
  targetY = 0;
  // Pointer-drag (touch / mouse) state.
  dragging = false;
  lastPointerY = 0;
  constructor(opts) {
    super();
    this.width = opts.width;
    this.height = opts.height;
    this.interactive = true;
    this.clipChildren = true;
    this.content = new class extends import_core8.Entity {
      isPointInside() {
        return false;
      }
      render() {
      }
      /** Snap to `y` without touching the spring — see `scrollToBottom`'s use. */
      jumpTo(y) {
        this.setImmediate("y", y);
      }
    }("ScrollViewContent");
    super.add(this.content);
    this.content.setTransition({ y: "spring" });
    this.on("wheel", (e) => {
      if (e.ctrlKey) return;
      e.preventDefault();
      this.targetY -= e.deltaY;
      this.clampTarget();
      this.content.y = this.targetY;
    });
    this.on("pointerdown", (e) => {
      if (e.localY === void 0) return;
      this.dragging = true;
      this.lastPointerY = e.localY;
    });
    this.on("pointermove", (e) => {
      if (!this.dragging || e.localY === void 0) return;
      const y = e.localY;
      this.targetY += y - this.lastPointerY;
      this.lastPointerY = y;
      this.clampTarget();
      this.content.y = this.targetY;
    });
    const endDrag = () => {
      this.dragging = false;
    };
    this.on("pointerup", endDrag);
    this.on("pointerleave", endDrag);
  }
  /** Clamp the scroll target to `[-maxScroll, 0]` (top and content-end edges). */
  clampTarget() {
    const maxScroll = Math.max(0, this.content.height - this.height);
    if (this.targetY > 0) this.targetY = 0;
    else if (this.targetY < -maxScroll) this.targetY = -maxScroll;
  }
  /**
   * Scroll to a specific Y offset (where 0 is top).
   *
   * @param y - The target scroll position in pixels.
   */
  scrollTo(y) {
    this.targetY = -Math.max(0, y);
    this.clampTarget();
    this.content.y = this.targetY;
  }
  /**
   * Scroll to the very bottom of the content.
   *
   * Snaps instantly rather than retargeting the spring: callers that track
   * growing content (e.g. streaming chat) call this on every update, often
   * many times a second — retargeting a spring that fast never lets it
   * settle, so the viewport visibly jitters instead of tracking the newest
   * content. Wheel/drag scrolling still springs (see above); only this
   * "pin to the end" path bypasses it.
   */
  scrollToBottom() {
    const maxScroll = Math.max(0, this.content.height - this.height);
    this.targetY = -maxScroll;
    this.content.jumpTo(this.targetY);
  }
  add(child) {
    this.content.add(child);
    this.updateContentSize();
    return this;
  }
  remove(child) {
    this.content.remove(child);
    this.updateContentSize();
    return this;
  }
  /**
   * Calculates the bounds of the content node to determine the max scroll area.
   */
  updateContentSize() {
    let maxW = 0;
    let maxH = 0;
    for (const child of this.content.children) {
      if (child.x + child.width > maxW) maxW = child.x + child.width;
      if (child.y + child.height > maxH) maxH = child.y + child.height;
    }
    this.content.width = maxW;
    this.content.height = maxH;
    const maxScroll = Math.max(0, this.content.height - this.height);
    if (this.targetY < -maxScroll) {
      this.targetY = -maxScroll;
      this.content.y = this.targetY;
    }
  }
  /**
   * Defensive re-clamp only — the actual scroll motion is driven by `content`'s
   * own spring transition, which the Scene tree walk ticks directly (calling
   * `content.update()` as a normal child node). Reassigning `content.y` here
   * unconditionally every frame would spawn a spurious (instantly-done) driver
   * even when nothing changed, permanently defeating the idle throttle this
   * fix restores — so only touch it when clamping actually moves the target.
   */
  update(dt, time) {
    super.update(dt, time);
    const before = this.targetY;
    this.clampTarget();
    if (this.targetY !== before) this.content.y = this.targetY;
  }
  render(_r) {
  }
};

// src/Markdown.ts
var import_core9 = require("@vectojs/core");
var import_marked = require("marked");
var import_mathjax = require("mathjax-full/js/mathjax.js");
var import_tex = require("mathjax-full/js/input/tex.js");
var import_svg = require("mathjax-full/js/output/svg.js");
var import_liteAdaptor = require("mathjax-full/js/adaptors/liteAdaptor.js");
var import_html = require("mathjax-full/js/handlers/html.js");
var import_AllPackages = require("mathjax-full/js/input/tex/AllPackages.js");

// src/Table.ts
var Table = class extends UIComponent {
  headers;
  rows;
  colWidths;
  rowHeights = [];
  headerHeight = 0;
  bg;
  headerBg;
  borderColor;
  headerTextColor;
  textColor;
  font;
  selectable;
  baseRowHeight;
  headerCells;
  bodyCells;
  constructor(opts) {
    super();
    if (opts.headers.length === 0) throw new RangeError("Table requires at least one column.");
    this.headers = opts.headers;
    this.rows = opts.rows;
    this.baseRowHeight = opts.rowHeight ?? 36;
    this.bg = opts.bg ?? "rgba(15, 15, 25, 0.4)";
    this.headerBg = opts.headerBg ?? "rgba(255, 255, 255, 0.08)";
    this.borderColor = opts.borderColor ?? "rgba(255, 255, 255, 0.15)";
    this.headerTextColor = opts.headerTextColor ?? "#ffffff";
    this.textColor = opts.textColor ?? "#e2e8f0";
    this.font = opts.font ?? "14px sans-serif";
    this.selectable = opts.selectable ?? true;
    this.width = opts.width ?? 600;
    this.colWidths = this.normalizeColumnWidths(opts.colWidths);
    const seen = /* @__PURE__ */ new Set();
    this.headerCells = this.headers.map((cell) => this.normalizeCell(cell, true, seen));
    this.bodyCells = this.rows.map(
      (row) => Array.from(
        { length: this.headers.length },
        (_, column) => this.normalizeCell(row[column] ?? "", false, seen)
      )
    );
    for (const cell of this.headerCells) this.add(cell);
    for (const row of this.bodyCells) for (const cell of row) this.add(cell);
    this.interactive = true;
    this.layout();
  }
  normalizeColumnWidths(widths) {
    const columns = this.headers.length;
    if (!widths || widths.length !== columns || widths.some((width) => !Number.isFinite(width) || width <= 0)) {
      return Array.from({ length: columns }, () => this.width / columns);
    }
    const sum = widths.reduce((total, width) => total + width, 0);
    const scale = this.width / sum;
    return widths.map((width) => width * scale);
  }
  normalizeCell(cell, header, seen) {
    if (typeof cell === "string") {
      return new Text(cell, {
        font: header ? `bold ${this.font}` : this.font,
        color: header ? this.headerTextColor : this.textColor,
        lineHeight: 20,
        selectable: this.selectable
      });
    }
    if (seen.has(cell)) {
      throw new Error("Table Entity cells must be unique instances.");
    }
    seen.add(cell);
    this.setCellSelectable(cell);
    return cell;
  }
  setCellSelectable(cell) {
    const candidate = cell;
    candidate.setSelectable?.(this.selectable);
  }
  fitCell(cell, column) {
    const maxWidth = Math.max(1, this.colWidths[column] - 24);
    const candidate = cell;
    candidate.setMaxWidth?.(maxWidth);
  }
  syncStringCell(entity, source) {
    if (typeof source === "string" && entity instanceof Text && entity.text !== source) {
      entity.setText(source);
    }
  }
  /**
   * Recompute cell wrapping, row heights, and child positions.
   *
   * Call after mutating an externally supplied Entity cell. String-backed
   * cells are owned by the Table and are already kept consistent.
   */
  layout() {
    this.headerHeight = this.baseRowHeight;
    for (let column = 0; column < this.headerCells.length; column++) {
      const cell = this.headerCells[column];
      this.syncStringCell(cell, this.headers[column]);
      this.fitCell(cell, column);
      this.headerHeight = Math.max(this.headerHeight, cell.height + 16);
    }
    this.rowHeights = this.bodyCells.map((row, rowIndex) => {
      let height = this.baseRowHeight;
      for (let column = 0; column < row.length; column++) {
        const cell = row[column];
        this.syncStringCell(cell, this.rows[rowIndex]?.[column]);
        this.fitCell(cell, column);
        height = Math.max(height, cell.height + 16);
      }
      return height;
    });
    let x = 0;
    for (let column = 0; column < this.headerCells.length; column++) {
      const cell = this.headerCells[column];
      cell.setPosition(x + 12, (this.headerHeight - cell.height) / 2);
      x += this.colWidths[column];
    }
    let y = this.headerHeight;
    for (let rowIndex = 0; rowIndex < this.bodyCells.length; rowIndex++) {
      const row = this.bodyCells[rowIndex];
      const rowHeight = this.rowHeights[rowIndex];
      x = 0;
      for (let column = 0; column < row.length; column++) {
        const cell = row[column];
        cell.setPosition(x + 12, y + (rowHeight - cell.height) / 2);
        x += this.colWidths[column];
      }
      y += rowHeight;
    }
    this.height = y;
    this.scene?.markDirty();
    return this;
  }
  /** Enable or disable browser-native selection for every selectable cell. */
  setSelectable(selectable) {
    this.selectable = selectable;
    for (const cell of this.headerCells) this.setCellSelectable(cell);
    for (const row of this.bodyCells) for (const cell of row) this.setCellSelectable(cell);
    this.scene?.markDirty();
    return this;
  }
  getA11yAttributes() {
    return {
      role: "grid",
      label: `Data table with ${this.headers.length} columns and ${this.rows.length} rows.`,
      pointerEvents: "none"
    };
  }
  /** Cell children own projected text, so the Table itself never duplicates it. */
  getContentProjection() {
    return null;
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 8);
    r.fill(this.bg);
    r.beginPath();
    r.roundRect(0, 0, this.width, this.headerHeight, [8, 8, 0, 0]);
    r.fill(this.headerBg);
    let x = 0;
    for (let column = 0; column < this.headers.length; column++) {
      if (column > 0) {
        r.beginPath();
        r.moveTo(x, 0);
        r.lineTo(x, this.height);
        r.stroke(this.borderColor, 1);
      }
      x += this.colWidths[column];
    }
    let y = this.headerHeight;
    for (const rowHeight of this.rowHeights) {
      r.beginPath();
      r.moveTo(0, y);
      r.lineTo(this.width, y);
      r.stroke(this.borderColor, 1);
      y += rowHeight;
    }
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 8);
    r.stroke(this.borderColor, 1.5);
  }
};

// src/MarkdownWorkerSource.ts
var WORKER_SOURCE_STRING = '"use strict";(()=>{function O(){return{async:!1,breaks:!1,extensions:null,gfm:!0,hooks:null,pedantic:!1,renderer:null,silent:!1,tokenizer:null,walkTokens:null}}var T=O();function ae(r){T=r}var R={exec:()=>null};function z(r){let e=[];return t=>{let s=Math.max(0,Math.min(3,t-1)),n=e[s];return n||(n=r(s),e[s]=n),n}}function u(r,e=""){let t=typeof r=="string"?r:r.source,s={replace:(n,i)=>{let l=typeof i=="string"?i:i.source;return l=l.replace(x.caret,"$1"),t=t.replace(n,l),s},getRegex:()=>new RegExp(t,e)};return s}var we=((r="")=>{try{return!!new RegExp("(?<=1)(?<!1)"+r)}catch{return!1}})(),x={codeRemoveIndent:/^(?: {1,4}| {0,3}\\t)/gm,outputLinkReplace:/\\\\([\\[\\]])/g,indentCodeCompensation:/^(\\s+)(?:```)/,beginningSpace:/^\\s+/,endingHash:/#$/,startingSpaceChar:/^ /,endingSpaceChar:/ $/,nonSpaceChar:/[^ ]/,newLineCharGlobal:/\\n/g,tabCharGlobal:/\\t/g,multipleSpaceGlobal:/\\s+/g,blankLine:/^[ \\t]*$/,doubleBlankLine:/\\n[ \\t]*\\n[ \\t]*$/,blockquoteStart:/^ {0,3}>/,blockquoteSetextReplace:/\\n {0,3}((?:=+|-+) *)(?=\\n|$)/g,blockquoteSetextReplace2:/^ {0,3}>[ \\t]?/gm,listReplaceNesting:/^ {1,4}(?=( {4})*[^ ])/g,listIsTask:/^\\[[ xX]\\] +\\S/,listReplaceTask:/^\\[[ xX]\\] +/,listTaskCheckbox:/\\[[ xX]\\]/,anyLine:/\\n.*\\n/,hrefBrackets:/^<(.*)>$/,tableDelimiter:/[:|]/,tableAlignChars:/^\\||\\| *$/g,tableRowBlankLine:/\\n[ \\t]*$/,tableAlignRight:/^ *-+: *$/,tableAlignCenter:/^ *:-+: *$/,tableAlignLeft:/^ *:-+ *$/,startATag:/^<a /i,endATag:/^<\\/a>/i,startPreScriptTag:/^<(pre|code|kbd|script)(\\s|>)/i,endPreScriptTag:/^<\\/(pre|code|kbd|script)(\\s|>)/i,startAngleBracket:/^</,endAngleBracket:/>$/,pedanticHrefTitle:/^([^\'"]*[^\\s])\\s+([\'"])(.*)\\2/,unicodeAlphaNumeric:/[\\p{L}\\p{N}]/u,escapeTest:/[&<>"\']/,escapeReplace:/[&<>"\']/g,escapeTestNoEncode:/[<>"\']|&(?!(#\\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\\w+);)/,escapeReplaceNoEncode:/[<>"\']|&(?!(#\\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\\w+);)/g,caret:/(^|[^\\[])\\^/g,percentDecode:/%25/g,findPipe:/\\|/g,splitPipe:/ \\|/,slashPipe:/\\\\\\|/g,carriageReturn:/\\r\\n|\\r/g,spaceLine:/^ +$/gm,notSpaceStart:/^\\S*/,endingNewline:/\\n$/,listItemRegex:r=>new RegExp(`^( {0,3}${r})((?:[	 ][^\\\\n]*)?(?:\\\\n|$))`),nextBulletRegex:z(r=>new RegExp(`^ {0,${r}}(?:[*+-]|\\\\d{1,9}[.)])((?:[ 	][^\\\\n]*)?(?:\\\\n|$))`)),hrRegex:z(r=>new RegExp(`^ {0,${r}}((?:- *){3,}|(?:_ *){3,}|(?:\\\\* *){3,})(?:\\\\n+|$)`)),fencesBeginRegex:z(r=>new RegExp(`^ {0,${r}}(?:\\`\\`\\`|~~~)`)),headingBeginRegex:z(r=>new RegExp(`^ {0,${r}}#`)),htmlBeginRegex:z(r=>new RegExp(`^ {0,${r}}<(?:[a-z].*>|!--)`,"i")),blockquoteBeginRegex:z(r=>new RegExp(`^ {0,${r}}>`))},me=/^(?:[ \\t]*(?:\\n|$))+/,ye=/^((?: {4}| {0,3}\\t)[^\\n]+(?:\\n(?:[ \\t]*(?:\\n|$))*)?)+/,$e=/^ {0,3}(`{3,}(?=[^`\\n]*(?:\\n|$))|~{3,})([^\\n]*)(?:\\n|$)(?:|([\\s\\S]*?)(?:\\n|$))(?: {0,3}\\1[~`]* *(?=\\n|$)|$)/,I=/^ {0,3}((?:-[\\t ]*){3,}|(?:_[ \\t]*){3,}|(?:\\*[ \\t]*){3,})(?:\\n+|$)/,Re=/^ {0,3}(#{1,6})(?=\\s|$)(.*)(?:\\n+|$)/,j=/ {0,3}(?:[*+-]|\\d{1,9}[.)])/,oe=/^(?!bull |blockCode|fences|blockquote|heading|html|table)((?:.|\\n(?!\\s*?\\n|bull |blockCode|fences|blockquote|heading|html|table))+?)\\n {0,3}(=+|-+) *(?:\\n+|$)/,ce=u(oe).replace(/bull/g,j).replace(/blockCode/g,/(?: {4}| {0,3}\\t)/).replace(/fences/g,/ {0,3}(?:`{3,}|~{3,})/).replace(/blockquote/g,/ {0,3}>/).replace(/heading/g,/ {0,3}#{1,6}/).replace(/html/g,/ {0,3}<[^\\n>]+>\\n/).replace(/\\|table/g,"").getRegex(),Se=u(oe).replace(/bull/g,j).replace(/blockCode/g,/(?: {4}| {0,3}\\t)/).replace(/fences/g,/ {0,3}(?:`{3,}|~{3,})/).replace(/blockquote/g,/ {0,3}>/).replace(/heading/g,/ {0,3}#{1,6}/).replace(/html/g,/ {0,3}<[^\\n>]+>\\n/).replace(/table/g,/ {0,3}\\|?(?:[:\\- ]*\\|)+[\\:\\- ]*\\n/).getRegex(),N=/^([^\\n]+(?:\\n(?!hr|heading|lheading|blockquote|fences|list|html|table| +\\n)[^\\n]+)*)/,Te=/^[^\\n]+/,G=/(?!\\s*\\])(?:\\\\[\\s\\S]|[^\\[\\]\\\\])+/,ze=u(/^ {0,3}\\[(label)\\]: *(?:\\n[ \\t]*)?([^<\\s][^\\s]*|<.*?>)(?:(?: +(?:\\n[ \\t]*)?| *\\n[ \\t]*)(title))? *(?:\\n+|$)/).replace("label",G).replace("title",/(?:"(?:\\\\"?|[^"\\\\])*"|\'[^\'\\n]*(?:\\n[^\'\\n]+)*\\n?\'|\\([^()]*\\))/).getRegex(),Ae=u(/^(bull)([ \\t][^\\n]*?)?(?:\\n|$)/).replace(/bull/g,j).getRegex(),q="address|article|aside|base|basefont|blockquote|body|caption|center|col|colgroup|dd|details|dialog|dir|div|dl|dt|fieldset|figcaption|figure|footer|form|frame|frameset|h[1-6]|head|header|hr|html|iframe|legend|li|link|main|menu|menuitem|meta|nav|noframes|ol|optgroup|option|p|param|search|section|summary|table|tbody|td|tfoot|th|thead|title|tr|track|ul",X=/<!--(?:-?>|[\\s\\S]*?(?:-->|$))/,Pe=u("^ {0,3}(?:<(script|pre|style|textarea)[\\\\s>][\\\\s\\\\S]*?(?:</\\\\1>[^\\\\n]*\\\\n+|$)|comment[^\\\\n]*(\\\\n+|$)|<\\\\?[\\\\s\\\\S]*?(?:\\\\?>[^\\\\n]*\\\\n+|$)|<![A-Z][\\\\s\\\\S]*?(?:>[^\\\\n]*\\\\n+|$)|<!\\\\[CDATA\\\\[[\\\\s\\\\S]*?(?:\\\\]\\\\]>[^\\\\n]*\\\\n+|$)|</?(tag)(?: +|\\\\n|/?>)[\\\\s\\\\S]*?(?:(?:\\\\n[ 	]*)+\\\\n|$)|<(?!script|pre|style|textarea)([a-z][\\\\w-]*)(?:attribute)*? */?>(?=[ \\\\t]*(?:\\\\n|$))[\\\\s\\\\S]*?(?:(?:\\\\n[ 	]*)+\\\\n|$)|</(?!script|pre|style|textarea)[a-z][\\\\w-]*\\\\s*>(?=[ \\\\t]*(?:\\\\n|$))[\\\\s\\\\S]*?(?:(?:\\\\n[ 	]*)+\\\\n|$))","i").replace("comment",X).replace("tag",q).replace("attribute",/ +[a-zA-Z:_][\\w.:-]*(?: *= *"[^"\\n]*"| *= *\'[^\'\\n]*\'| *= *[^\\s"\'=<>`]+)?/).getRegex(),he=r=>u(N).replace("hr",I).replace("heading"," {0,3}#{1,6}(?:\\\\s|$)").replace("|lheading","").replace("|table","").replace("blockquote"," {0,3}>").replace("fences"," {0,3}(?:`{3,}(?=[^`\\\\n]*\\\\n)|~{3,})[^\\\\n]*\\\\n").replace("list",r).replace("html","</?(?:tag)(?: +|\\\\n|/?>)|<(?:script|pre|style|textarea|!--)").replace("tag",q).getRegex(),_e=he(/ {0,3}(?:[*+-]|1[.)])[ \\t]+[^ \\t\\n]/),Le=he(/ {0,3}(?:[*+-]|\\d{1,9}[.)])[ \\t]+[^ \\t\\n]/),Ie=u(/^( {0,3}> ?(paragraph|[^\\n]*)(?:\\n|$))+/).replace("paragraph",Le).getRegex(),W={blockquote:Ie,code:ye,def:ze,fences:$e,heading:Re,hr:I,html:Pe,lheading:ce,list:Ae,newline:me,paragraph:_e,table:R,text:Te},ee=u("^ *([^\\\\n ].*)\\\\n {0,3}((?:\\\\| *)?:?-+:? *(?:\\\\| *:?-+:? *)*(?:\\\\| *)?)(?:\\\\n((?:(?! *\\\\n|hr|heading|blockquote|code|fences|list|html).*(?:\\\\n|$))*)\\\\n*|$)").replace("hr",I).replace("heading"," {0,3}#{1,6}(?:\\\\s|$)").replace("blockquote"," {0,3}>").replace("code","(?: {4}| {0,3}	)[^\\\\n]").replace("fences"," {0,3}(?:`{3,}(?=[^`\\\\n]*\\\\n)|~{3,})[^\\\\n]*\\\\n").replace("list"," {0,3}(?:[*+-]|1[.)])[ \\\\t]").replace("html","</?(?:tag)(?: +|\\\\n|/?>)|<(?:script|pre|style|textarea|!--)").replace("tag",q).getRegex(),Ee={...W,lheading:Se,table:ee,paragraph:u(N).replace("hr",I).replace("heading"," {0,3}#{1,6}(?:\\\\s|$)").replace("|lheading","").replace("table",ee).replace("blockquote"," {0,3}>").replace("fences"," {0,3}(?:`{3,}(?=[^`\\\\n]*\\\\n)|~{3,})[^\\\\n]*\\\\n").replace("list"," {0,3}(?:[*+-]|1[.)])[ \\\\t]+[^ \\\\t\\\\n]").replace("html","</?(?:tag)(?: +|\\\\n|/?>)|<(?:script|pre|style|textarea|!--)").replace("tag",q).getRegex()},Ce={...W,html:u(`^ *(?:comment *(?:\\\\n|\\\\s*$)|<(tag)[\\\\s\\\\S]+?</\\\\1> *(?:\\\\n{2,}|\\\\s*$)|<tag(?:"[^"]*"|\'[^\']*\'|\\\\s[^\'"/>\\\\s]*)*?/?> *(?:\\\\n{2,}|\\\\s*$))`).replace("comment",X).replace(/tag/g,"(?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img)\\\\b)\\\\w+(?!:|[^\\\\w\\\\s@]*@)\\\\b").getRegex(),def:/^ *\\[([^\\]]+)\\]: *<?([^\\s>]+)>?(?: +(["(][^\\n]+[")]))? *(?:\\n+|$)/,heading:/^(#{1,6})(.*)(?:\\n+|$)/,fences:R,lheading:/^(.+?)\\n {0,3}(=+|-+) *(?:\\n+|$)/,paragraph:u(N).replace("hr",I).replace("heading",` *#{1,6} *[^\n]`).replace("lheading",ce).replace("|table","").replace("blockquote"," {0,3}>").replace("|fences","").replace("|list","").replace("|html","").replace("|tag","").getRegex()},ve=/^\\\\([!"#$%&\'()*+,\\-./:;<=>?@\\[\\]\\\\^_`{|}~])/,Be=/^(`+)([^`]|[^`][\\s\\S]*?[^`])\\1(?!`)/,pe=/^( {2,}|\\\\)\\n(?!\\s*$)/,qe=/^(`+|[^`])(?:(?= {2,}\\n)|[\\s\\S]*?(?:(?=[\\\\<!\\[`*_]|\\b_|$)|[^ ](?= {2,}\\n)))/,A=/[\\p{P}\\p{S}]/u,Z=/[\\s\\p{P}\\p{S}]/u,F=/[^\\s\\p{P}\\p{S}]/u,Ze=u(/^((?![*_])punctSpace)/,"u").replace(/punctSpace/g,Z).getRegex(),ue=/(?!~)[\\p{P}\\p{S}]/u,De=/(?!~)[\\s\\p{P}\\p{S}]/u,Me=/(?:[^\\s\\p{P}\\p{S}]|~)/u,Qe=u(/link|precode-code|html/,"g").replace("link",/\\[(?:[^\\[\\]`]|(?<a>`+)[^`]+\\k<a>(?!`))*?\\]\\((?:\\\\[\\s\\S]|[^\\\\\\(\\)]|\\((?:\\\\[\\s\\S]|[^\\\\\\(\\)])*\\))*\\)/).replace("precode-",we?"(?<!`)()":"(^^|[^`])").replace("code",/(?<b>`+)[^`]+\\k<b>(?!`)/).replace("html",/<(?! )[^<>]*?>/).getRegex(),ge=/^(?:\\*+(?:((?!\\*)punct)|([^\\s*]))?)|^_+(?:((?!_)punct)|([^\\s_]))?/,He=u(ge,"u").replace(/punct/g,A).getRegex(),Oe=u(ge,"u").replace(/punct/g,ue).getRegex(),ke="^[^_*]*?__[^_*]*?\\\\*[^_*]*?(?=__)|[^*]+(?=[^*])|(?!\\\\*)punct(\\\\*+)(?=[\\\\s]|$)|notPunctSpace(\\\\*+)(?!\\\\*)(?=punctSpace|$)|(?!\\\\*)punctSpace(\\\\*+)(?=notPunctSpace)|[\\\\s](\\\\*+)(?!\\\\*)(?=punct)|(?!\\\\*)punct(\\\\*+)(?!\\\\*)(?=punct)|notPunctSpace(\\\\*+)(?=notPunctSpace)",je=u(ke,"gu").replace(/notPunctSpace/g,F).replace(/punctSpace/g,Z).replace(/punct/g,A).getRegex(),Ne=u(ke,"gu").replace(/notPunctSpace/g,Me).replace(/punctSpace/g,De).replace(/punct/g,ue).getRegex(),Ge=u("^[^_*]*?\\\\*\\\\*[^_*]*?_[^_*]*?(?=\\\\*\\\\*)|[^_]+(?=[^_])|(?!_)punct(_+)(?=[\\\\s]|$)|notPunctSpace(_+)(?!_)(?=punctSpace|$)|(?!_)punctSpace(_+)(?=notPunctSpace)|[\\\\s](_+)(?!_)(?=punct)|(?!_)punct(_+)(?!_)(?=punct)","gu").replace(/notPunctSpace/g,F).replace(/punctSpace/g,Z).replace(/punct/g,A).getRegex(),Xe=u(/^~~?(?:((?!~)punct)|[^\\s~])/,"u").replace(/punct/g,A).getRegex(),We="^[^~]+(?=[^~])|(?!~)punct(~~?)(?=[\\\\s]|$)|notPunctSpace(~~?)(?!~)(?=punctSpace|$)|(?!~)punctSpace(~~?)(?=notPunctSpace)|[\\\\s](~~?)(?!~)(?=punct)|(?!~)punct(~~?)(?!~)(?=punct)|notPunctSpace(~~?)(?=notPunctSpace)",Fe=u(We,"gu").replace(/notPunctSpace/g,F).replace(/punctSpace/g,Z).replace(/punct/g,A).getRegex(),Ue=u(/\\\\(punct)/,"gu").replace(/punct/g,A).getRegex(),Je=u(/^<(scheme:[^\\s\\x00-\\x1f<>]*|email)>/).replace("scheme",/[a-zA-Z][a-zA-Z0-9+.-]{1,31}/).replace("email",/[a-zA-Z0-9.!#$%&\'*+/=?^_`{|}~-]+(@)[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+(?![-_])/).getRegex(),Ke=u(X).replace("(?:-->|$)","-->").getRegex(),Ve=u("^comment|^</[a-zA-Z][\\\\w:-]*\\\\s*>|^<[a-zA-Z][\\\\w-]*(?:attribute)*?\\\\s*/?>|^<\\\\?[\\\\s\\\\S]*?\\\\?>|^<![a-zA-Z]+\\\\s[\\\\s\\\\S]*?>|^<!\\\\[CDATA\\\\[[\\\\s\\\\S]*?\\\\]\\\\]>").replace("comment",Ke).replace("attribute",/\\s+[a-zA-Z:_][\\w.:-]*(?:\\s*=\\s*"[^"]*"|\\s*=\\s*\'[^\']*\'|\\s*=\\s*[^\\s"\'=<>`]+)?/).getRegex(),C=/(?:\\[(?:\\\\[\\s\\S]|[^\\[\\]\\\\])*\\]|\\\\[\\s\\S]|`+(?!`)[^`]*?`+(?!`)|``+(?=\\])|[^\\[\\]\\\\`])*?/,Ye=u(/^!?\\[(label)\\]\\(\\s*(href)(?:(?:[ \\t]+(?:\\n[ \\t]*)?|\\n[ \\t]*)(title))?\\s*\\)/).replace("label",C).replace("href",/<(?:\\\\.|[^\\n<>\\\\])+>|[^ \\t\\n\\x00-\\x1f]+|(?=\\))/).replace("title",/"(?:\\\\"?|[^"\\\\])*"|\'(?:\\\\\'?|[^\'\\\\])*\'|\\((?:\\\\\\)?|[^)\\\\])*\\)/).getRegex(),de=u(/^!?\\[(label)\\]\\[(ref)\\]/).replace("label",C).replace("ref",G).getRegex(),fe=u(/^!?\\[(ref)\\](?:\\[\\])?/).replace("ref",G).getRegex(),et=u("reflink|nolink(?!\\\\()","g").replace("reflink",de).replace("nolink",fe).getRegex(),te=/[hH][tT][tT][pP][sS]?|[fF][tT][pP]/,U={_backpedal:R,anyPunctuation:Ue,autolink:Je,blockSkip:Qe,br:pe,code:Be,del:R,delLDelim:R,delRDelim:R,emStrongLDelim:He,emStrongRDelimAst:je,emStrongRDelimUnd:Ge,escape:ve,link:Ye,nolink:fe,punctuation:Ze,reflink:de,reflinkSearch:et,tag:Ve,text:qe,url:R},tt={...U,link:u(/^!?\\[(label)\\]\\((.*?)\\)/).replace("label",C).getRegex(),reflink:u(/^!?\\[(label)\\]\\s*\\[([^\\]]*)\\]/).replace("label",C).getRegex()},M={...U,emStrongRDelimAst:Ne,emStrongLDelim:Oe,delLDelim:Xe,delRDelim:Fe,url:u(/^((?:protocol):\\/\\/|www\\.)(?:[a-zA-Z0-9\\-]+\\.?)+[^\\s<]*|^email/).replace("protocol",te).replace("email",/[A-Za-z0-9._+-]+(@)[a-zA-Z0-9-_]+(?:\\.[a-zA-Z0-9-_]*[a-zA-Z0-9])+(?![-_])/).getRegex(),_backpedal:/(?:[^?!.,:;*_\'"~()&]+|\\([^)]*\\)|&(?![a-zA-Z0-9]+;$)|[?!.,:;*_\'"~)]+(?!$))+/,del:/^(~~?)(?=[^\\s~])((?:\\\\[\\s\\S]|[^\\\\])*?(?:\\\\[\\s\\S]|[^\\s~\\\\]))\\1(?=[^~]|$)/,text:u(/^([`~]+|[^`~])(?:(?= {2,}\\n)|(?=[a-zA-Z0-9.!#$%&\'*+\\/=?_`{\\|}~-]+@)|[\\s\\S]*?(?:(?=[\\\\<!\\[`*~_]|\\b_|protocol:\\/\\/|www\\.|$)|[^ ](?= {2,}\\n)|[^a-zA-Z0-9.!#$%&\'*+\\/=?_`{\\|}~-](?=[a-zA-Z0-9.!#$%&\'*+\\/=?_`{\\|}~-]+@)))/).replace("protocol",te).getRegex()},rt={...M,br:u(pe).replace("{2,}","*").getRegex(),text:u(M.text).replace("\\\\b_","\\\\b_| {2,}\\\\n").replace(/\\{2,\\}/g,"*").getRegex()},E={normal:W,gfm:Ee,pedantic:Ce},_={normal:U,gfm:M,breaks:rt,pedantic:tt},nt={"&":"&amp;","<":"&lt;",">":"&gt;",\'"\':"&quot;","\'":"&#39;"},re=r=>nt[r];function m(r,e){if(e){if(x.escapeTest.test(r))return r.replace(x.escapeReplace,re)}else if(x.escapeTestNoEncode.test(r))return r.replace(x.escapeReplaceNoEncode,re);return r}function ne(r){try{r=encodeURI(r).replace(x.percentDecode,"%")}catch{return null}return r}function se(r,e){let t=r.replace(x.findPipe,(i,l,a)=>{let c=!1,o=l;for(;--o>=0&&a[o]==="\\\\";)c=!c;return c?"|":" |"}),s=t.split(x.splitPipe),n=0;if(s[0].trim()||s.shift(),s.length>0&&!s.at(-1)?.trim()&&s.pop(),e)if(s.length>e)s.splice(e);else for(;s.length<e;)s.push("");for(;n<s.length;n++)s[n]=s[n].trim().replace(x.slashPipe,"|");return s}function $(r,e,t){let s=r.length;if(s===0)return"";let n=0;for(;n<s;){let i=r.charAt(s-n-1);if(i===e&&!t)n++;else if(i!==e&&t)n++;else break}return r.slice(0,s-n)}function ie(r){let e=r.split(`\n`),t=e.length-1;for(;t>=0&&x.blankLine.test(e[t]);)t--;return e.length-t<=2?r:e.slice(0,t+1).join(`\n`)}function st(r,e){if(r.indexOf(e[1])===-1)return-1;let t=0;for(let s=0;s<r.length;s++)if(r[s]==="\\\\")s++;else if(r[s]===e[0])t++;else if(r[s]===e[1]&&(t--,t<0))return s;return t>0?-2:-1}function it(r,e=0){let t=e,s="";for(let n of r)if(n==="	"){let i=4-t%4;s+=" ".repeat(i),t+=i}else s+=n,t++;return s}function le(r,e,t,s,n){let i=e.href,l=e.title||null,a=r[1].replace(n.other.outputLinkReplace,"$1");s.state.inLink=!0;let c={type:r[0].charAt(0)==="!"?"image":"link",raw:t,href:i,title:l,text:a,tokens:s.inlineTokens(a)};return s.state.inLink=!1,c}function lt(r,e,t){let s=r.match(t.other.indentCodeCompensation);if(s===null)return e;let n=s[1];return e.split(`\n`).map(i=>{let l=i.match(t.other.beginningSpace);if(l===null)return i;let[a]=l;return a.length>=n.length?i.slice(n.length):i}).join(`\n`)}var v=class{options;rules;lexer;constructor(r){this.options=r||T}space(r){let e=this.rules.block.newline.exec(r);if(e&&e[0].length>0)return{type:"space",raw:e[0]}}code(r){let e=this.rules.block.code.exec(r);if(e){let t=this.options.pedantic?e[0]:ie(e[0]),s=t.replace(this.rules.other.codeRemoveIndent,"");return{type:"code",raw:t,codeBlockStyle:"indented",text:s}}}fences(r){let e=this.rules.block.fences.exec(r);if(e){let t=e[0],s=lt(t,e[3]||"",this.rules);return{type:"code",raw:t,lang:e[2]?e[2].trim().replace(this.rules.inline.anyPunctuation,"$1"):e[2],text:s}}}heading(r){let e=this.rules.block.heading.exec(r);if(e){let t=e[2].trim();if(this.rules.other.endingHash.test(t)){let s=$(t,"#");(this.options.pedantic||!s||this.rules.other.endingSpaceChar.test(s))&&(t=s.trim())}return{type:"heading",raw:$(e[0],`\n`),depth:e[1].length,text:t,tokens:this.lexer.inline(t)}}}hr(r){let e=this.rules.block.hr.exec(r);if(e)return{type:"hr",raw:$(e[0],`\n`)}}blockquote(r){let e=this.rules.block.blockquote.exec(r);if(e){let t=$(e[0],`\n`).split(`\n`),s="",n="",i=[];for(;t.length>0;){let l=!1,a=[],c;for(c=0;c<t.length;c++)if(this.rules.other.blockquoteStart.test(t[c]))a.push(t[c]),l=!0;else if(!l)a.push(t[c]);else break;t=t.slice(c);let o=a.join(`\n`),p=o.replace(this.rules.other.blockquoteSetextReplace,`\n    $1`).replace(this.rules.other.blockquoteSetextReplace2,"");s=s?`${s}\n${o}`:o,n=n?`${n}\n${p}`:p;let h=this.lexer.state.top;if(this.lexer.state.top=!0,this.lexer.blockTokens(p,i,!0),this.lexer.state.top=h,t.length===0)break;let g=i.at(-1);if(g?.type==="code")break;if(g?.type==="blockquote"){let f=g,d=f.raw+`\n`+t.join(`\n`),y=this.blockquote(d);i[i.length-1]=y,s=s.substring(0,s.length-f.raw.length)+y.raw,n=n.substring(0,n.length-f.text.length)+y.text;break}else if(g?.type==="list"){let f=g,d=f.raw+`\n`+t.join(`\n`),y=this.list(d);i[i.length-1]=y,s=s.substring(0,s.length-g.raw.length)+y.raw,n=n.substring(0,n.length-f.raw.length)+y.raw,t=d.substring(i.at(-1).raw.length).split(`\n`);continue}}return{type:"blockquote",raw:s,tokens:i,text:n}}}list(r){let e=this.rules.block.list.exec(r);if(e){let t=e[1].trim(),s=t.length>1,n={type:"list",raw:"",ordered:s,start:s?+t.slice(0,-1):"",loose:!1,items:[]};t=s?`\\\\d{1,9}\\\\${t.slice(-1)}`:`\\\\${t}`,this.options.pedantic&&(t=s?t:"[*+-]");let i=this.rules.other.listItemRegex(t),l=!1;for(;r;){let c=!1,o="",p="";if(!(e=i.exec(r))||this.rules.block.hr.test(r))break;o=e[0],r=r.substring(o.length);let h=it(e[2].split(`\n`,1)[0],e[1].length),g=r.split(`\n`,1)[0],f=!h.trim(),d=0;if(this.options.pedantic?(d=2,p=h.trimStart()):f?d=e[1].length+1:(d=h.search(this.rules.other.nonSpaceChar),d=d>4?1:d,p=h.slice(d),d+=e[1].length),f&&this.rules.other.blankLine.test(g)&&(o+=g+`\n`,r=r.substring(g.length+1),c=!0),!c){let y=this.rules.other.nextBulletRegex(d),K=this.rules.other.hrRegex(d),V=this.rules.other.fencesBeginRegex(d),Y=this.rules.other.headingBeginRegex(d),xe=this.rules.other.htmlBeginRegex(d),be=this.rules.other.blockquoteBeginRegex(d);for(;r;){let D=r.split(`\n`,1)[0],P;if(g=D,this.options.pedantic?(g=g.replace(this.rules.other.listReplaceNesting,"  "),P=g):P=g.replace(this.rules.other.tabCharGlobal,"    "),V.test(g)||Y.test(g)||xe.test(g)||be.test(g)||y.test(g)||K.test(g))break;if(P.search(this.rules.other.nonSpaceChar)>=d||!g.trim())p+=`\n`+P.slice(d);else{if(f||h.replace(this.rules.other.tabCharGlobal,"    ").search(this.rules.other.nonSpaceChar)>=4||V.test(h)||Y.test(h)||K.test(h))break;p+=`\n`+g}f=!g.trim(),o+=D+`\n`,r=r.substring(D.length+1),h=P.slice(d)}}n.loose||(l?n.loose=!0:this.rules.other.doubleBlankLine.test(o)&&(l=!0)),n.items.push({type:"list_item",raw:o,task:!!this.options.gfm&&this.rules.other.listIsTask.test(p),loose:!1,text:p,tokens:[]}),n.raw+=o}let a=n.items.at(-1);if(a)a.raw=a.raw.trimEnd(),a.text=a.text.trimEnd();else return;n.raw=n.raw.trimEnd();for(let c of n.items){this.lexer.state.top=!1,c.tokens=this.lexer.blockTokens(c.text,[]);let o=c.tokens[0];if(c.task&&(o?.type==="text"||o?.type==="paragraph")){c.text=c.text.replace(this.rules.other.listReplaceTask,""),o.raw=o.raw.replace(this.rules.other.listReplaceTask,""),o.text=o.text.replace(this.rules.other.listReplaceTask,"");for(let h=this.lexer.inlineQueue.length-1;h>=0;h--)if(this.rules.other.listIsTask.test(this.lexer.inlineQueue[h].src)){this.lexer.inlineQueue[h].src=this.lexer.inlineQueue[h].src.replace(this.rules.other.listReplaceTask,"");break}let p=this.rules.other.listTaskCheckbox.exec(c.raw);if(p){let h={type:"checkbox",raw:p[0]+" ",checked:p[0]!=="[ ]"};c.checked=h.checked,n.loose?c.tokens[0]&&["paragraph","text"].includes(c.tokens[0].type)&&"tokens"in c.tokens[0]&&c.tokens[0].tokens?(c.tokens[0].raw=h.raw+c.tokens[0].raw,c.tokens[0].text=h.raw+c.tokens[0].text,c.tokens[0].tokens.unshift(h)):c.tokens.unshift({type:"paragraph",raw:h.raw,text:h.raw,tokens:[h]}):c.tokens.unshift(h)}}else c.task&&(c.task=!1);if(!n.loose){let p=c.tokens.filter(g=>g.type==="space"),h=p.length>0&&p.some(g=>this.rules.other.anyLine.test(g.raw));n.loose=h}}if(n.loose)for(let c of n.items){c.loose=!0;for(let o of c.tokens)o.type==="text"&&(o.type="paragraph")}return n}}html(r){let e=this.rules.block.html.exec(r);if(e){let t=ie(e[0]);return{type:"html",block:!0,raw:t,pre:e[1]==="pre"||e[1]==="script"||e[1]==="style",text:t}}}def(r){let e=this.rules.block.def.exec(r);if(e){let t=e[1].toLowerCase().replace(this.rules.other.multipleSpaceGlobal," "),s=e[2]?e[2].replace(this.rules.other.hrefBrackets,"$1").replace(this.rules.inline.anyPunctuation,"$1"):"",n=e[3]?e[3].substring(1,e[3].length-1).replace(this.rules.inline.anyPunctuation,"$1"):e[3];return{type:"def",tag:t,raw:$(e[0],`\n`),href:s,title:n}}}table(r){let e=this.rules.block.table.exec(r);if(!e||!this.rules.other.tableDelimiter.test(e[2]))return;let t=se(e[1]),s=e[2].replace(this.rules.other.tableAlignChars,"").split("|"),n=e[3]?.trim()?e[3].replace(this.rules.other.tableRowBlankLine,"").split(`\n`):[],i={type:"table",raw:$(e[0],`\n`),header:[],align:[],rows:[]};if(t.length===s.length){for(let l of s)this.rules.other.tableAlignRight.test(l)?i.align.push("right"):this.rules.other.tableAlignCenter.test(l)?i.align.push("center"):this.rules.other.tableAlignLeft.test(l)?i.align.push("left"):i.align.push(null);for(let l=0;l<t.length;l++)i.header.push({text:t[l],tokens:this.lexer.inline(t[l]),header:!0,align:i.align[l]});for(let l of n)i.rows.push(se(l,i.header.length).map((a,c)=>({text:a,tokens:this.lexer.inline(a),header:!1,align:i.align[c]})));return i}}lheading(r){let e=this.rules.block.lheading.exec(r);if(e){let t=e[1].trim();return{type:"heading",raw:$(e[0],`\n`),depth:e[2].charAt(0)==="="?1:2,text:t,tokens:this.lexer.inline(t)}}}paragraph(r){let e=this.rules.block.paragraph.exec(r);if(e){let t=e[1].charAt(e[1].length-1)===`\n`?e[1].slice(0,-1):e[1];return{type:"paragraph",raw:e[0],text:t,tokens:this.lexer.inline(t)}}}text(r){let e=this.rules.block.text.exec(r);if(e)return{type:"text",raw:e[0],text:e[0],tokens:this.lexer.inline(e[0])}}escape(r){let e=this.rules.inline.escape.exec(r);if(e)return{type:"escape",raw:e[0],text:e[1]}}tag(r){let e=this.rules.inline.tag.exec(r);if(e)return!this.lexer.state.inLink&&this.rules.other.startATag.test(e[0])?this.lexer.state.inLink=!0:this.lexer.state.inLink&&this.rules.other.endATag.test(e[0])&&(this.lexer.state.inLink=!1),!this.lexer.state.inRawBlock&&this.rules.other.startPreScriptTag.test(e[0])?this.lexer.state.inRawBlock=!0:this.lexer.state.inRawBlock&&this.rules.other.endPreScriptTag.test(e[0])&&(this.lexer.state.inRawBlock=!1),{type:"html",raw:e[0],inLink:this.lexer.state.inLink,inRawBlock:this.lexer.state.inRawBlock,block:!1,text:e[0]}}link(r){let e=this.rules.inline.link.exec(r);if(e){let t=e[2].trim();if(!this.options.pedantic&&this.rules.other.startAngleBracket.test(t)){if(!this.rules.other.endAngleBracket.test(t))return;let i=$(t.slice(0,-1),"\\\\");if((t.length-i.length)%2===0)return}else{let i=st(e[2],"()");if(i===-2)return;if(i>-1){let l=(e[0].indexOf("!")===0?5:4)+e[1].length+i;e[2]=e[2].substring(0,i),e[0]=e[0].substring(0,l).trim(),e[3]=""}}let s=e[2],n="";if(this.options.pedantic){let i=this.rules.other.pedanticHrefTitle.exec(s);i&&(s=i[1],n=i[3])}else n=e[3]?e[3].slice(1,-1):"";return s=s.trim(),this.rules.other.startAngleBracket.test(s)&&(this.options.pedantic&&!this.rules.other.endAngleBracket.test(t)?s=s.slice(1):s=s.slice(1,-1)),le(e,{href:s&&s.replace(this.rules.inline.anyPunctuation,"$1"),title:n&&n.replace(this.rules.inline.anyPunctuation,"$1")},e[0],this.lexer,this.rules)}}reflink(r,e){let t;if((t=this.rules.inline.reflink.exec(r))||(t=this.rules.inline.nolink.exec(r))){let s=(t[2]||t[1]).replace(this.rules.other.multipleSpaceGlobal," "),n=e[s.toLowerCase()];if(!n){let i=t[0].charAt(0);return{type:"text",raw:i,text:i}}return le(t,n,t[0],this.lexer,this.rules)}}emStrong(r,e,t=""){let s=this.rules.inline.emStrongLDelim.exec(r);if(!(!s||!s[1]&&!s[2]&&!s[3]&&!s[4]||s[4]&&t.match(this.rules.other.unicodeAlphaNumeric))&&(!(s[1]||s[3])||!t||this.rules.inline.punctuation.exec(t))){let n=[...s[0]].length-1,i,l,a=n,c=0,o=s[0][0]==="*"?this.rules.inline.emStrongRDelimAst:this.rules.inline.emStrongRDelimUnd;for(o.lastIndex=0,e=e.slice(-1*r.length+n);(s=o.exec(e))!==null;){if(i=s[1]||s[2]||s[3]||s[4]||s[5]||s[6],!i)continue;if(l=[...i].length,s[3]||s[4]){a+=l;continue}else if((s[5]||s[6])&&n%3&&!((n+l)%3)){c+=l;continue}if(a-=l,a>0)continue;l=Math.min(l,l+a+c);let p=[...s[0]][0].length,h=r.slice(0,n+s.index+p+l);if(Math.min(n,l)%2){let f=h.slice(1,-1);return{type:"em",raw:h,text:f,tokens:this.lexer.inlineTokens(f)}}let g=h.slice(2,-2);return{type:"strong",raw:h,text:g,tokens:this.lexer.inlineTokens(g)}}}}codespan(r){let e=this.rules.inline.code.exec(r);if(e){let t=e[2].replace(this.rules.other.newLineCharGlobal," "),s=this.rules.other.nonSpaceChar.test(t),n=this.rules.other.startingSpaceChar.test(t)&&this.rules.other.endingSpaceChar.test(t);return s&&n&&(t=t.substring(1,t.length-1)),{type:"codespan",raw:e[0],text:t}}}br(r){let e=this.rules.inline.br.exec(r);if(e)return{type:"br",raw:e[0]}}del(r,e,t=""){let s=this.rules.inline.delLDelim.exec(r);if(s&&(!s[1]||!t||this.rules.inline.punctuation.exec(t))){let n=[...s[0]].length-1,i,l,a=n,c=this.rules.inline.delRDelim;for(c.lastIndex=0,e=e.slice(-1*r.length+n);(s=c.exec(e))!==null;){if(i=s[1]||s[2]||s[3]||s[4]||s[5]||s[6],!i||(l=[...i].length,l!==n))continue;if(s[3]||s[4]){a+=l;continue}if(a-=l,a>0)continue;l=Math.min(l,l+a);let o=[...s[0]][0].length,p=r.slice(0,n+s.index+o+l),h=p.slice(n,-n);return{type:"del",raw:p,text:h,tokens:this.lexer.inlineTokens(h)}}}}autolink(r){let e=this.rules.inline.autolink.exec(r);if(e){let t,s;return e[2]==="@"?(t=e[1],s="mailto:"+t):(t=e[1],s=t),{type:"link",raw:e[0],text:t,href:s,tokens:[{type:"text",raw:t,text:t}]}}}url(r){let e;if(e=this.rules.inline.url.exec(r)){let t,s;if(e[2]==="@")t=e[0],s="mailto:"+t;else{let n;do n=e[0],e[0]=this.rules.inline._backpedal.exec(e[0])?.[0]??"";while(n!==e[0]);t=e[0],e[1]==="www."?s="http://"+e[0]:s=e[0]}return{type:"link",raw:e[0],text:t,href:s,tokens:[{type:"text",raw:t,text:t}]}}}inlineText(r){let e=this.rules.inline.text.exec(r);if(e){let t=this.lexer.state.inRawBlock;return{type:"text",raw:e[0],text:e[0],escaped:t}}}},b=class Q{tokens;options;state;inlineQueue;tokenizer;constructor(e){this.tokens=[],this.tokens.links=Object.create(null),this.options=e||T,this.options.tokenizer=this.options.tokenizer||new v,this.tokenizer=this.options.tokenizer,this.tokenizer.options=this.options,this.tokenizer.lexer=this,this.inlineQueue=[],this.state={inLink:!1,inRawBlock:!1,top:!0};let t={other:x,block:E.normal,inline:_.normal};this.options.pedantic?(t.block=E.pedantic,t.inline=_.pedantic):this.options.gfm&&(t.block=E.gfm,this.options.breaks?t.inline=_.breaks:t.inline=_.gfm),this.tokenizer.rules=t}static get rules(){return{block:E,inline:_}}static lex(e,t){return new Q(t).lex(e)}static lexInline(e,t){return new Q(t).inlineTokens(e)}lex(e){e=e.replace(x.carriageReturn,`\n`),this.blockTokens(e,this.tokens);for(let t=0;t<this.inlineQueue.length;t++){let s=this.inlineQueue[t];this.inlineTokens(s.src,s.tokens)}return this.inlineQueue=[],this.tokens}blockTokens(e,t=[],s=!1){this.tokenizer.lexer=this,this.options.pedantic&&(e=e.replace(x.tabCharGlobal,"    ").replace(x.spaceLine,""));let n=1/0;for(;e;){if(e.length<n)n=e.length;else{this.infiniteLoopError(e.charCodeAt(0));break}let i;if(this.options.extensions?.block?.some(a=>(i=a.call({lexer:this},e,t))?(e=e.substring(i.raw.length),t.push(i),!0):!1))continue;if(i=this.tokenizer.space(e)){e=e.substring(i.raw.length);let a=t.at(-1);i.raw.length===1&&a!==void 0?a.raw+=`\n`:t.push(i);continue}if(i=this.tokenizer.code(e)){e=e.substring(i.raw.length);let a=t.at(-1);a?.type==="paragraph"||a?.type==="text"?(a.raw+=(a.raw.endsWith(`\n`)?"":`\n`)+i.raw,a.text+=`\n`+i.text,this.inlineQueue.at(-1).src=a.text):t.push(i);continue}if(i=this.tokenizer.fences(e)){e=e.substring(i.raw.length),t.push(i);continue}if(i=this.tokenizer.heading(e)){e=e.substring(i.raw.length),t.push(i);continue}if(i=this.tokenizer.hr(e)){e=e.substring(i.raw.length),t.push(i);continue}if(i=this.tokenizer.blockquote(e)){e=e.substring(i.raw.length),t.push(i);continue}if(i=this.tokenizer.list(e)){e=e.substring(i.raw.length),t.push(i);continue}if(i=this.tokenizer.html(e)){e=e.substring(i.raw.length),t.push(i);continue}if(i=this.tokenizer.def(e)){e=e.substring(i.raw.length);let a=t.at(-1);a?.type==="paragraph"||a?.type==="text"?(a.raw+=(a.raw.endsWith(`\n`)?"":`\n`)+i.raw,a.text+=`\n`+i.raw,this.inlineQueue.at(-1).src=a.text):this.tokens.links[i.tag]||(this.tokens.links[i.tag]={href:i.href,title:i.title},t.push(i));continue}if(i=this.tokenizer.table(e)){e=e.substring(i.raw.length),t.push(i);continue}if(i=this.tokenizer.lheading(e)){e=e.substring(i.raw.length),t.push(i);continue}let l=e;if(this.options.extensions?.startBlock){let a=1/0,c=e.slice(1),o;this.options.extensions.startBlock.forEach(p=>{o=p.call({lexer:this},c),typeof o=="number"&&o>=0&&(a=Math.min(a,o))}),a<1/0&&a>=0&&(l=e.substring(0,a+1))}if(this.state.top&&(i=this.tokenizer.paragraph(l))){let a=t.at(-1);s&&a?.type==="paragraph"?(a.raw+=(a.raw.endsWith(`\n`)?"":`\n`)+i.raw,a.text+=`\n`+i.text,this.inlineQueue.pop(),this.inlineQueue.at(-1).src=a.text):t.push(i),s=l.length!==e.length,e=e.substring(i.raw.length);continue}if(i=this.tokenizer.text(e)){e=e.substring(i.raw.length);let a=t.at(-1);a?.type==="text"?(a.raw+=(a.raw.endsWith(`\n`)?"":`\n`)+i.raw,a.text+=`\n`+i.text,this.inlineQueue.pop(),this.inlineQueue.at(-1).src=a.text):t.push(i);continue}if(e){this.infiniteLoopError(e.charCodeAt(0));break}}return this.state.top=!0,t}inline(e,t=[]){return this.inlineQueue.push({src:e,tokens:t}),t}inlineTokens(e,t=[]){this.tokenizer.lexer=this;let s=e,n=null;if(this.tokens.links){let o=Object.keys(this.tokens.links);if(o.length>0)for(;(n=this.tokenizer.rules.inline.reflinkSearch.exec(s))!==null;)o.includes(n[0].slice(n[0].lastIndexOf("[")+1,-1))&&(s=s.slice(0,n.index)+"["+"a".repeat(n[0].length-2)+"]"+s.slice(this.tokenizer.rules.inline.reflinkSearch.lastIndex))}for(;(n=this.tokenizer.rules.inline.anyPunctuation.exec(s))!==null;)s=s.slice(0,n.index)+"++"+s.slice(this.tokenizer.rules.inline.anyPunctuation.lastIndex);let i;for(;(n=this.tokenizer.rules.inline.blockSkip.exec(s))!==null;)i=n[2]?n[2].length:0,s=s.slice(0,n.index+i)+"["+"a".repeat(n[0].length-i-2)+"]"+s.slice(this.tokenizer.rules.inline.blockSkip.lastIndex);s=this.options.hooks?.emStrongMask?.call({lexer:this},s)??s;let l=!1,a="",c=1/0;for(;e;){if(e.length<c)c=e.length;else{this.infiniteLoopError(e.charCodeAt(0));break}l||(a=""),l=!1;let o;if(this.options.extensions?.inline?.some(h=>(o=h.call({lexer:this},e,t))?(e=e.substring(o.raw.length),t.push(o),!0):!1))continue;if(o=this.tokenizer.escape(e)){e=e.substring(o.raw.length),t.push(o);continue}if(o=this.tokenizer.tag(e)){e=e.substring(o.raw.length),t.push(o);continue}if(o=this.tokenizer.link(e)){e=e.substring(o.raw.length),t.push(o);continue}if(o=this.tokenizer.reflink(e,this.tokens.links)){e=e.substring(o.raw.length);let h=t.at(-1);o.type==="text"&&h?.type==="text"?(h.raw+=o.raw,h.text+=o.text):t.push(o);continue}if(o=this.tokenizer.emStrong(e,s,a)){e=e.substring(o.raw.length),t.push(o);continue}if(o=this.tokenizer.codespan(e)){e=e.substring(o.raw.length),t.push(o);continue}if(o=this.tokenizer.br(e)){e=e.substring(o.raw.length),t.push(o);continue}if(o=this.tokenizer.del(e,s,a)){e=e.substring(o.raw.length),t.push(o);continue}if(o=this.tokenizer.autolink(e)){e=e.substring(o.raw.length),t.push(o);continue}if(!this.state.inLink&&(o=this.tokenizer.url(e))){e=e.substring(o.raw.length),t.push(o);continue}let p=e;if(this.options.extensions?.startInline){let h=1/0,g=e.slice(1),f;this.options.extensions.startInline.forEach(d=>{f=d.call({lexer:this},g),typeof f=="number"&&f>=0&&(h=Math.min(h,f))}),h<1/0&&h>=0&&(p=e.substring(0,h+1))}if(o=this.tokenizer.inlineText(p)){e=e.substring(o.raw.length),o.raw.slice(-1)!=="_"&&(a=o.raw.slice(-1)),l=!0;let h=t.at(-1);h?.type==="text"?(h.raw+=o.raw,h.text+=o.text):t.push(o);continue}if(e){this.infiniteLoopError(e.charCodeAt(0));break}}return t}infiniteLoopError(e){let t="Infinite loop on byte: "+e;if(this.options.silent)console.error(t);else throw new Error(t)}},B=class{options;parser;constructor(r){this.options=r||T}space(r){return""}code({text:r,lang:e,escaped:t}){let s=(e||"").match(x.notSpaceStart)?.[0],n=r.replace(x.endingNewline,"")+`\n`;return s?\'<pre><code class="language-\'+m(s)+\'">\'+(t?n:m(n,!0))+`</code></pre>\n`:"<pre><code>"+(t?n:m(n,!0))+`</code></pre>\n`}blockquote({tokens:r}){return`<blockquote>\n${this.parser.parse(r)}</blockquote>\n`}html({text:r}){return r}def(r){return""}heading({tokens:r,depth:e}){return`<h${e}>${this.parser.parseInline(r)}</h${e}>\n`}hr(r){return`<hr>\n`}list(r){let e=r.ordered,t=r.start,s="";for(let l=0;l<r.items.length;l++){let a=r.items[l];s+=this.listitem(a)}let n=e?"ol":"ul",i=e&&t!==1?\' start="\'+t+\'"\':"";return"<"+n+i+`>\n`+s+"</"+n+`>\n`}listitem(r){return`<li>${this.parser.parse(r.tokens)}</li>\n`}checkbox({checked:r}){return"<input "+(r?\'checked="" \':"")+\'disabled="" type="checkbox"> \'}paragraph({tokens:r}){return`<p>${this.parser.parseInline(r)}</p>\n`}table(r){let e="",t="";for(let n=0;n<r.header.length;n++)t+=this.tablecell(r.header[n]);e+=this.tablerow({text:t});let s="";for(let n=0;n<r.rows.length;n++){let i=r.rows[n];t="";for(let l=0;l<i.length;l++)t+=this.tablecell(i[l]);s+=this.tablerow({text:t})}return s&&(s=`<tbody>${s}</tbody>`),`<table>\n<thead>\n`+e+`</thead>\n`+s+`</table>\n`}tablerow({text:r}){return`<tr>\n${r}</tr>\n`}tablecell(r){let e=this.parser.parseInline(r.tokens),t=r.header?"th":"td";return(r.align?`<${t} align="${r.align}">`:`<${t}>`)+e+`</${t}>\n`}strong({tokens:r}){return`<strong>${this.parser.parseInline(r)}</strong>`}em({tokens:r}){return`<em>${this.parser.parseInline(r)}</em>`}codespan({text:r}){return`<code>${m(r,!0)}</code>`}br(r){return"<br>"}del({tokens:r}){return`<del>${this.parser.parseInline(r)}</del>`}link({href:r,title:e,tokens:t}){let s=this.parser.parseInline(t),n=ne(r);if(n===null)return s;r=n;let i=\'<a href="\'+r+\'"\';return e&&(i+=\' title="\'+m(e)+\'"\'),i+=">"+s+"</a>",i}image({href:r,title:e,text:t,tokens:s}){s&&(t=this.parser.parseInline(s,this.parser.textRenderer));let n=ne(r);if(n===null)return m(t);r=n;let i=`<img src="${r}" alt="${m(t)}"`;return e&&(i+=` title="${m(e)}"`),i+=">",i}text(r){return"tokens"in r&&r.tokens?this.parser.parseInline(r.tokens):"escaped"in r&&r.escaped?r.text:m(r.text)}},J=class{strong({text:r}){return r}em({text:r}){return r}codespan({text:r}){return r}del({text:r}){return r}html({text:r}){return r}text({text:r}){return r}link({text:r}){return""+r}image({text:r}){return""+r}br(){return""}checkbox({raw:r}){return r}},w=class H{options;renderer;textRenderer;constructor(e){this.options=e||T,this.options.renderer=this.options.renderer||new B,this.renderer=this.options.renderer,this.renderer.options=this.options,this.renderer.parser=this,this.textRenderer=new J}static parse(e,t){return new H(t).parse(e)}static parseInline(e,t){return new H(t).parseInline(e)}parse(e){this.renderer.parser=this;let t="";for(let s=0;s<e.length;s++){let n=e[s];if(this.options.extensions?.renderers?.[n.type]){let l=n,a=this.options.extensions.renderers[l.type].call({parser:this},l);if(a!==!1||!["space","hr","heading","code","table","blockquote","list","html","def","paragraph","text"].includes(l.type)){t+=a||"";continue}}let i=n;switch(i.type){case"space":{t+=this.renderer.space(i);break}case"hr":{t+=this.renderer.hr(i);break}case"heading":{t+=this.renderer.heading(i);break}case"code":{t+=this.renderer.code(i);break}case"table":{t+=this.renderer.table(i);break}case"blockquote":{t+=this.renderer.blockquote(i);break}case"list":{t+=this.renderer.list(i);break}case"checkbox":{t+=this.renderer.checkbox(i);break}case"html":{t+=this.renderer.html(i);break}case"def":{t+=this.renderer.def(i);break}case"paragraph":{t+=this.renderer.paragraph(i);break}case"text":{t+=this.renderer.text(i);break}default:{let l=\'Token with "\'+i.type+\'" type was not found.\';if(this.options.silent)return console.error(l),"";throw new Error(l)}}}return t}parseInline(e,t=this.renderer){this.renderer.parser=this;let s="";for(let n=0;n<e.length;n++){let i=e[n];if(this.options.extensions?.renderers?.[i.type]){let a=this.options.extensions.renderers[i.type].call({parser:this},i);if(a!==!1||!["escape","html","link","image","strong","em","codespan","br","del","text"].includes(i.type)){s+=a||"";continue}}let l=i;switch(l.type){case"escape":{s+=t.text(l);break}case"html":{s+=t.html(l);break}case"link":{s+=t.link(l);break}case"image":{s+=t.image(l);break}case"checkbox":{s+=t.checkbox(l);break}case"strong":{s+=t.strong(l);break}case"em":{s+=t.em(l);break}case"codespan":{s+=t.codespan(l);break}case"br":{s+=t.br(l);break}case"del":{s+=t.del(l);break}case"text":{s+=t.text(l);break}default:{let a=\'Token with "\'+l.type+\'" type was not found.\';if(this.options.silent)return console.error(a),"";throw new Error(a)}}}return s}},L=class{options;block;constructor(r){this.options=r||T}static passThroughHooks=new Set(["preprocess","postprocess","processAllTokens","emStrongMask"]);static passThroughHooksRespectAsync=new Set(["preprocess","postprocess","processAllTokens"]);preprocess(r){return r}postprocess(r){return r}processAllTokens(r){return r}emStrongMask(r){return r}provideLexer(r=this.block){return r?b.lex:b.lexInline}provideParser(r=this.block){return r?w.parse:w.parseInline}},at=class{defaults=O();options=this.setOptions;parse=this.parseMarkdown(!0);parseInline=this.parseMarkdown(!1);Parser=w;Renderer=B;TextRenderer=J;Lexer=b;Tokenizer=v;Hooks=L;constructor(...r){this.use(...r)}walkTokens(r,e){let t=[];for(let s of r)switch(t=t.concat(e.call(this,s)),s.type){case"table":{let n=s;for(let i of n.header)t=t.concat(this.walkTokens(i.tokens,e));for(let i of n.rows)for(let l of i)t=t.concat(this.walkTokens(l.tokens,e));break}case"list":{let n=s;t=t.concat(this.walkTokens(n.items,e));break}default:{let n=s;this.defaults.extensions?.childTokens?.[n.type]?this.defaults.extensions.childTokens[n.type].forEach(i=>{let l=n[i].flat(1/0);t=t.concat(this.walkTokens(l,e))}):n.tokens&&(t=t.concat(this.walkTokens(n.tokens,e)))}}return t}use(...r){let e=this.defaults.extensions||{renderers:{},childTokens:{}};return r.forEach(t=>{let s={...t};if(s.async=this.defaults.async||s.async||!1,t.extensions&&(t.extensions.forEach(n=>{if(!n.name)throw new Error("extension name required");if("renderer"in n){let i=e.renderers[n.name];i?e.renderers[n.name]=function(...l){let a=n.renderer.apply(this,l);return a===!1&&(a=i.apply(this,l)),a}:e.renderers[n.name]=n.renderer}if("tokenizer"in n){if(!n.level||n.level!=="block"&&n.level!=="inline")throw new Error("extension level must be \'block\' or \'inline\'");let i=e[n.level];i?i.unshift(n.tokenizer):e[n.level]=[n.tokenizer],n.start&&(n.level==="block"?e.startBlock?e.startBlock.push(n.start):e.startBlock=[n.start]:n.level==="inline"&&(e.startInline?e.startInline.push(n.start):e.startInline=[n.start]))}"childTokens"in n&&n.childTokens&&(e.childTokens[n.name]=n.childTokens)}),s.extensions=e),t.renderer){let n=this.defaults.renderer||new B(this.defaults);for(let i in t.renderer){if(!(i in n))throw new Error(`renderer \'${i}\' does not exist`);if(["options","parser"].includes(i))continue;let l=i,a=t.renderer[l],c=n[l];n[l]=(...o)=>{let p=a.apply(n,o);return p===!1&&(p=c.apply(n,o)),p||""}}s.renderer=n}if(t.tokenizer){let n=this.defaults.tokenizer||new v(this.defaults);for(let i in t.tokenizer){if(!(i in n))throw new Error(`tokenizer \'${i}\' does not exist`);if(["options","rules","lexer"].includes(i))continue;let l=i,a=t.tokenizer[l],c=n[l];n[l]=(...o)=>{let p=a.apply(n,o);return p===!1&&(p=c.apply(n,o)),p}}s.tokenizer=n}if(t.hooks){let n=this.defaults.hooks||new L;for(let i in t.hooks){if(!(i in n))throw new Error(`hook \'${i}\' does not exist`);if(["options","block"].includes(i))continue;let l=i,a=t.hooks[l],c=n[l];L.passThroughHooks.has(i)?n[l]=o=>{if(this.defaults.async&&L.passThroughHooksRespectAsync.has(i))return(async()=>{let h=await a.call(n,o);return c.call(n,h)})();let p=a.call(n,o);return c.call(n,p)}:n[l]=(...o)=>{if(this.defaults.async)return(async()=>{let h=await a.apply(n,o);return h===!1&&(h=await c.apply(n,o)),h})();let p=a.apply(n,o);return p===!1&&(p=c.apply(n,o)),p}}s.hooks=n}if(t.walkTokens){let n=this.defaults.walkTokens,i=t.walkTokens;s.walkTokens=function(l){let a=[];return a.push(i.call(this,l)),n&&(a=a.concat(n.call(this,l))),a}}this.defaults={...this.defaults,...s}}),this}setOptions(r){return this.defaults={...this.defaults,...r},this}lexer(r,e){return b.lex(r,e??this.defaults)}parser(r,e){return w.parse(r,e??this.defaults)}parseMarkdown(r){return(e,t)=>{let s={...t},n={...this.defaults,...s},i=this.onError(!!n.silent,!!n.async);if(this.defaults.async===!0&&s.async===!1)return i(new Error("marked(): The async option was set to true by an extension. Remove async: false from the parse options object to return a Promise."));if(typeof e>"u"||e===null)return i(new Error("marked(): input parameter is undefined or null"));if(typeof e!="string")return i(new Error("marked(): input parameter is of type "+Object.prototype.toString.call(e)+", string expected"));if(n.hooks&&(n.hooks.options=n,n.hooks.block=r),n.async)return(async()=>{let l=n.hooks?await n.hooks.preprocess(e):e,a=await(n.hooks?await n.hooks.provideLexer(r):r?b.lex:b.lexInline)(l,n),c=n.hooks?await n.hooks.processAllTokens(a):a;n.walkTokens&&await Promise.all(this.walkTokens(c,n.walkTokens));let o=await(n.hooks?await n.hooks.provideParser(r):r?w.parse:w.parseInline)(c,n);return n.hooks?await n.hooks.postprocess(o):o})().catch(i);try{n.hooks&&(e=n.hooks.preprocess(e));let l=(n.hooks?n.hooks.provideLexer(r):r?b.lex:b.lexInline)(e,n);n.hooks&&(l=n.hooks.processAllTokens(l)),n.walkTokens&&this.walkTokens(l,n.walkTokens);let a=(n.hooks?n.hooks.provideParser(r):r?w.parse:w.parseInline)(l,n);return n.hooks&&(a=n.hooks.postprocess(a)),a}catch(l){return i(l)}}}onError(r,e){return t=>{if(t.message+=`\nPlease report this to https://github.com/markedjs/marked.`,r){let s="<p>An error occurred:</p><pre>"+m(t.message+"",!0)+"</pre>";return e?Promise.resolve(s):s}if(e)return Promise.reject(t);throw t}}},S=new at;function k(r,e){return S.parse(r,e)}k.options=k.setOptions=function(r){return S.setOptions(r),k.defaults=S.defaults,ae(k.defaults),k};k.getDefaults=O;k.defaults=T;k.use=function(...r){return S.use(...r),k.defaults=S.defaults,ae(k.defaults),k};k.walkTokens=function(r,e){return S.walkTokens(r,e)};k.parseInline=S.parseInline;k.Parser=w;k.parser=w.parse;k.Renderer=B;k.TextRenderer=J;k.Lexer=b;k.lexer=b.lex;k.Tokenizer=v;k.Hooks=L;k.parse=k;var ot=k.options,ct=k.setOptions,ht=k.use,pt=k.walkTokens,ut=k.parseInline;var gt=w.parse,kt=b.lex;k.use({extensions:[{name:"inlineMath",level:"inline",start(r){return r.match(/\\$/)?.index},tokenizer(r){let e=/^\\$([^$]+)\\$/.exec(r);if(e)return{type:"inlineMath",raw:e[0],text:e[1].trim()}},renderer(r){return r.raw}}]});self.onmessage=r=>{let{id:e,text:t}=r.data;try{let s=k.lexer(t);self.postMessage({id:e,tokens:s})}catch(s){self.postMessage({id:e,error:String(s)})}};})();\n';

// src/Markdown.ts
import_marked.marked.use({
  extensions: [
    {
      name: "inlineMath",
      level: "inline",
      start(src) {
        return src.match(/\$/)?.index;
      },
      tokenizer(src) {
        const match = /^\$([^$]+)\$/.exec(src);
        if (match) {
          return {
            type: "inlineMath",
            raw: match[0],
            text: match[1].trim()
          };
        }
        return void 0;
      },
      renderer(token) {
        return token.raw;
      }
    }
  ]
});
var adaptor = (0, import_liteAdaptor.liteAdaptor)();
(0, import_html.RegisterHTMLHandler)(adaptor);
var tex = new import_tex.TeX({ packages: import_AllPackages.AllPackages });
var svg = new import_svg.SVG({ fontCache: "local" });
var htmlMathJax = import_mathjax.mathjax.document("", { InputJax: tex, OutputJax: svg });
function renderMathToSVGDataURI(formula, displayMode) {
  try {
    const node = htmlMathJax.convert(formula, { display: displayMode });
    const svgString = adaptor.innerHTML(node);
    const wMatch = svgString.match(/width="([^"]+)ex"/);
    const hMatch = svgString.match(/height="([^"]+)ex"/);
    const wEx = wMatch ? parseFloat(wMatch[1]) : 10;
    const hEx = hMatch ? parseFloat(hMatch[1]) : 2;
    const width = wEx * 8;
    const height = hEx * 8;
    const base64 = btoa(unescape(encodeURIComponent(svgString)));
    return { uri: `data:image/svg+xml;base64,${base64}`, width, height };
  } catch (e) {
    console.error("MathJax error", e);
    return null;
  }
}
var markdownWorker = null;
var workerIdCounter = 0;
var workerCallbacks = /* @__PURE__ */ new Map();
function runSyncFallback(entry) {
  try {
    entry.cb(import_marked.marked.lexer(entry.text));
  } catch (err) {
    console.warn("Markdown sync fallback parse failed", err);
  }
}
if (typeof Worker !== "undefined") {
  try {
    const blob = new Blob([WORKER_SOURCE_STRING], { type: "application/javascript" });
    markdownWorker = new Worker(URL.createObjectURL(blob));
    markdownWorker.onmessage = (e) => {
      const { id, tokens, error } = e.data;
      const entry = workerCallbacks.get(id);
      if (entry) {
        workerCallbacks.delete(id);
        if (!error) entry.cb(tokens);
        else runSyncFallback(entry);
      }
    };
    markdownWorker.onerror = () => {
      const pending = [...workerCallbacks.values()];
      workerCallbacks.clear();
      markdownWorker = null;
      for (const entry of pending) runSyncFallback(entry);
    };
  } catch (err) {
    console.warn("Failed to initialize MarkdownWorker", err);
  }
}
var DEFAULT_THEME = {
  textColor: "#e2e8f0",
  headingColor: "#f8fafc",
  codeColor: "#a5f3fc",
  codeBgColor: "rgba(30, 41, 59, 0.85)",
  quoteBorderColor: "#6366f1",
  quoteTextColor: "#94a3b8",
  hrColor: "rgba(148, 163, 184, 0.3)",
  tableBgColor: "rgba(15, 15, 25, 0.4)",
  tableHeaderBgColor: "rgba(255, 255, 255, 0.08)",
  bodyFont: "Inter, system-ui, sans-serif",
  codeFont: 'ui-monospace, "JetBrains Mono", "Fira Code", monospace',
  fontSize: 16
};
var HorizontalRule = class extends import_core9.Entity {
  color;
  constructor(w, color) {
    super();
    this.width = w;
    this.height = 1;
    this.color = color;
  }
  isPointInside() {
    return false;
  }
  render(r) {
    r.beginPath();
    r.moveTo(0, 0);
    r.lineTo(this.width, 0);
    r.stroke(this.color, 1);
  }
};
var QuoteBorder = class extends import_core9.Entity {
  color;
  constructor(height, color) {
    super();
    this.width = 4;
    this.height = height;
    this.color = color;
  }
  isPointInside() {
    return false;
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 2);
    r.fill(this.color);
  }
};
var MarkdownContainer = class extends import_core9.Entity {
  isPointInside(_globalX, _globalY) {
    return false;
  }
  render(_r) {
  }
};
var KEYWORD_SETS = {
  js: /* @__PURE__ */ new Set([
    "const",
    "let",
    "var",
    "function",
    "return",
    "if",
    "else",
    "for",
    "while",
    "class",
    "extends",
    "new",
    "this",
    "import",
    "export",
    "from",
    "default",
    "async",
    "await",
    "try",
    "catch",
    "throw",
    "of",
    "in",
    "typeof",
    "instanceof",
    "switch",
    "case",
    "break",
    "continue",
    "null",
    "undefined",
    "true",
    "false"
  ]),
  ts: /* @__PURE__ */ new Set([
    "const",
    "let",
    "var",
    "function",
    "return",
    "if",
    "else",
    "for",
    "while",
    "class",
    "extends",
    "new",
    "this",
    "import",
    "export",
    "from",
    "default",
    "async",
    "await",
    "try",
    "catch",
    "throw",
    "of",
    "in",
    "typeof",
    "instanceof",
    "switch",
    "case",
    "break",
    "continue",
    "null",
    "undefined",
    "true",
    "false",
    "type",
    "interface",
    "enum",
    "as",
    "is",
    "readonly",
    "implements",
    "abstract",
    "public",
    "private",
    "protected",
    "static",
    "void",
    "never",
    "any",
    "unknown"
  ]),
  py: /* @__PURE__ */ new Set([
    "def",
    "class",
    "return",
    "if",
    "elif",
    "else",
    "for",
    "while",
    "import",
    "from",
    "as",
    "with",
    "try",
    "except",
    "raise",
    "finally",
    "pass",
    "break",
    "continue",
    "and",
    "or",
    "not",
    "in",
    "is",
    "None",
    "True",
    "False",
    "yield",
    "lambda",
    "global",
    "nonlocal",
    "del",
    "assert",
    "async",
    "await"
  ]),
  rust: /* @__PURE__ */ new Set([
    "fn",
    "let",
    "mut",
    "const",
    "if",
    "else",
    "for",
    "while",
    "loop",
    "match",
    "return",
    "struct",
    "enum",
    "impl",
    "trait",
    "pub",
    "use",
    "mod",
    "crate",
    "self",
    "super",
    "where",
    "as",
    "in",
    "ref",
    "move",
    "async",
    "await",
    "true",
    "false",
    "type",
    "unsafe",
    "extern",
    "dyn",
    "static"
  ])
};
KEYWORD_SETS["javascript"] = KEYWORD_SETS["js"];
KEYWORD_SETS["typescript"] = KEYWORD_SETS["ts"];
KEYWORD_SETS["python"] = KEYWORD_SETS["py"];
KEYWORD_SETS["rs"] = KEYWORD_SETS["rust"];
function highlightLine(line, lang, theme) {
  const keywords = KEYWORD_SETS[lang];
  if (!keywords) {
    return [{ text: line, color: theme.codeColor }];
  }
  const segments = [];
  const KEYWORD_COLOR = "#c084fc";
  const STRING_COLOR = "#86efac";
  const COMMENT_COLOR = "#64748b";
  const NUMBER_COLOR = "#fbbf24";
  let i = 0;
  let buf = "";
  const flush = (color) => {
    if (buf) {
      segments.push({ text: buf, color });
      buf = "";
    }
  };
  while (i < line.length) {
    const ch = line[i];
    if (ch === "/" && line[i + 1] === "/") {
      flush(theme.codeColor);
      segments.push({ text: line.slice(i), color: COMMENT_COLOR });
      return segments;
    }
    if (ch === "#" && (lang === "py" || lang === "python" || lang === "rust" || lang === "rs")) {
      flush(theme.codeColor);
      segments.push({ text: line.slice(i), color: COMMENT_COLOR });
      return segments;
    }
    if (ch === '"' || ch === "'" || ch === "`") {
      flush(theme.codeColor);
      const quote = ch;
      let j = i + 1;
      while (j < line.length && line[j] !== quote) {
        if (line[j] === "\\") j++;
        j++;
      }
      j++;
      segments.push({ text: line.slice(i, j), color: STRING_COLOR });
      i = j;
      continue;
    }
    if (/\d/.test(ch) && (i === 0 || /[\s(,=+\-*/<>[\]{}:;]/.test(line[i - 1]))) {
      flush(theme.codeColor);
      let j = i;
      while (j < line.length && /[\d._xXa-fA-F]/.test(line[j])) j++;
      segments.push({ text: line.slice(i, j), color: NUMBER_COLOR });
      i = j;
      continue;
    }
    if (/[a-zA-Z_]/.test(ch)) {
      flush(theme.codeColor);
      let j = i;
      while (j < line.length && /[a-zA-Z0-9_]/.test(line[j])) j++;
      const word = line.slice(i, j);
      segments.push({ text: word, color: keywords.has(word) ? KEYWORD_COLOR : theme.codeColor });
      i = j;
      continue;
    }
    buf += ch;
    i++;
  }
  flush(theme.codeColor);
  return segments;
}
var CodeBlock = class extends UIComponent {
  lines;
  grid = null;
  cellWidth = 0;
  source;
  lang;
  theme;
  lineH = 24;
  pad = 18;
  codeFont;
  selectable;
  constructor(code, lang, maxWidth, theme, selectable = true) {
    super();
    this.source = code;
    this.lang = lang;
    this.theme = theme;
    this.codeFont = `15px ${theme.codeFont}`;
    this.selectable = selectable;
    this.lines = [];
    this.width = maxWidth;
    this.buildLines(code);
  }
  /** Re-parse code content (e.g. for live editing). */
  setCode(code, lang) {
    if (lang !== void 0) this.lang = lang;
    this.source = code;
    this.buildLines(code);
    this.scene?.markDirty();
    return this;
  }
  /** Enable or disable browser-native selection for this code block. */
  setSelectable(selectable) {
    this.selectable = selectable;
    this.scene?.markDirty();
    return this;
  }
  getContentProjection() {
    if (!this.source) return null;
    const grid = this.ensureGrid();
    return {
      text: this.source,
      font: this.codeFont,
      lineHeight: this.lineH,
      // Every row is absolutely positioned from the same local coordinates as
      // render(). A single pre-wrap DOM text node would introduce browser
      // wrapping for long source lines that canvas intentionally keeps intact.
      lines: grid.lines.map((line, row) => ({
        text: this.source.slice(line.sourceStart, line.sourceEnd),
        separatorAfter: this.source.slice(line.sourceEnd, line.nextSourceStart) || void 0,
        x: this.pad,
        y: this.pad + row * this.lineH,
        baseline: this.lineH * 0.75,
        font: this.codeFont,
        lineHeight: this.lineH
      })),
      selectable: this.selectable,
      // render() draws cell-by-cell (no ligatures can form); the DOM copy
      // must not ligate either or Firefox selection geometry drifts.
      ligatures: "none",
      grid
    };
  }
  buildLines(code) {
    const rawLines = code.split(/\r\n|\r|\n/);
    this.lines = rawLines.map((l) => highlightLine(l, this.lang, this.theme));
    this.grid = null;
    this.height = this.pad * 2 + rawLines.length * this.lineH;
  }
  ensureGrid() {
    const cellWidth = this.cellWidth || Math.max(1, measureText("M", this.codeFont));
    if (!this.grid || this.grid.source !== this.source || this.grid.font !== this.codeFont || this.grid.cellWidth !== cellWidth) {
      this.grid = (0, import_core9.prepareContentGrid)(this.source, {
        font: this.codeFont,
        cellWidth,
        lineHeight: this.lineH,
        baseline: this.lineH * 0.75
      });
    }
    return this.grid;
  }
  /** Code blocks are decorative — not interactive. */
  isPointInside() {
    return false;
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 8);
    r.fill(this.theme.codeBgColor);
    const grid = this.ensureGrid();
    for (let row = 0; row < grid.lines.length; row++) {
      const yBaseline = this.pad + row * this.lineH + this.lineH * 0.75;
      const segments = this.lines[row];
      let segmentIndex = 0;
      let segmentEnd = segments[0]?.text.length ?? 0;
      const lineStart = grid.lines[row].sourceStart;
      for (const cell of grid.lines[row].cells) {
        const localSourceStart = cell.sourceStart - lineStart;
        while (segmentIndex < segments.length - 1 && localSourceStart >= segmentEnd) {
          segmentIndex++;
          segmentEnd += segments[segmentIndex].text.length;
        }
        const sourceText = this.source.slice(cell.sourceStart, cell.sourceEnd);
        if (cell.advance <= 0 || sourceText === " " || sourceText === "	") continue;
        r.fillText(
          cell.glyph,
          this.pad + cell.x,
          yBaseline,
          this.codeFont,
          segments[segmentIndex]?.color ?? this.theme.codeColor
        );
      }
    }
  }
};
function decodeEntities(text) {
  return text.replace(/&lt;/g, "<").replace(/&gt;/g, ">").replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&amp;/g, "&");
}
function collectSpans(tokens, inherited, theme, out) {
  for (const token of tokens) {
    switch (token.type) {
      case "strong": {
        const t = token;
        if (t.tokens) {
          collectSpans(t.tokens, { ...inherited, bold: true }, theme, out);
        } else {
          out.push({ text: decodeEntities(t.text), style: { ...inherited, bold: true } });
        }
        break;
      }
      case "em": {
        const t = token;
        if (t.tokens) {
          collectSpans(t.tokens, { ...inherited, italic: true }, theme, out);
        } else {
          out.push({ text: decodeEntities(t.text), style: { ...inherited, italic: true } });
        }
        break;
      }
      case "codespan": {
        const t = token;
        out.push({ text: decodeEntities(t.text), style: { ...inherited, color: theme.codeColor } });
        break;
      }
      case "br": {
        out.push({ text: "\n" });
        break;
      }
      case "html": {
        const t = token;
        const raw = t.raw ?? t.text ?? "";
        const brCount = (raw.match(/<br\s*\/?>/gi) ?? []).length;
        for (let i = 0; i < brCount; i++) out.push({ text: "\n" });
        break;
      }
      case "inlineMath": {
        const t = token;
        out.push({ text: decodeEntities(t.raw), style: { ...inherited, color: "#fcd34d" } });
        break;
      }
      case "link": {
        const t = token;
        const linkStyle = { ...inherited, href: t.href, color: "#38bdf8" };
        if (t.tokens && t.tokens.length > 0) {
          collectSpans(t.tokens, linkStyle, theme, out);
        } else {
          out.push({ text: decodeEntities(t.text), style: linkStyle });
        }
        break;
      }
      case "text": {
        const t = token;
        if ("tokens" in t && t.tokens?.length) {
          collectSpans(t.tokens, inherited, theme, out);
        } else {
          const decoded = decodeEntities(t.text);
          if (decoded) {
            const style = Object.keys(inherited).length > 0 ? inherited : void 0;
            out.push({ text: decoded, style });
          }
        }
        break;
      }
      default: {
        if ("text" in token) {
          const decoded = decodeEntities(token.text);
          if (decoded) {
            const style = Object.keys(inherited).length > 0 ? inherited : void 0;
            out.push({ text: decoded, style });
          }
        }
        break;
      }
    }
  }
}
function renderInlineToRichText(tokens, fallbackText, font, color, maxWidth, theme, selectable, onLinkClick) {
  const spans = [];
  if (tokens && tokens.length > 0) {
    collectSpans(tokens, {}, theme, spans);
  }
  if (spans.length === 0) {
    spans.push({ text: decodeEntities(fallbackText) });
  }
  return new RichText(spans, {
    font,
    color,
    maxWidth,
    linkColor: "#38bdf8",
    selectable,
    onLinkClick
  });
}
var Markdown = class extends UIComponent {
  content;
  maxWidth;
  theme;
  onLinkClick;
  selectable;
  onLayoutUpdated;
  rawMarkdown;
  tokens;
  constructor(markdownText, opts = {}) {
    super();
    this.maxWidth = opts.maxWidth ?? 800;
    this.theme = { ...DEFAULT_THEME, ...opts.theme };
    this.onLinkClick = opts.onLinkClick;
    this.selectable = opts.selectable ?? true;
    this.content = new Stack({ direction: "vertical", gap: 16 });
    this.add(this.content);
    this.rawMarkdown = markdownText;
    this.tokens = [];
    this.renderMarkdown(markdownText);
  }
  renderMarkdown(text) {
    const tokens = import_marked.marked.lexer(text);
    this.tokens = tokens;
    for (const token of tokens) {
      const el = this.renderToken(token);
      if (el) {
        this.content.add(el);
      }
    }
    this.width = this.content.width;
    this.height = this.content.height;
  }
  /** Replace all markdown content (full rebuild). */
  setContent(markdown) {
    this.rawMarkdown = markdown;
    while (this.content.children.length > 0) {
      this.content.remove(this.content.children[this.content.children.length - 1]);
    }
    this.tokens = [];
    this.renderMarkdown(markdown);
    return this;
  }
  /** Enable or disable native selection for existing and future Markdown text. */
  setSelectable(selectable) {
    this.selectable = selectable;
    const apply = (entity) => {
      const candidate = entity;
      candidate.setSelectable?.(selectable);
      for (const child of entity.children) apply(child);
    };
    for (const child of this.content.children) apply(child);
    this.scene?.markDirty();
    return this;
  }
  /** Append a markdown chunk incrementally. Reuses unchanged prefix entities. */
  appendMarkdown(chunk) {
    this.rawMarkdown += chunk;
    if (markdownWorker) {
      const id = workerIdCounter++;
      workerCallbacks.set(id, {
        cb: (newTokens) => {
          this.updateTokens(newTokens);
        },
        text: this.rawMarkdown
      });
      markdownWorker.postMessage({ id, text: this.rawMarkdown });
    } else {
      const newTokens = import_marked.marked.lexer(this.rawMarkdown);
      this.updateTokens(newTokens);
    }
    return this;
  }
  updateTokens(newTokens) {
    const oldTokens = this.tokens;
    const oldChildren = [...this.content.children];
    let matchLen = 0;
    const minLen = Math.min(oldTokens.length, newTokens.length);
    let childIdx = 0;
    const oldTokenToChild = [];
    for (let i = 0; i < oldTokens.length; i++) {
      oldTokenToChild.push(childIdx);
      if (oldTokens[i].type !== "space") childIdx++;
    }
    for (let i = 0; i < minLen; i++) {
      if (oldTokens[i].raw === newTokens[i].raw) {
        matchLen++;
      } else {
        break;
      }
    }
    if (matchLen === oldTokens.length - 1 && matchLen < newTokens.length && oldTokens[matchLen]?.type === newTokens[matchLen]?.type && newTokens[matchLen]?.type === "paragraph") {
      const entityIdx = oldTokenToChild[matchLen];
      const existingEntity = oldChildren[entityIdx];
      if (existingEntity && "setSpans" in existingEntity) {
        const pToken = newTokens[matchLen];
        const t = this.theme;
        const spans = [];
        if (pToken.tokens && pToken.tokens.length > 0) {
          collectSpans(pToken.tokens, {}, t, spans);
        }
        if (spans.length === 0) {
          spans.push({ text: pToken.text });
        }
        existingEntity.setSpans(spans);
        matchLen++;
      }
    }
    for (let i = 0; i < oldTokens.length; i++) {
      if (i >= matchLen && oldTokens[i].type !== "space") {
        const idx = oldTokenToChild[i];
        if (idx < oldChildren.length) {
          this.content.remove(oldChildren[idx]);
        }
      }
    }
    for (let i = matchLen; i < newTokens.length; i++) {
      const el = this.renderToken(newTokens[i]);
      if (el) this.content.add(el);
    }
    this.tokens = newTokens;
    this.content.layout();
    this.width = this.content.width;
    this.height = this.content.height;
    this.scene?.markDirty();
    if (this.onLayoutUpdated) {
      this.onLayoutUpdated();
    }
  }
  renderToken(token) {
    const t = this.theme;
    const bodyFont = `${t.fontSize}px ${t.bodyFont}`;
    switch (token.type) {
      // ── Headings ─────────────────────────────────────────────────────
      case "heading": {
        const hToken = token;
        const sizes = [32, 28, 24, 20, 18, 16];
        const size = sizes[Math.min(hToken.depth - 1, 5)];
        const headingFont = `bold ${size}px ${t.bodyFont}`;
        return renderInlineToRichText(
          hToken.tokens,
          hToken.text,
          headingFont,
          t.headingColor,
          this.maxWidth,
          t,
          this.selectable,
          this.onLinkClick
        );
      }
      // ── Paragraphs ───────────────────────────────────────────────────
      case "paragraph": {
        const pToken = token;
        if (!pToken.tokens || !pToken.tokens.some((t2) => t2.type === "image")) {
          return renderInlineToRichText(
            pToken.tokens,
            pToken.text,
            bodyFont,
            t.textColor,
            this.maxWidth,
            t,
            this.selectable,
            this.onLinkClick
          );
        }
        const stack = new Stack({ direction: "vertical", gap: 16, maxWidth: this.maxWidth });
        let currentTokens = [];
        const flushText = () => {
          if (currentTokens.length > 0) {
            stack.add(
              renderInlineToRichText(
                currentTokens,
                "",
                bodyFont,
                t.textColor,
                this.maxWidth,
                t,
                this.selectable,
                this.onLinkClick
              )
            );
            currentTokens = [];
          }
        };
        for (const child of pToken.tokens) {
          if (child.type === "image") {
            flushText();
            const imgToken = child;
            const initialWidth = Math.min(800, this.maxWidth);
            const initialHeight = Math.round(initialWidth * 0.6);
            const img = new Image(imgToken.href, {
              width: initialWidth,
              height: initialHeight,
              alt: imgToken.text,
              radius: 8,
              onLoad: () => {
                const bmp = img.bitmap;
                if (bmp && bmp.naturalWidth && bmp.naturalHeight) {
                  const aspect = bmp.naturalHeight / bmp.naturalWidth;
                  img.width = Math.min(bmp.naturalWidth, this.maxWidth);
                  img.height = Math.round(img.width * aspect);
                  if (this.scene) this.scene.markDirty();
                }
              }
            });
            stack.add(img);
          } else {
            currentTokens.push(child);
          }
        }
        flushText();
        return stack;
      }
      // ── Code blocks ──────────────────────────────────────────────────
      case "code": {
        const codeToken = token;
        const lang = (codeToken.lang ?? "").toLowerCase();
        if (lang === "math" || lang === "latex" || lang === "tex") {
          const mathData = renderMathToSVGDataURI(codeToken.text, true);
          if (mathData) {
            const mathImg = new Image(mathData.uri, {
              width: Math.min(this.maxWidth, mathData.width),
              height: mathData.height * Math.min(1, this.maxWidth / mathData.width),
              alt: codeToken.text
            });
            const wrapper = new MarkdownContainer();
            mathImg.x = 16;
            mathImg.y = 8;
            wrapper.add(mathImg);
            wrapper.width = mathImg.width + 16;
            wrapper.height = mathImg.height + 16;
            return wrapper;
          }
        }
        return new CodeBlock(codeToken.text, lang, this.maxWidth, t, this.selectable);
      }
      // ── Blockquotes ──────────────────────────────────────────────────
      case "blockquote": {
        const bqToken = token;
        const innerStack = new Stack({ direction: "vertical", gap: 8 });
        if (bqToken.tokens) {
          for (const inner of bqToken.tokens) {
            const el = this.renderToken(inner);
            if (el) {
              const wrapper = new MarkdownContainer();
              el.x = 16;
              wrapper.add(el);
              wrapper.width = el.width + 16;
              wrapper.height = el.height;
              innerStack.add(wrapper);
            }
          }
        }
        const border = new QuoteBorder(innerStack.height || 20, t.quoteBorderColor);
        const container = new MarkdownContainer();
        border.x = 0;
        border.y = 0;
        container.add(border);
        innerStack.y = 0;
        innerStack.x = 0;
        container.add(innerStack);
        container.width = this.maxWidth;
        container.height = Math.max(border.height, innerStack.height);
        return container;
      }
      // ── Lists ────────────────────────────────────────────────────────
      case "list": {
        const listToken = token;
        const listStack = new Stack({ direction: "vertical", gap: 6 });
        for (let i = 0; i < listToken.items.length; i++) {
          const item = listToken.items[i];
          const bullet = listToken.ordered ? `${Number(listToken.start ?? 1) + i}. ` : "\u2022 ";
          const itemSpans = [{ text: bullet }];
          if (item.tokens && item.tokens.length > 0) {
            for (const inner of item.tokens) {
              if (inner.type === "text" && "tokens" in inner && inner.tokens?.length) {
                collectSpans(inner.tokens, {}, t, itemSpans);
              } else if ("tokens" in inner && inner.tokens?.length) {
                collectSpans(inner.tokens, {}, t, itemSpans);
              } else if ("text" in inner) {
                itemSpans.push({ text: decodeEntities(inner.text) });
              }
            }
          } else {
            itemSpans.push({ text: decodeEntities(item.text) });
          }
          const itemRt = new RichText(itemSpans, {
            font: bodyFont,
            color: t.textColor,
            maxWidth: this.maxWidth - 24,
            linkColor: "#38bdf8",
            selectable: this.selectable,
            onLinkClick: this.onLinkClick
          });
          itemRt.x = 12;
          listStack.add(itemRt);
        }
        return listStack;
      }
      // ── Table ────────────────────────────────────────────────────────
      case "table": {
        const tblToken = token;
        const buildCell = (cell, header) => {
          const spans = [];
          collectSpans(cell.tokens, {}, t, spans);
          if (spans.length === 0) return decodeEntities(cell.text);
          return new RichText(spans, {
            font: `${t.fontSize - 2}px ${t.bodyFont}`,
            color: header ? t.headingColor : t.textColor,
            baseStyle: header ? { bold: true } : void 0,
            linkColor: "#38bdf8",
            selectable: this.selectable,
            onLinkClick: this.onLinkClick
          });
        };
        const headers = tblToken.header.map((cell) => buildCell(cell, true));
        const rows = tblToken.rows.map((row) => row.map((cell) => buildCell(cell, false)));
        return new Table({
          headers,
          rows,
          width: this.maxWidth,
          textColor: t.textColor,
          headerTextColor: t.headingColor,
          font: `${t.fontSize - 2}px ${t.bodyFont}`,
          borderColor: t.hrColor,
          bg: t.tableBgColor,
          headerBg: t.tableHeaderBgColor,
          selectable: this.selectable
        });
      }
      // ── Horizontal rule ──────────────────────────────────────────────
      case "hr":
        return new HorizontalRule(this.maxWidth, t.hrColor);
      // ── Whitespace ───────────────────────────────────────────────────
      case "space":
        return null;
      // ── HTML (Support raw SVGs) ──────────────────────────────────────
      case "html": {
        const htmlToken = token;
        if (htmlToken.text.toLowerCase().includes("<svg") && htmlToken.text.toLowerCase().includes("</svg>")) {
          return new import_core9.SVGEntity(htmlToken.text);
        }
        return null;
      }
      // ── Fallback ─────────────────────────────────────────────────────
      default:
        if ("text" in token) {
          return new Text(token.text, {
            font: bodyFont,
            color: t.textColor,
            maxWidth: this.maxWidth,
            lineHeight: 24,
            selectable: this.selectable
          });
        }
        return null;
    }
  }
  /** Structural — children draw themselves. */
  render(_r) {
  }
};

// src/Dropdown.ts
var import_core10 = require("@vectojs/core");
var Dropdown = class extends UIComponent {
  options;
  selectedValue;
  button;
  activeMenu = null;
  activeBackdrop = null;
  highlightedIndex = -1;
  constructor(options, props = {}) {
    super();
    this.options = options;
    this.selectedValue = props.value ?? (options.length > 0 ? options[0] : "");
    this.interactive = true;
    this.width = props.width ?? 120;
    this.height = props.height ?? 36;
    this.button = new Button(this.selectedValue, {
      bg: props.bg ?? "rgba(30, 41, 59, 0.85)",
      color: props.color ?? "#fff",
      radius: props.radius ?? 8,
      font: props.font ?? "14px sans-serif"
    });
    this.button.width = this.width;
    this.button.height = this.height;
    this.button.interactive = false;
    this.add(this.button);
    this.on("focus", () => {
      this.button.focused = true;
      this.scene?.markDirty();
    });
    this.on("blur", () => {
      this.button.focused = false;
      this.scene?.markDirty();
    });
    this.on("click", () => this.toggleMenu());
    this.on("keydown", (e) => {
      const key = e.nativeEvent?.key;
      if (!key) return;
      if (key === "ArrowDown" || key === "ArrowUp") {
        e.preventDefault();
        e.nativeEvent?.stopImmediatePropagation();
        if (!this.activeMenu) {
          this.openMenu();
          this.highlightedIndex = this.options.indexOf(this.selectedValue);
          if (this.highlightedIndex === -1) this.highlightedIndex = 0;
        } else {
          const dir = key === "ArrowDown" ? 1 : -1;
          this.highlightedIndex = (this.highlightedIndex + dir + this.options.length) % this.options.length;
        }
        this.updateMenuHighlight();
        this.scene?.markDirty();
      } else if (key === "Enter" || key === " ") {
        e.preventDefault();
        e.nativeEvent?.stopImmediatePropagation();
        if (!this.activeMenu) {
          this.openMenu();
        } else {
          if (this.highlightedIndex >= 0 && this.highlightedIndex < this.options.length) {
            this.selectOption(this.options[this.highlightedIndex]);
          }
        }
      } else if (key === "Escape") {
        e.preventDefault();
        e.nativeEvent?.stopImmediatePropagation();
        if (this.activeMenu) {
          this.closeMenu();
        }
      }
    });
    this.on("change", (e) => {
      props.onChange?.(e.value);
    });
  }
  getValue() {
    return this.selectedValue;
  }
  getA11yAttributes() {
    return {
      role: "combobox",
      expanded: this.activeMenu !== null,
      controls: this.activeMenu ? this.activeMenu.id : void 0,
      haspopup: "listbox",
      value: this.selectedValue,
      activedescendant: this.activeMenu && this.highlightedIndex >= 0 && this.highlightedIndex < this.options.length ? `${this.id}-opt-${this.highlightedIndex}` : void 0
    };
  }
  toggleMenu() {
    if (this.activeMenu) {
      this.closeMenu();
    } else {
      this.openMenu();
    }
  }
  openMenu() {
    const scene = this.scene;
    if (!scene) return;
    const triggerBounds = this.getWorldBounds();
    const backdrop = new class Backdrop extends import_core10.Entity {
      isPointInside() {
        return true;
      }
      render() {
      }
      // Invisible
    }("dropdown-backdrop");
    backdrop.width = scene.width;
    backdrop.height = scene.height;
    backdrop.interactive = true;
    backdrop.on("click", (e) => {
      e.stopPropagation();
      this.closeMenu();
    });
    const menu = new Stack({ direction: "vertical", gap: 2 });
    menu.x = triggerBounds.x;
    menu.y = triggerBounds.y + triggerBounds.height + 4;
    menu.width = triggerBounds.width;
    menu.height = this.options.length * 36 + (this.options.length - 1) * 2;
    menu.interactive = true;
    menu.getA11yAttributes = () => ({
      role: "listbox",
      label: "Options"
    });
    this.options.forEach((opt, index) => {
      const item = new Button(opt, {
        bg: opt === this.selectedValue ? "rgba(0, 240, 255, 0.25)" : "rgba(15, 23, 42, 0.95)",
        color: "#fff",
        radius: 4,
        font: "13px sans-serif"
      });
      item.id = `${this.id}-opt-${index}`;
      item.width = menu.width;
      item.height = 36;
      item.interactive = true;
      item.getA11yAttributes = () => ({
        role: "option",
        label: opt,
        selected: opt === this.selectedValue
      });
      item.on("click", (e) => {
        e.stopPropagation();
        this.selectOption(opt);
      });
      menu.add(item);
    });
    scene.showOverlay(backdrop);
    scene.showOverlay(menu);
    this.activeBackdrop = backdrop;
    this.activeMenu = menu;
    this.highlightedIndex = this.options.indexOf(this.selectedValue);
    this.updateMenuHighlight();
  }
  updateMenuHighlight() {
    if (!this.activeMenu) return;
    this.activeMenu.children.forEach((child, idx) => {
      if (child instanceof Button) {
        child.bg = idx === this.highlightedIndex ? "rgba(0, 240, 255, 0.4)" : child.label === this.selectedValue ? "rgba(0, 240, 255, 0.25)" : "rgba(15, 23, 42, 0.95)";
      }
    });
  }
  selectOption(opt) {
    this.selectedValue = opt;
    this.button.label = opt;
    this.emit("change", { value: opt });
    this.closeMenu();
  }
  closeMenu() {
    const scene = this.scene;
    if (!scene) return;
    if (this.activeBackdrop) {
      scene.hideOverlay(this.activeBackdrop);
    }
    if (this.activeMenu) {
      scene.hideOverlay(this.activeMenu);
    }
    this.activeBackdrop = null;
    this.activeMenu = null;
    this.highlightedIndex = -1;
    this.scene?.markDirty();
  }
  render(_r) {
  }
};

// src/Slider.ts
var Slider = class extends UIComponent {
  min;
  max;
  value;
  /** Value granularity. Pointer and keyboard input snap to multiples of this. */
  step;
  isDragging = false;
  trackColor;
  progressColor;
  handleColor;
  constructor(props = {}) {
    super();
    this.min = props.min ?? 0;
    this.max = props.max ?? 100;
    this.value = props.value ?? this.min;
    this.step = props.step ?? 1;
    this.trackColor = props.trackColor ?? "rgba(255, 255, 255, 0.15)";
    this.progressColor = props.progressColor ?? "#00f0ff";
    this.handleColor = props.handleColor ?? "#fff";
    this.width = props.width ?? 200;
    this.height = props.height ?? 24;
    this.interactive = true;
    this.on("pointerdown", (e) => {
      this.isDragging = true;
      this.updateValueFromPointer(e.localX);
    });
    this.on("pointermove", (e) => {
      if (this.isDragging) {
        this.updateValueFromPointer(e.localX);
      }
    });
    this.on("pointerup", () => {
      this.isDragging = false;
    });
    this.on("keydown", (e) => {
      const key = e.nativeEvent?.key;
      if (!key) return;
      let next = null;
      switch (key) {
        case "ArrowRight":
        case "ArrowUp":
          next = this.value + this.step;
          break;
        case "ArrowLeft":
        case "ArrowDown":
          next = this.value - this.step;
          break;
        case "Home":
          next = this.min;
          break;
        case "End":
          next = this.max;
          break;
        default:
          return;
      }
      e.preventDefault?.();
      this.setValue(next);
    });
    this.on("change", (e) => {
      props.onChange?.(e.value);
    });
  }
  /** Snap to the step grid (anchored at `min`) and clamp into [min, max]. */
  snapToStep(raw) {
    const stepped = this.min + Math.round((raw - this.min) / this.step) * this.step;
    return Math.max(this.min, Math.min(this.max, stepped));
  }
  setValue(raw) {
    const nextValue = this.snapToStep(raw);
    if (nextValue === this.value) return;
    this.value = nextValue;
    this.emit("change", { value: this.value });
    this.scene?.markDirty();
  }
  updateValueFromPointer(localX) {
    if (localX === void 0) return;
    const relativeX = Math.max(0, Math.min(this.width, localX));
    const fraction = relativeX / this.width;
    this.setValue(this.min + fraction * (this.max - this.min));
  }
  getA11yAttributes() {
    return {
      role: "slider",
      value: String(this.value),
      valuemin: String(this.min),
      valuemax: String(this.max)
    };
  }
  render(r) {
    const thickness = 6;
    const progressFraction = (this.value - this.min) / (this.max - this.min);
    const progressWidth = this.width * progressFraction;
    const centerY = this.height / 2;
    r.beginPath();
    r.roundRect(0, centerY - thickness / 2, this.width, thickness, thickness / 2);
    r.fill(this.trackColor);
    r.beginPath();
    r.roundRect(0, centerY - thickness / 2, progressWidth, thickness, thickness / 2);
    r.fill(this.progressColor);
    const handleRadius = 8;
    r.beginPath();
    r.arc(progressWidth, centerY, handleRadius, 0, Math.PI * 2);
    r.fill(this.handleColor);
  }
};

// src/Modal.ts
var Modal = class extends UIComponent {
  card;
  backdropColor;
  constructor(title, props = {}) {
    super();
    this.width = props.width ?? (typeof window !== "undefined" ? window.innerWidth : 800);
    this.height = props.height ?? (typeof window !== "undefined" ? window.innerHeight : 600);
    this.interactive = true;
    this.backdropColor = props.backdropColor ?? "rgba(0, 0, 0, 0.5)";
    const modalW = props.modalWidth ?? 400;
    const modalH = props.modalHeight ?? 250;
    this.card = new Card({
      width: modalW,
      height: modalH,
      bg: props.cardBg ?? "rgba(15, 23, 42, 0.95)",
      border: props.cardBorder ?? "rgba(255, 255, 255, 0.15)",
      radius: 16
    });
    this.card.x = (this.width - modalW) / 2;
    this.card.y = (this.height - modalH) / 2;
    const titleText = new Text(title, {
      font: "600 20px sans-serif",
      color: "#fff"
    });
    this.card.add(titleText.setPosition(24, 24));
    const closeBtn = new Button("Close", {
      bg: "rgba(255, 255, 255, 0.1)",
      color: "#fff",
      radius: 8
    });
    closeBtn.width = 80;
    closeBtn.height = 36;
    closeBtn.on("click", (e) => {
      e.stopPropagation();
      void this.close();
    });
    this.card.add(closeBtn.setPosition(modalW - 104, modalH - 60));
    this.add(this.card);
    this.card.scaleX = 0;
    this.card.scaleY = 0;
    this.on("click", (e) => e.stopPropagation());
    this.on("pointerdown", (e) => e.stopPropagation());
  }
  onMounted() {
    void this.card.springTo({ scaleX: 1, scaleY: 1 }, { stiffness: 180, damping: 14 });
  }
  /** Animate the card out, then remove the modal from its overlay layer. */
  async close() {
    await this.card.springTo({ scaleX: 0, scaleY: 0 }, { stiffness: 220, damping: 20 });
    this.scene?.hideOverlay(this);
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 0);
    r.fill(this.backdropColor);
  }
};

// src/Overlay.ts
var Overlay = class extends UIComponent {
  placement;
  offset;
  visible = false;
  constructor(opts) {
    super();
    this.width = opts.width;
    this.height = opts.height;
    this.placement = opts.placement ?? "bottom";
    this.offset = opts.offset ?? 6;
    this.interactive = false;
    this.opacity = 0;
    this.scaleX = 0.92;
    this.scaleY = 0.92;
    this.setTransition({
      opacity: { duration: 160, easing: "easeOutQuad" },
      scaleX: "spring",
      scaleY: "spring"
    });
  }
  /**
   * Show the overlay anchored to a target entity.
   * Automatically mounts to `scene.overlayRoot` on first call.
   */
  showAt(target) {
    this._mount(target);
    this._position(target);
    this.visible = true;
    this.opacity = 1;
    this.scaleX = 1;
    this.scaleY = 1;
    this.scene?.markDirty();
  }
  /**
   * Show the overlay at an absolute canvas position.
   *
   * If the overlay has never been added to the scene tree (no `parent`), its
   * `scene` resolves to `null` (the `Entity.scene` getter walks the parent
   * chain). To make the documented "bare `new ContextMenu({...})` then
   * `showAtPoint`" pattern work without forcing every caller to pre-mount,
   * pass a `source` — either the `Scene` itself or any mounted `Entity` whose
   * `.scene` is set (e.g. the entity whose `pointerdown` listener is calling
   * you) — and the overlay will auto-mount to that scene's `overlayRoot`.
   * Without a `source` and no resolvable `scene`, this method silently
   * no-ops (preserved for backward compatibility; existing pre-mount-then-
   * showAtPoint callers are unchanged).
   */
  showAtPoint(x, y, source) {
    const scene = this.scene ?? this._sceneFromSource(source);
    if (!scene) return;
    if (!this.parent) scene.overlayRoot.add(this);
    this._placeAt(x, y);
    this.visible = true;
    this.opacity = 1;
    this.scaleX = 1;
    this.scaleY = 1;
    scene.markDirty();
  }
  /**
   * Resolve a {@link Scene} from the optional `showAtPoint` `source` arg.
   * Accepts a `Scene` passed directly, or any mounted `Entity` whose `.scene`
   * is the authoritative source (e.g. the entity whose `pointerdown` listener
   * is opening the menu).
   *
   * `protected`, not `private`: subclasses that override `showAtPoint` (e.g.
   * `ContextMenu`, to mount its outside-click backdrop) need this same
   * resolution — duplicating it drifts out of sync with the base
   * implementation, which is exactly how `ContextMenu.showAtPoint` silently
   * dropped the `source` arg it was handed (see its override for the fix).
   */
  _sceneFromSource(source) {
    if (!source) return null;
    if ("overlayRoot" in source && typeof source.markDirty === "function") {
      return source;
    }
    return source.scene ?? null;
  }
  /** Animate the overlay out (stays mounted; re-show via showAt). */
  hide() {
    this.visible = false;
    this.opacity = 0;
    this.scaleX = 0.92;
    this.scaleY = 0.92;
    this.scene?.markDirty();
  }
  _mount(target) {
    const scene = target.scene;
    if (!scene) return;
    if (this.parent) this.parent.remove(this);
    scene.overlayRoot.add(this);
  }
  _position(target) {
    const targetBounds = target.getWorldBounds();
    const gp = { x: targetBounds.x, y: targetBounds.y };
    const tw = targetBounds.width;
    const th = targetBounds.height;
    const sw = this.scene?.width ?? window.innerWidth;
    const sh = this.scene?.height ?? window.innerHeight;
    let pl = this.placement;
    if (pl === "auto") {
      const below = sh - (gp.y + th);
      const above = gp.y;
      pl = below >= this.height + this.offset ? "bottom" : "top";
      if (below < this.height + this.offset && above < this.height + this.offset)
        pl = below >= above ? "bottom" : "top";
    }
    let ax = gp.x;
    let ay = gp.y;
    switch (pl) {
      case "top":
        ax = gp.x + tw / 2 - this.width / 2;
        ay = gp.y - this.height - this.offset;
        break;
      case "top-start":
        ax = gp.x;
        ay = gp.y - this.height - this.offset;
        break;
      case "top-end":
        ax = gp.x + tw - this.width;
        ay = gp.y - this.height - this.offset;
        break;
      case "bottom":
        ax = gp.x + tw / 2 - this.width / 2;
        ay = gp.y + th + this.offset;
        break;
      case "bottom-start":
        ax = gp.x;
        ay = gp.y + th + this.offset;
        break;
      case "bottom-end":
        ax = gp.x + tw - this.width;
        ay = gp.y + th + this.offset;
        break;
      case "left":
        ax = gp.x - this.width - this.offset;
        ay = gp.y + th / 2 - this.height / 2;
        break;
      case "right":
        ax = gp.x + tw + this.offset;
        ay = gp.y + th / 2 - this.height / 2;
        break;
    }
    this._placeAt(ax, ay, sw, sh);
  }
  _placeAt(ax, ay, sw, sh) {
    const W = sw ?? this.scene?.width ?? window.innerWidth;
    const H = sh ?? this.scene?.height ?? window.innerHeight;
    this.x = Math.max(4, Math.min(ax, W - this.width - 4));
    this.y = Math.max(4, Math.min(ay, H - this.height - 4));
  }
  render(_r) {
  }
};

// src/VirtualList.ts
var VirtualList = class extends UIComponent {
  _items;
  _renderItem;
  _estH;
  _overscan;
  /** Measured height cache: item index → actual px height. */
  _hCache = /* @__PURE__ */ new Map();
  /** Currently rendered row entities keyed by item index. */
  _pool = /* @__PURE__ */ new Map();
  _scrollY = 0;
  _targetY = 0;
  _velY = 0;
  _drag = false;
  _lastPY = 0;
  constructor(opts) {
    super();
    this._items = opts.items;
    this._renderItem = opts.renderItem;
    this._estH = opts.estimatedRowHeight;
    this._overscan = opts.overscan ?? 3;
    this.width = opts.width;
    this.height = opts.height;
    this.interactive = true;
    this.clipChildren = true;
    this._bindEvents();
    this._reconcile();
  }
  /**
   * Replace the full item list.
   * Clears the height cache and resets scroll position to top.
   */
  setItems(items) {
    this._items = items;
    this._hCache.clear();
    this._targetY = 0;
    this._scrollY = 0;
    this._reconcile();
    this.scene?.markDirty();
  }
  /** Scroll to make the row at `index` visible. */
  scrollToIndex(index) {
    this._targetY = Math.min(this._rowTop(index), Math.max(0, this._totalH() - this.height));
    this.scene?.markDirty();
  }
  scrollToTop() {
    this._targetY = 0;
    this.scene?.markDirty();
  }
  scrollToBottom() {
    this._targetY = Math.max(0, this._totalH() - this.height);
    this.scene?.markDirty();
  }
  _totalH() {
    let h = 0;
    for (let i = 0; i < this._items.length; i++) h += this._hCache.get(i) ?? this._estH;
    return h;
  }
  _rowTop(index) {
    let y = 0;
    for (let i = 0; i < index; i++) y += this._hCache.get(i) ?? this._estH;
    return y;
  }
  _visibleRange() {
    const top = this._scrollY;
    const bot = this._scrollY + this.height;
    let y = 0;
    let start = -1;
    let end = 0;
    for (let i = 0; i < this._items.length; i++) {
      const h = this._hCache.get(i) ?? this._estH;
      if (start === -1 && y + h > top) start = i;
      if (y < bot) end = i;
      y += h;
    }
    if (start === -1) start = 0;
    return [
      Math.max(0, start - this._overscan),
      Math.min(this._items.length - 1, end + this._overscan)
    ];
  }
  _reconcile() {
    const [s, e] = this._visibleRange();
    const needed = /* @__PURE__ */ new Set();
    for (let i = s; i <= e; i++) needed.add(i);
    for (const [idx, ent] of this._pool) {
      if (!needed.has(idx)) {
        super.remove(ent);
        this._pool.delete(idx);
      }
    }
    let ry = this._rowTop(s);
    for (let i = s; i <= e; i++) {
      const h = this._hCache.get(i) ?? this._estH;
      if (!this._pool.has(i)) {
        const ent = this._renderItem(this._items[i], i);
        ent.x = 0;
        ent.y = ry - this._scrollY;
        ent.width = ent.width || this.width;
        super.add(ent);
        this._pool.set(i, ent);
        if (!this._hCache.has(i) && ent.height > 0) this._hCache.set(i, ent.height);
      } else {
        this._pool.get(i).y = ry - this._scrollY;
      }
      ry += h;
    }
  }
  _clamp() {
    const max = Math.max(0, this._totalH() - this.height);
    this._targetY = Math.max(0, Math.min(this._targetY, max));
  }
  _bindEvents() {
    this.on("wheel", (e) => {
      if (e.ctrlKey) return;
      e.preventDefault();
      this._targetY += e.deltaY;
      this._clamp();
      this.scene?.markDirty();
    });
    this.on("pointerdown", (e) => {
      if (e.localY === void 0) return;
      this._drag = true;
      this._lastPY = e.localY;
    });
    this.on("pointermove", (e) => {
      if (!this._drag || e.localY === void 0) return;
      const y = e.localY;
      this._targetY -= y - this._lastPY;
      this._lastPY = y;
      this._clamp();
      this.scene?.markDirty();
    });
    const end = () => {
      this._drag = false;
    };
    this.on("pointerup", end);
    this.on("pointerleave", end);
  }
  update(dt, time) {
    super.update(dt, time);
    const diff = this._targetY - this._scrollY;
    this._velY += diff * 0.12;
    this._velY *= 0.82;
    if (Math.abs(this._velY) > 0.05 || Math.abs(diff) > 0.05) {
      this._scrollY += this._velY;
      this._reconcile();
      this.scene?.markDirty();
    } else {
      this._scrollY = this._targetY;
      this._velY = 0;
    }
  }
  /**
   * The scroll integrator lives in update(), not a property driver, so without
   * this override it is invisible to the Scene's idle checks: markDirty() from
   * inside update() is wiped by the loop's own end-of-tick `dirty = false`, and
   * the animation then advances at the 2 FPS idle throttle (or stalls entirely
   * in onDemand mode). Same class of bug as the ScrollView 0.2.x fix.
   */
  hasPendingAnimations() {
    return super.hasPendingAnimations() || Math.abs(this._targetY - this._scrollY) > 0.05 || Math.abs(this._velY) > 0.05;
  }
  getA11yAttributes() {
    return { role: "list", label: `Virtual list with ${this._items.length} items` };
  }
  render(_r) {
  }
};

// src/Tree.ts
var TreeView = class extends UIComponent {
  _roots;
  _rows = [];
  _expanded = /* @__PURE__ */ new Set();
  _loaded = /* @__PURE__ */ new Map();
  _loading = /* @__PURE__ */ new Set();
  _selectedId = null;
  _hoverIdx = -1;
  _scrollY = 0;
  _targetY = 0;
  _velY = 0;
  _rh;
  _font;
  _color;
  _selColor;
  _hovColor;
  _onSelect;
  _onExpand;
  constructor(opts) {
    super();
    this._roots = opts.nodes;
    this.width = opts.width;
    this.height = opts.height;
    this._rh = opts.rowHeight ?? 28;
    this._font = opts.font ?? "13px monospace";
    this._color = opts.color ?? "#e2e8f0";
    this._selColor = opts.selectedColor ?? "rgba(0,240,255,0.18)";
    this._hovColor = opts.hoverColor ?? "rgba(255,255,255,0.06)";
    this._onSelect = opts.onSelect;
    this._onExpand = opts.onExpand;
    this.interactive = true;
    this.clipChildren = true;
    this._buildRows();
    this._bind();
  }
  /** Replace all root nodes and reset state. */
  setNodes(nodes) {
    this._roots = nodes;
    this._expanded.clear();
    this._loaded.clear();
    this._selectedId = null;
    this._buildRows();
    this.scene?.markDirty();
  }
  _buildRows() {
    this._rows = [];
    const walk = (nodes, depth) => {
      for (const node of nodes) {
        const lazy = typeof node.children === "function";
        const eager = Array.isArray(node.children) && node.children.length > 0;
        const hasChildren = lazy || eager || this._loaded.has(node.id);
        const expanded = this._expanded.has(node.id);
        const loading = this._loading.has(node.id);
        this._rows.push({ node, depth, expanded, loading, hasChildren });
        if (expanded) {
          const cached = this._loaded.get(node.id);
          if (cached) walk(cached, depth + 1);
          else if (Array.isArray(node.children)) walk(node.children, depth + 1);
        }
      }
    };
    walk(this._roots, 0);
  }
  async _toggle(idx) {
    const row = this._rows[idx];
    if (!row) return;
    if (!row.hasChildren) {
      this._selectedId = row.node.id;
      this._onSelect?.(row.node);
      this.scene?.markDirty();
      return;
    }
    const id = row.node.id;
    if (this._expanded.has(id)) {
      this._expanded.delete(id);
    } else {
      this._expanded.add(id);
      this._onExpand?.(row.node);
      if (typeof row.node.children === "function" && !this._loaded.has(id)) {
        this._loading.add(id);
        this._buildRows();
        this.scene?.markDirty();
        const children = await row.node.children();
        this._loaded.set(id, children);
        this._loading.delete(id);
      }
    }
    this._buildRows();
    this.scene?.markDirty();
  }
  _bind() {
    this.on("wheel", (e) => {
      e.preventDefault();
      this._targetY += e.deltaY;
      this._clamp();
      this.scene?.markDirty();
    });
    this.on("pointerdown", (e) => {
      if (e.localY === void 0) return;
      const ly = e.localY;
      const idx = Math.floor((ly + this._scrollY) / this._rh);
      if (idx >= 0 && idx < this._rows.length) void this._toggle(idx);
    });
    this.on("pointermove", (e) => {
      if (e.localY === void 0) return;
      const ly = e.localY;
      this._hoverIdx = Math.floor((ly + this._scrollY) / this._rh);
      this.scene?.markDirty();
    });
    this.on("pointerleave", () => {
      this._hoverIdx = -1;
      this.scene?.markDirty();
    });
  }
  _clamp() {
    const max = Math.max(0, this._rows.length * this._rh - this.height);
    this._targetY = Math.max(0, Math.min(this._targetY, max));
  }
  update(dt, time) {
    super.update(dt, time);
    const diff = this._targetY - this._scrollY;
    this._velY += diff * 0.12;
    this._velY *= 0.82;
    if (Math.abs(this._velY) > 0.05 || Math.abs(diff) > 0.05) {
      this._scrollY += this._velY;
      this.scene?.markDirty();
    } else {
      this._scrollY = this._targetY;
      this._velY = 0;
    }
  }
  /**
   * Keep the hand-rolled scroll integrator visible to the Scene's idle
   * throttle / onDemand skip — see the identical override in VirtualList.
   */
  hasPendingAnimations() {
    return super.hasPendingAnimations() || Math.abs(this._targetY - this._scrollY) > 0.05 || Math.abs(this._velY) > 0.05;
  }
  render(r) {
    const startIdx = Math.max(0, Math.floor(this._scrollY / this._rh) - 1);
    const endIdx = Math.min(
      this._rows.length - 1,
      Math.ceil((this._scrollY + this.height) / this._rh)
    );
    for (let i = startIdx; i <= endIdx; i++) {
      const row = this._rows[i];
      const y = i * this._rh - this._scrollY;
      const indent = row.depth * 16 + 8;
      if (row.node.id === this._selectedId) {
        r.beginPath();
        r.roundRect(0, y, this.width, this._rh, 4);
        r.fill(this._selColor);
      } else if (i === this._hoverIdx) {
        r.beginPath();
        r.roundRect(0, y, this.width, this._rh, 4);
        r.fill(this._hovColor);
      }
      if (row.hasChildren) {
        const arrow = row.loading ? "\u23F3" : row.expanded ? "\u25BE" : "\u25B8";
        r.fillText(arrow, indent, y + this._rh / 2 + 4, "11px monospace", "rgba(255,255,255,0.5)");
      }
      const lx = indent + (row.hasChildren ? 16 : 0);
      if (row.node.icon)
        r.fillText(
          row.node.icon,
          lx,
          y + this._rh / 2 + 4,
          this._font,
          row.node.iconColor ?? this._color
        );
      r.fillText(
        row.node.label,
        lx + (row.node.icon ? 20 : 0),
        y + this._rh / 2 + 4,
        this._font,
        this._color
      );
    }
  }
  getA11yAttributes() {
    return { role: "tree", label: "Tree view" };
  }
};

// src/ResizablePanel.ts
var PanelResizeHandle = class extends UIComponent {
  direction;
  _drag = false;
  _lastPos = 0;
  _hovered = false;
  _color;
  _hoverColor;
  constructor(dir, size, color, hoverColor, onResize) {
    super();
    this.direction = dir;
    this._color = color;
    this._hoverColor = hoverColor;
    this.width = dir === "horizontal" ? size : 0;
    this.height = dir === "vertical" ? size : 0;
    this.interactive = true;
    const posOf = (e) => {
      const scenePos = dir === "horizontal" ? e.sceneX : e.sceneY;
      if (scenePos !== void 0) return scenePos;
      return dir === "horizontal" ? e.localX : e.localY;
    };
    this.on("pointerdown", (e) => {
      const pos = posOf(e);
      if (pos === void 0) return;
      this._drag = true;
      this._lastPos = pos;
    });
    this.on("pointermove", (e) => {
      if (!this._drag) return;
      const pos = posOf(e);
      if (pos === void 0) return;
      const delta = pos - this._lastPos;
      this._lastPos = pos;
      if (delta !== 0) onResize(delta);
      this.scene?.markDirty();
    });
    const end = () => {
      this._drag = false;
    };
    this.on("pointerup", end);
    this.on("hover", () => {
      this._hovered = true;
      this.scene?.markDirty();
    });
    this.on("pointerleave", () => {
      this._hovered = false;
    });
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 2);
    r.fill(this._hovered || this._drag ? this._hoverColor : this._color);
  }
};
var Panel = class extends UIComponent {
  minSize;
  defaultSize;
  _content = null;
  _fitWidth = false;
  _fitHeight = false;
  constructor(opts = {}) {
    super();
    this.minSize = opts.minSize ?? 60;
    this.defaultSize = opts.defaultSize;
    this.clipChildren = true;
    this.interactive = false;
  }
  /**
   * Replace the panel's content entity.
   *
   * `fit` (default `true`, both axes) keeps the content's `width`/`height`
   * synced to the panel's own box — on this call and on every subsequent
   * resize of the panel (e.g. from a {@link PanelGroup} divider drag or
   * {@link PanelGroup.resize}). A viewport that doesn't track the thing it's
   * a viewport onto is the surprising case, not the default one — pass
   * `false` (or a per-axis object) to restore the old position-only
   * behavior, e.g. for content deliberately larger than its viewport and
   * scrolled via `clipChildren`.
   *
   * Panel has no setter hook on `width`/`height` to intercept a resize from
   * the outside (`Entity.width`/`height` are plain fields — see the
   * container-sizing-contract design doc), so the fit is re-applied once per
   * frame from {@link update} rather than reactively — the same approach
   * `Tabs` already uses for its own hosted content.
   */
  setContent(content, fit = true) {
    if (this._content) super.remove(this._content);
    this._content = content;
    content.x = 0;
    content.y = 0;
    if (fit === false) {
      this._fitWidth = false;
      this._fitHeight = false;
    } else if (fit === true) {
      this._fitWidth = true;
      this._fitHeight = true;
    } else {
      this._fitWidth = fit.width ?? true;
      this._fitHeight = fit.height ?? true;
    }
    this._applyFit();
    super.add(content);
    return this;
  }
  _applyFit() {
    if (!this._content) return;
    if (this._fitWidth) this._content.width = this.width;
    if (this._fitHeight) this._content.height = this.height;
  }
  update(dt, time) {
    super.update(dt, time);
    this._applyFit();
  }
  render(_r) {
  }
};
var PanelGroup = class extends UIComponent {
  direction;
  _panels = [];
  _handles = [];
  _sizes = [];
  _hSize;
  _hColor;
  _hHoverColor;
  constructor(opts) {
    super();
    this.width = opts.width;
    this.height = opts.height;
    this.direction = opts.direction;
    this._hSize = opts.handleSize ?? 4;
    this._hColor = opts.handleColor ?? "rgba(255,255,255,0.12)";
    this._hHoverColor = opts.handleHoverColor ?? "rgba(0,240,255,0.4)";
  }
  /**
   * Add a panel.
   * A {@link PanelResizeHandle} is inserted automatically before it
   * (except for the first panel).
   */
  addPanel(panel) {
    if (this._panels.length > 0) {
      const idx = this._panels.length - 1;
      const handle = new PanelResizeHandle(
        this.direction,
        this._hSize,
        this._hColor,
        this._hHoverColor,
        (delta) => this._onResize(idx, delta)
      );
      this._handles.push(handle);
      super.add(handle);
    }
    this._panels.push(panel);
    super.add(panel);
    this._initSizes();
    this._layout();
    return this;
  }
  /** Update the group's canvas dimensions (e.g., on window resize). */
  resize(w, h) {
    const oldAvail = this._avail();
    this.width = w;
    this.height = h;
    const avail = this._avail();
    if (this._sizes.length === this._panels.length && oldAvail > 0) {
      const total = this._sizes.reduce((sum, size) => sum + size, 0);
      const basis = total > 0 ? total : oldAvail;
      this._sizes = this._sizes.map((size) => size / basis * avail);
      this._normalizeSizes(avail);
    } else {
      this._initSizes();
    }
    this._layout();
    this.scene?.markDirty();
  }
  _avail() {
    return (this.direction === "horizontal" ? this.width : this.height) - this._handles.length * this._hSize;
  }
  _initSizes() {
    const avail = this._avail();
    const n = this._panels.length;
    if (this._sizes.length === n) return;
    this._sizes = [];
    let rem = avail;
    const assigned = Array(n).fill(false);
    for (let i = 0; i < n; i++) {
      if (this._panels[i].defaultSize !== void 0) {
        this._sizes[i] = Math.max(this._panels[i].minSize, this._panels[i].defaultSize * avail);
        rem -= this._sizes[i];
        assigned[i] = true;
      }
    }
    const unassigned = assigned.filter((v) => !v).length;
    const share = unassigned > 0 ? rem / unassigned : 0;
    for (let i = 0; i < n; i++) {
      if (!assigned[i]) this._sizes[i] = Math.max(this._panels[i].minSize, share);
    }
    this._normalizeSizes(avail);
  }
  _normalizeSizes(avail) {
    const n = this._panels.length;
    if (n === 0) return;
    this._sizes = this._sizes.map((size, i) => Math.max(this._panels[i].minSize, size));
    const minTotal = this._panels.reduce((sum, panel) => sum + panel.minSize, 0);
    if (avail <= minTotal) return;
    const total = this._sizes.reduce((sum, size) => sum + size, 0);
    if (Math.abs(total - avail) < 0.01) return;
    if (total > avail) {
      const excess = total - avail;
      const adjustable = this._sizes.map((size, i) => Math.max(0, size - this._panels[i].minSize));
      const adjustableTotal = adjustable.reduce((sum, amount) => sum + amount, 0);
      if (adjustableTotal <= 0) return;
      this._sizes = this._sizes.map((size, i) => size - excess * (adjustable[i] / adjustableTotal));
      return;
    }
    const extra = avail - total;
    const basis = total > 0 ? total : n;
    this._sizes = this._sizes.map((size) => size + extra * (size / basis));
  }
  _onResize(idx, delta) {
    const a = idx;
    const b = idx + 1;
    this._sizes[a] = Math.max(this._panels[a].minSize, this._sizes[a] + delta);
    this._sizes[b] = Math.max(this._panels[b].minSize, this._sizes[b] - delta);
    this._normalizeSizes(this._avail());
    this._layout();
    this.scene?.markDirty();
  }
  _layout() {
    const isH = this.direction === "horizontal";
    const cross = isH ? this.height : this.width;
    let pos = 0;
    for (let i = 0; i < this._panels.length; i++) {
      const p = this._panels[i];
      const sz = this._sizes[i] ?? 0;
      if (isH) {
        p.x = pos;
        p.y = 0;
        p.width = sz;
        p.height = cross;
      } else {
        p.x = 0;
        p.y = pos;
        p.width = cross;
        p.height = sz;
      }
      pos += sz;
      if (i < this._handles.length) {
        const h = this._handles[i];
        if (isH) {
          h.x = pos;
          h.y = 0;
          h.width = this._hSize;
          h.height = cross;
        } else {
          h.x = 0;
          h.y = pos;
          h.width = cross;
          h.height = this._hSize;
        }
        pos += this._hSize;
      }
    }
  }
  render(_r) {
  }
};

// src/Tooltip.ts
var Tooltip = class extends Overlay {
  _content;
  _font;
  _textColor;
  _bg;
  _delay;
  _timer = null;
  _target;
  _onHover;
  _onLeave;
  constructor(opts) {
    const estimatedW = Math.min(opts.content.length * 7.5 + 20, 320);
    super({
      width: estimatedW,
      height: 30,
      placement: opts.placement ?? "top",
      offset: 8
    });
    this._content = opts.content;
    this._font = opts.font ?? "12px sans-serif";
    this._textColor = opts.color ?? "#e2e8f0";
    this._bg = opts.bg ?? "rgba(15,15,30,0.92)";
    this._delay = opts.delay ?? 320;
    this._target = opts.target;
    this._onHover = () => {
      if (this._timer) clearTimeout(this._timer);
      this._timer = setTimeout(() => {
        this._timer = null;
        this.showAt(opts.target);
      }, this._delay);
    };
    this._onLeave = () => {
      if (this._timer) {
        clearTimeout(this._timer);
        this._timer = null;
      }
      this.hide();
    };
    opts.target.on("hover", this._onHover);
    opts.target.on("pointerleave", this._onLeave);
  }
  destroy() {
    if (this._timer) {
      clearTimeout(this._timer);
      this._timer = null;
    }
    this._target.off("hover", this._onHover);
    this._target.off("pointerleave", this._onLeave);
    super.destroy();
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 6);
    r.fill(this._bg);
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 6);
    r.stroke("rgba(255,255,255,0.12)", 1);
    r.fillText(this._content, 10, this.height / 2 + 4, this._font, this._textColor);
  }
};

// src/Popover.ts
var Popover = class extends Overlay {
  _bg;
  _border;
  _target;
  _onClick;
  constructor(opts) {
    super({
      width: opts.width,
      height: opts.height,
      placement: opts.placement ?? "bottom-start",
      offset: 6
    });
    this._bg = opts.bg ?? "rgba(15,15,30,0.96)";
    this._border = opts.borderColor ?? "rgba(255,255,255,0.14)";
    this.interactive = true;
    this._target = opts.target;
    this._onClick = () => {
      if (this.visible) this.hide();
      else this.showAt(opts.target);
    };
    opts.target.on("click", this._onClick);
  }
  destroy() {
    this._target.off("click", this._onClick);
    super.destroy();
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 8);
    r.fill(this._bg);
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 8);
    r.stroke(this._border, 1);
  }
};

// src/ContextMenu.ts
var import_core11 = require("@vectojs/core");
var nextContextMenuId = 1;
var ContextMenu = class _ContextMenu extends Overlay {
  _items;
  _font;
  _textColor;
  _disColor;
  _bg;
  _hoverBg;
  _border;
  _iH;
  _sH;
  _hoverIdx = -1;
  _submenu = null;
  _parentMenu = null;
  /** Which item's `children` `_submenu` currently represents, if any. */
  _submenuFor = null;
  _opts;
  /** Full-screen invisible entity mounted behind the menu while open, so a
   * click anywhere outside it closes the menu — the way every native context
   * menu behaves. Without this the menu only ever closed by selecting one of
   * its own (non-disabled) items. */
  _backdrop = null;
  constructor(opts) {
    const iH = opts.itemHeight ?? 32;
    const sH = opts.separatorHeight ?? 9;
    const totalH = (opts.items ?? []).reduce((acc, it) => acc + (it.separator ? sH : iH), 0);
    super({ width: opts.width ?? 220, height: totalH + 8, placement: "auto", offset: 2 });
    this.id = `context-menu-${nextContextMenuId++}`;
    this._opts = opts;
    this._items = opts.items ?? [];
    this._font = opts.font ?? "13px sans-serif";
    this._textColor = opts.color ?? "#e2e8f0";
    this._disColor = opts.disabledColor ?? "rgba(255,255,255,0.3)";
    this._bg = opts.bg ?? "rgba(18,18,32,0.97)";
    this._hoverBg = opts.hoverBg ?? "rgba(0,240,255,0.14)";
    this._border = opts.borderColor ?? "rgba(255,255,255,0.12)";
    this._iH = iH;
    this._sH = sH;
    this.interactive = false;
    this.on("pointermove", (e) => {
      this._hoverIdx = e.localY === void 0 ? -1 : this._idxAt(e.localY);
      this.scene?.markDirty();
    });
    this.on("pointerleave", () => {
      this._hoverIdx = -1;
      this.scene?.markDirty();
    });
    this.on("pointerdown", (e) => {
      if (e.localY === void 0) return;
      const idx = this._idxAt(e.localY);
      const item = this._items[idx];
      if (!item || item.separator || item.disabled) return;
      if (item.children && item.children.length > 0) {
        if (!this._submenu || this._submenuFor !== item) {
          if (this._submenu) this._submenu.destroy();
          this._submenu = new _ContextMenu({ ...this._opts, items: item.children });
          this._submenu._parentMenu = this;
          this._submenuFor = item;
          if (this.scene) this.scene.overlayRoot.add(this._submenu);
        }
        this._submenu.showAtPoint(this.x + this.width, this.y + this._rowTop(idx));
      } else {
        item.onClick?.();
        this._rootMenu().hide();
      }
    });
  }
  showAtPoint(x, y, source) {
    const scene = this.scene ?? this._sceneFromSource(source);
    if (scene) this.interactive = true;
    if (this._parentMenu === null && !this._backdrop && scene) {
      const backdrop = new class ContextMenuBackdrop extends import_core11.Entity {
        isPointInside() {
          return true;
        }
        render() {
        }
      }(`${this.id}-backdrop`);
      backdrop.width = scene.width;
      backdrop.height = scene.height;
      backdrop.interactive = true;
      const dismiss = (e) => {
        e.stopPropagation();
        this.hide();
      };
      backdrop.on("pointerdown", dismiss);
      backdrop.on("click", dismiss);
      const parent = this.parent ?? scene.overlayRoot;
      if (this.parent) this.parent.remove(this);
      scene.overlayRoot.add(backdrop);
      parent.add(this);
      this._backdrop = backdrop;
    }
    super.showAtPoint(x, y, source);
  }
  hide() {
    const scene = this.scene;
    if (this._backdrop) {
      this._backdrop.destroy();
      this._backdrop = null;
    }
    if (this._submenu) this._submenu.hide();
    super.hide();
    this.interactive = false;
    scene?.detachA11y(this);
  }
  destroy() {
    if (this._backdrop) {
      this._backdrop.destroy();
      this._backdrop = null;
    }
    if (this._submenu) {
      this._submenu.destroy();
      this._submenu = null;
    }
    this._parentMenu = null;
    super.destroy();
  }
  _rootMenu() {
    return this._parentMenu?._rootMenu() ?? this;
  }
  _idxAt(localY) {
    let y = 4;
    for (let i = 0; i < this._items.length; i++) {
      const h = this._items[i].separator ? this._sH : this._iH;
      if (localY >= y && localY < y + h) return i;
      y += h;
    }
    return -1;
  }
  _rowTop(idx) {
    let y = 4;
    for (let i = 0; i < idx; i++) y += this._items[i].separator ? this._sH : this._iH;
    return y;
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 8);
    r.fill(this._bg);
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, 8);
    r.stroke(this._border, 1);
    let y = 4;
    for (let i = 0; i < this._items.length; i++) {
      const item = this._items[i];
      if (item.separator) {
        const mid = y + this._sH / 2;
        r.beginPath();
        r.moveTo(8, mid);
        r.lineTo(this.width - 8, mid);
        r.stroke("rgba(255,255,255,0.1)", 1);
        y += this._sH;
        continue;
      }
      const col = item.disabled ? this._disColor : this._textColor;
      if (i === this._hoverIdx && !item.disabled) {
        r.beginPath();
        r.roundRect(4, y, this.width - 8, this._iH, 4);
        r.fill(this._hoverBg);
      }
      const ty = y + this._iH / 2 + 4;
      let lx = 12;
      if (item.icon) {
        r.fillText(item.icon, lx, ty, this._font, col);
        lx += 22;
      }
      r.fillText(item.label ?? "", lx, ty, this._font, col);
      if (item.shortcut) {
        r.fillText(
          item.shortcut,
          this.width - 12 - measureText(item.shortcut, this._font),
          ty,
          this._font,
          item.disabled ? this._disColor : "rgba(255,255,255,0.4)"
        );
      }
      if (item.children && item.children.length > 0) {
        r.fillText("\u25B8", this.width - 16, ty, "10px sans-serif", "rgba(255,255,255,0.5)");
      }
      y += this._iH;
    }
  }
  getA11yAttributes() {
    return { role: "menu", label: "Context menu" };
  }
};

// src/RadioGroup.ts
var RadioGroup = class extends UIComponent {
  options;
  value;
  direction;
  gap;
  size;
  font;
  color;
  accent;
  border;
  _hoverIdx = -1;
  constructor(opts) {
    super();
    this.options = opts.options;
    this.value = opts.value ?? (opts.options.length > 0 ? opts.options[0].value : "");
    this.direction = opts.direction ?? "vertical";
    this.gap = opts.gap ?? 12;
    this.size = opts.size ?? 18;
    this.font = opts.font ?? "16px sans-serif";
    this.color = opts.color ?? "#e2e8f0";
    this.accent = opts.accent ?? "#2563eb";
    this.border = opts.border ?? "#475569";
    this.interactive = true;
    this._layout();
    this.on("pointerdown", (e) => {
      if (e.localX === void 0 || e.localY === void 0) return;
      const idx = this._idxAt(e.localX, e.localY);
      if (idx !== -1) {
        const opt = this.options[idx];
        if (!opt.disabled && opt.value !== this.value) {
          this.emit("change", { value: opt.value });
        }
      }
    });
    this.on("pointermove", (e) => {
      this._hoverIdx = e.localX === void 0 || e.localY === void 0 ? -1 : this._idxAt(e.localX, e.localY);
      this.scene?.markDirty();
    });
    this.on("pointerleave", () => {
      this._hoverIdx = -1;
      this.scene?.markDirty();
    });
    this.on("change", (e) => {
      if (this.value === e.value) return;
      this.value = e.value;
      opts.onChange?.(this.value);
      this.scene?.markDirty();
    });
  }
  _layout() {
    let totalW = 0;
    let totalH = 0;
    const isH = this.direction === "horizontal";
    for (let i = 0; i < this.options.length; i++) {
      const labelW = measureText(this.options[i].label, this.font);
      const itemW = this.size + 8 + labelW;
      if (isH) {
        totalW += itemW + (i > 0 ? this.gap : 0);
        totalH = Math.max(totalH, this.size);
      } else {
        totalW = Math.max(totalW, itemW);
        totalH += this.size + (i > 0 ? this.gap : 0);
      }
    }
    this.width = totalW;
    this.height = totalH;
  }
  _idxAt(lx, ly) {
    let current = 0;
    const isH = this.direction === "horizontal";
    for (let i = 0; i < this.options.length; i++) {
      const labelW = measureText(this.options[i].label, this.font);
      const itemW = this.size + 8 + labelW;
      if (isH) {
        if (lx >= current && lx <= current + itemW && ly >= 0 && ly <= this.size) {
          return i;
        }
        current += itemW + this.gap;
      } else {
        if (ly >= current && ly <= current + this.size && lx >= 0 && lx <= itemW) {
          return i;
        }
        current += this.size + this.gap;
      }
    }
    return -1;
  }
  render(r) {
    let current = 0;
    const isH = this.direction === "horizontal";
    for (let i = 0; i < this.options.length; i++) {
      const opt = this.options[i];
      const selected = opt.value === this.value;
      const x = isH ? current : 0;
      const y = isH ? 0 : current;
      const centerY = y + this.size / 2;
      const cX = x + this.size / 2;
      r.beginPath();
      r.arc(cX, centerY, this.size / 2 - 1, 0, Math.PI * 2);
      r.stroke(selected ? this.accent : this.border, 2);
      if (selected) {
        r.beginPath();
        r.arc(cX, centerY, this.size / 4, 0, Math.PI * 2);
        r.fill(this.accent);
      }
      const disabledColor = "rgba(255,255,255,0.3)";
      const color = opt.disabled ? disabledColor : this.color;
      r.fillText(opt.label, x + this.size + 8, centerY + 4, this.font, color);
      if (i === this._hoverIdx && !opt.disabled) {
        r.beginPath();
        r.arc(cX, centerY, this.size / 2 + 4, 0, Math.PI * 2);
        r.stroke("rgba(0,240,255,0.15)", 1);
      }
      const labelW = measureText(opt.label, this.font);
      const itemW = this.size + 8 + labelW;
      current += (isH ? itemW : this.size) + this.gap;
    }
  }
  getA11yAttributes() {
    return {
      role: "radiogroup",
      label: "Radio group"
    };
  }
};

// src/Tabs.ts
var Tabs = class extends UIComponent {
  tabs;
  value;
  tabHeight;
  font;
  color;
  selectedColor;
  borderColor;
  closable;
  tabWidth;
  minTabWidth;
  autoHideTabBar;
  _hoverIdx = -1;
  _hoverClose = false;
  _scrollX = 0;
  _closeBox = 14;
  // px hit region for the × glyph
  constructor(opts) {
    super();
    this.tabs = opts.tabs;
    this.value = opts.value ?? (opts.tabs.length > 0 ? opts.tabs[0].id : "");
    this.width = opts.width;
    this.height = opts.height;
    this.tabHeight = opts.tabHeight ?? 40;
    this.font = opts.font ?? "14px sans-serif";
    this.color = opts.color ?? "#94a3b8";
    this.selectedColor = opts.selectedColor ?? "#00f0ff";
    this.borderColor = opts.borderColor ?? "rgba(255,255,255,0.1)";
    this.closable = opts.closable ?? false;
    this.tabWidth = opts.tabWidth ?? 160;
    this.minTabWidth = opts.minTabWidth ?? 96;
    this.autoHideTabBar = opts.autoHideTabBar ?? false;
    this.interactive = true;
    this._updateContentVisibility();
    this.on("pointerdown", (e) => {
      const { localX: lx, localY: ly } = e;
      if (lx === void 0 || ly === void 0) return;
      const barH = this._barHeight();
      if (barH === 0 || ly < 0 || ly > barH) return;
      const idx = this._tabIdxAt(lx);
      if (idx === -1) return;
      const tab = this.tabs[idx];
      if (this.closable && this._isOverClose(lx, idx)) {
        opts.onClose?.(tab.id);
        return;
      }
      if (tab.id !== this.value) {
        this.emit("change", { value: tab.id });
      }
    });
    this.on("pointermove", (e) => {
      const { localX: lx, localY: ly } = e;
      if (lx === void 0 || ly === void 0) return;
      const barH = this._barHeight();
      if (barH > 0 && ly >= 0 && ly <= barH) {
        this._hoverIdx = this._tabIdxAt(lx);
        this._hoverClose = this.closable && this._hoverIdx !== -1 && this._isOverClose(lx, this._hoverIdx);
      } else {
        this._hoverIdx = -1;
        this._hoverClose = false;
      }
      this.scene?.markDirty();
    });
    this.on("wheel", (e) => {
      const max = this._maxScroll();
      if (max <= 0) return;
      const d = Math.abs(e.deltaX ?? 0) > Math.abs(e.deltaY ?? 0) ? e.deltaX ?? 0 : e.deltaY ?? 0;
      const next = Math.max(0, Math.min(max, this._scrollX + d));
      if (next !== this._scrollX) {
        this._scrollX = next;
        e.nativeEvent?.preventDefault?.();
        this.scene?.markDirty();
      }
    });
    this.on("pointerleave", () => {
      this._hoverIdx = -1;
      this._hoverClose = false;
      this.scene?.markDirty();
    });
    this.on("change", (e) => {
      if (this.value === e.value) return;
      this.value = e.value;
      this._ensureActiveVisible();
      this._updateContentVisibility();
      opts.onChange?.(this.value);
      this.scene?.markDirty();
    });
  }
  /**
   * The height the tab bar currently occupies — `0` when `autoHideTabBar`
   * has hidden it. Owners that lay out siblings around the bar (status
   * lines, gutters) should read this instead of assuming `tabHeight`.
   */
  get effectiveTabBarHeight() {
    return this._barHeight();
  }
  _barHeight() {
    return this.autoHideTabBar && this.tabs.length < 2 ? 0 : this.tabHeight;
  }
  _tabW() {
    if (this.tabs.length === 0) return this.tabWidth;
    const even = this.width / this.tabs.length;
    if (even >= this.tabWidth) return this.tabWidth;
    return Math.max(this.minTabWidth, Math.min(this.tabWidth, even));
  }
  _contentWidth() {
    return this._tabW() * this.tabs.length;
  }
  _maxScroll() {
    return Math.max(0, this._contentWidth() - this.width);
  }
  _ensureActiveVisible() {
    const idx = this.tabs.findIndex((t) => t.id === this.value);
    if (idx === -1) return;
    const w = this._tabW();
    const left = idx * w;
    const right = left + w;
    if (left < this._scrollX) this._scrollX = left;
    else if (right > this._scrollX + this.width) this._scrollX = right - this.width;
    this._scrollX = Math.max(0, Math.min(this._maxScroll(), this._scrollX));
  }
  _tabIdxAt(lx) {
    const w = this._tabW();
    const idx = Math.floor((lx + this._scrollX) / w);
    if (idx >= 0 && idx < this.tabs.length) return idx;
    return -1;
  }
  _closeCenterX(idx) {
    const w = this._tabW();
    return idx * w + w - 12 - this._scrollX;
  }
  _isOverClose(lx, idx) {
    return Math.abs(lx - this._closeCenterX(idx)) <= this._closeBox / 2;
  }
  _updateContentVisibility() {
    const contentY = this._barHeight();
    const contentH = this.height - contentY;
    for (const tab of this.tabs) {
      if (tab.id === this.value) {
        tab.content.x = 0;
        tab.content.y = contentY;
        tab.content.width = this.width;
        tab.content.height = contentH;
        if (!this.children.includes(tab.content)) {
          this.add(tab.content);
        }
      } else {
        if (this.children.includes(tab.content)) {
          this.remove(tab.content);
        }
      }
    }
  }
  /**
   * Re-derive content geometry every frame: `tabs` is a public field owners
   * reassign directly (no setter to intercept), and the bar height itself is
   * dynamic under `autoHideTabBar` — the active content must follow both
   * without requiring a `change` emit.
   */
  update(dt, time) {
    super.update(dt, time);
    this._updateContentVisibility();
  }
  render(r) {
    if (this._barHeight() === 0) return;
    const tabW = this._tabW();
    this._scrollX = Math.max(0, Math.min(this._maxScroll(), this._scrollX));
    r.beginPath();
    r.moveTo(0, this.tabHeight);
    r.lineTo(this.width, this.tabHeight);
    r.stroke(this.borderColor, 1);
    r.save();
    r.clip(0, 0, this.width, this.tabHeight);
    const closePad = this.closable ? this._closeBox + 6 : 8;
    for (let i = 0; i < this.tabs.length; i++) {
      const tab = this.tabs[i];
      const x = i * tabW - this._scrollX;
      if (x + tabW < 0 || x > this.width) continue;
      const active = tab.id === this.value;
      const avail = tabW - 16 - closePad;
      const label = this._truncate(tab.label, avail);
      const textColor = active ? this.selectedColor : this.color;
      r.fillText(label, x + 12, this.tabHeight / 2 + 4, this.font, textColor);
      if (this.closable) {
        const cx = this._closeCenterX(i);
        const overThis = this._hoverIdx === i && this._hoverClose;
        if (overThis) {
          r.beginPath();
          r.roundRect(cx - 8, this.tabHeight / 2 - 8, 16, 16, 4);
          r.fill("rgba(255,255,255,0.12)");
        }
        r.fillText(
          "\xD7",
          cx - 4,
          this.tabHeight / 2 + 4,
          this.font,
          overThis ? "#f87171" : this.color
        );
      }
      if (active) {
        r.beginPath();
        r.moveTo(x + 12, this.tabHeight - 2);
        r.lineTo(x + tabW - 12, this.tabHeight - 2);
        r.stroke(this.selectedColor, 3);
      } else if (i === this._hoverIdx && !this._hoverClose) {
        r.beginPath();
        r.moveTo(x + 20, this.tabHeight - 2);
        r.lineTo(x + tabW - 20, this.tabHeight - 2);
        r.stroke("rgba(255,255,255,0.2)", 2);
      }
      r.beginPath();
      r.moveTo(x + tabW, 6);
      r.lineTo(x + tabW, this.tabHeight - 6);
      r.stroke(this.borderColor, 1);
    }
    r.restore();
    if (this._scrollX > 0) this._edgeFade(r, true);
    if (this._scrollX < this._maxScroll()) this._edgeFade(r, false);
  }
  _truncate(label, avail) {
    if (avail <= 0) return "";
    if (measureText(label, this.font) <= avail) return label;
    let lo = 0;
    let hi = label.length;
    while (lo < hi) {
      const mid = lo + hi + 1 >> 1;
      if (measureText(label.slice(0, mid) + "\u2026", this.font) <= avail) lo = mid;
      else hi = mid - 1;
    }
    return lo > 0 ? label.slice(0, lo) + "\u2026" : "\u2026";
  }
  _edgeFade(r, left) {
    const w = 18;
    const x = left ? 0 : this.width - w;
    r.beginPath();
    r.roundRect(x, 0, w, this.tabHeight - 2, 0);
    r.fill(left ? "rgba(0,0,0,0.28)" : "rgba(0,0,0,0.28)");
  }
  getA11yAttributes() {
    return {
      role: "tablist",
      label: "Tab switching panel"
    };
  }
};

// src/ProgressBar.ts
var ProgressBar = class extends UIComponent {
  value;
  radius;
  bg;
  accent;
  showText;
  font;
  color;
  constructor(opts = { value: 0 }) {
    super();
    this.value = opts.value;
    this.width = opts.width ?? 200;
    this.height = opts.height ?? 16;
    this.radius = opts.radius ?? 8;
    this.bg = opts.bg ?? "rgba(255, 255, 255, 0.1)";
    this.accent = opts.accent ?? "#00f0ff";
    this.showText = opts.showText ?? false;
    this.font = opts.font ?? "12px sans-serif";
    this.color = opts.color ?? "#e2e8f0";
    this.interactive = false;
  }
  /** Set new progress value (clamped between `0` and `1`). */
  setValue(val) {
    const clamped = Math.max(0, Math.min(val, 1));
    if (this.value === clamped) return;
    this.value = clamped;
    this.scene?.markDirty();
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.fill(this.bg);
    if (this.value > 0) {
      r.beginPath();
      r.roundRect(0, 0, this.width * this.value, this.height, this.radius);
      r.fill(this.accent);
    }
    if (this.showText) {
      const pct = `${Math.round(this.value * 100)}%`;
      const textW = measureText(pct, this.font);
      const textX = this.width / 2 - textW / 2;
      r.fillText(pct, textX, this.height / 2 + 4, this.font, this.color);
    }
  }
  getA11yAttributes() {
    return {
      role: "progressbar",
      value: String(Math.round(this.value * 100))
    };
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Button,
  Card,
  Checkbox,
  CodeBlock,
  ContextMenu,
  Dropdown,
  Flow,
  Image,
  Input,
  Link,
  Markdown,
  Modal,
  Overlay,
  Panel,
  PanelGroup,
  PanelResizeHandle,
  Popover,
  ProgressBar,
  RadioGroup,
  RichText,
  ScrollView,
  Slider,
  Stack,
  Table,
  Tabs,
  Text,
  TextArea,
  Toggle,
  Tooltip,
  TreeView,
  UIComponent,
  VirtualList,
  measureText,
  wrapLines,
  wrapText
});
