// @vectojs/ui/dist/Input.mjs
// src/Input.ts
import { cssLineBoxBaseline, LayoutEngine } from "@vectojs/core";

// src/UIComponent.ts
import { Entity } from "@vectojs/core";
var UIComponent = class extends Entity {
  /** Inner padding in pixels, used by box-style components. */
  padding = 0;
  /** Played automatically when the component mounts to a live scene, if set. */
  enterMotion;
  /** Played by {@link dismiss} before the component removes itself, if set. */
  exitMotion;
  onMounted() {
    if (this.enterMotion) void this.playMotion(this.enterMotion);
  }
  /** Seed each prop to `from`, register the transition, then drive to `to`. */
  playMotion(spec) {
    const cfg = spec.config ?? "spring";
    const targets = {};
    const trans = {};
    for (const [k, pair] of Object.entries(spec.props)) {
      const prop = k;
      const [from, to] = pair;
      this.setImmediate(prop, from);
      targets[prop] = to;
      trans[prop] = cfg;
    }
    this.setTransition(trans);
    return this.driveMotion(targets, cfg);
  }
  driveMotion(targets, cfg) {
    if (typeof cfg === "object" && "duration" in cfg) {
      return this.animateTo(targets, cfg);
    }
    return this.springTo(targets, cfg === "spring" ? {} : cfg);
  }
  /** Play the exit motion (if any), then remove self from the tree. */
  async dismiss() {
    if (this.exitMotion) await this.playMotion(this.exitMotion);
    this.parent?.remove(this);
  }
  /** Wakes the render loop at the next caret-blink phase boundary, or `null`. */
  caretBlinkTimer = null;
  /**
   * Keep a focused text component's caret blink visible to the idle throttle.
   * The blink phase is derived from `Date.now()` inside `render()`, so nothing
   * marks the scene dirty when the phase flips — an idle `onDemand` scene
   * would freeze the caret solid (the ScrollView 0.2.x regression class).
   * One timeout per 500 ms phase boundary costs ~2 renders/s while focused;
   * reporting a permanent pending animation instead would pin the scene at
   * full frame rate for as long as the field holds focus.
   */
  startCaretBlinkWake() {
    this.stopCaretBlinkWake();
    const schedule = () => {
      this.caretBlinkTimer = setTimeout(
        () => {
          this.scene?.markDirty();
          schedule();
        },
        500 - Date.now() % 500
      );
    };
    schedule();
    this.scene?.markDirty();
  }
  /** Stop the caret-blink wake-up (on blur; destroy also clears it). */
  stopCaretBlinkWake() {
    if (this.caretBlinkTimer !== null) {
      clearTimeout(this.caretBlinkTimer);
      this.caretBlinkTimer = null;
    }
    this.scene?.markDirty();
  }
  destroy() {
    if (this.caretBlinkTimer !== null) {
      clearTimeout(this.caretBlinkTimer);
      this.caretBlinkTimer = null;
    }
    super.destroy();
  }
  /**
   * Axis-aligned hit-test against the component's box in global space.
   *
   * @param globalX - World-space X coordinate.
   * @param globalY - World-space Y coordinate.
   * @returns Whether the point lies within `[0, width] x [0, height]` locally.
   */
  isPointInside(globalX, globalY) {
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x >= 0 && local.x <= this.width && local.y >= 0 && local.y <= this.height;
  }
  /** Local-space box, enabling viewport culling by {@link Scene}. */
  getBounds() {
    return { x: 0, y: 0, width: this.width, height: this.height };
  }
};

// src/measure.ts
import { ArabicShaper } from "@vectojs/core";
var sharedCtx;
function getCtx() {
  if (sharedCtx !== void 0) return sharedCtx;
  sharedCtx = typeof document !== "undefined" ? document.createElement("canvas").getContext("2d") : null;
  return sharedCtx;
}
function fontSizePx(font) {
  const pxIndex = font.indexOf("px");
  if (pxIndex <= 0) return 16;
  let start = pxIndex - 1;
  while (start >= 0) {
    const ch = font[start];
    if (ch >= "0" && ch <= "9" || ch === ".") {
      start--;
    } else {
      break;
    }
  }
  const raw = font.slice(start + 1, pxIndex);
  if (raw === "") return 16;
  const size = Number.parseFloat(raw);
  return Number.isFinite(size) ? size : 16;
}
var MEASURE_CACHE_MAX = 1e3;
var measureCache = /* @__PURE__ */ new Map();
if (typeof document !== "undefined" && document.fonts) {
  const clearCache = () => measureCache.clear();
  document.fonts.ready.then(clearCache);
  document.fonts.addEventListener("loadingdone", clearCache);
}
function measureText(text, font) {
  const shaped = ArabicShaper.shapeArabic(text).shapedText;
  const key = `${font} ${shaped}`;
  const cached = measureCache.get(key);
  if (cached !== void 0) {
    measureCache.delete(key);
    measureCache.set(key, cached);
    return cached;
  }
  const ctx = getCtx();
  let width;
  if (!ctx) {
    width = shaped.length * fontSizePx(font) * 0.5;
  } else {
    ctx.font = font;
    width = ctx.measureText(shaped).width;
  }
  measureCache.set(key, width);
  if (measureCache.size > MEASURE_CACHE_MAX) {
    measureCache.delete(measureCache.keys().next().value);
  }
  return width;
}

// src/Input.ts
var Input = class extends UIComponent {
  value;
  placeholder;
  font;
  color;
  placeholderColor;
  bg;
  border;
  selectionColor;
  radius;
  /** Caret / selection anchor & focus offsets, mirrored from the real input. */
  selectionStart;
  selectionEnd;
  /** Active IME pre-edit range `[start, start+length)`, or `null`. */
  composition = null;
  /** Whether the shadow input currently holds focus (drives caret blink). */
  focused = false;
  /** Horizontal scroll offset so the caret stays in view (scroll-to-caret). */
  scrollLeft = 0;
  constructor(opts) {
    super();
    this.width = opts.width;
    this.height = opts.height ?? 40;
    this.value = opts.value ?? "";
    this.placeholder = opts.placeholder ?? "";
    this.font = opts.font ?? "16px sans-serif";
    this.color = opts.color ?? "#e2e8f0";
    this.placeholderColor = opts.placeholderColor ?? "#64748b";
    this.bg = opts.bg ?? "#0f172a";
    this.border = opts.border ?? "#334155";
    this.selectionColor = opts.selectionColor ?? "rgba(56, 189, 248, 0.35)";
    this.radius = opts.radius ?? 6;
    this.padding = opts.padding ?? 10;
    this.selectionStart = this.value.length;
    this.selectionEnd = this.value.length;
    this.interactive = true;
    this.on(
      "change",
      (e) => {
        this.value = e.value;
        this.selectionStart = e.selectionStart ?? this.value.length;
        this.selectionEnd = e.selectionEnd ?? this.value.length;
        this.composition = e.composition ?? null;
        opts.onChange?.(this.value);
      }
    );
    this.on("focus", () => {
      this.focused = true;
      this.startCaretBlinkWake();
    });
    this.on("blur", () => {
      this.focused = false;
      this.stopCaretBlinkWake();
    });
  }
  getA11yAttributes() {
    return {
      tag: "input",
      inputType: "text",
      placeholder: this.placeholder,
      value: this.value,
      label: this.placeholder,
      textInputStyle: {
        font: this.font,
        lineHeight: this.height - 2 * this.padding,
        padding: this.padding
      }
    };
  }
  cachedValue = "";
  cachedFont = "";
  cachedLayout = null;
  _rtlCacheValue = null;
  _rtlCacheResult = false;
  /**
   * Whether `value` contains any RTL-script character. Cached per distinct
   * `value` — otherwise this O(n) scan re-runs from scratch on every
   * `charOffset()` call, and a single render/caret-blink tick calls it several
   * times (caret, selection start, selection end, composition bounds).
   */
  hasRTL() {
    if (this._rtlCacheValue === this.value) return this._rtlCacheResult;
    let result = false;
    for (let j = 0; j < this.value.length; j++) {
      const code = this.value.charCodeAt(j);
      if (code >= 1424 && code <= 1535 || code >= 1536 && code <= 1791 || code >= 64336 && code <= 65279) {
        result = true;
        break;
      }
    }
    this._rtlCacheValue = this.value;
    this._rtlCacheResult = result;
    return result;
  }
  getLayout() {
    if (this.value === this.cachedValue && this.font === this.cachedFont && this.cachedLayout) {
      return this.cachedLayout;
    }
    const fSize = fontSizePx(this.font);
    const engine = new LayoutEngine(1e6, 1e3, {
      measure: (char) => measureText(char, this.font)
    });
    engine.preserveLeadingSpaces = true;
    const layout = engine.layoutText(this.value, {}, fSize);
    this.cachedValue = this.value;
    this.cachedFont = this.font;
    this.cachedLayout = layout;
    return layout;
  }
  /** Measured x of the glyph boundary before char index `i` (text-relative). */
  charOffset(i) {
    if (this.value.length === 0) return 0;
    if (!this.hasRTL()) {
      return measureText(this.value.slice(0, i), this.font);
    }
    const layout = this.getLayout();
    if (layout.nodes.length === 0) return 0;
    let targetNode = null;
    let isRTL = false;
    for (const node of layout.nodes) {
      const start2 = node.sourceIndex ?? 0;
      const len2 = node.sourceLength ?? 0;
      if (i >= start2 && i <= start2 + len2) {
        targetNode = node;
        isRTL = !!node.isRTL;
        if (i > start2 && i < start2 + len2) {
          break;
        }
      }
    }
    if (!targetNode) {
      let maxNode = layout.nodes[0];
      for (const node of layout.nodes) {
        if ((node.sourceIndex ?? 0) + (node.sourceLength ?? 0) > (maxNode.sourceIndex ?? 0) + (maxNode.sourceLength ?? 0)) {
          maxNode = node;
        }
      }
      targetNode = maxNode;
      isRTL = !!maxNode.isRTL;
    }
    const start = targetNode.sourceIndex ?? 0;
    const len = targetNode.sourceLength ?? 0;
    const fraction = len > 0 ? (i - start) / len : 0;
    if (isRTL) {
      return targetNode.x + targetNode.width * (1 - fraction);
    } else {
      return targetNode.x + targetNode.width * fraction;
    }
  }
  /** Screen-space x of the caret (after scroll), in the component's box coords. */
  caretScreenX() {
    return this.padding - this.scrollLeft + this.charOffset(this.selectionStart);
  }
  /** Keep the caret within the padded inner box by adjusting `scrollLeft`. */
  updateScroll() {
    const innerWidth = this.width - 2 * this.padding;
    const caretAbs = this.charOffset(this.selectionStart);
    if (caretAbs - this.scrollLeft > innerWidth) this.scrollLeft = caretAbs - innerWidth;
    if (caretAbs - this.scrollLeft < 0) this.scrollLeft = caretAbs;
    if (this.scrollLeft < 0) this.scrollLeft = 0;
  }
  /** Caret blink "on" phase (500ms cadence). */
  caretOn() {
    return Math.floor(Date.now() / 500) % 2 === 0;
  }
  render(r) {
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.fill(this.bg);
    r.beginPath();
    r.roundRect(0, 0, this.width, this.height, this.radius);
    r.stroke(this.border, 1);
    this.updateScroll();
    const innerWidth = this.width - 2 * this.padding;
    const editorLineHeight = this.height - 2 * this.padding;
    const baseline = this.padding + cssLineBoxBaseline(this.font, editorLineHeight);
    const textOriginX = this.padding - this.scrollLeft;
    r.save();
    r.clip(this.padding, this.padding, innerWidth, editorLineHeight);
    if (this.selectionStart !== this.selectionEnd) {
      const a = Math.min(this.selectionStart, this.selectionEnd);
      const b = Math.max(this.selectionStart, this.selectionEnd);
      if (!this.hasRTL()) {
        const sx = textOriginX + this.charOffset(a);
        const ex = textOriginX + this.charOffset(b);
        r.beginPath();
        r.roundRect(sx, this.padding, ex - sx, editorLineHeight, 0);
        r.fill(this.selectionColor);
      } else {
        const layout = this.getLayout();
        r.beginPath();
        for (const node of layout.nodes) {
          const start = node.sourceIndex ?? 0;
          const len = node.sourceLength ?? 0;
          if (!(start + len <= a || start >= b)) {
            r.roundRect(textOriginX + node.x, this.padding, node.width, editorLineHeight, 0);
          }
        }
        r.fill(this.selectionColor);
      }
    }
    const text = this.value || this.placeholder;
    const color = this.value ? this.color : this.placeholderColor;
    r.fillText(text, textOriginX, baseline, this.font, color);
    if (this.composition && this.composition.length > 0) {
      const ux0 = textOriginX + this.charOffset(this.composition.start);
      const ux1 = textOriginX + this.charOffset(this.composition.start + this.composition.length);
      const uy = baseline + 2;
      r.beginPath();
      r.moveTo(ux0, uy);
      r.lineTo(ux1, uy);
      r.stroke(this.color, 1);
    }
    if (this.focused && this.caretOn()) {
      const cx = this.caretScreenX();
      r.beginPath();
      r.moveTo(cx, this.padding + 2);
      r.lineTo(cx, this.padding + editorLineHeight - 2);
      r.stroke(this.color, 1);
    }
    r.restore();
  }
};
export {
  Input
};
