// @vectojs/ui/dist/Text.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/Text.ts
var Text_exports = {};
__export(Text_exports, {
  Text: () => Text
});
module.exports = __toCommonJS(Text_exports);
var import_core3 = require("@vectojs/core");

// src/UIComponent.ts
var import_core = require("@vectojs/core");
var UIComponent = class extends import_core.Entity {
  /** Inner padding in pixels, used by box-style components. */
  padding = 0;
  /** Played automatically when the component mounts to a live scene, if set. */
  enterMotion;
  /** Played by {@link dismiss} before the component removes itself, if set. */
  exitMotion;
  onMounted() {
    if (this.enterMotion) void this.playMotion(this.enterMotion);
  }
  /** Seed each prop to `from`, register the transition, then drive to `to`. */
  playMotion(spec) {
    const cfg = spec.config ?? "spring";
    const targets = {};
    const trans = {};
    for (const [k, pair] of Object.entries(spec.props)) {
      const prop = k;
      const [from, to] = pair;
      this.setImmediate(prop, from);
      targets[prop] = to;
      trans[prop] = cfg;
    }
    this.setTransition(trans);
    return this.driveMotion(targets, cfg);
  }
  driveMotion(targets, cfg) {
    if (typeof cfg === "object" && "duration" in cfg) {
      return this.animateTo(targets, cfg);
    }
    return this.springTo(targets, cfg === "spring" ? {} : cfg);
  }
  /** Play the exit motion (if any), then remove self from the tree. */
  async dismiss() {
    if (this.exitMotion) await this.playMotion(this.exitMotion);
    this.parent?.remove(this);
  }
  /** Wakes the render loop at the next caret-blink phase boundary, or `null`. */
  caretBlinkTimer = null;
  /**
   * Keep a focused text component's caret blink visible to the idle throttle.
   * The blink phase is derived from `Date.now()` inside `render()`, so nothing
   * marks the scene dirty when the phase flips — an idle `onDemand` scene
   * would freeze the caret solid (the ScrollView 0.2.x regression class).
   * One timeout per 500 ms phase boundary costs ~2 renders/s while focused;
   * reporting a permanent pending animation instead would pin the scene at
   * full frame rate for as long as the field holds focus.
   */
  startCaretBlinkWake() {
    this.stopCaretBlinkWake();
    const schedule = () => {
      this.caretBlinkTimer = setTimeout(
        () => {
          this.scene?.markDirty();
          schedule();
        },
        500 - Date.now() % 500
      );
    };
    schedule();
    this.scene?.markDirty();
  }
  /** Stop the caret-blink wake-up (on blur; destroy also clears it). */
  stopCaretBlinkWake() {
    if (this.caretBlinkTimer !== null) {
      clearTimeout(this.caretBlinkTimer);
      this.caretBlinkTimer = null;
    }
    this.scene?.markDirty();
  }
  destroy() {
    if (this.caretBlinkTimer !== null) {
      clearTimeout(this.caretBlinkTimer);
      this.caretBlinkTimer = null;
    }
    super.destroy();
  }
  /**
   * Axis-aligned hit-test against the component's box in global space.
   *
   * @param globalX - World-space X coordinate.
   * @param globalY - World-space Y coordinate.
   * @returns Whether the point lies within `[0, width] x [0, height]` locally.
   */
  isPointInside(globalX, globalY) {
    const local = this.worldToLocal(globalX, globalY);
    if (!local) return false;
    return local.x >= 0 && local.x <= this.width && local.y >= 0 && local.y <= this.height;
  }
  /** Local-space box, enabling viewport culling by {@link Scene}. */
  getBounds() {
    return { x: 0, y: 0, width: this.width, height: this.height };
  }
};

// src/measure.ts
var import_core2 = require("@vectojs/core");
function fontSizePx(font) {
  const pxIndex = font.indexOf("px");
  if (pxIndex <= 0) return 16;
  let start = pxIndex - 1;
  while (start >= 0) {
    const ch = font[start];
    if (ch >= "0" && ch <= "9" || ch === ".") {
      start--;
    } else {
      break;
    }
  }
  const raw = font.slice(start + 1, pxIndex);
  if (raw === "") return 16;
  const size = Number.parseFloat(raw);
  return Number.isFinite(size) ? size : 16;
}
var measureCache = /* @__PURE__ */ new Map();
if (typeof document !== "undefined" && document.fonts) {
  const clearCache = () => measureCache.clear();
  document.fonts.ready.then(clearCache);
  document.fonts.addEventListener("loadingdone", clearCache);
}

// src/Text.ts
function fontMeasurer(font) {
  if (typeof document === "undefined") return null;
  const ctx = document.createElement("canvas").getContext("2d");
  if (!ctx) return null;
  const cache = /* @__PURE__ */ new Map();
  return {
    measure(char) {
      let w = cache.get(char);
      if (w === void 0) {
        ctx.font = font;
        w = ctx.measureText(char).width;
        cache.set(char, w);
      }
      return w;
    }
  };
}
var Text = class extends UIComponent {
  text;
  font;
  color;
  maxWidth;
  lineHeight;
  selectable;
  engine;
  prepared;
  fontSize;
  lines = [];
  lineSourceRanges = [];
  constructor(text, opts = {}) {
    super();
    this.text = text;
    this.font = opts.font ?? "16px sans-serif";
    this.color = opts.color ?? "#e2e8f0";
    this.maxWidth = opts.maxWidth;
    this.lineHeight = opts.lineHeight ?? 20;
    this.selectable = opts.selectable ?? true;
    this.fontSize = fontSizePx(this.font);
    this.engine = new import_core3.LayoutEngine(this.maxWidth ?? 1e9, 1e9, fontMeasurer(this.font));
    if (opts.preserveLeadingSpaces) {
      this.engine.preserveLeadingSpaces = true;
    }
    this.prepared = this.engine.prepare(this.text, {}, this.fontSize);
    this.interactive = false;
    this.applyLayout();
  }
  /**
   * Replace the text. Runs the **cold** pass (re-segment + re-measure), then re-lays out.
   *
   * @param text - The new text content.
   * @returns `this` for chaining.
   */
  /** Mirror the rendered text into the DOM content layer (find-in-page, SR, SEO). */
  getContentProjection() {
    if (!this.text) return null;
    const lines = this.lines.map((visualText, index) => {
      const range = this.lineSourceRanges[index] ?? { start: 0, end: 0 };
      const nextStart = this.lineSourceRanges[index + 1]?.start ?? this.text.length;
      return {
        // Canvas keeps its visual glyph order; the semantic layer keeps logical
        // source order so native copy and RTL text remain correct.
        text: this.text.slice(range.start, range.end) || visualText,
        separatorAfter: this.text.slice(range.end, Math.max(range.end, nextStart)),
        x: 0,
        y: index * this.lineHeight,
        baseline: this.lineHeight * 0.8,
        font: this.font,
        lineHeight: this.lineHeight
      };
    });
    return {
      text: this.text,
      font: this.font,
      lineHeight: this.lineHeight,
      // The canvas has already decided the exact line breaks. Project visual
      // rows independently so browser whitespace/wrapping can never create a
      // different selection grid at a fractional zoom level.
      lines,
      selectable: this.selectable
    };
  }
  /** Enable or disable browser-native drag selection without rebuilding the entity. */
  setSelectable(selectable) {
    this.selectable = selectable;
    this.scene?.markDirty();
    return this;
  }
  setText(text) {
    this.text = text;
    this.prepared = this.engine.prepare(this.text, {}, this.fontSize);
    this.applyLayout();
    this.scene?.markDirty();
    return this;
  }
  /**
   * Append text — the streaming / typewriter path. Goes through the same cold
   * pass as {@link setText}, but the engine's paragraph memo reuses every
   * untouched leading paragraph, so only the changed (last) paragraph is
   * re-segmented + re-measured.
   *
   * @returns `this` for chaining.
   */
  append(text) {
    return this.setText(this.text + text);
  }
  /**
   * Change the wrap width and reflow via the cheap **hot** path (reuses the cached
   * measured text — no re-segmentation or re-measurement).
   *
   * @returns `this` for chaining.
   */
  setMaxWidth(maxWidth) {
    this.maxWidth = maxWidth;
    this.engine.maxWidth = maxWidth;
    this.applyLayout();
    this.scene?.markDirty();
    return this;
  }
  /** Hot pass: place the cached prepared text and regroup glyphs into lines. */
  applyLayout() {
    const result = this.engine.layoutPrepared(this.prepared);
    const lineQuantum = this.fontSize * 1.5;
    const byLine = /* @__PURE__ */ new Map();
    const nodesByLine = /* @__PURE__ */ new Map();
    let maxIdx = -1;
    for (const node of result.nodes) {
      const idx = Math.round(node.y / lineQuantum);
      byLine.set(idx, (byLine.get(idx) ?? "") + node.char);
      const nodes = nodesByLine.get(idx) ?? [];
      nodes.push(node);
      nodesByLine.set(idx, nodes);
      if (idx > maxIdx) maxIdx = idx;
    }
    this.lines = [];
    this.lineSourceRanges = [];
    let previousEnd = 0;
    for (let i = 0; i <= maxIdx; i++) {
      this.lines.push(byLine.get(i) ?? "");
      const nodes = nodesByLine.get(i) ?? [];
      let start = nodes.length > 0 ? Math.min(...nodes.map((node) => node.sourceIndex ?? previousEnd)) : previousEnd;
      const end = nodes.length > 0 ? Math.max(
        ...nodes.map((node) => (node.sourceIndex ?? previousEnd) + (node.sourceLength ?? 0))
      ) : start;
      if (i === 0) start = 0;
      this.lineSourceRanges.push({ start, end: Math.max(start, end) });
      previousEnd = Math.max(previousEnd, end);
    }
    if (this.lines.length === 0 && this.text) {
      this.lines = [""];
      this.lineSourceRanges = [{ start: 0, end: 0 }];
    }
    this.width = result.totalWidth;
    this.height = Math.max(maxIdx + 1, 1) * this.lineHeight;
  }
  getA11yAttributes() {
    return { label: this.text };
  }
  render(r) {
    for (let i = 0; i < this.lines.length; i++) {
      if (this.lines[i])
        r.fillText(this.lines[i], 0, (i + 0.8) * this.lineHeight, this.font, this.color);
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Text
});
