// @vectojs/devtools/dist/headless.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/headless.ts
var headless_exports = {};
__export(headless_exports, {
  EventTrace: () => EventTrace,
  auditScene: () => auditScene,
  auditTree: () => auditTree,
  buildTreeModel: () => buildTreeModel,
  captureSnapshot: () => captureSnapshot,
  createEventTrace: () => createEventTrace,
  describeEntity: () => describeEntity,
  diffSnapshots: () => diffSnapshots,
  entityPath: () => entityPath,
  findEntityAt: () => findEntityAt,
  inspectEntity: () => inspectEntity,
  pickInScene: () => pickInScene,
  textPreviewOf: () => textPreviewOf
});
module.exports = __toCommonJS(headless_exports);

// src/model.ts
function buildTreeModel(root) {
  const index = /* @__PURE__ */ new Map();
  const toNode2 = (entity) => {
    index.set(entity.id, entity);
    const type = entity.constructor.name;
    const size = entity.width > 0 || entity.height > 0 ? ` ${Math.round(entity.width)}\xD7${Math.round(entity.height)}` : "";
    const badges = `${entity.interactive ? " \u26A1" : ""}${entity.hasPendingAnimations() ? " \u25B6" : ""}`;
    return {
      id: entity.id,
      label: `${type} (${Math.round(entity.x)},${Math.round(entity.y)})${size}${badges}`,
      children: entity.children.length > 0 ? entity.children.map(toNode2) : void 0
    };
  };
  return { nodes: root.children.map(toNode2), index };
}
function findEntityAt(root, x, y) {
  for (let i = root.children.length - 1; i >= 0; i--) {
    const hit = findEntityAt(root.children[i], x, y);
    if (hit) return hit;
  }
  if (root.isPointInside && root.isPointInside(x, y)) return root;
  if (root.width > 0 && root.height > 0) {
    const local = root.worldToLocal(x, y);
    if (local && local.x >= 0 && local.x <= root.width && local.y >= 0 && local.y <= root.height) {
      return root;
    }
  }
  return null;
}
function describeEntity(entity) {
  const { a, b, c, d, e, f } = entity.getWorldTransform();
  const r = (n) => Math.round(n * 100) / 100;
  return [
    `${entity.constructor.name} #${entity.id}`,
    `x ${r(entity.x)}  y ${r(entity.y)}  w ${r(entity.width)}  h ${r(entity.height)}`,
    `scale ${r(entity.scaleX)},${r(entity.scaleY)}  rot ${r(entity.rotation)}  op ${r(entity.opacity)}`,
    `world [${r(a)} ${r(b)} ${r(c)} ${r(d)} ${r(e)} ${r(f)}]`,
    `interactive ${entity.interactive}  animating ${entity.hasPendingAnimations()}`,
    `children ${entity.children.length}`
  ];
}
function pickInScene(scene, sceneX, sceneY) {
  return findEntityAt(scene.overlayRootEntity, sceneX, sceneY) ?? findEntityAt(scene.rootEntity, sceneX, sceneY);
}

// src/inspect.ts
var round2 = (n) => Math.round(n * 100) / 100;
var TEXT_PREVIEW_MAX = 80;
function textPreviewOf(entity) {
  const candidate = typeof entity.text === "string" ? entity.text : typeof entity.value === "string" ? entity.value : void 0;
  if (candidate === void 0) return void 0;
  return candidate.length > TEXT_PREVIEW_MAX ? `${candidate.slice(0, TEXT_PREVIEW_MAX)}\u2026` : candidate;
}
function entityPath(entity) {
  const segments = [];
  let node = entity;
  while (node) {
    segments.unshift(
      node.parent === null ? "Scene" : `${node.constructor.name}#${node.id.slice(0, 8)}`
    );
    node = node.parent;
  }
  return segments.join(" > ");
}
function roundBounds(b) {
  return { x: round2(b.x), y: round2(b.y), width: round2(b.width), height: round2(b.height) };
}
function inspectEntity(entity) {
  const { a, b, c, d, e, f } = entity.getWorldTransform();
  const info = {
    id: entity.id,
    type: entity.constructor.name,
    path: entityPath(entity),
    x: round2(entity.x),
    y: round2(entity.y),
    width: round2(entity.width),
    height: round2(entity.height),
    scaleX: round2(entity.scaleX),
    scaleY: round2(entity.scaleY),
    rotation: round2(entity.rotation),
    opacity: round2(entity.opacity),
    worldTransform: {
      a: round2(a),
      b: round2(b),
      c: round2(c),
      d: round2(d),
      e: round2(e),
      f: round2(f)
    },
    worldBounds: roundBounds(entity.getWorldBounds()),
    interactive: entity.interactive,
    animating: entity.hasPendingAnimations(),
    clipChildren: entity.clipChildren,
    childCount: entity.children.length
  };
  const text = textPreviewOf(entity);
  if (text !== void 0) info.text = text;
  const a11y = entity.getA11yAttributes();
  if (a11y && (a11y.tag || a11y.role || a11y.label)) {
    info.a11y = { tag: a11y.tag, role: a11y.role, label: a11y.label };
  }
  return info;
}

// src/eventTrace.ts
var TRACE_TYPES = [
  "pointerdown",
  "pointerup",
  "pointercancel",
  "pointermove",
  "wheel",
  "keydown",
  "keyup"
];
var isKeyboardType = (type) => type === "keydown" || type === "keyup";
var EventTrace = class {
  constructor(scene, options = {}) {
    this.scene = scene;
    this.capacity = Math.max(1, Math.floor(options.capacity ?? 50));
    this.includeGlobalKeyboard = options.includeGlobalKeyboard ?? true;
    if (typeof document === "undefined") return;
    for (const type of TRACE_TYPES) {
      document.addEventListener(type, this.onEvent, true);
      document.addEventListener(type, this.onEventBubbled);
    }
  }
  scene;
  capacity;
  includeGlobalKeyboard;
  records = [];
  listeners = /* @__PURE__ */ new Set();
  pendingDefaults = /* @__PURE__ */ new WeakMap();
  destroyed = false;
  onEventBubbled = (event) => {
    const pending = this.pendingDefaults.get(event);
    if (!pending) return;
    pending.prevented ||= event.defaultPrevented;
    queueMicrotask(pending.finalize);
  };
  onEvent = (event) => {
    const type = event.type;
    if (!TRACE_TYPES.includes(type)) return;
    const context = this.resolveContext(event, type);
    if (!context) return;
    const target = event.target;
    let finalized = false;
    const pending = {
      prevented: event.defaultPrevented,
      finalize: () => {
        if (finalized) return;
        finalized = true;
        target?.removeEventListener(type, observeTargetResult);
        this.pendingDefaults.delete(event);
        if (this.destroyed) return;
        const entry = this.createEntry(event, type, context, pending.prevented);
        this.records.push(entry);
        if (this.records.length > this.capacity) this.records.shift();
        for (const listener of this.listeners) listener(entry);
      }
    };
    const observeTargetResult = () => {
      pending.prevented ||= event.defaultPrevented;
    };
    this.pendingDefaults.set(event, pending);
    target?.addEventListener(type, observeTargetResult, { once: true });
    setTimeout(pending.finalize, 0);
  };
  /** Retained records ordered from oldest to newest. */
  get entries() {
    return this.records;
  }
  /** Subscribe to completed records. Returns an idempotent unsubscribe function. */
  subscribe(listener) {
    if (this.destroyed) return () => {
    };
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }
  /** Remove retained records without affecting observation. */
  clear() {
    this.records.length = 0;
  }
  /** Stop observing and release subscribers. Safe to call more than once. */
  destroy() {
    if (this.destroyed) return;
    this.destroyed = true;
    if (typeof document !== "undefined") {
      for (const type of TRACE_TYPES) {
        document.removeEventListener(type, this.onEvent, true);
        document.removeEventListener(type, this.onEventBubbled);
      }
    }
    this.listeners.clear();
  }
  resolveContext(event, type) {
    const projected = this.projectedTarget(event.target);
    if (projected) return projected;
    const target = event.target;
    if (target === this.scene.canvas) {
      const point = scenePointOf(this.scene, event);
      return { source: "canvas", target: point ? pickInScene(this.scene, point.x, point.y) : null };
    }
    if (isKeyboardType(type) && this.includeGlobalKeyboard)
      return { source: "document", target: null };
    return null;
  }
  projectedTarget(target) {
    if (!(target instanceof Element)) return null;
    const element = target.closest("[data-vecto-content], [data-vecto-id]");
    if (!element) return null;
    const source = element.dataset.vectoContent ? "content" : "a11y";
    const id = element.dataset.vectoContent ?? element.dataset.vectoId;
    if (!id) return null;
    const projectedElement = source === "content" ? typeof this.scene.getContentElement === "function" ? this.scene.getContentElement(id) : element.isConnected ? element : void 0 : this.scene.getA11yElement(id);
    if (projectedElement !== element) return null;
    const entity = findEntityById(this.scene.rootEntity, id) ?? findEntityById(this.scene.overlayRootEntity, id);
    if (!entity || source === "content" && !entity.getContentProjection()) return null;
    return { source, target: entity };
  }
  createEntry(event, type, context, defaultPrevented) {
    const pointer = scenePointOf(this.scene, event);
    const local = context.target && pointer ? context.target.worldToLocal(pointer.x, pointer.y) : null;
    const keyboard = event;
    const wheel = event;
    return Object.freeze({
      type,
      source: context.source,
      timeStamp: event.timeStamp,
      ...context.target ? { targetId: context.target.id, targetPath: entityPath(context.target) } : {},
      ...pointer ? { sceneX: pointer.x, sceneY: pointer.y } : {},
      ...local ? { localX: local.x, localY: local.y } : {},
      ...isKeyboardType(type) ? { key: keyboard.key } : {},
      modifiers: {
        shift: keyboard.shiftKey,
        ctrl: keyboard.ctrlKey,
        alt: keyboard.altKey,
        meta: keyboard.metaKey
      },
      ...type === "wheel" ? { deltaX: wheel.deltaX, deltaY: wheel.deltaY } : {},
      defaultPrevented
    });
  }
};
function scenePointOf(scene, event) {
  const candidate = event;
  if (typeof candidate.clientX !== "number" || typeof candidate.clientY !== "number") return null;
  return scene.clientToScene(candidate.clientX, candidate.clientY);
}
function findEntityById(root, id) {
  if (root.id === id) return root;
  for (const child of root.children) {
    const found = findEntityById(child, id);
    if (found) return found;
  }
  return null;
}
function createEventTrace(scene, options) {
  return new EventTrace(scene, options);
}

// src/snapshot.ts
var round22 = (n) => Math.round(n * 100) / 100;
function toNode(entity) {
  const wb = entity.getWorldBounds();
  const node = {
    type: entity.constructor.name,
    id: entity.id,
    x: round22(entity.x),
    y: round22(entity.y),
    width: round22(entity.width),
    height: round22(entity.height),
    worldBounds: {
      x: round22(wb.x),
      y: round22(wb.y),
      width: round22(wb.width),
      height: round22(wb.height)
    }
  };
  if (entity.opacity !== 1) node.opacity = round22(entity.opacity);
  if (entity.interactive) node.interactive = true;
  if (entity.hasPendingAnimations()) node.animating = true;
  if (entity.clipChildren) node.clipChildren = true;
  const text = textPreviewOf(entity);
  if (text !== void 0) node.text = text;
  if (entity.children.length > 0) node.children = entity.children.map(toNode);
  return node;
}
function captureSnapshot(scene) {
  return {
    width: round22(scene.width),
    height: round22(scene.height),
    root: scene.rootEntity.children.map(toNode),
    overlay: scene.overlayRootEntity.children.map(toNode)
  };
}
var COMPARED_KEYS = [
  "type",
  "x",
  "y",
  "width",
  "height",
  "worldBounds",
  "opacity",
  "interactive",
  "animating",
  "clipChildren",
  "text"
];
function pathOf(parent, node, index) {
  return `${parent} > ${node.type}[${index}]`;
}
function diffNodes(a, b, parent, out) {
  const max = Math.max(a.length, b.length);
  for (let i = 0; i < max; i++) {
    const na = a[i];
    const nb = b[i];
    if (na && !nb) {
      out.push({ path: pathOf(parent, na, i), kind: "removed" });
      continue;
    }
    if (!na && nb) {
      out.push({ path: pathOf(parent, nb, i), kind: "added" });
      continue;
    }
    if (!na || !nb) continue;
    const path = pathOf(parent, nb, i);
    const changes = {};
    for (const key of COMPARED_KEYS) {
      const va = na[key];
      const vb = nb[key];
      const same = key === "worldBounds" ? JSON.stringify(va) === JSON.stringify(vb) : va === vb;
      if (!same) changes[key] = { from: va, to: vb };
    }
    if (Object.keys(changes).length > 0) out.push({ path, kind: "changed", changes });
    diffNodes(na.children ?? [], nb.children ?? [], path, out);
  }
}
function diffSnapshots(a, b) {
  const out = [];
  diffNodes(a.root, b.root, "root", out);
  diffNodes(a.overlay, b.overlay, "overlay", out);
  return out;
}

// src/audit.ts
var DEFAULT_SCROLLABLE = ["ScrollView", "VirtualList", "TreeView", "Tree"];
var round23 = (n) => Math.round(n * 100) / 100;
function roundBounds2(b) {
  return { x: round23(b.x), y: round23(b.y), width: round23(b.width), height: round23(b.height) };
}
function worldBox(entity) {
  const { a, b, c, d, e, f } = entity.getWorldTransform();
  const xs = [
    e,
    a * entity.width + e,
    c * entity.height + e,
    a * entity.width + c * entity.height + e
  ];
  const ys = [
    f,
    b * entity.width + f,
    d * entity.height + f,
    b * entity.width + d * entity.height + f
  ];
  const minX = Math.min(...xs);
  const minY = Math.min(...ys);
  return { x: minX, y: minY, width: Math.max(...xs) - minX, height: Math.max(...ys) - minY };
}
function escapes(inner, outer) {
  return {
    left: Math.max(0, outer.x - inner.x),
    right: Math.max(0, inner.x + inner.width - (outer.x + outer.width)),
    top: Math.max(0, outer.y - inner.y),
    bottom: Math.max(0, inner.y + inner.height - (outer.y + outer.height))
  };
}
function intersect(a, b) {
  const x = Math.max(a.x, b.x);
  const y = Math.max(a.y, b.y);
  const right = Math.min(a.x + a.width, b.x + b.width);
  const bottom = Math.min(a.y + a.height, b.y + b.height);
  if (right <= x || bottom <= y) return null;
  return { x, y, width: right - x, height: bottom - y };
}
var isTextLike = (entity) => textPreviewOf(entity) !== void 0;
function auditTree(root, sceneBounds, opts = {}) {
  const tolerance = opts.tolerance ?? 0.5;
  const scrollable = new Set(opts.scrollableTypes ?? DEFAULT_SCROLLABLE);
  const findings = [];
  const overflowFinding = (kind, entity, bounds, container, over) => {
    const sides = ["left", "right", "top", "bottom"].filter((s) => over[s] > tolerance).map((s) => `${s} ${round23(over[s])}px`).join(", ");
    findings.push({
      kind,
      entityId: entity.id,
      entityPath: entityPath(entity),
      worldBounds: roundBounds2(bounds),
      message: `${entity.constructor.name} escapes ${container ? container.entity.constructor.name : "scene viewport"} by ${sides}`,
      ...container ? {
        containerId: container.entity.id,
        containerPath: entityPath(container.entity),
        containerBounds: roundBounds2(container.box)
      } : {},
      overflow: {
        left: round23(over.left),
        right: round23(over.right),
        top: round23(over.top),
        bottom: round23(over.bottom)
      }
    });
  };
  const beyond = (over) => over.left > tolerance || over.right > tolerance || over.top > tolerance || over.bottom > tolerance;
  const checkSiblingOverlaps = (kids) => {
    for (let i = 0; i < kids.length; i++) {
      const a = kids[i];
      if (a.opacity <= 0 || a.width <= 0 || a.height <= 0 || opts.ignore?.(a)) continue;
      const boxA = worldBox(a);
      for (let j = i + 1; j < kids.length; j++) {
        const b = kids[j];
        if (b.opacity <= 0 || b.width <= 0 || b.height <= 0 || opts.ignore?.(b)) continue;
        if (opts.ignoreOverlap?.(a, b)) continue;
        const inter = intersect(boxA, worldBox(b));
        if (inter && inter.width > tolerance && inter.height > tolerance) {
          findings.push({
            kind: "overlap",
            entityId: a.id,
            entityPath: entityPath(a),
            worldBounds: roundBounds2(boxA),
            otherId: b.id,
            otherPath: entityPath(b),
            otherBounds: roundBounds2(worldBox(b)),
            intersection: roundBounds2(inter),
            message: `${a.constructor.name} overlaps sibling ${b.constructor.name} by ${round23(inter.width)}\xD7${round23(inter.height)}px`
          });
        }
      }
    }
  };
  const walk = (entity, ancestors) => {
    if (opts.ignore?.(entity)) return;
    if (entity.opacity <= 0) return;
    const bounds = entity.getWorldBounds();
    const hasSize = entity.width > 0 && entity.height > 0;
    if (hasSize) {
      if (isTextLike(entity)) {
        const container = [...ancestors].reverse().find((an) => an.sized && !an.textLike);
        if (container) {
          const over = escapes(bounds, container.box);
          if (container.scrollable) over.top = over.bottom = 0;
          if (beyond(over)) overflowFinding("text-overflow", entity, bounds, container, over);
        }
      }
      const clipper = [...ancestors].reverse().find((an) => an.clips);
      if (clipper) {
        const over = escapes(bounds, clipper.box);
        if (clipper.scrollable) over.top = over.bottom = 0;
        if (beyond(over)) overflowFinding("clip-overflow", entity, bounds, clipper, over);
      } else if (sceneBounds && ancestors.every((an) => !an.sized)) {
        const over = escapes(bounds, sceneBounds);
        if (beyond(over)) overflowFinding("viewport-overflow", entity, bounds, null, over);
      }
    }
    checkSiblingOverlaps(entity.children);
    const self = {
      entity,
      box: worldBox(entity),
      clips: entity.clipChildren,
      scrollable: entity.clipChildren && scrollable.has(entity.constructor.name),
      sized: hasSize,
      textLike: isTextLike(entity)
    };
    ancestors.push(self);
    for (const child of entity.children) walk(child, ancestors);
    ancestors.pop();
  };
  checkSiblingOverlaps(root.children);
  for (const child of root.children) walk(child, []);
  findings.sort(
    (a, b) => a.kind !== b.kind ? a.kind.localeCompare(b.kind) : a.entityPath !== b.entityPath ? a.entityPath.localeCompare(b.entityPath) : (a.otherPath ?? "").localeCompare(b.otherPath ?? "")
  );
  return findings;
}
function auditScene(scene, opts = {}) {
  const sceneBounds = { x: 0, y: 0, width: scene.width, height: scene.height };
  const findings = auditTree(scene.rootEntity, sceneBounds, opts);
  if (opts.includeOverlay) {
    findings.push(...auditTree(scene.overlayRootEntity, sceneBounds, opts));
  }
  return findings;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EventTrace,
  auditScene,
  auditTree,
  buildTreeModel,
  captureSnapshot,
  createEventTrace,
  describeEntity,
  diffSnapshots,
  entityPath,
  findEntityAt,
  inspectEntity,
  pickInScene,
  textPreviewOf
});
