// @vectojs/devtools/dist/index.mjs
// src/panel.ts
import { Scene, Entity } from "@vectojs/core";
import { TreeView, Text, Button } from "@vectojs/ui";

// src/model.ts
function buildTreeModel(root) {
  const index = /* @__PURE__ */ new Map();
  const toNode2 = (entity) => {
    index.set(entity.id, entity);
    const type = entity.constructor.name;
    const size = entity.width > 0 || entity.height > 0 ? ` ${Math.round(entity.width)}\xD7${Math.round(entity.height)}` : "";
    const badges = `${entity.interactive ? " \u26A1" : ""}${entity.hasPendingAnimations() ? " \u25B6" : ""}`;
    return {
      id: entity.id,
      label: `${type} (${Math.round(entity.x)},${Math.round(entity.y)})${size}${badges}`,
      children: entity.children.length > 0 ? entity.children.map(toNode2) : void 0
    };
  };
  return { nodes: root.children.map(toNode2), index };
}
function findEntityAt(root, x, y) {
  for (let i = root.children.length - 1; i >= 0; i--) {
    const hit = findEntityAt(root.children[i], x, y);
    if (hit) return hit;
  }
  if (root.isPointInside && root.isPointInside(x, y)) return root;
  if (root.width > 0 && root.height > 0) {
    const local = root.worldToLocal(x, y);
    if (local && local.x >= 0 && local.x <= root.width && local.y >= 0 && local.y <= root.height) {
      return root;
    }
  }
  return null;
}
function describeEntity(entity) {
  const { a, b, c, d, e, f } = entity.getWorldTransform();
  const r = (n) => Math.round(n * 100) / 100;
  return [
    `${entity.constructor.name} #${entity.id}`,
    `x ${r(entity.x)}  y ${r(entity.y)}  w ${r(entity.width)}  h ${r(entity.height)}`,
    `scale ${r(entity.scaleX)},${r(entity.scaleY)}  rot ${r(entity.rotation)}  op ${r(entity.opacity)}`,
    `world [${r(a)} ${r(b)} ${r(c)} ${r(d)} ${r(e)} ${r(f)}]`,
    `interactive ${entity.interactive}  animating ${entity.hasPendingAnimations()}`,
    `children ${entity.children.length}`
  ];
}
function pickInScene(scene, sceneX, sceneY) {
  return findEntityAt(scene.overlayRootEntity, sceneX, sceneY) ?? findEntityAt(scene.rootEntity, sceneX, sceneY);
}

// src/inspect.ts
var round2 = (n) => Math.round(n * 100) / 100;
var TEXT_PREVIEW_MAX = 80;
function textPreviewOf(entity) {
  const candidate = typeof entity.text === "string" ? entity.text : typeof entity.value === "string" ? entity.value : void 0;
  if (candidate === void 0) return void 0;
  return candidate.length > TEXT_PREVIEW_MAX ? `${candidate.slice(0, TEXT_PREVIEW_MAX)}\u2026` : candidate;
}
function entityPath(entity) {
  const segments = [];
  let node = entity;
  while (node) {
    segments.unshift(
      node.parent === null ? "Scene" : `${node.constructor.name}#${node.id.slice(0, 8)}`
    );
    node = node.parent;
  }
  return segments.join(" > ");
}
function roundBounds(b) {
  return { x: round2(b.x), y: round2(b.y), width: round2(b.width), height: round2(b.height) };
}
function inspectEntity(entity) {
  const { a, b, c, d, e, f } = entity.getWorldTransform();
  const info = {
    id: entity.id,
    type: entity.constructor.name,
    path: entityPath(entity),
    x: round2(entity.x),
    y: round2(entity.y),
    width: round2(entity.width),
    height: round2(entity.height),
    scaleX: round2(entity.scaleX),
    scaleY: round2(entity.scaleY),
    rotation: round2(entity.rotation),
    opacity: round2(entity.opacity),
    worldTransform: {
      a: round2(a),
      b: round2(b),
      c: round2(c),
      d: round2(d),
      e: round2(e),
      f: round2(f)
    },
    worldBounds: roundBounds(entity.getWorldBounds()),
    interactive: entity.interactive,
    animating: entity.hasPendingAnimations(),
    clipChildren: entity.clipChildren,
    childCount: entity.children.length
  };
  const text = textPreviewOf(entity);
  if (text !== void 0) info.text = text;
  const a11y = entity.getA11yAttributes();
  if (a11y && (a11y.tag || a11y.role || a11y.label)) {
    info.a11y = { tag: a11y.tag, role: a11y.role, label: a11y.label };
  }
  return info;
}

// src/audit.ts
var DEFAULT_SCROLLABLE = ["ScrollView", "VirtualList", "TreeView", "Tree"];
var round22 = (n) => Math.round(n * 100) / 100;
function roundBounds2(b) {
  return { x: round22(b.x), y: round22(b.y), width: round22(b.width), height: round22(b.height) };
}
function worldBox(entity) {
  const { a, b, c, d, e, f } = entity.getWorldTransform();
  const xs = [
    e,
    a * entity.width + e,
    c * entity.height + e,
    a * entity.width + c * entity.height + e
  ];
  const ys = [
    f,
    b * entity.width + f,
    d * entity.height + f,
    b * entity.width + d * entity.height + f
  ];
  const minX = Math.min(...xs);
  const minY = Math.min(...ys);
  return { x: minX, y: minY, width: Math.max(...xs) - minX, height: Math.max(...ys) - minY };
}
function escapes(inner, outer) {
  return {
    left: Math.max(0, outer.x - inner.x),
    right: Math.max(0, inner.x + inner.width - (outer.x + outer.width)),
    top: Math.max(0, outer.y - inner.y),
    bottom: Math.max(0, inner.y + inner.height - (outer.y + outer.height))
  };
}
function intersect(a, b) {
  const x = Math.max(a.x, b.x);
  const y = Math.max(a.y, b.y);
  const right = Math.min(a.x + a.width, b.x + b.width);
  const bottom = Math.min(a.y + a.height, b.y + b.height);
  if (right <= x || bottom <= y) return null;
  return { x, y, width: right - x, height: bottom - y };
}
var isTextLike = (entity) => textPreviewOf(entity) !== void 0;
function auditTree(root, sceneBounds, opts = {}) {
  const tolerance = opts.tolerance ?? 0.5;
  const scrollable = new Set(opts.scrollableTypes ?? DEFAULT_SCROLLABLE);
  const findings = [];
  const overflowFinding = (kind, entity, bounds, container, over) => {
    const sides = ["left", "right", "top", "bottom"].filter((s) => over[s] > tolerance).map((s) => `${s} ${round22(over[s])}px`).join(", ");
    findings.push({
      kind,
      entityId: entity.id,
      entityPath: entityPath(entity),
      worldBounds: roundBounds2(bounds),
      message: `${entity.constructor.name} escapes ${container ? container.entity.constructor.name : "scene viewport"} by ${sides}`,
      ...container ? {
        containerId: container.entity.id,
        containerPath: entityPath(container.entity),
        containerBounds: roundBounds2(container.box)
      } : {},
      overflow: {
        left: round22(over.left),
        right: round22(over.right),
        top: round22(over.top),
        bottom: round22(over.bottom)
      }
    });
  };
  const beyond = (over) => over.left > tolerance || over.right > tolerance || over.top > tolerance || over.bottom > tolerance;
  const checkSiblingOverlaps = (kids) => {
    for (let i = 0; i < kids.length; i++) {
      const a = kids[i];
      if (a.opacity <= 0 || a.width <= 0 || a.height <= 0 || opts.ignore?.(a)) continue;
      const boxA = worldBox(a);
      for (let j = i + 1; j < kids.length; j++) {
        const b = kids[j];
        if (b.opacity <= 0 || b.width <= 0 || b.height <= 0 || opts.ignore?.(b)) continue;
        if (opts.ignoreOverlap?.(a, b)) continue;
        const inter = intersect(boxA, worldBox(b));
        if (inter && inter.width > tolerance && inter.height > tolerance) {
          findings.push({
            kind: "overlap",
            entityId: a.id,
            entityPath: entityPath(a),
            worldBounds: roundBounds2(boxA),
            otherId: b.id,
            otherPath: entityPath(b),
            otherBounds: roundBounds2(worldBox(b)),
            intersection: roundBounds2(inter),
            message: `${a.constructor.name} overlaps sibling ${b.constructor.name} by ${round22(inter.width)}\xD7${round22(inter.height)}px`
          });
        }
      }
    }
  };
  const walk = (entity, ancestors) => {
    if (opts.ignore?.(entity)) return;
    if (entity.opacity <= 0) return;
    const bounds = entity.getWorldBounds();
    const hasSize = entity.width > 0 && entity.height > 0;
    if (hasSize) {
      if (isTextLike(entity)) {
        const container = [...ancestors].reverse().find((an) => an.sized && !an.textLike);
        if (container) {
          const over = escapes(bounds, container.box);
          if (container.scrollable) over.top = over.bottom = 0;
          if (beyond(over)) overflowFinding("text-overflow", entity, bounds, container, over);
        }
      }
      const clipper = [...ancestors].reverse().find((an) => an.clips);
      if (clipper) {
        const over = escapes(bounds, clipper.box);
        if (clipper.scrollable) over.top = over.bottom = 0;
        if (beyond(over)) overflowFinding("clip-overflow", entity, bounds, clipper, over);
      } else if (sceneBounds && ancestors.every((an) => !an.sized)) {
        const over = escapes(bounds, sceneBounds);
        if (beyond(over)) overflowFinding("viewport-overflow", entity, bounds, null, over);
      }
    }
    checkSiblingOverlaps(entity.children);
    const self = {
      entity,
      box: worldBox(entity),
      clips: entity.clipChildren,
      scrollable: entity.clipChildren && scrollable.has(entity.constructor.name),
      sized: hasSize,
      textLike: isTextLike(entity)
    };
    ancestors.push(self);
    for (const child of entity.children) walk(child, ancestors);
    ancestors.pop();
  };
  checkSiblingOverlaps(root.children);
  for (const child of root.children) walk(child, []);
  findings.sort(
    (a, b) => a.kind !== b.kind ? a.kind.localeCompare(b.kind) : a.entityPath !== b.entityPath ? a.entityPath.localeCompare(b.entityPath) : (a.otherPath ?? "").localeCompare(b.otherPath ?? "")
  );
  return findings;
}
function auditScene(scene, opts = {}) {
  const sceneBounds = { x: 0, y: 0, width: scene.width, height: scene.height };
  const findings = auditTree(scene.rootEntity, sceneBounds, opts);
  if (opts.includeOverlay) {
    findings.push(...auditTree(scene.overlayRootEntity, sceneBounds, opts));
  }
  return findings;
}

// src/eventTrace.ts
var TRACE_TYPES = [
  "pointerdown",
  "pointerup",
  "pointercancel",
  "pointermove",
  "wheel",
  "keydown",
  "keyup"
];
var isKeyboardType = (type) => type === "keydown" || type === "keyup";
var EventTrace = class {
  constructor(scene, options = {}) {
    this.scene = scene;
    this.capacity = Math.max(1, Math.floor(options.capacity ?? 50));
    this.includeGlobalKeyboard = options.includeGlobalKeyboard ?? true;
    if (typeof document === "undefined") return;
    for (const type of TRACE_TYPES) {
      document.addEventListener(type, this.onEvent, true);
      document.addEventListener(type, this.onEventBubbled);
    }
  }
  scene;
  capacity;
  includeGlobalKeyboard;
  records = [];
  listeners = /* @__PURE__ */ new Set();
  pendingDefaults = /* @__PURE__ */ new WeakMap();
  destroyed = false;
  onEventBubbled = (event) => {
    const pending = this.pendingDefaults.get(event);
    if (!pending) return;
    pending.prevented ||= event.defaultPrevented;
    queueMicrotask(pending.finalize);
  };
  onEvent = (event) => {
    const type = event.type;
    if (!TRACE_TYPES.includes(type)) return;
    const context = this.resolveContext(event, type);
    if (!context) return;
    const target = event.target;
    let finalized = false;
    const pending = {
      prevented: event.defaultPrevented,
      finalize: () => {
        if (finalized) return;
        finalized = true;
        target?.removeEventListener(type, observeTargetResult);
        this.pendingDefaults.delete(event);
        if (this.destroyed) return;
        const entry = this.createEntry(event, type, context, pending.prevented);
        this.records.push(entry);
        if (this.records.length > this.capacity) this.records.shift();
        for (const listener of this.listeners) listener(entry);
      }
    };
    const observeTargetResult = () => {
      pending.prevented ||= event.defaultPrevented;
    };
    this.pendingDefaults.set(event, pending);
    target?.addEventListener(type, observeTargetResult, { once: true });
    setTimeout(pending.finalize, 0);
  };
  /** Retained records ordered from oldest to newest. */
  get entries() {
    return this.records;
  }
  /** Subscribe to completed records. Returns an idempotent unsubscribe function. */
  subscribe(listener) {
    if (this.destroyed) return () => {
    };
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }
  /** Remove retained records without affecting observation. */
  clear() {
    this.records.length = 0;
  }
  /** Stop observing and release subscribers. Safe to call more than once. */
  destroy() {
    if (this.destroyed) return;
    this.destroyed = true;
    if (typeof document !== "undefined") {
      for (const type of TRACE_TYPES) {
        document.removeEventListener(type, this.onEvent, true);
        document.removeEventListener(type, this.onEventBubbled);
      }
    }
    this.listeners.clear();
  }
  resolveContext(event, type) {
    const projected = this.projectedTarget(event.target);
    if (projected) return projected;
    const target = event.target;
    if (target === this.scene.canvas) {
      const point = scenePointOf(this.scene, event);
      return { source: "canvas", target: point ? pickInScene(this.scene, point.x, point.y) : null };
    }
    if (isKeyboardType(type) && this.includeGlobalKeyboard)
      return { source: "document", target: null };
    return null;
  }
  projectedTarget(target) {
    if (!(target instanceof Element)) return null;
    const element = target.closest("[data-vecto-content], [data-vecto-id]");
    if (!element) return null;
    const source = element.dataset.vectoContent ? "content" : "a11y";
    const id = element.dataset.vectoContent ?? element.dataset.vectoId;
    if (!id) return null;
    const projectedElement = source === "content" ? typeof this.scene.getContentElement === "function" ? this.scene.getContentElement(id) : element.isConnected ? element : void 0 : this.scene.getA11yElement(id);
    if (projectedElement !== element) return null;
    const entity = findEntityById(this.scene.rootEntity, id) ?? findEntityById(this.scene.overlayRootEntity, id);
    if (!entity || source === "content" && !entity.getContentProjection()) return null;
    return { source, target: entity };
  }
  createEntry(event, type, context, defaultPrevented) {
    const pointer = scenePointOf(this.scene, event);
    const local = context.target && pointer ? context.target.worldToLocal(pointer.x, pointer.y) : null;
    const keyboard = event;
    const wheel = event;
    return Object.freeze({
      type,
      source: context.source,
      timeStamp: event.timeStamp,
      ...context.target ? { targetId: context.target.id, targetPath: entityPath(context.target) } : {},
      ...pointer ? { sceneX: pointer.x, sceneY: pointer.y } : {},
      ...local ? { localX: local.x, localY: local.y } : {},
      ...isKeyboardType(type) ? { key: keyboard.key } : {},
      modifiers: {
        shift: keyboard.shiftKey,
        ctrl: keyboard.ctrlKey,
        alt: keyboard.altKey,
        meta: keyboard.metaKey
      },
      ...type === "wheel" ? { deltaX: wheel.deltaX, deltaY: wheel.deltaY } : {},
      defaultPrevented
    });
  }
};
function scenePointOf(scene, event) {
  const candidate = event;
  if (typeof candidate.clientX !== "number" || typeof candidate.clientY !== "number") return null;
  return scene.clientToScene(candidate.clientX, candidate.clientY);
}
function findEntityById(root, id) {
  if (root.id === id) return root;
  for (const child of root.children) {
    const found = findEntityById(child, id);
    if (found) return found;
  }
  return null;
}
function createEventTrace(scene, options) {
  return new EventTrace(scene, options);
}

// src/panel.ts
var PANEL_BG = "rgba(10, 14, 24, 0.96)";
var PANEL_FG = "#cbd5e1";
var ACCENT = "#38bdf8";
var DevtoolsPanel = class {
  host;
  container;
  panelScene;
  tree;
  detailLines = [];
  traceLines = [];
  eventTrace = null;
  index = /* @__PURE__ */ new Map();
  selected = null;
  highlight = null;
  refreshTimer = null;
  pickArmed = false;
  width;
  destroyed = false;
  findings = [];
  onHostPick = (ev) => {
    if (!this.pickArmed) return;
    this.pickArmed = false;
    const point = this.host.clientToScene(ev.clientX, ev.clientY);
    const hit = pickInScene(this.host, point.x, point.y);
    if (hit) this.select(hit);
    ev.stopPropagation();
    ev.preventDefault();
  };
  onKeyNudge = (ev) => {
    if (!this.selected) return;
    const target = ev.target;
    if (target && (target.tagName === "INPUT" || target.tagName === "TEXTAREA")) return;
    const step = ev.shiftKey ? 10 : 1;
    let handled = true;
    switch (ev.key) {
      case "ArrowLeft":
        this.selected.x -= step;
        break;
      case "ArrowRight":
        this.selected.x += step;
        break;
      case "ArrowUp":
        this.selected.y -= step;
        break;
      case "ArrowDown":
        this.selected.y += step;
        break;
      case "+":
        this.selected.opacity = Math.min(1, this.selected.opacity + 0.1);
        break;
      case "-":
        this.selected.opacity = Math.max(0, this.selected.opacity - 0.1);
        break;
      default:
        handled = false;
    }
    if (handled) {
      ev.preventDefault();
      this.host.markDirty();
      this.select(this.selected);
    }
  };
  constructor(host, options = {}) {
    this.host = host;
    this.width = options.width ?? 320;
    const height = typeof window !== "undefined" ? window.innerHeight : 600;
    this.container = document.createElement("div");
    this.container.setAttribute("data-vecto-devtools", "");
    const cs = this.container.style;
    cs.position = "fixed";
    cs.top = "0";
    cs.right = "0";
    cs.width = `${this.width}px`;
    cs.height = "100%";
    cs.zIndex = "99999";
    cs.background = PANEL_BG;
    cs.borderLeft = `1px solid ${ACCENT}`;
    cs.pointerEvents = "none";
    const canvas = document.createElement("canvas");
    canvas.width = this.width;
    canvas.height = height;
    canvas.style.pointerEvents = "none";
    this.container.appendChild(canvas);
    document.body.appendChild(this.container);
    this.panelScene = new Scene(canvas, {
      disableWindowResize: true,
      contentProjection: false
    });
    this.panelScene.renderMode = "onDemand";
    const title = new Text("VectoJS DevTools", { font: "bold 14px monospace" });
    title.setPosition(12, 10);
    this.panelScene.add(title);
    const pickBtn = new Button("Pick", { width: 64, height: 24 });
    pickBtn.setPosition(12, 34);
    pickBtn.on("click", () => this.armPick());
    this.panelScene.add(pickBtn);
    const refreshBtn = new Button("Refresh", { width: 84, height: 24 });
    refreshBtn.setPosition(84, 34);
    refreshBtn.on("click", () => this.refresh());
    this.panelScene.add(refreshBtn);
    const auditBtn = new Button("Audit", { width: 64, height: 24 });
    auditBtn.setPosition(176, 34);
    auditBtn.on("click", () => this.audit());
    this.panelScene.add(auditBtn);
    const treeHeight = Math.max(160, Math.floor(height * (options.traceEvents ? 0.42 : 0.55)));
    this.tree = new TreeView({
      nodes: [],
      width: this.width - 16,
      height: treeHeight,
      rowHeight: 20,
      font: "12px monospace",
      color: PANEL_FG,
      selectedColor: ACCENT,
      onSelect: (node) => {
        const findingMatch = /^finding:(\d+)$/.exec(node.id);
        if (findingMatch) {
          this.selectFinding(Number(findingMatch[1]));
          return;
        }
        const entity = this.index.get(node.id);
        if (entity) this.select(entity);
      }
    });
    this.tree.setPosition(8, 68);
    this.panelScene.add(this.tree);
    const detailTop = 68 + treeHeight + 12;
    for (let i = 0; i < 8; i++) {
      const line = new Text("", { font: "11px monospace" });
      line.setPosition(12, detailTop + i * 16);
      this.detailLines.push(line);
      this.panelScene.add(line);
    }
    if (options.traceEvents) {
      this.eventTrace = createEventTrace(this.host, { capacity: options.traceCapacity });
      const traceTop = detailTop + this.detailLines.length * 16 + 12;
      for (let i = 0; i < 4; i++) {
        const line = new Text("", { font: "11px monospace" });
        line.setPosition(12, traceTop + i * 16);
        this.traceLines.push(line);
        this.panelScene.add(line);
      }
      this.eventTrace.subscribe(() => {
        this.writeTrace();
        this.panelScene.markDirty();
      });
      this.writeTrace();
    }
    document.addEventListener("click", this.onHostPick, true);
    document.addEventListener("keydown", this.onKeyNudge);
    const interval = options.refreshInterval ?? 500;
    if (interval > 0) {
      this.refreshTimer = setInterval(() => this.refresh(), interval);
    }
    this.panelScene.start();
    this.refresh();
  }
  /** Rebuild the tree model from the host scene. */
  refresh() {
    if (this.destroyed) return;
    const { nodes, index } = buildTreeModel(this.host.rootEntity);
    const overlay = buildTreeModel(this.host.overlayRootEntity);
    for (const [id, entity] of overlay.index) index.set(id, entity);
    this.index = index;
    this.tree.setNodes([...nodes, ...overlay.nodes]);
    if (this.selected) this.writeDetails(this.selected);
    this.panelScene.markDirty();
  }
  /** Arm one-shot pick mode: the next click on the page selects the entity under it. */
  armPick() {
    this.pickArmed = true;
  }
  /**
   * Run the layout audit on the host scene and list the findings in place of
   * the entity tree (hit Refresh to restore it). Selecting a finding selects
   * and highlights the offending entity. Returns the findings so agents and
   * tests can drive the panel programmatically.
   */
  audit() {
    if (this.destroyed) return [];
    const { index } = buildTreeModel(this.host.rootEntity);
    const overlay = buildTreeModel(this.host.overlayRootEntity);
    for (const [id, entity] of overlay.index) index.set(id, entity);
    this.index = index;
    this.findings = auditScene(this.host);
    this.tree.setNodes(
      this.findings.map((f, i) => ({
        id: `finding:${i}`,
        label: `\u26A0 ${f.kind}: ${f.message}`
      }))
    );
    this.detailLines[0]?.setText(
      this.findings.length === 0 ? "audit clean" : `${this.findings.length} finding(s)`
    );
    this.panelScene.markDirty();
    return this.findings;
  }
  /** Select and highlight the entity behind finding `i` from the last {@link audit} run. */
  selectFinding(i) {
    const finding = this.findings[i];
    const entity = finding ? this.index.get(finding.entityId) : void 0;
    if (entity) this.select(entity);
  }
  /** Select an entity: highlight it on the host scene and show its state. */
  select(entity) {
    this.selected = entity;
    if (!this.highlight) {
      this.highlight = new HighlightEntity();
      this.host.showOverlay(this.highlight);
    }
    this.highlight.track(entity);
    this.host.markDirty();
    this.writeDetails(entity);
    this.panelScene.markDirty();
  }
  /** The currently selected entity, if any. */
  get selection() {
    return this.selected;
  }
  /** Optional generic routing trace, enabled with `traceEvents`. */
  get trace() {
    return this.eventTrace;
  }
  writeDetails(entity) {
    const lines = describeEntity(entity);
    for (let i = 0; i < this.detailLines.length; i++) {
      this.detailLines[i].setText(lines[i] ?? "");
    }
  }
  writeTrace() {
    if (!this.eventTrace) return;
    const recent = this.eventTrace.entries.slice(-3).reverse();
    this.traceLines[0]?.setText(`trace ${this.eventTrace.entries.length} event(s)`);
    for (let i = 1; i < this.traceLines.length; i++) {
      const entry = recent[i - 1];
      this.traceLines[i].setText(
        entry ? `${entry.type} ${entry.source} ${entry.targetId?.slice(0, 8) ?? entry.key ?? ""}` : ""
      );
    }
  }
  /** Tear down the panel, host highlight, listeners, and timers. */
  destroy() {
    if (this.destroyed) return;
    this.destroyed = true;
    if (this.refreshTimer) clearInterval(this.refreshTimer);
    this.eventTrace?.destroy();
    document.removeEventListener("click", this.onHostPick, true);
    document.removeEventListener("keydown", this.onKeyNudge);
    if (this.highlight) {
      this.host.hideOverlay(this.highlight);
      this.highlight.destroy();
      this.highlight = null;
    }
    this.panelScene.destroy();
    this.container.remove();
  }
};
var HighlightEntity = class extends Entity {
  target = null;
  track(target) {
    this.target = target;
  }
  isPointInside() {
    return false;
  }
  getContentProjection() {
    return null;
  }
  render(r) {
    const t = this.target;
    if (!t) return;
    const { a, b, c, d, e, f } = t.getWorldTransform();
    const w = t.width || 8;
    const h = t.height || 8;
    let minX = Infinity;
    let minY = Infinity;
    let maxX = -Infinity;
    let maxY = -Infinity;
    for (let i = 0; i < 4; i++) {
      const lx = i & 1 ? w : 0;
      const ly = i & 2 ? h : 0;
      const wx = a * lx + c * ly + e;
      const wy = b * lx + d * ly + f;
      if (wx < minX) minX = wx;
      if (wx > maxX) maxX = wx;
      if (wy < minY) minY = wy;
      if (wy > maxY) maxY = wy;
    }
    r.beginPath();
    r.roundRect(minX - 1, minY - 1, maxX - minX + 2, maxY - minY + 2, 2);
    r.stroke(ACCENT, 2);
  }
};

// src/snapshot.ts
var round23 = (n) => Math.round(n * 100) / 100;
function toNode(entity) {
  const wb = entity.getWorldBounds();
  const node = {
    type: entity.constructor.name,
    id: entity.id,
    x: round23(entity.x),
    y: round23(entity.y),
    width: round23(entity.width),
    height: round23(entity.height),
    worldBounds: {
      x: round23(wb.x),
      y: round23(wb.y),
      width: round23(wb.width),
      height: round23(wb.height)
    }
  };
  if (entity.opacity !== 1) node.opacity = round23(entity.opacity);
  if (entity.interactive) node.interactive = true;
  if (entity.hasPendingAnimations()) node.animating = true;
  if (entity.clipChildren) node.clipChildren = true;
  const text = textPreviewOf(entity);
  if (text !== void 0) node.text = text;
  if (entity.children.length > 0) node.children = entity.children.map(toNode);
  return node;
}
function captureSnapshot(scene) {
  return {
    width: round23(scene.width),
    height: round23(scene.height),
    root: scene.rootEntity.children.map(toNode),
    overlay: scene.overlayRootEntity.children.map(toNode)
  };
}
var COMPARED_KEYS = [
  "type",
  "x",
  "y",
  "width",
  "height",
  "worldBounds",
  "opacity",
  "interactive",
  "animating",
  "clipChildren",
  "text"
];
function pathOf(parent, node, index) {
  return `${parent} > ${node.type}[${index}]`;
}
function diffNodes(a, b, parent, out) {
  const max = Math.max(a.length, b.length);
  for (let i = 0; i < max; i++) {
    const na = a[i];
    const nb = b[i];
    if (na && !nb) {
      out.push({ path: pathOf(parent, na, i), kind: "removed" });
      continue;
    }
    if (!na && nb) {
      out.push({ path: pathOf(parent, nb, i), kind: "added" });
      continue;
    }
    if (!na || !nb) continue;
    const path = pathOf(parent, nb, i);
    const changes = {};
    for (const key of COMPARED_KEYS) {
      const va = na[key];
      const vb = nb[key];
      const same = key === "worldBounds" ? JSON.stringify(va) === JSON.stringify(vb) : va === vb;
      if (!same) changes[key] = { from: va, to: vb };
    }
    if (Object.keys(changes).length > 0) out.push({ path, kind: "changed", changes });
    diffNodes(na.children ?? [], nb.children ?? [], path, out);
  }
}
function diffSnapshots(a, b) {
  const out = [];
  diffNodes(a.root, b.root, "root", out);
  diffNodes(a.overlay, b.overlay, "overlay", out);
  return out;
}

// src/index.ts
function attachDevtools(scene, options) {
  const panel = new DevtoolsPanel(scene, options);
  panel.detach = () => panel.destroy();
  return panel;
}
export {
  DevtoolsPanel,
  EventTrace,
  attachDevtools,
  auditScene,
  auditTree,
  buildTreeModel,
  captureSnapshot,
  createEventTrace,
  describeEntity,
  diffSnapshots,
  entityPath,
  findEntityAt,
  inspectEntity,
  pickInScene,
  textPreviewOf
};
